{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f798bf75-0dc4-40a8-ba00-be26e6b1c558",
   "metadata": {},
   "source": [
    "# 环境设置"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1be0a1cf-8c48-462e-b6b1-099aa307d671",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "                                                        \n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "import random\n",
    "\n",
    "import torch\n",
    "\n",
    "import sys\n",
    "\n",
    "\n",
    "\n",
    "os.chdir(os.path.join(os.getcwd(), '..'))\n",
    "\n",
    "if \"..\" not in sys.path:\n",
    "    sys.path.insert(0, \"..\")\n",
    "\n",
    "data_dir = '../data'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b85a5c9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "536f7330",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/maindata/data/user/user_wan/yul/ExpSPT'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "os.getcwd()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f437a21-7c0c-4c05-b722-ecd5f0068ae5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "580.5550000000001"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "1733*0.335\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e752ecc1-f592-4a98-9ad1-48d7a38174db",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe87a043-d5da-4e15-8282-b3a02f17dc60",
   "metadata": {},
   "source": [
    "# 数据"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84d09852-859f-49f2-9949-c3b6c5e0ff12",
   "metadata": {},
   "source": [
    "## proteinatlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6cda1f8b-e18c-499a-b1a9-5b9ce478d825",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>IHC tissue name</th>\n",
       "      <th>Cell type</th>\n",
       "      <th>Level</th>\n",
       "      <th>Reliability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Adipose tissue</td>\n",
       "      <td>Adipose tissue</td>\n",
       "      <td>adipocytes</td>\n",
       "      <td>Not detected</td>\n",
       "      <td>Approved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Adrenal gland</td>\n",
       "      <td>Adrenal gland</td>\n",
       "      <td>glandular cells</td>\n",
       "      <td>Not detected</td>\n",
       "      <td>Approved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Appendix</td>\n",
       "      <td>Appendix</td>\n",
       "      <td>glandular cells</td>\n",
       "      <td>Medium</td>\n",
       "      <td>Approved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Appendix</td>\n",
       "      <td>Appendix</td>\n",
       "      <td>lymphoid tissue</td>\n",
       "      <td>Not detected</td>\n",
       "      <td>Approved</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>Bone marrow</td>\n",
       "      <td>Bone marrow</td>\n",
       "      <td>hematopoietic cells</td>\n",
       "      <td>Not detected</td>\n",
       "      <td>Approved</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene Gene name          Tissue IHC tissue name  \\\n",
       "0  ENSG00000000003    TSPAN6  Adipose tissue  Adipose tissue   \n",
       "1  ENSG00000000003    TSPAN6   Adrenal gland   Adrenal gland   \n",
       "2  ENSG00000000003    TSPAN6        Appendix        Appendix   \n",
       "3  ENSG00000000003    TSPAN6        Appendix        Appendix   \n",
       "4  ENSG00000000003    TSPAN6     Bone marrow     Bone marrow   \n",
       "\n",
       "             Cell type         Level Reliability  \n",
       "0           adipocytes  Not detected    Approved  \n",
       "1      glandular cells  Not detected    Approved  \n",
       "2      glandular cells        Medium    Approved  \n",
       "3      lymphoid tissue  Not detected    Approved  \n",
       "4  hematopoietic cells  Not detected    Approved  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "proteinatlas = pd.read_csv(os.path.join('data/proteinatlas', 'normal_ihc_data.tsv'), sep='\\t')\n",
    "\n",
    "proteinatlas.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b4ecaff7-a69c-4c11-9bc9-e985c962de87",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Cell type\n",
       "glandular cells              259573\n",
       "squamous epithelial cells     65648\n",
       "glial cells                   40409\n",
       "neuronal cells                40198\n",
       "fibroblasts                   39998\n",
       "                              ...  \n",
       "ductal cells                      9\n",
       "cells in posterior                6\n",
       "cells in dentate nucleus          4\n",
       "ependymal cells                   4\n",
       "atretic follicle                  1\n",
       "Name: count, Length: 148, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "proteinatlas['Level'].value_counts()\n",
    "proteinatlas['Tissue'].value_counts()\n",
    "\n",
    "proteinatlas['Cell type'].value_counts().shape\n",
    "\n",
    "# proteinatlas.shape\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f5e9391d-8272-4b26-9b2e-42abf9a7646e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>TPM</th>\n",
       "      <th>pTPM</th>\n",
       "      <th>nTPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>28.7</td>\n",
       "      <td>35.6</td>\n",
       "      <td>28.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>15.2</td>\n",
       "      <td>18.4</td>\n",
       "      <td>17.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>amygdala</td>\n",
       "      <td>7.2</td>\n",
       "      <td>9.1</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>basal ganglia</td>\n",
       "      <td>7.8</td>\n",
       "      <td>9.6</td>\n",
       "      <td>12.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>breast</td>\n",
       "      <td>32.5</td>\n",
       "      <td>40.8</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene Gene name          Tissue   TPM  pTPM  nTPM\n",
       "0  ENSG00000000003    TSPAN6  adipose tissue  28.7  35.6  28.6\n",
       "1  ENSG00000000003    TSPAN6   adrenal gland  15.2  18.4  17.6\n",
       "2  ENSG00000000003    TSPAN6        amygdala   7.2   9.1  13.0\n",
       "3  ENSG00000000003    TSPAN6   basal ganglia   7.8   9.6  12.4\n",
       "4  ENSG00000000003    TSPAN6          breast  32.5  40.8  33.0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "proteinatlas_gtex = pd.read_csv(os.path.join('data/proteinatlas', 'rna_tissue_gtex.tsv'), sep='\\t')\n",
    "\n",
    "proteinatlas_gtex.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "629a5f46-6b71-427d-bb24-61d37082a1b1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(675206, 6)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteinatlas_gtex['Tissue'].value_counts()\n",
    "\n",
    "proteinatlas_gtex.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e6f73e88-c590-4202-bc47-a90a4bb584b6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "62"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_set = set(proteinatlas['Tissue'].value_counts().reset_index()['index'].apply(lambda x: x.lower()).values) & set(proteinatlas_gtex['Tissue'].value_counts().index)\n",
    "\n",
    "len(common_set)\n",
    "\n",
    "union_set = set(proteinatlas['Tissue'].value_counts().reset_index()['index'].apply(lambda x: x.lower()).values) | set(proteinatlas_gtex['Tissue'].value_counts().index)\n",
    "\n",
    "len(union_set)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c1034896-3e20-49f9-b7f7-0da2f77022b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "set(proteinatlas['Tissue'].value_counts().index) & set(proteinatlas_gtex['Tissue'].value_counts().index)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "05cee27f-88b7-4d8a-bee3-68ff0747fe59",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hpa_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>celline</th>\n",
       "      <th>summary_expression_value</th>\n",
       "      <th>tissue</th>\n",
       "      <th>intensity</th>\n",
       "      <th>fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAB000001</td>\n",
       "      <td>ENSG00000169083</td>\n",
       "      <td>A-431</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAB000001</td>\n",
       "      <td>ENSG00000169083</td>\n",
       "      <td>A-549</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAB000001</td>\n",
       "      <td>ENSG00000169083</td>\n",
       "      <td>AN3-CA</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAB000001</td>\n",
       "      <td>ENSG00000169083</td>\n",
       "      <td>BEWO</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAB000001</td>\n",
       "      <td>ENSG00000169083</td>\n",
       "      <td>CACO-2</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      hpa_id  ensembl_gene_id celline summary_expression_value tissue  \\\n",
       "0  CAB000001  ENSG00000169083   A-431                 negative    NaN   \n",
       "1  CAB000001  ENSG00000169083   A-549                 negative    NaN   \n",
       "2  CAB000001  ENSG00000169083  AN3-CA                 negative    NaN   \n",
       "3  CAB000001  ENSG00000169083    BEWO                 negative    NaN   \n",
       "4  CAB000001  ENSG00000169083  CACO-2                 negative    NaN   \n",
       "\n",
       "  intensity fraction  \n",
       "0       NaN      NaN  \n",
       "1       NaN      NaN  \n",
       "2       NaN      NaN  \n",
       "3       NaN      NaN  \n",
       "4       NaN      NaN  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# 导入数据\n",
    "\n",
    "CMA_expression_data = pd.read_csv(os.path.join('data/proteinatlas', 'expression_data.CMA.csv'), sep=',')\n",
    "\n",
    "TMA_expression_data = pd.read_csv(os.path.join('data/proteinatlas', 'expression_data.TMA.csv'), sep=',')\n",
    "\n",
    "IF_expression_data = pd.read_csv(os.path.join('data/proteinatlas', 'expression_data.IF.csv'), sep=',')\n",
    "\n",
    "\n",
    "# 合并\n",
    "whole_expression_data =  pd.DataFrame(columns=['hpa_id', 'ensembl_gene_id', 'celline', 'summary_expression_value'])\n",
    "\n",
    "whole_expression_data = pd.concat([whole_expression_data, CMA_expression_data, TMA_expression_data, IF_expression_data])\n",
    "\n",
    "whole_expression_data.head()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8e0cdd83-246c-4305-bce7-2b06b1eba140",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(45,)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "whole_expression_data['celline'].value_counts().shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "abf40ae2-044b-435a-8100-d21eacb5ed46",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "218"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Tissues: 62\n",
    "# Cellines: 45\n",
    "# Seq: 101\n",
    "# RNA_Exp:3\n",
    "# Protein_Exp: 7\n",
    "# special tokens:\n",
    "    \n",
    "\n",
    "62+45+101+3+7\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edba9ab2-47b5-4567-b6c1-61b6f4bac543",
   "metadata": {},
   "source": [
    "## GTEx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6fa71f59-1378-453d-9f63-780eb8dd3d11",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "GTEx = pd.read_csv(os.path.join('data/GTEx', 'GTEx_Analysis_v10_RNASeQCv2.4.2_gene_tpm.gct.tsv'), sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0f908866-4326-4edb-99f5-e1f4212e3d82",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "      <th>GTEX-1117F-0005-SM-HL9SH</th>\n",
       "      <th>GTEX-1117F-0011-R10b-SM-GI4VE</th>\n",
       "      <th>GTEX-1117F-0011-R11b-SM-GIN8R</th>\n",
       "      <th>GTEX-1117F-0011-R2b-SM-GI4VL</th>\n",
       "      <th>GTEX-1117F-0011-R3a-SM-GJ3PJ</th>\n",
       "      <th>GTEX-1117F-0011-R4b-SM-GI4VM</th>\n",
       "      <th>GTEX-1117F-0011-R5a-SM-GI4VW</th>\n",
       "      <th>GTEX-1117F-0011-R6a-SM-GI4VX</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZZPU-1326-SM-5GZWS</th>\n",
       "      <th>GTEX-ZZPU-1426-SM-5GZZ6</th>\n",
       "      <th>GTEX-ZZPU-1826-SM-5E43L</th>\n",
       "      <th>GTEX-ZZPU-2126-SM-5EGIU</th>\n",
       "      <th>GTEX-ZZPU-2226-SM-5EGIV</th>\n",
       "      <th>GTEX-ZZPU-2326-SM-GOQYU</th>\n",
       "      <th>GTEX-ZZPU-2426-SM-5E44I</th>\n",
       "      <th>GTEX-ZZPU-2526-SM-GOQZ3</th>\n",
       "      <th>GTEX-ZZPU-2626-SM-5E45Y</th>\n",
       "      <th>GTEX-ZZPU-2726-SM-5NQ8O</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.012397</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.029395</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.017791</td>\n",
       "      <td>0.020221</td>\n",
       "      <td>0.025713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1.33343</td>\n",
       "      <td>3.579280</td>\n",
       "      <td>10.189300</td>\n",
       "      <td>2.96650</td>\n",
       "      <td>3.6828</td>\n",
       "      <td>3.599210</td>\n",
       "      <td>3.51443</td>\n",
       "      <td>4.286380</td>\n",
       "      <td>...</td>\n",
       "      <td>5.464380</td>\n",
       "      <td>2.3599</td>\n",
       "      <td>2.427010</td>\n",
       "      <td>3.87837</td>\n",
       "      <td>1.85892</td>\n",
       "      <td>2.383760</td>\n",
       "      <td>2.886100</td>\n",
       "      <td>5.095810</td>\n",
       "      <td>0.842465</td>\n",
       "      <td>2.209520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.093825</td>\n",
       "      <td>0.034191</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.024750</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.062071</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.086553</td>\n",
       "      <td>0.14685</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.041073</td>\n",
       "      <td>0.053323</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.01766</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19618 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Name  Description  GTEX-1117F-0005-SM-HL9SH  \\\n",
       "0  ENSG00000223972.5      DDX11L1                   0.00000   \n",
       "1  ENSG00000227232.5       WASH7P                   1.33343   \n",
       "2  ENSG00000278267.1    MIR6859-1                   0.00000   \n",
       "3  ENSG00000243485.5  MIR1302-2HG                   0.00000   \n",
       "4  ENSG00000237613.2      FAM138A                   0.00000   \n",
       "\n",
       "   GTEX-1117F-0011-R10b-SM-GI4VE  GTEX-1117F-0011-R11b-SM-GIN8R  \\\n",
       "0                       0.000000                       0.000000   \n",
       "1                       3.579280                      10.189300   \n",
       "2                       0.000000                       0.000000   \n",
       "3                       0.093825                       0.034191   \n",
       "4                       0.000000                       0.000000   \n",
       "\n",
       "   GTEX-1117F-0011-R2b-SM-GI4VL  GTEX-1117F-0011-R3a-SM-GJ3PJ  \\\n",
       "0                       0.00000                        0.0000   \n",
       "1                       2.96650                        3.6828   \n",
       "2                       0.00000                        0.0000   \n",
       "3                       0.00000                        0.0000   \n",
       "4                       0.01766                        0.0000   \n",
       "\n",
       "   GTEX-1117F-0011-R4b-SM-GI4VM  GTEX-1117F-0011-R5a-SM-GI4VW  \\\n",
       "0                      0.012397                       0.00000   \n",
       "1                      3.599210                       3.51443   \n",
       "2                      0.000000                       0.00000   \n",
       "3                      0.024750                       0.00000   \n",
       "4                      0.000000                       0.00000   \n",
       "\n",
       "   GTEX-1117F-0011-R6a-SM-GI4VX  ...  GTEX-ZZPU-1326-SM-5GZWS  \\\n",
       "0                      0.029395  ...                 0.000000   \n",
       "1                      4.286380  ...                 5.464380   \n",
       "2                      0.000000  ...                 0.000000   \n",
       "3                      0.000000  ...                 0.062071   \n",
       "4                      0.000000  ...                 0.000000   \n",
       "\n",
       "   GTEX-ZZPU-1426-SM-5GZZ6  GTEX-ZZPU-1826-SM-5E43L  GTEX-ZZPU-2126-SM-5EGIU  \\\n",
       "0                   0.0000                 0.000000                  0.00000   \n",
       "1                   2.3599                 2.427010                  3.87837   \n",
       "2                   0.0000                 0.000000                  0.00000   \n",
       "3                   0.0000                 0.086553                  0.14685   \n",
       "4                   0.0000                 0.000000                  0.00000   \n",
       "\n",
       "   GTEX-ZZPU-2226-SM-5EGIV  GTEX-ZZPU-2326-SM-GOQYU  GTEX-ZZPU-2426-SM-5E44I  \\\n",
       "0                  0.00000                 0.000000                 0.000000   \n",
       "1                  1.85892                 2.383760                 2.886100   \n",
       "2                  0.00000                 0.000000                 0.000000   \n",
       "3                  0.00000                 0.041073                 0.053323   \n",
       "4                  0.00000                 0.000000                 0.000000   \n",
       "\n",
       "   GTEX-ZZPU-2526-SM-GOQZ3  GTEX-ZZPU-2626-SM-5E45Y  GTEX-ZZPU-2726-SM-5NQ8O  \n",
       "0                 0.017791                 0.020221                 0.025713  \n",
       "1                 5.095810                 0.842465                 2.209520  \n",
       "2                 0.000000                 0.000000                 0.000000  \n",
       "3                 0.000000                 0.000000                 0.000000  \n",
       "4                 0.000000                 0.000000                 0.000000  \n",
       "\n",
       "[5 rows x 19618 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GTEx.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "96be650d-5d78-455d-91cc-f7da803273ff",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(51970, 19618)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "GTEx.shape\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d4a749a-e24b-4d9c-9c74-9cf537822384",
   "metadata": {},
   "source": [
    "## public cellines "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09927b50-13ab-41c7-8031-ed5af3107d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c61a506-fdd2-4da5-8d14-324cd284f6ea",
   "metadata": {},
   "source": [
    "## Modern  MRL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1c06529-48be-4b99-b874-67744407ed18",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2917fc28-7426-4d8e-a1fd-e71308fa5ad0",
   "metadata": {},
   "source": [
    "## annotation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e687bfd0-2db9-48ec-a319-0557e0f25067",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "from dataset.tokenizer import ntTokenizer, CDSTokenizer, Vocabulary\n",
    "\n",
    "from dataset.data_prepare import _load_mRNA\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "72ae3f9b-dd1a-4476-ab73-dbb614c36fed",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/processed/mRNAfna/Human_gencode.v40.pc_transcripts.fa.norm.fa:\t88659\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "ncRNA_fn = 'data/processed/mRNAfna/Human_gencode.v40.pc_transcripts.fa.norm.fa'\n",
    "\n",
    "\n",
    "mRNAs = _load_mRNA(ncRNA_fn, compress=False)\n",
    "\n",
    "print('{}:\\t{}'.format(ncRNA_fn, len(mRNAs)))\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bcbc4f95-5967-4300-9704-aad0ed3a11eb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21.19252000252406"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1271.5512001514435/60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9fb382c9-632d-4956-8520-6404bee0ef0b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "      <th>utr5</th>\n",
       "      <th>cds</th>\n",
       "      <th>utr3</th>\n",
       "      <th>aa</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...</td>\n",
       "      <td>CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...</td>\n",
       "      <td>CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...</td>\n",
       "      <td>ATGAAGAAGGTAACTGCAGAGGCTATTTCCTGGAATGAATCAACGA...</td>\n",
       "      <td>ATCTTATATAACTGTGAGATTAATCTCAGATAATGACACAAAATAT...</td>\n",
       "      <td>MKKVTAEAISWNESTSETNNSMVTEFIFLGLSDSQELQTFLFMLFF...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;ENST00000426406.4|ENSG00000284733.2|OTTHUMG00...</td>\n",
       "      <td>ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...</td>\n",
       "      <td></td>\n",
       "      <td>ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...</td>\n",
       "      <td></td>\n",
       "      <td>MDGENHSVVSEFLFLGLTHSWEIQLLLLVFSSVLYVASITGNILIV...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;ENST00000332831.5|ENSG00000284662.2|OTTHUMG00...</td>\n",
       "      <td>ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...</td>\n",
       "      <td></td>\n",
       "      <td>ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...</td>\n",
       "      <td></td>\n",
       "      <td>MDGENHSVVSEFLFLGLTHSWEIQLLLLVFSSVLYVASITGNILIV...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;ENST00000616016.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>GGCGGCGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTG...</td>\n",
       "      <td>GGCGGCGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTG...</td>\n",
       "      <td>ATGCCGGCGGTCAAGAAGGAGTTCCCGGGCCGCGAGGACCTGGCCC...</td>\n",
       "      <td>GGTTGCCGGGGGTAGGGGTGGGGCCACACAAATCTCCAGGAGCCAC...</td>\n",
       "      <td>MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAAL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;ENST00000618323.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>GGCGGCGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTG...</td>\n",
       "      <td>GGCGGCGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTG...</td>\n",
       "      <td>ATGCCGGCGGTCAAGAAGGAGTTCCCGGGCCGCGAGGACCTGGCCC...</td>\n",
       "      <td>GGTTGCCGGGGGTAGGGGTGGGGCCACACAAATCTCCAGGAGCCAC...</td>\n",
       "      <td>MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAAL...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ids  \\\n",
       "0  >ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...   \n",
       "1  >ENST00000426406.4|ENSG00000284733.2|OTTHUMG00...   \n",
       "2  >ENST00000332831.5|ENSG00000284662.2|OTTHUMG00...   \n",
       "3  >ENST00000616016.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "4  >ENST00000618323.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "\n",
       "                                                 seq  \\\n",
       "0  CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...   \n",
       "1  ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...   \n",
       "2  ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...   \n",
       "3  GGCGGCGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTG...   \n",
       "4  GGCGGCGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTG...   \n",
       "\n",
       "                                                utr5  \\\n",
       "0  CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...   \n",
       "1                                                      \n",
       "2                                                      \n",
       "3  GGCGGCGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTG...   \n",
       "4  GGCGGCGGAGTCTCCCAAGTCCCCGCCGGGCGGGCGCGCGCCAGTG...   \n",
       "\n",
       "                                                 cds  \\\n",
       "0  ATGAAGAAGGTAACTGCAGAGGCTATTTCCTGGAATGAATCAACGA...   \n",
       "1  ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...   \n",
       "2  ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...   \n",
       "3  ATGCCGGCGGTCAAGAAGGAGTTCCCGGGCCGCGAGGACCTGGCCC...   \n",
       "4  ATGCCGGCGGTCAAGAAGGAGTTCCCGGGCCGCGAGGACCTGGCCC...   \n",
       "\n",
       "                                                utr3  \\\n",
       "0  ATCTTATATAACTGTGAGATTAATCTCAGATAATGACACAAAATAT...   \n",
       "1                                                      \n",
       "2                                                      \n",
       "3  GGTTGCCGGGGGTAGGGGTGGGGCCACACAAATCTCCAGGAGCCAC...   \n",
       "4  GGTTGCCGGGGGTAGGGGTGGGGCCACACAAATCTCCAGGAGCCAC...   \n",
       "\n",
       "                                                  aa label  \n",
       "0  MKKVTAEAISWNESTSETNNSMVTEFIFLGLSDSQELQTFLFMLFF...  None  \n",
       "1  MDGENHSVVSEFLFLGLTHSWEIQLLLLVFSSVLYVASITGNILIV...  None  \n",
       "2  MDGENHSVVSEFLFLGLTHSWEIQLLLLVFSSVLYVASITGNILIV...  None  \n",
       "3  MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAAL...  None  \n",
       "4  MPAVKKEFPGREDLALALATFHPTLAALPLPPLPGYLAPLPAAAAL...  None  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "RNA_data = pd.DataFrame(mRNAs, columns=['ids', 'seq', 'utr5', 'cds', 'utr3', 'aa', 'label'])\n",
    "\n",
    "\n",
    "RNA_data.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b6a72fd7-8305-40dd-90f5-aad3685e5a5d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "RNA_aa_data = RNA_data.drop_duplicates(subset=['aa'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bb375093-4d0c-449f-990a-aeba61b7da36",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(72415, 7)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "RNA_aa_data.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c1590909-717e-45d9-a690-ad286a2fdfe8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/maindata/data/user/user_wan/yul/bioSPT2'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d61fd061-a628-4a63-88e8-141d41d5496a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "with open('data/processed/mRNAfna/human_aa.fa', 'w') as f:\n",
    "    for i,v in RNA_aa_data.iterrows():\n",
    "\n",
    "        f.write(v['ids']+ '\\n' + v['aa']+ '\\n')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37cdb2a0-edf2-4944-8ec8-318f3cd53d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7eb9cc-5bae-4f72-864a-4acc565dd5ab",
   "metadata": {},
   "source": [
    "## 湿实验数据"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e01083ae-fe52-4032-bc6e-dfd5e9562207",
   "metadata": {},
   "source": [
    "### CDS 下限数据生成"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "28385776-992a-4fc5-ace7-33566373d22c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-06 15:20:06.749404: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2024-06-06 15:20:06.880818: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2024-06-06 15:20:07.639168: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2024-06-06 15:20:07.643621: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-06-06 15:20:09.625983: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 80 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "from datasets import Dataset\n",
    "from torch.utils.data import TensorDataset  \n",
    "from datasets import load_dataset\n",
    "from datasets import DatasetDict\n",
    "from datasets import load_dataset, load_from_disk\n",
    "\n",
    "from pandarallel import pandarallel\n",
    " \n",
    "# Initialization\n",
    "pandarallel.initialize(nb_workers=80, progress_bar=True)\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0c66b103",
   "metadata": {},
   "outputs": [],
   "source": [
    "import shlex, subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "9e56f1ab",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Usage:\tdocker [OPTIONS] COMMAND\n",
      "\n",
      "A self-sufficient runtime for containers\n",
      "\n",
      "Options:\n",
      "      --config string      Location of client config files (default\n",
      "                           \"/home/yul/.docker\")\n",
      "  -c, --context string     Name of the context to use to connect to the\n",
      "                           daemon (overrides DOCKER_HOST env var and\n",
      "                           default context set with \"docker context use\")\n",
      "  -D, --debug              Enable debug mode\n",
      "  -H, --host list          Daemon socket(s) to connect to\n",
      "  -l, --log-level string   Set the logging level\n",
      "                           (\"debug\"|\"info\"|\"warn\"|\"error\"|\"fatal\")\n",
      "                           (default \"info\")\n",
      "      --tls                Use TLS; implied by --tlsverify\n",
      "      --tlscacert string   Trust certs signed only by this CA (default\n",
      "                           \"/home/yul/.docker/ca.pem\")\n",
      "      --tlscert string     Path to TLS certificate file (default\n",
      "                           \"/home/yul/.docker/cert.pem\")\n",
      "      --tlskey string      Path to TLS key file (default\n",
      "                           \"/home/yul/.docker/key.pem\")\n",
      "      --tlsverify          Use TLS and verify the remote\n",
      "  -v, --version            Print version information and quit\n",
      "\n",
      "Management Commands:\n",
      "  builder     Manage builds\n",
      "  config      Manage Docker configs\n",
      "  container   Manage containers\n",
      "  context     Manage contexts\n",
      "  engine      Manage the docker engine\n",
      "  image       Manage images\n",
      "  network     Manage networks\n",
      "  node        Manage Swarm nodes\n",
      "  plugin      Manage plugins\n",
      "  secret      Manage Docker secrets\n",
      "  service     Manage services\n",
      "  stack       Manage Docker stacks\n",
      "  swarm       Manage Swarm\n",
      "  system      Manage Docker\n",
      "  trust       Manage trust on Docker images\n",
      "  volume      Manage volumes\n",
      "\n",
      "Commands:\n",
      "  attach      Attach local standard input, output, and error streams to a running container\n",
      "  build       Build an image from a Dockerfile\n",
      "  commit      Create a new image from a container's changes\n",
      "  cp          Copy files/folders between a container and the local filesystem\n",
      "  create      Create a new container\n",
      "  diff        Inspect changes to files or directories on a container's filesystem\n",
      "  events      Get real time events from the server\n",
      "  exec        Run a command in a running container\n",
      "  export      Export a container's filesystem as a tar archive\n",
      "  history     Show the history of an image\n",
      "  images      List images\n",
      "  import      Import the contents from a tarball to create a filesystem image\n",
      "  info        Display system-wide information\n",
      "  inspect     Return low-level information on Docker objects\n",
      "  kill        Kill one or more running containers\n",
      "  load        Load an image from a tar archive or STDIN\n",
      "  login       Log in to a Docker registry\n",
      "  logout      Log out from a Docker registry\n",
      "  logs        Fetch the logs of a container\n",
      "  pause       Pause all processes within one or more containers\n",
      "  port        List port mappings or a specific mapping for the container\n",
      "  ps          List containers\n",
      "  pull        Pull an image or a repository from a registry\n",
      "  push        Push an image or a repository to a registry\n",
      "  rename      Rename a container\n",
      "  restart     Restart one or more containers\n",
      "  rm          Remove one or more containers\n",
      "  rmi         Remove one or more images\n",
      "  run         Run a command in a new container\n",
      "  save        Save one or more images to a tar archive (streamed to STDOUT by default)\n",
      "  search      Search the Docker Hub for images\n",
      "  start       Start one or more stopped containers\n",
      "  stats       Display a live stream of container(s) resource usage statistics\n",
      "  stop        Stop one or more running containers\n",
      "  tag         Create a tag TARGET_IMAGE that refers to SOURCE_IMAGE\n",
      "  top         Display the running processes of a container\n",
      "  unpause     Unpause all processes within one or more containers\n",
      "  update      Update configuration of one or more containers\n",
      "  version     Show the Docker version information\n",
      "  wait        Block until one or more containers stop, then print their exit codes\n",
      "\n",
      "Run 'docker COMMAND --help' for more information on a command.\n"
     ]
    }
   ],
   "source": [
    "seq = 'MNDTEAI'\n",
    "# p1 = subprocess.Popen(['echo', seq ], stdout=subprocess.PIPE, encoding='utf8')\n",
    "p2= subprocess.Popen([\"docker\", \"run\", '-w', '/home/cyq/LinearDesign/','-it',  \"jysgro/lineardesign:v1\", \n",
    "                      \"sh -c\" + \" '\" + 'echo ' + seq + '| /home/cyq/LinearDesign/lineardesign --lambda 4' + \"'\"], stdout=subprocess.PIPE, shell=True, encoding='utf8')\n",
    "\n",
    "# p1.communicate()[0].split('\\n')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "d3591f53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('', None)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = 'MNDTEAI'\n",
    "p3= subprocess.Popen([\"docker run jysgro/lineardesign:v1\", \n",
    "                       \"sh -c\" + \" '\" + 'echo ' + seq + ' | /home/cyq/LinearDesign/lineardesign --lambda 4' + \"'\"], \n",
    "                     stdout=subprocess.PIPE, encoding='utf8', shell=True)\n",
    "\n",
    "\n",
    "p3.communicate()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "020772b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['docker run jysgro/lineardesign:v1', \"sh -c 'echo MNDTEAI | /home/cyq/LinearDesign/lineardesign --lambda 4'\"], returncode=0, stdout=b'', stderr=b'')"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "subprocess.run([\"docker run jysgro/lineardesign:v1\", \n",
    "                       \"sh -c\" + \" '\" + 'echo ' + seq + ' | /home/cyq/LinearDesign/lineardesign --lambda 4' + \"'\"], capture_output=True,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c7249b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "templates_v3_NO38.parallel_apply(lambda x: genetic_evolution_parallel(x,  iterations=10, pop_size=10, \n",
    "                                                              outdir=os.path.join(os.getcwd(), 'data/processed/designed_seqs_v3_NO38')), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8d4757",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CAI > 0.95 是一个不错的CDS\n",
    "\n",
    "# docker pull jysgro/lineardesign:v1\n",
    "\n",
    "# using linear design as basic cds designer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ca1e940",
   "metadata": {},
   "source": [
    "#### 汇总"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "040f35cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data1/genomics'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "810fc91a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from itertools import islice\n",
    "\n",
    "# with open('/data1/genomics/yul/bioSPT2/scripts/BigLinearDesign_LAMBDA_merged.txt', 'r') as f:\n",
    "with open('/data1/genomics/yul/bioSPT2/scripts/LinearDesign_LAMBDA_merged.txt', 'r') as f:\n",
    "    lines = f.readlines()\n",
    "    # lines=f.readlines()[0:15]\n",
    "    \n",
    "    cds_gens = []\n",
    "    for line in lines:\n",
    "        if line.startswith('>'):\n",
    "            mRNA_seq = None\n",
    "            mRNA_structure = None\n",
    "            mfe=None\n",
    "            cai=None\n",
    "            ids = line.strip()\n",
    "        elif line.startswith('mRNA sequence: '):\n",
    "            # mRNA_seq = line.lstrip('mRNA sequence: ')\n",
    "            mRNA_seq = line.split(':')[-1].strip()\n",
    "        elif line.startswith('mRNA folding free energy:'):\n",
    "            v_line= line.lstrip('mRNA folding free energy:').strip()\n",
    "            mfe = v_line.split(' ')[0].strip()\n",
    "            cai = v_line.split(' ')[-1].strip()\n",
    "        elif line.startswith('mRNA structure:'):\n",
    "                mRNA_structure = line.lstrip('mRNA structure:').strip()\n",
    "        else:\n",
    "            cds_gens.append([ids, mRNA_seq, mRNA_structure, mfe, cai])\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "5ba248cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['>ENST00000641515.2|ENSG00000186092.7|OTTHUMG00000001094.4|OTTHUMT00000003223.4|OR4F5-201|OR4F5|2618|UTR5:1-60|CDS:61-1041|UTR3:1042-2618|',\n",
       " 'UGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCAGCGAGACCAACAACUCCAUGGUGACCGAGUUCAUCUUCCUGGGGCUGAGCGACUCCCAGGAGCUGCAGACCUUCCUGUUCAUGCUGUUCUUCGUGUUCUACGGGGGCAUCGUGUUCGGCAACCUGCUGAUCGUGAUCACCGUGGUGUCCGACAGCCACCUGCACAGCCCCAUGUACUUCCUGCUGGCCAACCUGAGCCUGAUCGACCUGAGCCUGUCCAGCGUGACAGCCCCCAAGAUGAUCACCGACUUCUUCAGCCAGAGGAAGGUGAUCAGCUUCAAGGGCUGUCUGGUGCAGAUCUUCCUGCUGCACUUCUUCGGGGGCAGCGAGAUGGUGAUCCUGAUCGCCAUGGGCUUCGAUCGGUACAUCGCCAUCUGCAAGCCCCUGCACUACACCACCAUCAUGUGCGGCAACGCCUGCGUGGGCAUCAUGGCCGUGACCUGGGGCAUCGGCUUCCUGCACAGCGUGAGCCAGCUGGCCUUCGCUGUGCACCUGCUGUUCUGCGGCCCCAACGAGGUGGACAGCUUCUACUGCGAUCUGCCUCGGGUGAUCAAGCUGGCCUGCACCGACACCUACAGGCUGGACAUCAUGGUGAUCGCCAACAGCGGGGUGCUGACCGUGUGCUCCUUCGUGCUGCUGAUCAUCAGCUACACCAUCAUCCUGAUGACCAUCCAGCACAGGCCCCUGGACAAGAGCAGCAAGGCCCUGAGCACCCUGACAGCCCACAUCACCGUGGUGCUGCUGUUCUUCGGGCCUUGCGUGUUCAUCUACGCCUGGCCCUUCCCCAUCAAGAGCCUGGACAAGUUCCUGGCCGUGUUCUACAGCGUGAUCACCCCCCUGCUGAACCCCAUCAUCUACACCCUGCGGAACAAGGACAUGAAGACAGCCAUCAGGCAGCUGCGGAAGUGGGACGCCCACAGCUCCGUGAAGUUCUGA\\n',\n",
       " '(((((...(((.((((..(((....((((((((..(....)..)))).)))).......)))..)))).)))...)))))((((.((((((((.(((..(((((....((((((..((.((((.....(((((.(((((((((((..((((((((((.((.(((((.(((((((...((((((((((((((((((...((((...............((((..((.(((.((((((((((((.(.(((((....(((.(((((((.(..((((((((..(((.((((((((..(((((......)))))..)))))))).)))...)))))))).(((((((.........)))))))....(((((((.((.((((((((((.(((((((((...)))...))))))...))))))))))))..))))))).....(((..((((.((((.(((((.....))).)).)))).))))..)))...((((((..((((....(((((((((...(((....)))...)))))))))...)))).......))))))....).))))))).))).....))))).).)))))))........))))).)))..))....))))))))))))))))))))))))))...))))))).))))))))).))))))))..(((..(((((...)))))..))).....((.(((((........((.(((((...((((...((((((((((((((.(((((....(.((((((((......)))).))))))))))...))))))))).))))).)))).))))))).......))))).))(((.(((.....))).)))..((((((.(((.(((.............................))).))).))))))))))))))))))))))...)))).))))))))...))))))))...)))))))).))))......',\n",
       " '-420.60',\n",
       " '0.987']"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "cds_gens[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "f67c5469",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>mRNA_seq</th>\n",
       "      <th>mRNA_structure</th>\n",
       "      <th>mfe</th>\n",
       "      <th>cai</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...</td>\n",
       "      <td>AUGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCA...</td>\n",
       "      <td>(((((...(((.((((..(((....((((((((..(....)..)))...</td>\n",
       "      <td>-420.60</td>\n",
       "      <td>0.987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;ENST00000458637.6|ENSG00000137474.23|OTTHUMG0...</td>\n",
       "      <td>AUGGUGAUCCUGCAGCAGGGCGACCACGUGUGGAUGGACCUGCGGC...</td>\n",
       "      <td>.((.((.(((((((.((((((((....(((((((.((...(((.((...</td>\n",
       "      <td>-3096.05</td>\n",
       "      <td>0.989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;ENST00000409619.6|ENSG00000137474.23|OTTHUMG0...</td>\n",
       "      <td>AUGGACCUGCGCCUGGGCCAGGAGUUCGACGUGCCCAUCGGCGCCG...</td>\n",
       "      <td>....((((((((((((.((.((((((..(((.((.(((((.........</td>\n",
       "      <td>-3078.55</td>\n",
       "      <td>0.989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;ENST00000660626.1|ENSG00000137474.23|OTTHUMG0...</td>\n",
       "      <td>AUGGAGGACCCAGCCGACCCCGCCGACCCCGCCCGGAUCCCCCAGC...</td>\n",
       "      <td>.....((.((((((((.(....(((...((((.(((((((((((((...</td>\n",
       "      <td>-108.90</td>\n",
       "      <td>0.990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;ENST00000409893.6|ENSG00000137474.23|OTTHUMG0...</td>\n",
       "      <td>CUGUUCGACAGGCACCUGUGUGUGCGCCAGCUGCGGUACUCCGGCA...</td>\n",
       "      <td>........(((((.((((.((((.((((.((((((((.(((((......</td>\n",
       "      <td>-824.00</td>\n",
       "      <td>0.991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ids  \\\n",
       "0  >ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...   \n",
       "1  >ENST00000458637.6|ENSG00000137474.23|OTTHUMG0...   \n",
       "2  >ENST00000409619.6|ENSG00000137474.23|OTTHUMG0...   \n",
       "3  >ENST00000660626.1|ENSG00000137474.23|OTTHUMG0...   \n",
       "4  >ENST00000409893.6|ENSG00000137474.23|OTTHUMG0...   \n",
       "\n",
       "                                            mRNA_seq  \\\n",
       "0  AUGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCA...   \n",
       "1  AUGGUGAUCCUGCAGCAGGGCGACCACGUGUGGAUGGACCUGCGGC...   \n",
       "2  AUGGACCUGCGCCUGGGCCAGGAGUUCGACGUGCCCAUCGGCGCCG...   \n",
       "3  AUGGAGGACCCAGCCGACCCCGCCGACCCCGCCCGGAUCCCCCAGC...   \n",
       "4  CUGUUCGACAGGCACCUGUGUGUGCGCCAGCUGCGGUACUCCGGCA...   \n",
       "\n",
       "                                      mRNA_structure       mfe    cai  \n",
       "0  (((((...(((.((((..(((....((((((((..(....)..)))...   -420.60  0.987  \n",
       "1  .((.((.(((((((.((((((((....(((((((.((...(((.((...  -3096.05  0.989  \n",
       "2  ....((((((((((((.((.((((((..(((.((.(((((.........  -3078.55  0.989  \n",
       "3  .....((.((((((((.(....(((...((((.(((((((((((((...   -108.90  0.990  \n",
       "4  ........(((((.((((.((((.((((.((((((((.(((((......   -824.00  0.991  "
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "BigLinearDesign_LAMBDA_merged_CDS = pd.DataFrame(cds_gens, columns=['ids', 'mRNA_seq', 'mRNA_structure', 'mfe', 'cai'])\n",
    "BigLinearDesign_LAMBDA_merged_CDS.shape\n",
    "BigLinearDesign_LAMBDA_merged_CDS.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "8aba93cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>mRNA_seq</th>\n",
       "      <th>mRNA_structure</th>\n",
       "      <th>mfe</th>\n",
       "      <th>cai</th>\n",
       "      <th>aa</th>\n",
       "      <th>aa_valid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...</td>\n",
       "      <td>AUGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCA...</td>\n",
       "      <td>(((((...(((.((((..(((....((((((((..(....)..)))...</td>\n",
       "      <td>-420.60</td>\n",
       "      <td>0.987</td>\n",
       "      <td>MKKVTAEAISWNESTSETNNSMVTEFIFLGLSDSQELQTFLFMLFF...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;ENST00000458637.6|ENSG00000137474.23|OTTHUMG0...</td>\n",
       "      <td>AUGGUGAUCCUGCAGCAGGGCGACCACGUGUGGAUGGACCUGCGGC...</td>\n",
       "      <td>.((.((.(((((((.((((((((....(((((((.((...(((.((...</td>\n",
       "      <td>-3096.05</td>\n",
       "      <td>0.989</td>\n",
       "      <td>MVILQQGDHVWMDLRLGQEFDVPIGAVVKLCDSGQVQVVDDEDNEH...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;ENST00000409619.6|ENSG00000137474.23|OTTHUMG0...</td>\n",
       "      <td>AUGGACCUGCGCCUGGGCCAGGAGUUCGACGUGCCCAUCGGCGCCG...</td>\n",
       "      <td>....((((((((((((.((.((((((..(((.((.(((((.........</td>\n",
       "      <td>-3078.55</td>\n",
       "      <td>0.989</td>\n",
       "      <td>MDLRLGQEFDVPIGAVVKLCDSGQVQVVDDEDNEHWISPQNATHIK...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;ENST00000660626.1|ENSG00000137474.23|OTTHUMG0...</td>\n",
       "      <td>AUGGAGGACCCAGCCGACCCCGCCGACCCCGCCCGGAUCCCCCAGC...</td>\n",
       "      <td>.....((.((((((((.(....(((...((((.(((((((((((((...</td>\n",
       "      <td>-108.90</td>\n",
       "      <td>0.990</td>\n",
       "      <td>MEDPADPADPARIPQPGKWALVRSLNQAFKTYAVLPGDHVWMDLRL...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;ENST00000409893.6|ENSG00000137474.23|OTTHUMG0...</td>\n",
       "      <td>CUGUUCGACAGGCACCUGUGUGUGCGCCAGCUGCGGUACUCCGGCA...</td>\n",
       "      <td>........(((((.((((.((((.((((.((((((((.(((((......</td>\n",
       "      <td>-824.00</td>\n",
       "      <td>0.991</td>\n",
       "      <td>LFDRHLCVRQLRYSGMMETIRIRRAGYPIRYSFVEFVERYRVLLPG...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ids  \\\n",
       "0  >ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...   \n",
       "1  >ENST00000458637.6|ENSG00000137474.23|OTTHUMG0...   \n",
       "2  >ENST00000409619.6|ENSG00000137474.23|OTTHUMG0...   \n",
       "3  >ENST00000660626.1|ENSG00000137474.23|OTTHUMG0...   \n",
       "4  >ENST00000409893.6|ENSG00000137474.23|OTTHUMG0...   \n",
       "\n",
       "                                            mRNA_seq  \\\n",
       "0  AUGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCA...   \n",
       "1  AUGGUGAUCCUGCAGCAGGGCGACCACGUGUGGAUGGACCUGCGGC...   \n",
       "2  AUGGACCUGCGCCUGGGCCAGGAGUUCGACGUGCCCAUCGGCGCCG...   \n",
       "3  AUGGAGGACCCAGCCGACCCCGCCGACCCCGCCCGGAUCCCCCAGC...   \n",
       "4  CUGUUCGACAGGCACCUGUGUGUGCGCCAGCUGCGGUACUCCGGCA...   \n",
       "\n",
       "                                      mRNA_structure       mfe    cai  \\\n",
       "0  (((((...(((.((((..(((....((((((((..(....)..)))...   -420.60  0.987   \n",
       "1  .((.((.(((((((.((((((((....(((((((.((...(((.((...  -3096.05  0.989   \n",
       "2  ....((((((((((((.((.((((((..(((.((.(((((.........  -3078.55  0.989   \n",
       "3  .....((.((((((((.(....(((...((((.(((((((((((((...   -108.90  0.990   \n",
       "4  ........(((((.((((.((((.((((.((((((((.(((((......   -824.00  0.991   \n",
       "\n",
       "                                                  aa  aa_valid  \n",
       "0  MKKVTAEAISWNESTSETNNSMVTEFIFLGLSDSQELQTFLFMLFF...      True  \n",
       "1  MVILQQGDHVWMDLRLGQEFDVPIGAVVKLCDSGQVQVVDDEDNEH...      True  \n",
       "2  MDLRLGQEFDVPIGAVVKLCDSGQVQVVDDEDNEHWISPQNATHIK...      True  \n",
       "3  MEDPADPADPARIPQPGKWALVRSLNQAFKTYAVLPGDHVWMDLRL...     False  \n",
       "4  LFDRHLCVRQLRYSGMMETIRIRRAGYPIRYSFVEFVERYRVLLPG...      True  "
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "BigLinearDesign_LAMBDA_merged_CDS['aa'] = BigLinearDesign_LAMBDA_merged_CDS['mRNA_seq'].apply(lambda x: ''.join(Seq(x).translate()))\n",
    "\n",
    "\n",
    "BigLinearDesign_LAMBDA_merged_CDS['aa_valid'] = BigLinearDesign_LAMBDA_merged_CDS['aa'].apply(lambda x: True if x.find('*') == len(x)-1 else False)\n",
    "\n",
    "BigLinearDesign_LAMBDA_merged_CDS['aa_valid'].sum()\n",
    "BigLinearDesign_LAMBDA_merged_CDS.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "c2b289df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2278, 5)"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BigLinearDesign_LAMBDA_merged_CDS.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "eb4a9f56",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS = pd.DataFrame(cds_gens, columns=['ids', 'mRNA_seq', 'mRNA_structure', 'mfe', 'cai'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "4e58c1b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>mRNA_seq</th>\n",
       "      <th>mRNA_structure</th>\n",
       "      <th>mfe</th>\n",
       "      <th>cai</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...</td>\n",
       "      <td>AUGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCA...</td>\n",
       "      <td>(((((...(((.((((..(((....((((((((..(....)..)))...</td>\n",
       "      <td>-420.60</td>\n",
       "      <td>0.987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;ENST00000426406.4|ENSG00000284733.2|OTTHUMG00...</td>\n",
       "      <td>AUGGACGGCGAGAACCACAGCGUGGUGUCCGAGUUCCUGUUCCUGG...</td>\n",
       "      <td>....((((((.(((((.((((((((.(((..((((((.......))...</td>\n",
       "      <td>-416.90</td>\n",
       "      <td>0.986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;ENST00000616016.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>AUGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCC...</td>\n",
       "      <td>..(((((..((..((((((.(((((((((((((((((.(((((((....</td>\n",
       "      <td>-1578.40</td>\n",
       "      <td>0.975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;ENST00000618323.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>AUGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCC...</td>\n",
       "      <td>..(((((..((..((((((.(((((((((((((((((.(((((((....</td>\n",
       "      <td>-1584.70</td>\n",
       "      <td>0.975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;ENST00000437963.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>AUGAGCAAGGGCAUCCUGCAGGUGCACCCCCCCAUCUGCGACUGUC...</td>\n",
       "      <td>........(((((...((((((((........)))))))).(((((...</td>\n",
       "      <td>-173.00</td>\n",
       "      <td>0.986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ids  \\\n",
       "0  >ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...   \n",
       "1  >ENST00000426406.4|ENSG00000284733.2|OTTHUMG00...   \n",
       "2  >ENST00000616016.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "3  >ENST00000618323.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "4  >ENST00000437963.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "\n",
       "                                            mRNA_seq  \\\n",
       "0  AUGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCA...   \n",
       "1  AUGGACGGCGAGAACCACAGCGUGGUGUCCGAGUUCCUGUUCCUGG...   \n",
       "2  AUGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCC...   \n",
       "3  AUGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCC...   \n",
       "4  AUGAGCAAGGGCAUCCUGCAGGUGCACCCCCCCAUCUGCGACUGUC...   \n",
       "\n",
       "                                      mRNA_structure       mfe    cai  \n",
       "0  (((((...(((.((((..(((....((((((((..(....)..)))...   -420.60  0.987  \n",
       "1  ....((((((.(((((.((((((((.(((..((((((.......))...   -416.90  0.986  \n",
       "2  ..(((((..((..((((((.(((((((((((((((((.(((((((....  -1578.40  0.975  \n",
       "3  ..(((((..((..((((((.(((((((((((((((((.(((((((....  -1584.70  0.975  \n",
       "4  ........(((((...((((((((........)))))))).(((((...   -173.00  0.986  "
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS.shape\n",
    "LinearDesign_LAMBDA_merged_CDS.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "9be79527",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5139, 5)"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens = pd.concat([LinearDesign_LAMBDA_merged_CDS, BigLinearDesign_LAMBDA_merged_CDS], axis=0)\n",
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "cdc1f447",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens.isnull().sum()\n",
    "\n",
    "# LinearDesign_LAMBDA_merged_CDS['mRNA_seq'].value_counts()\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique = LinearDesign_LAMBDA_merged_CDS_gens.drop_duplicates(subset=['ids','mRNA_seq'], keep='first', inplace=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "cdb24e02",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique.to_csv('/data1/genomics/yul/bioSPT2/data/interiem/processed/LinearDesign_LAMBDA_merged_CDS_gens_unique.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "93fac2ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1320,)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique['ids'].value_counts().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "7a3b9361",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>mRNA_seq</th>\n",
       "      <th>mRNA_structure</th>\n",
       "      <th>mfe</th>\n",
       "      <th>cai</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...</td>\n",
       "      <td>AUGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCA...</td>\n",
       "      <td>(((((...(((.((((..(((....((((((((..(....)..)))...</td>\n",
       "      <td>-420.6</td>\n",
       "      <td>0.987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;ENST00000426406.4|ENSG00000284733.2|OTTHUMG00...</td>\n",
       "      <td>AUGGACGGCGAGAACCACAGCGUGGUGUCCGAGUUCCUGUUCCUGG...</td>\n",
       "      <td>....((((((.(((((.((((((((.(((..((((((.......))...</td>\n",
       "      <td>-416.9</td>\n",
       "      <td>0.986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;ENST00000616016.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>AUGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCC...</td>\n",
       "      <td>..(((((..((..((((((.(((((((((((((((((.(((((((....</td>\n",
       "      <td>-1578.4</td>\n",
       "      <td>0.975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;ENST00000618323.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>AUGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCC...</td>\n",
       "      <td>..(((((..((..((((((.(((((((((((((((((.(((((((....</td>\n",
       "      <td>-1584.7</td>\n",
       "      <td>0.975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;ENST00000437963.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>AUGAGCAAGGGCAUCCUGCAGGUGCACCCCCCCAUCUGCGACUGUC...</td>\n",
       "      <td>........(((((...((((((((........)))))))).(((((...</td>\n",
       "      <td>-173.0</td>\n",
       "      <td>0.986</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ids  \\\n",
       "0  >ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...   \n",
       "1  >ENST00000426406.4|ENSG00000284733.2|OTTHUMG00...   \n",
       "2  >ENST00000616016.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "3  >ENST00000618323.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "4  >ENST00000437963.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "\n",
       "                                            mRNA_seq  \\\n",
       "0  AUGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCA...   \n",
       "1  AUGGACGGCGAGAACCACAGCGUGGUGUCCGAGUUCCUGUUCCUGG...   \n",
       "2  AUGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCC...   \n",
       "3  AUGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCC...   \n",
       "4  AUGAGCAAGGGCAUCCUGCAGGUGCACCCCCCCAUCUGCGACUGUC...   \n",
       "\n",
       "                                      mRNA_structure     mfe    cai  \n",
       "0  (((((...(((.((((..(((....((((((((..(....)..)))...  -420.6  0.987  \n",
       "1  ....((((((.(((((.((((((((.(((..((((((.......))...  -416.9  0.986  \n",
       "2  ..(((((..((..((((((.(((((((((((((((((.(((((((.... -1578.4  0.975  \n",
       "3  ..(((((..((..((((((.(((((((((((((((((.(((((((.... -1584.7  0.975  \n",
       "4  ........(((((...((((((((........)))))))).(((((...  -173.0  0.986  "
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique = pd.read_csv('/data1/genomics/yul/bioSPT2/data/interiem/processed/LinearDesign_LAMBDA_merged_CDS_gens_unique.csv')\n",
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "5a3a3071",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "4241ab6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yul/softwares/anaconda3/envs/ai/lib/python3.8/site-packages/Bio/Seq.py:2334: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique['aa'] = LinearDesign_LAMBDA_merged_CDS_gens_unique['mRNA_seq'].apply(lambda x: ''.join(Seq(x).translate()))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "7b21ce76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>mRNA_seq</th>\n",
       "      <th>mRNA_structure</th>\n",
       "      <th>mfe</th>\n",
       "      <th>cai</th>\n",
       "      <th>aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...</td>\n",
       "      <td>UGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCAG...</td>\n",
       "      <td>(((((...(((.((((..(((....((((((((..(....)..)))...</td>\n",
       "      <td>-420.6</td>\n",
       "      <td>0.987</td>\n",
       "      <td>*RR*PPRPSPGTRAPARPTTPW*PSSSSWG*ATPRSCRPSCSCCSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;ENST00000426406.4|ENSG00000284733.2|OTTHUMG00...</td>\n",
       "      <td>UGGACGGCGAGAACCACAGCGUGGUGUCCGAGUUCCUGUUCCUGGG...</td>\n",
       "      <td>....((((((.(((((.((((((((.(((..((((((.......))...</td>\n",
       "      <td>-416.9</td>\n",
       "      <td>0.986</td>\n",
       "      <td>WTARTTAWCPSSCSWD*PTAGRSSCCCWCSAACCTWPASPATS*SC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;ENST00000616016.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>UGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCCU...</td>\n",
       "      <td>..(((((..((..((((((.(((((((((((((((((.(((((((....</td>\n",
       "      <td>-1578.4</td>\n",
       "      <td>0.975</td>\n",
       "      <td>CQP*RRSSPGGRIWPWPWPPSTPPWPPCPCPPCLATWPPCQPPPPC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;ENST00000618323.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>UGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCCU...</td>\n",
       "      <td>..(((((..((..((((((.(((((((((((((((((.(((((((....</td>\n",
       "      <td>-1584.7</td>\n",
       "      <td>0.975</td>\n",
       "      <td>CQP*RRSSPGGRIWPWPWPPSTPPWPPCPCPPCLATWPLCQPPPPC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;ENST00000437963.5|ENSG00000187634.13|OTTHUMG0...</td>\n",
       "      <td>UGAGCAAGGGCAUCCUGCAGGUGCACCCCCCCAUCUGCGACUGUCC...</td>\n",
       "      <td>........(((((...((((((((........)))))))).(((((...</td>\n",
       "      <td>-173.0</td>\n",
       "      <td>0.986</td>\n",
       "      <td>*ARASCRCTPPSATVLAAGSALP*TGAAWPTSGPWPCPPPGT*RRS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ids  \\\n",
       "0  >ENST00000641515.2|ENSG00000186092.7|OTTHUMG00...   \n",
       "1  >ENST00000426406.4|ENSG00000284733.2|OTTHUMG00...   \n",
       "2  >ENST00000616016.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "3  >ENST00000618323.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "4  >ENST00000437963.5|ENSG00000187634.13|OTTHUMG0...   \n",
       "\n",
       "                                            mRNA_seq  \\\n",
       "0  UGAAGAAGGUGACCGCCGAGGCCAUCUCCUGGAACGAGAGCACCAG...   \n",
       "1  UGGACGGCGAGAACCACAGCGUGGUGUCCGAGUUCCUGUUCCUGGG...   \n",
       "2  UGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCCU...   \n",
       "3  UGCCAGCCGUGAAGAAGGAGUUCCCCGGGCGGGAGGAUCUGGCCCU...   \n",
       "4  UGAGCAAGGGCAUCCUGCAGGUGCACCCCCCCAUCUGCGACUGUCC...   \n",
       "\n",
       "                                      mRNA_structure     mfe    cai  \\\n",
       "0  (((((...(((.((((..(((....((((((((..(....)..)))...  -420.6  0.987   \n",
       "1  ....((((((.(((((.((((((((.(((..((((((.......))...  -416.9  0.986   \n",
       "2  ..(((((..((..((((((.(((((((((((((((((.(((((((.... -1578.4  0.975   \n",
       "3  ..(((((..((..((((((.(((((((((((((((((.(((((((.... -1584.7  0.975   \n",
       "4  ........(((((...((((((((........)))))))).(((((...  -173.0  0.986   \n",
       "\n",
       "                                                  aa  \n",
       "0  *RR*PPRPSPGTRAPARPTTPW*PSSSSWG*ATPRSCRPSCSCCSS...  \n",
       "1  WTARTTAWCPSSCSWD*PTAGRSSCCCWCSAACCTWPASPATS*SC...  \n",
       "2  CQP*RRSSPGGRIWPWPWPPSTPPWPPCPCPPCLATWPPCQPPPPC...  \n",
       "3  CQP*RRSSPGGRIWPWPWPPSTPPWPPCPCPPCLATWPLCQPPPPC...  \n",
       "4  *ARASCRCTPPSATVLAAGSALP*TGAAWPTSGPWPCPPPGT*RRS...  "
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique['aa'].head()\n",
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "028e7d12-997f-4231-8286-1c909c287a05",
   "metadata": {},
   "source": [
    "## UTR 下限数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72758969-522b-49ca-9fb8-744fdf3bc1a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "去多细胞多组学数据中寻找"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "fec758b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data1/genomics'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a9d913ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id_raw</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>UTR5_regs</th>\n",
       "      <th>CDS_regs</th>\n",
       "      <th>UTR3_regs</th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>UTR3</th>\n",
       "      <th>Esemble</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>ENSG00000000003.15</td>\n",
       "      <td>&gt;ENST00000373020.9|ENSG00000000003.15|OTTHUMG0...</td>\n",
       "      <td>AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>UTR5:1-112</td>\n",
       "      <td>CDS:113-850</td>\n",
       "      <td>UTR3:851-3768</td>\n",
       "      <td>AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...</td>\n",
       "      <td>ATGGCGTCCCCGTCTCGGAGACTGCAGACTAAACCAGTCATTACTT...</td>\n",
       "      <td>CCCAATGTATCTGTGGGCCTATTCCTCTCTACCTTTAAGGACATTT...</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000373031</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>&gt;ENST00000373031.5|ENSG00000000005.6|OTTHUMG00...</td>\n",
       "      <td>AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...</td>\n",
       "      <td>ENST00000373031.5</td>\n",
       "      <td>ENST00000373031.5</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>UTR5:1-83</td>\n",
       "      <td>CDS:84-1037</td>\n",
       "      <td>UTR3:1038-1205</td>\n",
       "      <td>AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...</td>\n",
       "      <td>ATGGCAAAGAATCCTCCAGAGAATTGTGAAGACTGTCACATTCTAA...</td>\n",
       "      <td>TAGGAGGTTTGAGCTCAAATGCTTAAACTGCTGGCAACATATAATA...</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000371588</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>&gt;ENST00000371588.10|ENSG00000000419.14|OTTHUMG...</td>\n",
       "      <td>AGTTCCGCCATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGT...</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>UTR5:1-9</td>\n",
       "      <td>CDS:10-792</td>\n",
       "      <td>UTR3:793-1054</td>\n",
       "      <td>AGTTCCGCC</td>\n",
       "      <td>ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...</td>\n",
       "      <td>AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000367772</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>&gt;ENST00000367772.8|ENSG00000000457.14|OTTHUMG0...</td>\n",
       "      <td>TGTCCCGTTTCCGGACCCGTCTCTATGGTGTAGGAGAAACCCGGCC...</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>UTR5:1-222</td>\n",
       "      <td>CDS:223-2451</td>\n",
       "      <td>UTR3:2452-3090</td>\n",
       "      <td>TGTCCCGTTTCCGGACCCGTCTCTATGGTGTAGGAGAAACCCGGCC...</td>\n",
       "      <td>ATGGGATCAGAGAACAGTGCTTTAAAGAGCTATACACTGAGAGAAC...</td>\n",
       "      <td>CAATAGATGTGAGTTAAACTTTAGGAAAAAGGATTCCCTTTTTTTA...</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000413811</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>&gt;ENST00000413811.3|ENSG00000000460.17|OTTHUMG0...</td>\n",
       "      <td>ACATTCTTGATTGGTCTCATGCCTTTGTGGTTGTAAATGTGCCTGG...</td>\n",
       "      <td>ENST00000413811.3</td>\n",
       "      <td>ENST00000413811.3</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>UTR5:1-1018</td>\n",
       "      <td>CDS:1019-2611</td>\n",
       "      <td>UTR3:2612-2661</td>\n",
       "      <td>ACATTCTTGATTGGTCTCATGCCTTTGTGGTTGTAAATGTGCCTGG...</td>\n",
       "      <td>ATGCAGGTGGTTTTGGACAGTAAATTAGACCTGCCATGTGAACTGC...</td>\n",
       "      <td>AACTTATCACTAGGCAGAACTGGGTTTGATGCTTTGTCAACTGAAA...</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Transcript             gene_id  \\\n",
       "0  ENST00000373020  ENSG00000000003.15   \n",
       "1  ENST00000373031   ENSG00000000005.6   \n",
       "2  ENST00000371588  ENSG00000000419.14   \n",
       "3  ENST00000367772  ENSG00000000457.14   \n",
       "4  ENST00000413811  ENSG00000000460.17   \n",
       "\n",
       "                                                name  \\\n",
       "0  >ENST00000373020.9|ENSG00000000003.15|OTTHUMG0...   \n",
       "1  >ENST00000373031.5|ENSG00000000005.6|OTTHUMG00...   \n",
       "2  >ENST00000371588.10|ENSG00000000419.14|OTTHUMG...   \n",
       "3  >ENST00000367772.8|ENSG00000000457.14|OTTHUMG0...   \n",
       "4  >ENST00000413811.3|ENSG00000000460.17|OTTHUMG0...   \n",
       "\n",
       "                                                 seq       transcript_id  \\\n",
       "0  AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...   ENST00000373020.9   \n",
       "1  AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...   ENST00000373031.5   \n",
       "2  AGTTCCGCCATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGT...  ENST00000371588.10   \n",
       "3  TGTCCCGTTTCCGGACCCGTCTCTATGGTGTAGGAGAAACCCGGCC...   ENST00000367772.8   \n",
       "4  ACATTCTTGATTGGTCTCATGCCTTTGTGGTTGTAAATGTGCCTGG...   ENST00000413811.3   \n",
       "\n",
       "          gene_id_raw gene_name    UTR5_regs       CDS_regs       UTR3_regs  \\\n",
       "0   ENST00000373020.9    TSPAN6   UTR5:1-112    CDS:113-850   UTR3:851-3768   \n",
       "1   ENST00000373031.5      TNMD    UTR5:1-83    CDS:84-1037  UTR3:1038-1205   \n",
       "2  ENST00000371588.10      DPM1     UTR5:1-9     CDS:10-792   UTR3:793-1054   \n",
       "3   ENST00000367772.8     SCYL3   UTR5:1-222   CDS:223-2451  UTR3:2452-3090   \n",
       "4   ENST00000413811.3  C1orf112  UTR5:1-1018  CDS:1019-2611  UTR3:2612-2661   \n",
       "\n",
       "                                                UTR5  \\\n",
       "0  AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...   \n",
       "1  AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...   \n",
       "2                                          AGTTCCGCC   \n",
       "3  TGTCCCGTTTCCGGACCCGTCTCTATGGTGTAGGAGAAACCCGGCC...   \n",
       "4  ACATTCTTGATTGGTCTCATGCCTTTGTGGTTGTAAATGTGCCTGG...   \n",
       "\n",
       "                                                 CDS  \\\n",
       "0  ATGGCGTCCCCGTCTCGGAGACTGCAGACTAAACCAGTCATTACTT...   \n",
       "1  ATGGCAAAGAATCCTCCAGAGAATTGTGAAGACTGTCACATTCTAA...   \n",
       "2  ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...   \n",
       "3  ATGGGATCAGAGAACAGTGCTTTAAAGAGCTATACACTGAGAGAAC...   \n",
       "4  ATGCAGGTGGTTTTGGACAGTAAATTAGACCTGCCATGTGAACTGC...   \n",
       "\n",
       "                                                UTR3          Esemble  \n",
       "0  CCCAATGTATCTGTGGGCCTATTCCTCTCTACCTTTAAGGACATTT...  ENSG00000000003  \n",
       "1  TAGGAGGTTTGAGCTCAAATGCTTAAACTGCTGGCAACATATAATA...  ENSG00000000005  \n",
       "2  AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...  ENSG00000000419  \n",
       "3  CAATAGATGTGAGTTAAACTTTAGGAAAAAGGATTCCCTTTTTTTA...  ENSG00000000457  \n",
       "4  AACTTATCACTAGGCAGAACTGGGTTTGATGCTTTGTCAACTGAAA...  ENSG00000000460  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 加载公共注释数据\n",
    "dataset_dir = '/data1/genomics/yul/bioSPT2/data/interiem/annotation'\n",
    "\n",
    "dominant_Txs_select_rank1_top = pd.read_csv(os.path.join(dataset_dir, 'genecode_human_dominant_Txs_select_rank1_top_processed.tsv'), sep='\\t')\n",
    "\n",
    "dominant_Txs_select_rank1_top.head()\n",
    "\n",
    "\n",
    "human_ts = pd.read_csv(os.path.join(dataset_dir, 'genecode_human_annotation_dominant_transcript_cdna.tsv'), sep='\\t')\n",
    "\n",
    "human_ts.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4d48d94d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>iBAQ A549_exp</th>\n",
       "      <th>iBAQ HEK293_exp</th>\n",
       "      <th>iBAQ HeLa_exp</th>\n",
       "      <th>iBAQ K562_exp</th>\n",
       "      <th>iBAQ_TransEffect_A549_exp</th>\n",
       "      <th>iBAQ_TransEffect_HEK293_exp</th>\n",
       "      <th>iBAQ_TransEffect_HeLa_exp</th>\n",
       "      <th>iBAQ_TransEffect_K562_exp</th>\n",
       "      <th>iBAQ_TransEffect_A549_rank</th>\n",
       "      <th>iBAQ_TransEffect_HEK293_rank</th>\n",
       "      <th>...</th>\n",
       "      <th>K562_GR_2017_rank</th>\n",
       "      <th>K562_elife_rank</th>\n",
       "      <th>K562_timeLapse_rank</th>\n",
       "      <th>ESC_GR_2017_label</th>\n",
       "      <th>Hela_GSE102113_label</th>\n",
       "      <th>K562_GR_2017_label</th>\n",
       "      <th>K562_elife_label</th>\n",
       "      <th>K562_timeLapse_label</th>\n",
       "      <th>consistent_label_HalfLife</th>\n",
       "      <th>consistent_label_all</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>6.527933</td>\n",
       "      <td>6.796727</td>\n",
       "      <td>5.517671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.134571</td>\n",
       "      <td>5.195596</td>\n",
       "      <td>4.181992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1074.0</td>\n",
       "      <td>2159.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>6.989328</td>\n",
       "      <td>6.970460</td>\n",
       "      <td>6.817138</td>\n",
       "      <td>6.404391</td>\n",
       "      <td>5.275929</td>\n",
       "      <td>5.256304</td>\n",
       "      <td>4.987699</td>\n",
       "      <td>4.315847</td>\n",
       "      <td>756.0</td>\n",
       "      <td>1958.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1926.0</td>\n",
       "      <td>5030.0</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>4.331123</td>\n",
       "      <td>5.157944</td>\n",
       "      <td>3.871246</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.588284</td>\n",
       "      <td>4.523821</td>\n",
       "      <td>3.200597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5498.0</td>\n",
       "      <td>4664.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9413.0</td>\n",
       "      <td>8183.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>5.157366</td>\n",
       "      <td>5.762855</td>\n",
       "      <td>4.583522</td>\n",
       "      <td>5.264936</td>\n",
       "      <td>3.871606</td>\n",
       "      <td>4.495252</td>\n",
       "      <td>3.234262</td>\n",
       "      <td>3.624279</td>\n",
       "      <td>4839.0</td>\n",
       "      <td>4755.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5547.0</td>\n",
       "      <td>4374.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.211548</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 iBAQ A549_exp  iBAQ HEK293_exp  iBAQ HeLa_exp  iBAQ K562_exp  \\\n",
       "ensembl_gene_id                                                                 \n",
       "ENSG00000000003       6.527933         6.796727       5.517671            NaN   \n",
       "ENSG00000000419       6.989328         6.970460       6.817138       6.404391   \n",
       "ENSG00000000457       4.331123         5.157944       3.871246            NaN   \n",
       "ENSG00000000460       5.157366         5.762855       4.583522       5.264936   \n",
       "ENSG00000000938            NaN         4.211548            NaN            NaN   \n",
       "\n",
       "                 iBAQ_TransEffect_A549_exp  iBAQ_TransEffect_HEK293_exp  \\\n",
       "ensembl_gene_id                                                           \n",
       "ENSG00000000003                   5.134571                     5.195596   \n",
       "ENSG00000000419                   5.275929                     5.256304   \n",
       "ENSG00000000457                   3.588284                     4.523821   \n",
       "ENSG00000000460                   3.871606                     4.495252   \n",
       "ENSG00000000938                        NaN                          NaN   \n",
       "\n",
       "                 iBAQ_TransEffect_HeLa_exp  iBAQ_TransEffect_K562_exp  \\\n",
       "ensembl_gene_id                                                         \n",
       "ENSG00000000003                   4.181992                        NaN   \n",
       "ENSG00000000419                   4.987699                   4.315847   \n",
       "ENSG00000000457                   3.200597                        NaN   \n",
       "ENSG00000000460                   3.234262                   3.624279   \n",
       "ENSG00000000938                        NaN                        NaN   \n",
       "\n",
       "                 iBAQ_TransEffect_A549_rank  iBAQ_TransEffect_HEK293_rank  \\\n",
       "ensembl_gene_id                                                             \n",
       "ENSG00000000003                      1074.0                        2159.0   \n",
       "ENSG00000000419                       756.0                        1958.0   \n",
       "ENSG00000000457                      5498.0                        4664.0   \n",
       "ENSG00000000460                      4839.0                        4755.0   \n",
       "ENSG00000000938                         NaN                           NaN   \n",
       "\n",
       "                 ...  K562_GR_2017_rank  K562_elife_rank  K562_timeLapse_rank  \\\n",
       "ensembl_gene_id  ...                                                            \n",
       "ENSG00000000003  ...                NaN              NaN                  NaN   \n",
       "ENSG00000000419  ...             1926.0           5030.0               1362.0   \n",
       "ENSG00000000457  ...             9413.0           8183.0                  NaN   \n",
       "ENSG00000000460  ...             5547.0           4374.0                  NaN   \n",
       "ENSG00000000938  ...                NaN              NaN                  NaN   \n",
       "\n",
       "                 ESC_GR_2017_label  Hela_GSE102113_label  K562_GR_2017_label  \\\n",
       "ensembl_gene_id                                                                \n",
       "ENSG00000000003                NaN                   1.0                 NaN   \n",
       "ENSG00000000419                1.0                   1.0                 1.0   \n",
       "ENSG00000000457               -1.0                   NaN                -1.0   \n",
       "ENSG00000000460                0.0                   0.0                 0.0   \n",
       "ENSG00000000938                NaN                   NaN                 NaN   \n",
       "\n",
       "                 K562_elife_label  K562_timeLapse_label  \\\n",
       "ensembl_gene_id                                           \n",
       "ENSG00000000003               NaN                   NaN   \n",
       "ENSG00000000419               0.0                   0.0   \n",
       "ENSG00000000457              -1.0                   NaN   \n",
       "ENSG00000000460               0.0                   NaN   \n",
       "ENSG00000000938               NaN                   NaN   \n",
       "\n",
       "                 consistent_label_HalfLife  consistent_label_all  \n",
       "ensembl_gene_id                                                   \n",
       "ENSG00000000003                        NaN                   NaN  \n",
       "ENSG00000000419                        NaN                   NaN  \n",
       "ENSG00000000457                        NaN                   NaN  \n",
       "ENSG00000000460                        NaN                   NaN  \n",
       "ENSG00000000938                        NaN                   NaN  \n",
       "\n",
       "[5 rows x 54 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 多组学多细胞系数据\n",
    "\n",
    "all_genes_omics = pd.read_csv('/data1/genomics/yul/RNA_optimization/ExpUTR5/data/external/cleandata/ALL_gene_omics.csv', sep=',')\n",
    "\n",
    "all_genes_omics = all_genes_omics.set_index('ensembl_gene_id')\n",
    "\n",
    "all_genes_omics.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ad5a0074",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>data size</th>\n",
       "      <th>top 10 genes</th>\n",
       "      <th>top 20 genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>iBAQ_TransEffect_A549_label</td>\n",
       "      <td>1698</td>\n",
       "      <td>[ENSG00000203811, ENSG00000203852, ENSG0000018...</td>\n",
       "      <td>[ENSG00000203811, ENSG00000203852, ENSG0000018...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>iBAQ_TransEffect_HEK293_label</td>\n",
       "      <td>1978</td>\n",
       "      <td>[ENSG00000203811, ENSG00000203852, ENSG0000018...</td>\n",
       "      <td>[ENSG00000203811, ENSG00000203852, ENSG0000018...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>iBAQ_TransEffect_HeLa_label</td>\n",
       "      <td>1805</td>\n",
       "      <td>[ENSG00000204389, ENSG00000158373, ENSG0000027...</td>\n",
       "      <td>[ENSG00000204389, ENSG00000158373, ENSG0000027...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>iBAQ_TransEffect_K562_label</td>\n",
       "      <td>1571</td>\n",
       "      <td>[ENSG00000203811, ENSG00000183598, ENSG0000020...</td>\n",
       "      <td>[ENSG00000203811, ENSG00000183598, ENSG0000020...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TransEffect_Hela_Nagaraj_label</td>\n",
       "      <td>1247</td>\n",
       "      <td>[ENSG00000254618, ENSG00000243678, ENSG0000013...</td>\n",
       "      <td>[ENSG00000254618, ENSG00000243678, ENSG0000013...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TransEffect_Hela_Zecha_label</td>\n",
       "      <td>1485</td>\n",
       "      <td>[ENSG00000131323, ENSG00000131171, ENSG0000010...</td>\n",
       "      <td>[ENSG00000131323, ENSG00000131171, ENSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A549-SRR611121-SRR3286543-zhanggong_label</td>\n",
       "      <td>3015</td>\n",
       "      <td>[ENSG00000277027, ENSG00000284459, ENSG0000025...</td>\n",
       "      <td>[ENSG00000277027, ENSG00000284459, ENSG0000025...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hela-SRR8443314-SRR8443308_label</td>\n",
       "      <td>2914</td>\n",
       "      <td>[ENSG00000200169, ENSG00000238825, ENSG0000027...</td>\n",
       "      <td>[ENSG00000200169, ENSG00000238825, ENSG0000027...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HEK293-SRR1795409_10_label</td>\n",
       "      <td>3557</td>\n",
       "      <td>[ENSG00000178977, ENSG00000138606, ENSG0000011...</td>\n",
       "      <td>[ENSG00000178977, ENSG00000138606, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ESC_GR_2017_label</td>\n",
       "      <td>2367</td>\n",
       "      <td>[ENSG00000105618, ENSG00000172340, ENSG0000010...</td>\n",
       "      <td>[ENSG00000105618, ENSG00000172340, ENSG0000010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Hela_GSE102113_label</td>\n",
       "      <td>2888</td>\n",
       "      <td>[ENSG00000000003, ENSG00000165672, ENSG0000006...</td>\n",
       "      <td>[ENSG00000000003, ENSG00000165672, ENSG0000006...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>K562_GR_2017_label</td>\n",
       "      <td>2785</td>\n",
       "      <td>[ENSG00000161016, ENSG00000244038, ENSG0000011...</td>\n",
       "      <td>[ENSG00000161016, ENSG00000244038, ENSG0000011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>K562_elife_label</td>\n",
       "      <td>2215</td>\n",
       "      <td>[ENSG00000117592, ENSG00000176894, ENSG0000019...</td>\n",
       "      <td>[ENSG00000117592, ENSG00000176894, ENSG0000019...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>K562_timeLapse_label</td>\n",
       "      <td>1323</td>\n",
       "      <td>[ENSG00000196565, ENSG00000142534, ENSG0000013...</td>\n",
       "      <td>[ENSG00000196565, ENSG00000142534, ENSG0000013...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      dataset  data size  \\\n",
       "0                 iBAQ_TransEffect_A549_label       1698   \n",
       "1               iBAQ_TransEffect_HEK293_label       1978   \n",
       "2                 iBAQ_TransEffect_HeLa_label       1805   \n",
       "3                 iBAQ_TransEffect_K562_label       1571   \n",
       "4              TransEffect_Hela_Nagaraj_label       1247   \n",
       "5                TransEffect_Hela_Zecha_label       1485   \n",
       "6   A549-SRR611121-SRR3286543-zhanggong_label       3015   \n",
       "7            hela-SRR8443314-SRR8443308_label       2914   \n",
       "8                  HEK293-SRR1795409_10_label       3557   \n",
       "9                           ESC_GR_2017_label       2367   \n",
       "10                       Hela_GSE102113_label       2888   \n",
       "11                         K562_GR_2017_label       2785   \n",
       "12                           K562_elife_label       2215   \n",
       "13                       K562_timeLapse_label       1323   \n",
       "\n",
       "                                         top 10 genes  \\\n",
       "0   [ENSG00000203811, ENSG00000203852, ENSG0000018...   \n",
       "1   [ENSG00000203811, ENSG00000203852, ENSG0000018...   \n",
       "2   [ENSG00000204389, ENSG00000158373, ENSG0000027...   \n",
       "3   [ENSG00000203811, ENSG00000183598, ENSG0000020...   \n",
       "4   [ENSG00000254618, ENSG00000243678, ENSG0000013...   \n",
       "5   [ENSG00000131323, ENSG00000131171, ENSG0000010...   \n",
       "6   [ENSG00000277027, ENSG00000284459, ENSG0000025...   \n",
       "7   [ENSG00000200169, ENSG00000238825, ENSG0000027...   \n",
       "8   [ENSG00000178977, ENSG00000138606, ENSG0000011...   \n",
       "9   [ENSG00000105618, ENSG00000172340, ENSG0000010...   \n",
       "10  [ENSG00000000003, ENSG00000165672, ENSG0000006...   \n",
       "11  [ENSG00000161016, ENSG00000244038, ENSG0000011...   \n",
       "12  [ENSG00000117592, ENSG00000176894, ENSG0000019...   \n",
       "13  [ENSG00000196565, ENSG00000142534, ENSG0000013...   \n",
       "\n",
       "                                         top 20 genes  \n",
       "0   [ENSG00000203811, ENSG00000203852, ENSG0000018...  \n",
       "1   [ENSG00000203811, ENSG00000203852, ENSG0000018...  \n",
       "2   [ENSG00000204389, ENSG00000158373, ENSG0000027...  \n",
       "3   [ENSG00000203811, ENSG00000183598, ENSG0000020...  \n",
       "4   [ENSG00000254618, ENSG00000243678, ENSG0000013...  \n",
       "5   [ENSG00000131323, ENSG00000131171, ENSG0000010...  \n",
       "6   [ENSG00000277027, ENSG00000284459, ENSG0000025...  \n",
       "7   [ENSG00000200169, ENSG00000238825, ENSG0000027...  \n",
       "8   [ENSG00000178977, ENSG00000138606, ENSG0000011...  \n",
       "9   [ENSG00000105618, ENSG00000172340, ENSG0000010...  \n",
       "10  [ENSG00000000003, ENSG00000165672, ENSG0000006...  \n",
       "11  [ENSG00000161016, ENSG00000244038, ENSG0000011...  \n",
       "12  [ENSG00000117592, ENSG00000176894, ENSG0000019...  \n",
       "13  [ENSG00000196565, ENSG00000142534, ENSG0000013...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "celllines_cols = [item  for item in all_genes_omics.columns if 'label' in item and 'consistent_label' not in item]\n",
    "\n",
    "celllines_high_genes, celllines_counts, cellines_top_10_genes, cellines_top_20_genes = [], [], [], []\n",
    "for col in celllines_cols:\n",
    "    celllines_high_genes.append(list(all_genes_omics[col][all_genes_omics[col]==1].index))\n",
    "    cellines_top_10_genes.append(list(all_genes_omics[all_genes_omics[col]==1\n",
    "                                                     ][col.replace('label', 'exp')].sort_values(ascending=False).head(10).index))\n",
    "    cellines_top_20_genes.append(list(all_genes_omics[all_genes_omics[col]==1\n",
    "                                                     ][col.replace('label', 'exp')].sort_values(ascending=False).head(20).index))\n",
    "    \n",
    "    celllines_counts.append(all_genes_omics[col][all_genes_omics[col]==1].shape[0])\n",
    "    \n",
    "cellines_pd = pd.DataFrame()\n",
    "\n",
    "cellines_pd['dataset']  = celllines_cols\n",
    "cellines_pd['data size']  = celllines_counts\n",
    "cellines_pd['top 10 genes']  = cellines_top_10_genes\n",
    "cellines_pd['top 20 genes']  = cellines_top_20_genes\n",
    "\n",
    "cellines_pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e95a6ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " top 10 total: 113\n",
      " top 20 total: 216\n"
     ]
    }
   ],
   "source": [
    "from itertools import chain\n",
    "\n",
    "all_cellines_top_10_genes = set(list(chain.from_iterable(cellines_top_10_genes)))\n",
    "all_cellines_top_20_genes = set(list(chain.from_iterable(cellines_top_20_genes)))\n",
    "print(f\" top 10 total: {len(all_cellines_top_10_genes)}\")\n",
    "\n",
    "print(f\" top 20 total: {len(all_cellines_top_20_genes)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0a94310",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "all_cells_omics_mRNA = os.path.join('/home/yul/project/nas_project/bioSPT/data/processed/mRNAfna', 'all_cells_omics_mRNA.train.fa')\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d4e85db7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ENSG00000196565', 'ENSG00000142534', 'ENSG00000137571',\n",
       "       'ENSG00000142168', 'ENSG00000137818', 'ENSG00000168653',\n",
       "       'ENSG00000182774', 'ENSG00000118181', 'ENSG00000231500',\n",
       "       'ENSG00000084207'],\n",
       "      dtype='object', name='ensembl_gene_id')"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# all_genes_omics\n",
    "\n",
    "all_genes_omics[celllines_cols]\n",
    "\n",
    "all_genes_omics[all_genes_omics[col]==1][col.replace('label', 'exp')].sort_values(ascending=False).head(10).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "54520aaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 所有细胞系共用\n",
    "all_genes_omics_consistent_genes = all_genes_omics[(all_genes_omics[celllines_cols] == 1).all(axis=1)].index\n",
    "\n",
    "all_cellines_top_10_genes & set(all_genes_omics_consistent_genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4e377951",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000000003',\n",
       " 'ENSG00000013306',\n",
       " 'ENSG00000013583',\n",
       " 'ENSG00000060762',\n",
       " 'ENSG00000067177',\n",
       " 'ENSG00000069849',\n",
       " 'ENSG00000075945',\n",
       " 'ENSG00000079819',\n",
       " 'ENSG00000083457',\n",
       " 'ENSG00000084207',\n",
       " 'ENSG00000084754',\n",
       " 'ENSG00000087086',\n",
       " 'ENSG00000087191',\n",
       " 'ENSG00000089280',\n",
       " 'ENSG00000101082',\n",
       " 'ENSG00000101152',\n",
       " 'ENSG00000101444',\n",
       " 'ENSG00000102007',\n",
       " 'ENSG00000102178',\n",
       " 'ENSG00000102317',\n",
       " 'ENSG00000103202',\n",
       " 'ENSG00000103275',\n",
       " 'ENSG00000103534',\n",
       " 'ENSG00000105193',\n",
       " 'ENSG00000105583',\n",
       " 'ENSG00000105618',\n",
       " 'ENSG00000108294',\n",
       " 'ENSG00000108298',\n",
       " 'ENSG00000109107',\n",
       " 'ENSG00000109846',\n",
       " 'ENSG00000110107',\n",
       " 'ENSG00000110651',\n",
       " 'ENSG00000111229',\n",
       " 'ENSG00000111911',\n",
       " 'ENSG00000112335',\n",
       " 'ENSG00000112378',\n",
       " 'ENSG00000114023',\n",
       " 'ENSG00000114942',\n",
       " 'ENSG00000115112',\n",
       " 'ENSG00000115541',\n",
       " 'ENSG00000115685',\n",
       " 'ENSG00000116209',\n",
       " 'ENSG00000116288',\n",
       " 'ENSG00000116455',\n",
       " 'ENSG00000117395',\n",
       " 'ENSG00000117419',\n",
       " 'ENSG00000117450',\n",
       " 'ENSG00000117592',\n",
       " 'ENSG00000118181',\n",
       " 'ENSG00000119396',\n",
       " 'ENSG00000122566',\n",
       " 'ENSG00000124164',\n",
       " 'ENSG00000124299',\n",
       " 'ENSG00000124562',\n",
       " 'ENSG00000124942',\n",
       " 'ENSG00000126067',\n",
       " 'ENSG00000128422',\n",
       " 'ENSG00000128534',\n",
       " 'ENSG00000129235',\n",
       " 'ENSG00000129562',\n",
       " 'ENSG00000130600',\n",
       " 'ENSG00000130707',\n",
       " 'ENSG00000131171',\n",
       " 'ENSG00000131238',\n",
       " 'ENSG00000131323',\n",
       " 'ENSG00000131400',\n",
       " 'ENSG00000131469',\n",
       " 'ENSG00000131828',\n",
       " 'ENSG00000132963',\n",
       " 'ENSG00000134184',\n",
       " 'ENSG00000134709',\n",
       " 'ENSG00000135046',\n",
       " 'ENSG00000137571',\n",
       " 'ENSG00000137818',\n",
       " 'ENSG00000138175',\n",
       " 'ENSG00000138606',\n",
       " 'ENSG00000138772',\n",
       " 'ENSG00000139684',\n",
       " 'ENSG00000142168',\n",
       " 'ENSG00000142534',\n",
       " 'ENSG00000142657',\n",
       " 'ENSG00000142669',\n",
       " 'ENSG00000142937',\n",
       " 'ENSG00000143321',\n",
       " 'ENSG00000143632',\n",
       " 'ENSG00000145335',\n",
       " 'ENSG00000147140',\n",
       " 'ENSG00000147164',\n",
       " 'ENSG00000148303',\n",
       " 'ENSG00000151304',\n",
       " 'ENSG00000157110',\n",
       " 'ENSG00000158104',\n",
       " 'ENSG00000158373',\n",
       " 'ENSG00000158406',\n",
       " 'ENSG00000161016',\n",
       " 'ENSG00000161970',\n",
       " 'ENSG00000163159',\n",
       " 'ENSG00000164040',\n",
       " 'ENSG00000164111',\n",
       " 'ENSG00000164587',\n",
       " 'ENSG00000165410',\n",
       " 'ENSG00000165629',\n",
       " 'ENSG00000165672',\n",
       " 'ENSG00000165678',\n",
       " 'ENSG00000165704',\n",
       " 'ENSG00000166165',\n",
       " 'ENSG00000167515',\n",
       " 'ENSG00000167645',\n",
       " 'ENSG00000168653',\n",
       " 'ENSG00000169230',\n",
       " 'ENSG00000169241',\n",
       " 'ENSG00000169567',\n",
       " 'ENSG00000169689',\n",
       " 'ENSG00000170144',\n",
       " 'ENSG00000170860',\n",
       " 'ENSG00000171262',\n",
       " 'ENSG00000171858',\n",
       " 'ENSG00000172007',\n",
       " 'ENSG00000172340',\n",
       " 'ENSG00000173113',\n",
       " 'ENSG00000174021',\n",
       " 'ENSG00000174444',\n",
       " 'ENSG00000174748',\n",
       " 'ENSG00000175792',\n",
       " 'ENSG00000176894',\n",
       " 'ENSG00000177156',\n",
       " 'ENSG00000178741',\n",
       " 'ENSG00000178977',\n",
       " 'ENSG00000178982',\n",
       " 'ENSG00000181588',\n",
       " 'ENSG00000182054',\n",
       " 'ENSG00000182718',\n",
       " 'ENSG00000182774',\n",
       " 'ENSG00000182809',\n",
       " 'ENSG00000183598',\n",
       " 'ENSG00000183684',\n",
       " 'ENSG00000184260',\n",
       " 'ENSG00000184678',\n",
       " 'ENSG00000185305',\n",
       " 'ENSG00000187109',\n",
       " 'ENSG00000187837',\n",
       " 'ENSG00000196465',\n",
       " 'ENSG00000196565',\n",
       " 'ENSG00000196747',\n",
       " 'ENSG00000196787',\n",
       " 'ENSG00000196976',\n",
       " 'ENSG00000197061',\n",
       " 'ENSG00000198034',\n",
       " 'ENSG00000198830',\n",
       " 'ENSG00000199024',\n",
       " 'ENSG00000199094',\n",
       " 'ENSG00000199753',\n",
       " 'ENSG00000199961',\n",
       " 'ENSG00000199990',\n",
       " 'ENSG00000200084',\n",
       " 'ENSG00000200169',\n",
       " 'ENSG00000201659',\n",
       " 'ENSG00000202314',\n",
       " 'ENSG00000202529',\n",
       " 'ENSG00000203811',\n",
       " 'ENSG00000203852',\n",
       " 'ENSG00000203989',\n",
       " 'ENSG00000204389',\n",
       " 'ENSG00000204390',\n",
       " 'ENSG00000206602',\n",
       " 'ENSG00000206630',\n",
       " 'ENSG00000206754',\n",
       " 'ENSG00000206775',\n",
       " 'ENSG00000207116',\n",
       " 'ENSG00000207280',\n",
       " 'ENSG00000207405',\n",
       " 'ENSG00000207468',\n",
       " 'ENSG00000207500',\n",
       " 'ENSG00000207757',\n",
       " 'ENSG00000210825',\n",
       " 'ENSG00000212452',\n",
       " 'ENSG00000213931',\n",
       " 'ENSG00000214253',\n",
       " 'ENSG00000221740',\n",
       " 'ENSG00000228474',\n",
       " 'ENSG00000231500',\n",
       " 'ENSG00000232119',\n",
       " 'ENSG00000236362',\n",
       " 'ENSG00000238825',\n",
       " 'ENSG00000238835',\n",
       " 'ENSG00000241837',\n",
       " 'ENSG00000243678',\n",
       " 'ENSG00000244038',\n",
       " 'ENSG00000246705',\n",
       " 'ENSG00000251733',\n",
       " 'ENSG00000252316',\n",
       " 'ENSG00000252712',\n",
       " 'ENSG00000254618',\n",
       " 'ENSG00000263764',\n",
       " 'ENSG00000263776',\n",
       " 'ENSG00000264229',\n",
       " 'ENSG00000264294',\n",
       " 'ENSG00000270276',\n",
       " 'ENSG00000270882',\n",
       " 'ENSG00000272015',\n",
       " 'ENSG00000273213',\n",
       " 'ENSG00000273542',\n",
       " 'ENSG00000273544',\n",
       " 'ENSG00000274978',\n",
       " 'ENSG00000274997',\n",
       " 'ENSG00000275143',\n",
       " 'ENSG00000275221',\n",
       " 'ENSG00000275996',\n",
       " 'ENSG00000276180',\n",
       " 'ENSG00000276903',\n",
       " 'ENSG00000276966',\n",
       " 'ENSG00000277027',\n",
       " 'ENSG00000277157',\n",
       " 'ENSG00000277370',\n",
       " 'ENSG00000277443',\n",
       " 'ENSG00000277586',\n",
       " 'ENSG00000278637',\n",
       " 'ENSG00000278677',\n",
       " 'ENSG00000278705',\n",
       " 'ENSG00000281394',\n",
       " 'ENSG00000283203',\n",
       " 'ENSG00000284179',\n",
       " 'ENSG00000284459',\n",
       " 'ENSG00000284485'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "UTR_genes = (all_cellines_top_20_genes |  set(all_genes_omics_consistent_genes))\n",
    "\n",
    "\n",
    "UTR_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "20b2773b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "224"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "len(UTR_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a3542fdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "UTR_genes_human_ts = human_ts.set_index('Esemble').reindex(UTR_genes)\n",
    "\n",
    "\n",
    "UTR_genes_human_ts_clean = UTR_genes_human_ts[\n",
    "    ~UTR_genes_human_ts[\"UTR5\"].isnull() & ~UTR_genes_human_ts[\"UTR3\"].isnull() ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f5f90b05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>utr</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>r5</th>\n",
       "      <th>r6</th>\n",
       "      <th>r7</th>\n",
       "      <th>r8</th>\n",
       "      <th>r9</th>\n",
       "      <th>r10</th>\n",
       "      <th>r11</th>\n",
       "      <th>r12</th>\n",
       "      <th>r13</th>\n",
       "      <th>rl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120605</th>\n",
       "      <td>CCACTCGATTAACATGTTAACAACATACTCGTCCGGCCGATCAGCG...</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033645</td>\n",
       "      <td>0.020343</td>\n",
       "      <td>0.019850</td>\n",
       "      <td>0.017988</td>\n",
       "      <td>0.019283</td>\n",
       "      <td>0.033252</td>\n",
       "      <td>0.033252</td>\n",
       "      <td>0.027581</td>\n",
       "      <td>0.035877</td>\n",
       "      <td>3.039939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11605</th>\n",
       "      <td>CAAATCATGTGCAGCCCTGGCGACCGTACTGCGGTACAAGAAAGTA...</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.049711</td>\n",
       "      <td>0.028540</td>\n",
       "      <td>0.027431</td>\n",
       "      <td>0.024766</td>\n",
       "      <td>0.024241</td>\n",
       "      <td>0.039457</td>\n",
       "      <td>0.039457</td>\n",
       "      <td>0.039840</td>\n",
       "      <td>0.038785</td>\n",
       "      <td>3.895109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128224</th>\n",
       "      <td>GTTATACTAGAAGAAACTTGAGATTATGGAGCAGTCCGTCAAGGAC...</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.042436</td>\n",
       "      <td>0.022460</td>\n",
       "      <td>0.025015</td>\n",
       "      <td>0.024275</td>\n",
       "      <td>0.021591</td>\n",
       "      <td>0.028353</td>\n",
       "      <td>0.028353</td>\n",
       "      <td>0.028963</td>\n",
       "      <td>0.041985</td>\n",
       "      <td>3.334524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239107</th>\n",
       "      <td>CTTAGACAAAAACAACGCGCTTTCCAGTATGCGGAGCCTTGACGGT...</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041241</td>\n",
       "      <td>0.025793</td>\n",
       "      <td>0.022751</td>\n",
       "      <td>0.022632</td>\n",
       "      <td>0.026617</td>\n",
       "      <td>0.038302</td>\n",
       "      <td>0.038302</td>\n",
       "      <td>0.032788</td>\n",
       "      <td>0.031043</td>\n",
       "      <td>3.575082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59082</th>\n",
       "      <td>GTATCAAATCACGGCCAACCCGACGGAGTACCCCGCGTCGATGGTC...</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>...</td>\n",
       "      <td>0.071306</td>\n",
       "      <td>0.040719</td>\n",
       "      <td>0.033335</td>\n",
       "      <td>0.030367</td>\n",
       "      <td>0.033080</td>\n",
       "      <td>0.051449</td>\n",
       "      <td>0.051449</td>\n",
       "      <td>0.046052</td>\n",
       "      <td>0.036447</td>\n",
       "      <td>4.593712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      utr         0         1  \\\n",
       "120605  CCACTCGATTAACATGTTAACAACATACTCGTCCGGCCGATCAGCG...  0.000137  0.000109   \n",
       "11605   CAAATCATGTGCAGCCCTGGCGACCGTACTGCGGTACAAGAAAGTA...  0.000067  0.000070   \n",
       "128224  GTTATACTAGAAGAAACTTGAGATTATGGAGCAGTCCGTCAAGGAC...  0.000088  0.000081   \n",
       "239107  CTTAGACAAAAACAACGCGCTTTCCAGTATGCGGAGCCTTGACGGT...  0.000078  0.000071   \n",
       "59082   GTATCAAATCACGGCCAACCCGACGGAGTACCCCGCGTCGATGGTC...  0.000044  0.000045   \n",
       "\n",
       "               2         3         4         5         6         7         8  \\\n",
       "120605  0.000057  0.000033  0.000015  0.000016  0.000009  0.000009  0.000008   \n",
       "11605   0.000065  0.000048  0.000023  0.000020  0.000011  0.000011  0.000010   \n",
       "128224  0.000059  0.000035  0.000017  0.000016  0.000009  0.000009  0.000009   \n",
       "239107  0.000060  0.000038  0.000025  0.000016  0.000010  0.000009  0.000009   \n",
       "59082   0.000050  0.000051  0.000033  0.000026  0.000015  0.000012  0.000011   \n",
       "\n",
       "        ...        r5        r6        r7        r8        r9       r10  \\\n",
       "120605  ...  0.033645  0.020343  0.019850  0.017988  0.019283  0.033252   \n",
       "11605   ...  0.049711  0.028540  0.027431  0.024766  0.024241  0.039457   \n",
       "128224  ...  0.042436  0.022460  0.025015  0.024275  0.021591  0.028353   \n",
       "239107  ...  0.041241  0.025793  0.022751  0.022632  0.026617  0.038302   \n",
       "59082   ...  0.071306  0.040719  0.033335  0.030367  0.033080  0.051449   \n",
       "\n",
       "             r11       r12       r13        rl  \n",
       "120605  0.033252  0.027581  0.035877  3.039939  \n",
       "11605   0.039457  0.039840  0.038785  3.895109  \n",
       "128224  0.028353  0.028963  0.041985  3.334524  \n",
       "239107  0.038302  0.032788  0.031043  3.575082  \n",
       "59082   0.051449  0.046052  0.036447  4.593712  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gzip\n",
    "\n",
    "# with gzip.open('/data1/genomics/yul/bioSPT2/data/interiem/moderna_polysome/GSM3130435_egfp_unmod_1.csv.gz', 'r') as f:\n",
    "    \n",
    "egfp_unmod_1 = pd.read_csv('/data1/genomics/yul/bioSPT2/data/interiem/moderna_polysome/GSM3130435_egfp_unmod_1.csv.gz', compression='gzip', sep=',', index_col=0)\n",
    "\n",
    "egfp_unmod_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fe02c2e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "UTR_genes_human_ts_clean_all = pd.concat([UTR_genes_human_ts_clean, pd.DataFrame(egfp_unmod_1.sort_values(by='rl', ascending=False).head(50)['utr']).rename(columns={'utr':'UTR5'})], axis=0)\n",
    "\n",
    "UTR_genes_human_ts_clean_all.to_csv('/data1/genomics/yul/bioSPT2/data/interiem/processed/UTR_high_genes_human_ts_clean_all.csv', sep=',')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "51c1b08d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id_raw</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>UTR5_regs</th>\n",
       "      <th>CDS_regs</th>\n",
       "      <th>UTR3_regs</th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>UTR3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000087086</th>\n",
       "      <td>ENST00000331825</td>\n",
       "      <td>ENSG00000087086.15</td>\n",
       "      <td>&gt;ENST00000331825.11|ENSG00000087086.15|OTTHUMG...</td>\n",
       "      <td>GCAGTTCGGCGGTCCCGCGGGTCTGTCTCTTGCTTCAACAGTGTTT...</td>\n",
       "      <td>ENST00000331825.11</td>\n",
       "      <td>ENST00000331825.11</td>\n",
       "      <td>FTL</td>\n",
       "      <td>UTR5:1-199</td>\n",
       "      <td>CDS:200-727</td>\n",
       "      <td>UTR3:728-871</td>\n",
       "      <td>GCAGTTCGGCGGTCCCGCGGGTCTGTCTCTTGCTTCAACAGTGTTT...</td>\n",
       "      <td>ATGAGCTCCCAGATTCGTCAGAATTATTCCACCGACGTGGAGGCAG...</td>\n",
       "      <td>GAGCCTTCTGAGCCCAGCGACTTCTGAAGGGCCCCTTGCAAAGTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000138606</th>\n",
       "      <td>ENST00000458022</td>\n",
       "      <td>ENSG00000138606.20</td>\n",
       "      <td>&gt;ENST00000458022.6|ENSG00000138606.20|OTTHUMG0...</td>\n",
       "      <td>GAGTTGCCACCTGGGTGCCCATGGGCACTGCAGGCTAGAGCTGAGG...</td>\n",
       "      <td>ENST00000458022.6</td>\n",
       "      <td>ENST00000458022.6</td>\n",
       "      <td>SHF</td>\n",
       "      <td>UTR5:1-209</td>\n",
       "      <td>CDS:210-929</td>\n",
       "      <td>UTR3:930-1163</td>\n",
       "      <td>GAGTTGCCACCTGGGTGCCCATGGGCACTGCAGGCTAGAGCTGAGG...</td>\n",
       "      <td>ATGGAGCCCTATGAGGCTCAAAAGATGATGGCCGAGATCCGGGGCT...</td>\n",
       "      <td>ATGTGAAGCCAGGGCACTGTGATAGACCTGTACCCAGCCCTGTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000105618</th>\n",
       "      <td>ENST00000321030</td>\n",
       "      <td>ENSG00000105618.14</td>\n",
       "      <td>&gt;ENST00000321030.9|ENSG00000105618.14|OTTHUMG0...</td>\n",
       "      <td>GGTGAGCGACTAACGCTAGAAACAGTGGTGCGCGGAGAGGAGAGGC...</td>\n",
       "      <td>ENST00000321030.9</td>\n",
       "      <td>ENST00000321030.9</td>\n",
       "      <td>PRPF31</td>\n",
       "      <td>UTR5:1-52</td>\n",
       "      <td>CDS:53-1552</td>\n",
       "      <td>UTR3:1553-1833</td>\n",
       "      <td>GGTGAGCGACTAACGCTAGAAACAGTGGTGCGCGGAGAGGAGAGGC...</td>\n",
       "      <td>ATGTCTCTGGCAGATGAGCTCTTAGCTGATCTCGAAGAGGCAGCAG...</td>\n",
       "      <td>ATGACTGCGTGTGTCCAAGGTGGCTTCCCACTGAAGGGACACAGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000164111</th>\n",
       "      <td>ENST00000296511</td>\n",
       "      <td>ENSG00000164111.15</td>\n",
       "      <td>&gt;ENST00000296511.10|ENSG00000164111.15|OTTHUMG...</td>\n",
       "      <td>AGTCTAGGTGCAGCTGCCGGATCCTTCAGCGTCTGCATCTCGGCGT...</td>\n",
       "      <td>ENST00000296511.10</td>\n",
       "      <td>ENST00000296511.10</td>\n",
       "      <td>ANXA5</td>\n",
       "      <td>UTR5:1-153</td>\n",
       "      <td>CDS:154-1116</td>\n",
       "      <td>UTR3:1117-1638</td>\n",
       "      <td>AGTCTAGGTGCAGCTGCCGGATCCTTCAGCGTCTGCATCTCGGCGT...</td>\n",
       "      <td>ATGGCACAGGTTCTCAGAGGCACTGTGACTGACTTCCCTGGATTTG...</td>\n",
       "      <td>CGTGTCACGGGGAAGAGCTCCCTGCTGTGTGCCTGCACCACCCCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278637</th>\n",
       "      <td>ENST00000617569</td>\n",
       "      <td>ENSG00000278637.2</td>\n",
       "      <td>&gt;ENST00000617569.2|ENSG00000278637.2|OTTHUMG00...</td>\n",
       "      <td>ACTTGCTCTTGGTTCACTTCTTGGGAAGTCATGTCTGGACGTGGTA...</td>\n",
       "      <td>ENST00000617569.2</td>\n",
       "      <td>ENST00000617569.2</td>\n",
       "      <td>H4C1</td>\n",
       "      <td>UTR5:1-30</td>\n",
       "      <td>CDS:31-342</td>\n",
       "      <td>UTR3:343-402</td>\n",
       "      <td>ACTTGCTCTTGGTTCACTTCTTGGGAAGTC</td>\n",
       "      <td>ATGTCTGGACGTGGTAAGGGCGGGAAGGGTTTGGGTAAGGGGGGTG...</td>\n",
       "      <td>GGTTGCTGATTTCTCCACAGCTTGCATTTCTGAACCAAAGGCCCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325682</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TAATATCATTCGTCTACACAACTACGATATGGAAGCTATTTCGCAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325969</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCGGGAGATGTACTGGGAGAGATTTCCTTCAAGCGTAGAAGATCAT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325933</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CATCTATGCATAACGATATAAAACATTAGAGTCAACCTTTAAGCAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325719</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGTGCCGCACAACGCCGGTAAATAGTAGTAGGGTCGTGCACATCAC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325728</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCGCCTAGGGAAACAAACACCTTATCACATATAATGCAGGCGGCCA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Transcript             gene_id  \\\n",
       "ENSG00000087086  ENST00000331825  ENSG00000087086.15   \n",
       "ENSG00000138606  ENST00000458022  ENSG00000138606.20   \n",
       "ENSG00000105618  ENST00000321030  ENSG00000105618.14   \n",
       "ENSG00000164111  ENST00000296511  ENSG00000164111.15   \n",
       "ENSG00000278637  ENST00000617569   ENSG00000278637.2   \n",
       "...                          ...                 ...   \n",
       "325682                       NaN                 NaN   \n",
       "325969                       NaN                 NaN   \n",
       "325933                       NaN                 NaN   \n",
       "325719                       NaN                 NaN   \n",
       "325728                       NaN                 NaN   \n",
       "\n",
       "                                                              name  \\\n",
       "ENSG00000087086  >ENST00000331825.11|ENSG00000087086.15|OTTHUMG...   \n",
       "ENSG00000138606  >ENST00000458022.6|ENSG00000138606.20|OTTHUMG0...   \n",
       "ENSG00000105618  >ENST00000321030.9|ENSG00000105618.14|OTTHUMG0...   \n",
       "ENSG00000164111  >ENST00000296511.10|ENSG00000164111.15|OTTHUMG...   \n",
       "ENSG00000278637  >ENST00000617569.2|ENSG00000278637.2|OTTHUMG00...   \n",
       "...                                                            ...   \n",
       "325682                                                         NaN   \n",
       "325969                                                         NaN   \n",
       "325933                                                         NaN   \n",
       "325719                                                         NaN   \n",
       "325728                                                         NaN   \n",
       "\n",
       "                                                               seq  \\\n",
       "ENSG00000087086  GCAGTTCGGCGGTCCCGCGGGTCTGTCTCTTGCTTCAACAGTGTTT...   \n",
       "ENSG00000138606  GAGTTGCCACCTGGGTGCCCATGGGCACTGCAGGCTAGAGCTGAGG...   \n",
       "ENSG00000105618  GGTGAGCGACTAACGCTAGAAACAGTGGTGCGCGGAGAGGAGAGGC...   \n",
       "ENSG00000164111  AGTCTAGGTGCAGCTGCCGGATCCTTCAGCGTCTGCATCTCGGCGT...   \n",
       "ENSG00000278637  ACTTGCTCTTGGTTCACTTCTTGGGAAGTCATGTCTGGACGTGGTA...   \n",
       "...                                                            ...   \n",
       "325682                                                         NaN   \n",
       "325969                                                         NaN   \n",
       "325933                                                         NaN   \n",
       "325719                                                         NaN   \n",
       "325728                                                         NaN   \n",
       "\n",
       "                      transcript_id         gene_id_raw gene_name   UTR5_regs  \\\n",
       "ENSG00000087086  ENST00000331825.11  ENST00000331825.11       FTL  UTR5:1-199   \n",
       "ENSG00000138606   ENST00000458022.6   ENST00000458022.6       SHF  UTR5:1-209   \n",
       "ENSG00000105618   ENST00000321030.9   ENST00000321030.9    PRPF31   UTR5:1-52   \n",
       "ENSG00000164111  ENST00000296511.10  ENST00000296511.10     ANXA5  UTR5:1-153   \n",
       "ENSG00000278637   ENST00000617569.2   ENST00000617569.2      H4C1   UTR5:1-30   \n",
       "...                             ...                 ...       ...         ...   \n",
       "325682                          NaN                 NaN       NaN         NaN   \n",
       "325969                          NaN                 NaN       NaN         NaN   \n",
       "325933                          NaN                 NaN       NaN         NaN   \n",
       "325719                          NaN                 NaN       NaN         NaN   \n",
       "325728                          NaN                 NaN       NaN         NaN   \n",
       "\n",
       "                     CDS_regs       UTR3_regs  \\\n",
       "ENSG00000087086   CDS:200-727    UTR3:728-871   \n",
       "ENSG00000138606   CDS:210-929   UTR3:930-1163   \n",
       "ENSG00000105618   CDS:53-1552  UTR3:1553-1833   \n",
       "ENSG00000164111  CDS:154-1116  UTR3:1117-1638   \n",
       "ENSG00000278637    CDS:31-342    UTR3:343-402   \n",
       "...                       ...             ...   \n",
       "325682                    NaN             NaN   \n",
       "325969                    NaN             NaN   \n",
       "325933                    NaN             NaN   \n",
       "325719                    NaN             NaN   \n",
       "325728                    NaN             NaN   \n",
       "\n",
       "                                                              UTR5  \\\n",
       "ENSG00000087086  GCAGTTCGGCGGTCCCGCGGGTCTGTCTCTTGCTTCAACAGTGTTT...   \n",
       "ENSG00000138606  GAGTTGCCACCTGGGTGCCCATGGGCACTGCAGGCTAGAGCTGAGG...   \n",
       "ENSG00000105618  GGTGAGCGACTAACGCTAGAAACAGTGGTGCGCGGAGAGGAGAGGC...   \n",
       "ENSG00000164111  AGTCTAGGTGCAGCTGCCGGATCCTTCAGCGTCTGCATCTCGGCGT...   \n",
       "ENSG00000278637                     ACTTGCTCTTGGTTCACTTCTTGGGAAGTC   \n",
       "...                                                            ...   \n",
       "325682           TAATATCATTCGTCTACACAACTACGATATGGAAGCTATTTCGCAG...   \n",
       "325969           GCGGGAGATGTACTGGGAGAGATTTCCTTCAAGCGTAGAAGATCAT...   \n",
       "325933           CATCTATGCATAACGATATAAAACATTAGAGTCAACCTTTAAGCAG...   \n",
       "325719           GGTGCCGCACAACGCCGGTAAATAGTAGTAGGGTCGTGCACATCAC...   \n",
       "325728           CCGCCTAGGGAAACAAACACCTTATCACATATAATGCAGGCGGCCA...   \n",
       "\n",
       "                                                               CDS  \\\n",
       "ENSG00000087086  ATGAGCTCCCAGATTCGTCAGAATTATTCCACCGACGTGGAGGCAG...   \n",
       "ENSG00000138606  ATGGAGCCCTATGAGGCTCAAAAGATGATGGCCGAGATCCGGGGCT...   \n",
       "ENSG00000105618  ATGTCTCTGGCAGATGAGCTCTTAGCTGATCTCGAAGAGGCAGCAG...   \n",
       "ENSG00000164111  ATGGCACAGGTTCTCAGAGGCACTGTGACTGACTTCCCTGGATTTG...   \n",
       "ENSG00000278637  ATGTCTGGACGTGGTAAGGGCGGGAAGGGTTTGGGTAAGGGGGGTG...   \n",
       "...                                                            ...   \n",
       "325682                                                         NaN   \n",
       "325969                                                         NaN   \n",
       "325933                                                         NaN   \n",
       "325719                                                         NaN   \n",
       "325728                                                         NaN   \n",
       "\n",
       "                                                              UTR3  \n",
       "ENSG00000087086  GAGCCTTCTGAGCCCAGCGACTTCTGAAGGGCCCCTTGCAAAGTAA...  \n",
       "ENSG00000138606  ATGTGAAGCCAGGGCACTGTGATAGACCTGTACCCAGCCCTGTGCC...  \n",
       "ENSG00000105618  ATGACTGCGTGTGTCCAAGGTGGCTTCCCACTGAAGGGACACAGAG...  \n",
       "ENSG00000164111  CGTGTCACGGGGAAGAGCTCCCTGCTGTGTGCCTGCACCACCCCAC...  \n",
       "ENSG00000278637  GGTTGCTGATTTCTCCACAGCTTGCATTTCTGAACCAAAGGCCCTT...  \n",
       "...                                                            ...  \n",
       "325682                                                         NaN  \n",
       "325969                                                         NaN  \n",
       "325933                                                         NaN  \n",
       "325719                                                         NaN  \n",
       "325728                                                         NaN  \n",
       "\n",
       "[215 rows x 13 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "UTR_genes_human_ts_clean_all\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dd5fe76",
   "metadata": {},
   "source": [
    "## mRNA 构造"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "d2904e46",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# load from file\n",
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique = pd.read_csv('/data1/genomics/yul/bioSPT2/data/interiem/processed/LinearDesign_LAMBDA_merged_CDS_gens_unique.csv')\n",
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique.shape\n",
    "\n",
    "\n",
    "raw_genes_ids = LinearDesign_LAMBDA_merged_CDS_gens_unique.drop_duplicates(subset=['ids'], keep='first', inplace=False)['ids'].apply(lambda x: x.split('|')[1].split('.')[0]).value_counts().index\n",
    "\n",
    "raw_genes_ids\n",
    "raw_genes_mRNAs = human_ts.set_index('Esemble').reindex(raw_genes_ids).reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "46704a41",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "raw_high_genes_mRNAs = UTR_genes_human_ts_clean_all[\n",
    "    \n",
    "    ~UTR_genes_human_ts_clean_all['UTR5'].isnull() &\n",
    "    ~UTR_genes_human_ts_clean_all['UTR3'].isnull() &\n",
    "    ~UTR_genes_human_ts_clean_all['CDS'].isnull()  \n",
    "                             ]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "54435856",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSG00000087086    ENST00000331825\n",
       "ENSG00000138606    ENST00000458022\n",
       "ENSG00000105618    ENST00000321030\n",
       "ENSG00000164111    ENST00000296511\n",
       "ENSG00000278637    ENST00000617569\n",
       "                        ...       \n",
       "325682                   MRL_TOP46\n",
       "325969                   MRL_TOP47\n",
       "325933                   MRL_TOP48\n",
       "325719                   MRL_TOP49\n",
       "325728                   MRL_TOP50\n",
       "Name: Transcript, Length: 215, dtype: object"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cds、utr配对\n",
    "\n",
    "# CDS\n",
    "\n",
    "UTR_genes_human_ts_clean_all['Transcript'].apply(lambda x: False if isinstance(x, str)  else True)\n",
    "# UTR_genes_human_ts_clean_all['Transcript_MRL']\n",
    "\n",
    "\n",
    "UTR_genes_human_ts_clean_all[\n",
    "UTR_genes_human_ts_clean_all['Transcript'].apply(lambda x: False if isinstance(x, str)  else True)].reset_index().index\n",
    "\n",
    "UTR_genes_human_ts_clean_all.loc[UTR_genes_human_ts_clean_all['Transcript'].apply(lambda x: False if isinstance(x, str)  else True)].index = list(range(50))\n",
    "\n",
    "UTR_genes_human_ts_clean_all['Transcript'] = UTR_genes_human_ts_clean_all.loc[~UTR_genes_human_ts_clean_all['Transcript'].apply(lambda x: False if isinstance(x, str)  else True)]['Transcript'].values.tolist() + [f'MRL_TOP{a}' for a in range(1, 51)]\n",
    "\n",
    "UTR_genes_human_ts_clean_all['Transcript']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "f75ebdc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mRNA_seq</th>\n",
       "      <th>mRNA_structure</th>\n",
       "      <th>mfe</th>\n",
       "      <th>cai</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ids</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000000233.10|ENSG00000004059.11|OTTHUMG00000023246.7|OTTHUMT00000059567.3|ARF5-201|ARF5|1032|UTR5:1-88|CDS:89-631|UTR3:632-1032|</th>\n",
       "      <td>[UGGGCCUGACCGUGUCCGCCCUGUUCAGCAGGAUCUUCGGCAAGA...</td>\n",
       "      <td>[....(((.(.(.(((((((((((..(((.(((((((((((((......</td>\n",
       "      <td>[-255.0]</td>\n",
       "      <td>[0.991]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000001146.7|ENSG00000003137.9|OTTHUMG00000129756.5|OTTHUMT00000251969.1|CYP26B1-201|CYP26B1|4556|UTR5:1-28|CDS:29-1567|UTR3:1568-4556|</th>\n",
       "      <td>[UGCUGUUCGAGGGCCUGGACCUGGUGAGCGCCCUGGCCACCCUGG...</td>\n",
       "      <td>[.((((((((..(((((((..(((((((.(((...((((((...))...</td>\n",
       "      <td>[-825.7, -866.0]</td>\n",
       "      <td>[0.984, 0.974]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000037869.7|ENSG00000035141.8|OTTHUMG00000129668.4|OTTHUMT00000251869.2|FAM136A-201|FAM136A|1810|UTR5:1-79|CDS:80-496|UTR3:497-1810|</th>\n",
       "      <td>[UGGCCGAGCUGCAGCAGCUGCGGGUGCAGGAGGCCGUGGAGUCCA...</td>\n",
       "      <td>[...((((((((.((((((.(((((((((((((((((..(((((((...</td>\n",
       "      <td>[-195.2, -207.5]</td>\n",
       "      <td>[0.994, 0.982]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000054650.9|ENSG00000041988.16|OTTHUMG00000001440.3|OTTHUMT00000004203.1|THAP3-201|THAP3|1361|UTR5:1-156|CDS:157-876|UTR3:877-1361|</th>\n",
       "      <td>[UGCCCAAGAGCUGUGCUGCUCGGCAGUGCUGCAACCGGUACUCCA...</td>\n",
       "      <td>[.........((((.((((((((((.(((((((....))))))).....</td>\n",
       "      <td>[-382.3, -424.3, -406.7, -403.5, -400.3, -391.6]</td>\n",
       "      <td>[0.987, 0.962, 0.975, 0.977, 0.979, 0.983]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000054666.11|ENSG00000049245.13|OTTHUMG00000001225.4|OTTHUMT00000003625.2|VAMP3-201|VAMP3|2178|UTR5:1-88|CDS:89-391|UTR3:392-2178|</th>\n",
       "      <td>[UGUCCACCGGGCCUACUGCCGCCACCGGCAGCAACCGGCGGCUGC...</td>\n",
       "      <td>[.(((((((((((.((.((((.(((.((((......)))).))).)...</td>\n",
       "      <td>[-132.0, -144.7, -140.4, -136.7]</td>\n",
       "      <td>[0.988, 0.972, 0.978, 0.983]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000695960.1|ENSG00000119844.17|OTTHUMG00000129539.6|-|AFTPH-212|AFTPH|3009|UTR5:1-270|CDS:271-2691|UTR3:2692-3009|</th>\n",
       "      <td>[UGGAGCCCGACAUCAUCAGGAUGUACAGCUCCAGCCCUCCACCCC...</td>\n",
       "      <td>[................(((((((((((((((((.(...(((((.(...</td>\n",
       "      <td>[-1139.7, -1266.0, -1249.6, -1193.7, -1157.0]</td>\n",
       "      <td>[0.982, 0.963, 0.966, 0.975, 0.98]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000696500.1|ENSG00000054523.20|OTTHUMG00000001817.2|-|KIF1B-215|KIF1B|7839|UTR5:1-388|CDS:389-1255|UTR3:1256-7839|</th>\n",
       "      <td>[UGUCAGGCGCCAGCGUGAAGGUGGCCGUGCGCGUGAGGCCCUUCA...</td>\n",
       "      <td>[..(((((((((((((((..((((..((.(.(((((((((((((.....</td>\n",
       "      <td>[-394.9, -378.5]</td>\n",
       "      <td>[0.977, 0.983]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000696501.1|ENSG00000054523.20|OTTHUMG00000001817.2|-|KIF1B-216|KIF1B|890|UTR5:1-161|CDS:162-617|UTR3:618-890|</th>\n",
       "      <td>[UGAGCGGGGCCAGCGUGAAGGUGGCCGUGCGGGUGAGGCCCUUCA...</td>\n",
       "      <td>[....(((.(((((.(.....).))))).)))((((((.((.(((....</td>\n",
       "      <td>[-193.7, -185.4]</td>\n",
       "      <td>[0.98, 0.986]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000696502.1|ENSG00000054523.20|OTTHUMG00000001817.2|-|KIF1B-217|KIF1B|7017|UTR5:1-153|CDS:154-5667|UTR3:5668-7017|</th>\n",
       "      <td>[UGUCUGGCGCCAGCGUGAAGGUGGCCGUGCGGGUGAGGCCCUUCA...</td>\n",
       "      <td>[((((((((((((((.((..(((..(((((((((((.(((.........</td>\n",
       "      <td>[-2904.0, -2789.0]</td>\n",
       "      <td>[0.974, 0.982]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt;ENST00000696503.1|ENSG00000054523.20|OTTHUMG00000001817.2|-|KIF1B-218|KIF1B|6879|UTR5:1-153|CDS:154-5529|UTR3:5530-6879|</th>\n",
       "      <td>[UGUCUGGCGCCAGCGUGAAGGUGGCCGUGCGGGUGAGGCCCUUCA...</td>\n",
       "      <td>[((((((((((((((.((..(((..(((((((((((.(((.........</td>\n",
       "      <td>[-2829.6]</td>\n",
       "      <td>[0.974]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1320 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                             mRNA_seq  \\\n",
       "ids                                                                                                     \n",
       ">ENST00000000233.10|ENSG00000004059.11|OTTHUMG0...  [UGGGCCUGACCGUGUCCGCCCUGUUCAGCAGGAUCUUCGGCAAGA...   \n",
       ">ENST00000001146.7|ENSG00000003137.9|OTTHUMG000...  [UGCUGUUCGAGGGCCUGGACCUGGUGAGCGCCCUGGCCACCCUGG...   \n",
       ">ENST00000037869.7|ENSG00000035141.8|OTTHUMG000...  [UGGCCGAGCUGCAGCAGCUGCGGGUGCAGGAGGCCGUGGAGUCCA...   \n",
       ">ENST00000054650.9|ENSG00000041988.16|OTTHUMG00...  [UGCCCAAGAGCUGUGCUGCUCGGCAGUGCUGCAACCGGUACUCCA...   \n",
       ">ENST00000054666.11|ENSG00000049245.13|OTTHUMG0...  [UGUCCACCGGGCCUACUGCCGCCACCGGCAGCAACCGGCGGCUGC...   \n",
       "...                                                                                               ...   \n",
       ">ENST00000695960.1|ENSG00000119844.17|OTTHUMG00...  [UGGAGCCCGACAUCAUCAGGAUGUACAGCUCCAGCCCUCCACCCC...   \n",
       ">ENST00000696500.1|ENSG00000054523.20|OTTHUMG00...  [UGUCAGGCGCCAGCGUGAAGGUGGCCGUGCGCGUGAGGCCCUUCA...   \n",
       ">ENST00000696501.1|ENSG00000054523.20|OTTHUMG00...  [UGAGCGGGGCCAGCGUGAAGGUGGCCGUGCGGGUGAGGCCCUUCA...   \n",
       ">ENST00000696502.1|ENSG00000054523.20|OTTHUMG00...  [UGUCUGGCGCCAGCGUGAAGGUGGCCGUGCGGGUGAGGCCCUUCA...   \n",
       ">ENST00000696503.1|ENSG00000054523.20|OTTHUMG00...  [UGUCUGGCGCCAGCGUGAAGGUGGCCGUGCGGGUGAGGCCCUUCA...   \n",
       "\n",
       "                                                                                       mRNA_structure  \\\n",
       "ids                                                                                                     \n",
       ">ENST00000000233.10|ENSG00000004059.11|OTTHUMG0...  [....(((.(.(.(((((((((((..(((.(((((((((((((......   \n",
       ">ENST00000001146.7|ENSG00000003137.9|OTTHUMG000...  [.((((((((..(((((((..(((((((.(((...((((((...))...   \n",
       ">ENST00000037869.7|ENSG00000035141.8|OTTHUMG000...  [...((((((((.((((((.(((((((((((((((((..(((((((...   \n",
       ">ENST00000054650.9|ENSG00000041988.16|OTTHUMG00...  [.........((((.((((((((((.(((((((....))))))).....   \n",
       ">ENST00000054666.11|ENSG00000049245.13|OTTHUMG0...  [.(((((((((((.((.((((.(((.((((......)))).))).)...   \n",
       "...                                                                                               ...   \n",
       ">ENST00000695960.1|ENSG00000119844.17|OTTHUMG00...  [................(((((((((((((((((.(...(((((.(...   \n",
       ">ENST00000696500.1|ENSG00000054523.20|OTTHUMG00...  [..(((((((((((((((..((((..((.(.(((((((((((((.....   \n",
       ">ENST00000696501.1|ENSG00000054523.20|OTTHUMG00...  [....(((.(((((.(.....).))))).)))((((((.((.(((....   \n",
       ">ENST00000696502.1|ENSG00000054523.20|OTTHUMG00...  [((((((((((((((.((..(((..(((((((((((.(((.........   \n",
       ">ENST00000696503.1|ENSG00000054523.20|OTTHUMG00...  [((((((((((((((.((..(((..(((((((((((.(((.........   \n",
       "\n",
       "                                                                                                 mfe  \\\n",
       "ids                                                                                                    \n",
       ">ENST00000000233.10|ENSG00000004059.11|OTTHUMG0...                                          [-255.0]   \n",
       ">ENST00000001146.7|ENSG00000003137.9|OTTHUMG000...                                  [-825.7, -866.0]   \n",
       ">ENST00000037869.7|ENSG00000035141.8|OTTHUMG000...                                  [-195.2, -207.5]   \n",
       ">ENST00000054650.9|ENSG00000041988.16|OTTHUMG00...  [-382.3, -424.3, -406.7, -403.5, -400.3, -391.6]   \n",
       ">ENST00000054666.11|ENSG00000049245.13|OTTHUMG0...                  [-132.0, -144.7, -140.4, -136.7]   \n",
       "...                                                                                              ...   \n",
       ">ENST00000695960.1|ENSG00000119844.17|OTTHUMG00...     [-1139.7, -1266.0, -1249.6, -1193.7, -1157.0]   \n",
       ">ENST00000696500.1|ENSG00000054523.20|OTTHUMG00...                                  [-394.9, -378.5]   \n",
       ">ENST00000696501.1|ENSG00000054523.20|OTTHUMG00...                                  [-193.7, -185.4]   \n",
       ">ENST00000696502.1|ENSG00000054523.20|OTTHUMG00...                                [-2904.0, -2789.0]   \n",
       ">ENST00000696503.1|ENSG00000054523.20|OTTHUMG00...                                         [-2829.6]   \n",
       "\n",
       "                                                                                           cai  \n",
       "ids                                                                                             \n",
       ">ENST00000000233.10|ENSG00000004059.11|OTTHUMG0...                                     [0.991]  \n",
       ">ENST00000001146.7|ENSG00000003137.9|OTTHUMG000...                              [0.984, 0.974]  \n",
       ">ENST00000037869.7|ENSG00000035141.8|OTTHUMG000...                              [0.994, 0.982]  \n",
       ">ENST00000054650.9|ENSG00000041988.16|OTTHUMG00...  [0.987, 0.962, 0.975, 0.977, 0.979, 0.983]  \n",
       ">ENST00000054666.11|ENSG00000049245.13|OTTHUMG0...                [0.988, 0.972, 0.978, 0.983]  \n",
       "...                                                                                        ...  \n",
       ">ENST00000695960.1|ENSG00000119844.17|OTTHUMG00...          [0.982, 0.963, 0.966, 0.975, 0.98]  \n",
       ">ENST00000696500.1|ENSG00000054523.20|OTTHUMG00...                              [0.977, 0.983]  \n",
       ">ENST00000696501.1|ENSG00000054523.20|OTTHUMG00...                               [0.98, 0.986]  \n",
       ">ENST00000696502.1|ENSG00000054523.20|OTTHUMG00...                              [0.974, 0.982]  \n",
       ">ENST00000696503.1|ENSG00000054523.20|OTTHUMG00...                                     [0.974]  \n",
       "\n",
       "[1320 rows x 4 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "agg_LinearDesign_LAMBDA_merged_CDS_gens_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "80b31ab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "agg_LinearDesign_LAMBDA_merged_CDS_gens_unique = LinearDesign_LAMBDA_merged_CDS_gens_unique.groupby('ids').agg(list)\n",
    "\n",
    "agg_LinearDesign_LAMBDA_merged_CDS_gens_unique['ids'] = agg_LinearDesign_LAMBDA_merged_CDS_gens_unique.index\n",
    "\n",
    "agg_LinearDesign_LAMBDA_merged_CDS_gens_unique['id'] = agg_LinearDesign_LAMBDA_merged_CDS_gens_unique.apply(lambda x:  [ f\"{x['ids'].split('|')[1].split('.')[0]}_NUM{v}\" for v in range(1, len(x['mfe'])+1)], axis=1)\n",
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique_new = agg_LinearDesign_LAMBDA_merged_CDS_gens_unique.explode(column=['id', 'mRNA_seq', 'mRNA_structure', 'mfe', 'cai'])\n",
    "# \n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique_new.reset_index(drop=True, inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "4209633e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['>ENST00000000233.10|ENSG00000004059.11|OTTHUMG00000023246.7|OTTHUMT00000059567.3|ARF5-201|ARF5|1032|UTR5:1-88|CDS:89-631|UTR3:632-1032|'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique_new.head(1)['ids'].values\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "766e4525",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "LinearDesign_LAMBDA_merged_CDS_gens_unique_new.to_csv('/home/yul/project/nas_project/bioSPT2/data/interiem/processed/LinearDesign_LAMBDA_merged_CDS_gens_unique_new.csv', \n",
    "                                                      index=False,\n",
    "                                                      sep=',',\n",
    "                                                      \n",
    "                                                      )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "d4ec5a02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(215, 13)"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "UTR_genes_human_ts_clean_all.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "589c5e1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "UTR_genes_human_ts_clean_all.to_csv('/home/yul/project/nas_project/bioSPT2/data/interiem/processed/UTR_genes_human_ts_clean_all.csv',\n",
    "                                    index=False,\n",
    "                                    sep=',',\n",
    "                                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "c7b6a686",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing record:  1000\n",
      "processing record:  2000\n",
      "processing record:  3000\n",
      "processing record:  4000\n",
      "processing record:  5000\n",
      "processing record:  6000\n",
      "processing record:  7000\n",
      "processing record:  8000\n",
      "processing record:  9000\n",
      "processing record:  10000\n",
      "processing record:  11000\n",
      "processing record:  12000\n",
      "processing record:  13000\n",
      "processing record:  14000\n",
      "processing record:  15000\n",
      "processing record:  16000\n",
      "processing record:  17000\n",
      "processing record:  18000\n",
      "processing record:  19000\n",
      "processing record:  20000\n",
      "processing record:  21000\n",
      "processing record:  22000\n",
      "processing record:  23000\n",
      "processing record:  24000\n",
      "processing record:  25000\n",
      "processing record:  26000\n",
      "processing record:  27000\n",
      "processing record:  28000\n",
      "processing record:  29000\n",
      "processing record:  30000\n",
      "processing record:  31000\n",
      "processing record:  32000\n",
      "processing record:  33000\n",
      "processing record:  34000\n",
      "processing record:  35000\n",
      "processing record:  36000\n",
      "processing record:  37000\n",
      "processing record:  38000\n",
      "processing record:  39000\n",
      "processing record:  40000\n",
      "processing record:  41000\n",
      "processing record:  42000\n",
      "processing record:  43000\n",
      "processing record:  44000\n",
      "processing record:  45000\n",
      "processing record:  46000\n",
      "processing record:  47000\n",
      "processing record:  48000\n",
      "processing record:  49000\n",
      "processing record:  50000\n",
      "processing record:  51000\n",
      "processing record:  52000\n",
      "processing record:  53000\n",
      "processing record:  54000\n",
      "processing record:  55000\n",
      "processing record:  56000\n",
      "processing record:  57000\n",
      "processing record:  58000\n",
      "processing record:  59000\n",
      "processing record:  60000\n",
      "processing record:  61000\n",
      "processing record:  62000\n",
      "processing record:  63000\n",
      "processing record:  64000\n",
      "processing record:  65000\n",
      "processing record:  66000\n",
      "processing record:  67000\n",
      "processing record:  68000\n",
      "processing record:  69000\n",
      "processing record:  70000\n",
      "processing record:  71000\n",
      "processing record:  72000\n",
      "processing record:  73000\n",
      "processing record:  74000\n",
      "processing record:  75000\n",
      "processing record:  76000\n",
      "processing record:  77000\n",
      "processing record:  78000\n",
      "processing record:  79000\n",
      "processing record:  80000\n",
      "processing record:  81000\n",
      "processing record:  82000\n",
      "processing record:  83000\n",
      "processing record:  84000\n",
      "processing record:  85000\n",
      "processing record:  86000\n",
      "processing record:  87000\n",
      "processing record:  88000\n",
      "processing record:  89000\n",
      "processing record:  90000\n",
      "processing record:  91000\n",
      "processing record:  92000\n",
      "processing record:  93000\n",
      "processing record:  94000\n",
      "processing record:  95000\n",
      "processing record:  96000\n",
      "processing record:  97000\n",
      "processing record:  98000\n",
      "processing record:  99000\n",
      "processing record:  100000\n",
      "processing record:  101000\n",
      "processing record:  102000\n",
      "processing record:  103000\n",
      "processing record:  104000\n",
      "processing record:  105000\n",
      "processing record:  106000\n",
      "processing record:  107000\n",
      "processing record:  108000\n",
      "processing record:  109000\n",
      "processing record:  110000\n",
      "processing record:  111000\n",
      "processing record:  112000\n",
      "processing record:  113000\n",
      "processing record:  114000\n",
      "processing record:  115000\n",
      "processing record:  116000\n",
      "processing record:  117000\n",
      "processing record:  118000\n",
      "processing record:  119000\n",
      "processing record:  120000\n",
      "processing record:  121000\n",
      "processing record:  122000\n",
      "processing record:  123000\n",
      "processing record:  124000\n",
      "processing record:  125000\n",
      "processing record:  126000\n",
      "processing record:  127000\n",
      "processing record:  128000\n",
      "processing record:  129000\n",
      "processing record:  130000\n",
      "processing record:  131000\n",
      "processing record:  132000\n",
      "processing record:  133000\n",
      "processing record:  134000\n",
      "processing record:  135000\n",
      "processing record:  136000\n",
      "processing record:  137000\n",
      "processing record:  138000\n",
      "processing record:  139000\n",
      "processing record:  140000\n",
      "processing record:  141000\n",
      "processing record:  142000\n",
      "processing record:  143000\n",
      "processing record:  144000\n",
      "processing record:  145000\n",
      "processing record:  146000\n",
      "processing record:  147000\n",
      "processing record:  148000\n",
      "processing record:  149000\n",
      "processing record:  150000\n",
      "processing record:  151000\n",
      "processing record:  152000\n",
      "processing record:  153000\n",
      "processing record:  154000\n",
      "processing record:  155000\n",
      "processing record:  156000\n",
      "processing record:  157000\n",
      "processing record:  158000\n",
      "processing record:  159000\n",
      "processing record:  160000\n",
      "processing record:  161000\n",
      "processing record:  162000\n",
      "processing record:  163000\n",
      "processing record:  164000\n",
      "processing record:  165000\n",
      "processing record:  166000\n",
      "processing record:  167000\n",
      "processing record:  168000\n",
      "processing record:  169000\n",
      "processing record:  170000\n",
      "processing record:  171000\n",
      "processing record:  172000\n",
      "processing record:  173000\n",
      "processing record:  174000\n",
      "processing record:  175000\n",
      "processing record:  176000\n",
      "processing record:  177000\n",
      "processing record:  178000\n",
      "processing record:  179000\n",
      "processing record:  180000\n",
      "processing record:  181000\n",
      "processing record:  182000\n",
      "processing record:  183000\n",
      "processing record:  184000\n",
      "processing record:  185000\n",
      "processing record:  186000\n",
      "processing record:  187000\n",
      "processing record:  188000\n",
      "processing record:  189000\n",
      "processing record:  190000\n",
      "processing record:  191000\n",
      "processing record:  192000\n",
      "processing record:  193000\n",
      "processing record:  194000\n",
      "processing record:  195000\n",
      "processing record:  196000\n",
      "processing record:  197000\n",
      "processing record:  198000\n",
      "processing record:  199000\n",
      "processing record:  200000\n",
      "processing record:  201000\n",
      "processing record:  202000\n",
      "processing record:  203000\n",
      "processing record:  204000\n",
      "processing record:  205000\n",
      "processing record:  206000\n",
      "processing record:  207000\n",
      "processing record:  208000\n",
      "processing record:  209000\n",
      "processing record:  210000\n",
      "processing record:  211000\n",
      "processing record:  212000\n",
      "processing record:  213000\n",
      "processing record:  214000\n",
      "processing record:  215000\n",
      "processing record:  216000\n",
      "processing record:  217000\n",
      "processing record:  218000\n",
      "processing record:  219000\n",
      "processing record:  220000\n",
      "processing record:  221000\n",
      "processing record:  222000\n",
      "processing record:  223000\n",
      "processing record:  224000\n",
      "processing record:  225000\n",
      "processing record:  226000\n",
      "processing record:  227000\n",
      "processing record:  228000\n",
      "processing record:  229000\n",
      "processing record:  230000\n",
      "processing record:  231000\n",
      "processing record:  232000\n",
      "processing record:  233000\n",
      "processing record:  234000\n",
      "processing record:  235000\n",
      "processing record:  236000\n",
      "processing record:  237000\n",
      "processing record:  238000\n",
      "processing record:  239000\n",
      "processing record:  240000\n",
      "processing record:  241000\n",
      "processing record:  242000\n",
      "processing record:  243000\n",
      "processing record:  244000\n",
      "processing record:  245000\n",
      "processing record:  246000\n",
      "processing record:  247000\n",
      "processing record:  248000\n",
      "processing record:  249000\n",
      "processing record:  250000\n",
      "processing record:  251000\n",
      "processing record:  252000\n",
      "processing record:  253000\n",
      "processing record:  254000\n",
      "processing record:  255000\n",
      "processing record:  256000\n",
      "processing record:  257000\n",
      "processing record:  258000\n",
      "processing record:  259000\n",
      "processing record:  260000\n",
      "processing record:  261000\n",
      "processing record:  262000\n",
      "processing record:  263000\n",
      "processing record:  264000\n",
      "processing record:  265000\n",
      "processing record:  266000\n",
      "processing record:  267000\n",
      "processing record:  268000\n",
      "processing record:  269000\n",
      "processing record:  270000\n",
      "processing record:  271000\n",
      "processing record:  272000\n",
      "processing record:  273000\n",
      "processing record:  274000\n",
      "processing record:  275000\n",
      "processing record:  276000\n",
      "processing record:  277000\n",
      "processing record:  278000\n",
      "processing record:  279000\n",
      "processing record:  280000\n",
      "processing record:  281000\n",
      "processing record:  282000\n",
      "processing record:  283000\n",
      "processing record:  284000\n",
      "processing record:  285000\n",
      "processing record:  286000\n",
      "processing record:  287000\n",
      "processing record:  288000\n",
      "processing record:  289000\n",
      "processing record:  290000\n",
      "processing record:  291000\n",
      "processing record:  292000\n",
      "processing record:  293000\n",
      "processing record:  294000\n",
      "processing record:  295000\n",
      "processing record:  296000\n",
      "processing record:  297000\n",
      "processing record:  298000\n",
      "processing record:  299000\n",
      "processing record:  300000\n",
      "processing record:  301000\n",
      "processing record:  302000\n",
      "processing record:  303000\n",
      "processing record:  304000\n",
      "processing record:  305000\n",
      "processing record:  306000\n",
      "processing record:  307000\n",
      "processing record:  308000\n",
      "processing record:  309000\n",
      "processing record:  310000\n",
      "processing record:  311000\n",
      "processing record:  312000\n",
      "processing record:  313000\n",
      "processing record:  314000\n",
      "processing record:  315000\n",
      "processing record:  316000\n",
      "processing record:  317000\n",
      "processing record:  318000\n",
      "processing record:  319000\n",
      "processing record:  320000\n",
      "processing record:  321000\n",
      "processing record:  322000\n",
      "processing record:  323000\n",
      "processing record:  324000\n",
      "processing record:  325000\n",
      "processing record:  326000\n",
      "processing record:  327000\n",
      "processing record:  328000\n",
      "processing record:  329000\n",
      "processing record:  330000\n",
      "processing record:  331000\n",
      "processing record:  332000\n",
      "processing record:  333000\n",
      "processing record:  334000\n",
      "processing record:  335000\n",
      "processing record:  336000\n",
      "processing record:  337000\n",
      "processing record:  338000\n",
      "processing record:  339000\n",
      "processing record:  340000\n",
      "processing record:  341000\n",
      "processing record:  342000\n",
      "processing record:  343000\n",
      "processing record:  344000\n",
      "processing record:  345000\n",
      "processing record:  346000\n",
      "processing record:  347000\n",
      "processing record:  348000\n",
      "processing record:  349000\n",
      "processing record:  350000\n",
      "processing record:  351000\n",
      "processing record:  352000\n",
      "processing record:  353000\n",
      "processing record:  354000\n",
      "processing record:  355000\n",
      "processing record:  356000\n",
      "processing record:  357000\n",
      "processing record:  358000\n",
      "processing record:  359000\n",
      "processing record:  360000\n",
      "processing record:  361000\n",
      "processing record:  362000\n",
      "processing record:  363000\n",
      "processing record:  364000\n",
      "processing record:  365000\n",
      "processing record:  366000\n",
      "processing record:  367000\n",
      "processing record:  368000\n",
      "processing record:  369000\n",
      "processing record:  370000\n",
      "processing record:  371000\n",
      "processing record:  372000\n",
      "processing record:  373000\n",
      "processing record:  374000\n",
      "processing record:  375000\n",
      "processing record:  376000\n",
      "processing record:  377000\n",
      "processing record:  378000\n",
      "processing record:  379000\n",
      "processing record:  380000\n",
      "processing record:  381000\n",
      "processing record:  382000\n",
      "processing record:  383000\n",
      "processing record:  384000\n",
      "processing record:  385000\n",
      "processing record:  386000\n",
      "processing record:  387000\n",
      "processing record:  388000\n",
      "processing record:  389000\n",
      "processing record:  390000\n",
      "processing record:  391000\n",
      "processing record:  392000\n",
      "processing record:  393000\n",
      "processing record:  394000\n",
      "processing record:  395000\n",
      "processing record:  396000\n",
      "processing record:  397000\n",
      "processing record:  398000\n",
      "processing record:  399000\n",
      "processing record:  400000\n",
      "processing record:  401000\n",
      "processing record:  402000\n",
      "processing record:  403000\n",
      "processing record:  404000\n",
      "processing record:  405000\n",
      "processing record:  406000\n",
      "processing record:  407000\n",
      "processing record:  408000\n",
      "processing record:  409000\n",
      "processing record:  410000\n",
      "processing record:  411000\n",
      "processing record:  412000\n",
      "processing record:  413000\n",
      "processing record:  414000\n",
      "processing record:  415000\n",
      "processing record:  416000\n",
      "processing record:  417000\n",
      "processing record:  418000\n",
      "processing record:  419000\n",
      "processing record:  420000\n",
      "processing record:  421000\n",
      "processing record:  422000\n",
      "processing record:  423000\n",
      "processing record:  424000\n",
      "processing record:  425000\n",
      "processing record:  426000\n",
      "processing record:  427000\n",
      "processing record:  428000\n",
      "processing record:  429000\n",
      "processing record:  430000\n",
      "processing record:  431000\n",
      "processing record:  432000\n",
      "processing record:  433000\n",
      "processing record:  434000\n",
      "processing record:  435000\n",
      "processing record:  436000\n",
      "processing record:  437000\n",
      "processing record:  438000\n",
      "processing record:  439000\n",
      "processing record:  440000\n",
      "processing record:  441000\n",
      "processing record:  442000\n",
      "processing record:  443000\n",
      "processing record:  444000\n",
      "processing record:  445000\n",
      "processing record:  446000\n",
      "processing record:  447000\n",
      "processing record:  448000\n",
      "processing record:  449000\n",
      "processing record:  450000\n",
      "processing record:  451000\n",
      "processing record:  452000\n",
      "processing record:  453000\n",
      "processing record:  454000\n",
      "processing record:  455000\n",
      "processing record:  456000\n",
      "processing record:  457000\n",
      "processing record:  458000\n",
      "processing record:  459000\n",
      "processing record:  460000\n",
      "processing record:  461000\n",
      "processing record:  462000\n",
      "processing record:  463000\n",
      "processing record:  464000\n",
      "processing record:  465000\n",
      "processing record:  466000\n",
      "processing record:  467000\n",
      "processing record:  468000\n",
      "processing record:  469000\n",
      "processing record:  470000\n",
      "processing record:  471000\n",
      "processing record:  472000\n",
      "processing record:  473000\n",
      "processing record:  474000\n",
      "processing record:  475000\n",
      "processing record:  476000\n",
      "processing record:  477000\n",
      "processing record:  478000\n",
      "processing record:  479000\n",
      "processing record:  480000\n",
      "processing record:  481000\n",
      "processing record:  482000\n",
      "processing record:  483000\n",
      "processing record:  484000\n",
      "processing record:  485000\n",
      "processing record:  486000\n",
      "processing record:  487000\n",
      "processing record:  488000\n",
      "processing record:  489000\n",
      "processing record:  490000\n",
      "processing record:  491000\n",
      "processing record:  492000\n",
      "processing record:  493000\n",
      "processing record:  494000\n",
      "processing record:  495000\n",
      "processing record:  496000\n",
      "processing record:  497000\n",
      "processing record:  498000\n",
      "processing record:  499000\n",
      "processing record:  500000\n",
      "processing record:  501000\n",
      "processing record:  502000\n",
      "processing record:  503000\n",
      "processing record:  504000\n",
      "processing record:  505000\n",
      "processing record:  506000\n",
      "processing record:  507000\n",
      "processing record:  508000\n",
      "processing record:  509000\n",
      "processing record:  510000\n",
      "processing record:  511000\n",
      "processing record:  512000\n",
      "processing record:  513000\n",
      "processing record:  514000\n",
      "processing record:  515000\n",
      "processing record:  516000\n",
      "processing record:  517000\n",
      "processing record:  518000\n",
      "processing record:  519000\n",
      "processing record:  520000\n",
      "processing record:  521000\n",
      "processing record:  522000\n",
      "processing record:  523000\n",
      "processing record:  524000\n",
      "processing record:  525000\n",
      "processing record:  526000\n",
      "processing record:  527000\n",
      "processing record:  528000\n",
      "processing record:  529000\n",
      "processing record:  530000\n",
      "processing record:  531000\n",
      "processing record:  532000\n",
      "processing record:  533000\n",
      "processing record:  534000\n",
      "processing record:  535000\n",
      "processing record:  536000\n",
      "processing record:  537000\n",
      "processing record:  538000\n",
      "processing record:  539000\n",
      "processing record:  540000\n",
      "processing record:  541000\n",
      "processing record:  542000\n",
      "processing record:  543000\n",
      "processing record:  544000\n",
      "processing record:  545000\n",
      "processing record:  546000\n",
      "processing record:  547000\n",
      "processing record:  548000\n",
      "processing record:  549000\n",
      "processing record:  550000\n",
      "processing record:  551000\n",
      "processing record:  552000\n",
      "processing record:  553000\n",
      "processing record:  554000\n",
      "processing record:  555000\n",
      "processing record:  556000\n",
      "processing record:  557000\n",
      "processing record:  558000\n",
      "processing record:  559000\n",
      "processing record:  560000\n",
      "processing record:  561000\n",
      "processing record:  562000\n",
      "processing record:  563000\n",
      "processing record:  564000\n",
      "processing record:  565000\n",
      "processing record:  566000\n",
      "processing record:  567000\n",
      "processing record:  568000\n",
      "processing record:  569000\n",
      "processing record:  570000\n",
      "processing record:  571000\n",
      "processing record:  572000\n",
      "processing record:  573000\n",
      "processing record:  574000\n",
      "processing record:  575000\n",
      "processing record:  576000\n",
      "processing record:  577000\n",
      "processing record:  578000\n",
      "processing record:  579000\n",
      "processing record:  580000\n",
      "processing record:  581000\n",
      "processing record:  582000\n",
      "processing record:  583000\n",
      "processing record:  584000\n",
      "processing record:  585000\n",
      "processing record:  586000\n",
      "processing record:  587000\n",
      "processing record:  588000\n",
      "processing record:  589000\n",
      "processing record:  590000\n",
      "processing record:  591000\n",
      "processing record:  592000\n",
      "processing record:  593000\n",
      "processing record:  594000\n",
      "processing record:  595000\n",
      "processing record:  596000\n",
      "processing record:  597000\n",
      "processing record:  598000\n",
      "processing record:  599000\n",
      "processing record:  600000\n",
      "processing record:  601000\n",
      "processing record:  602000\n",
      "processing record:  603000\n",
      "processing record:  604000\n",
      "processing record:  605000\n",
      "processing record:  606000\n",
      "processing record:  607000\n",
      "processing record:  608000\n",
      "processing record:  609000\n",
      "processing record:  610000\n",
      "processing record:  611000\n",
      "processing record:  612000\n",
      "processing record:  613000\n",
      "processing record:  614000\n",
      "processing record:  615000\n",
      "processing record:  616000\n",
      "processing record:  617000\n",
      "processing record:  618000\n",
      "processing record:  619000\n",
      "processing record:  620000\n",
      "processing record:  621000\n",
      "processing record:  622000\n",
      "processing record:  623000\n",
      "processing record:  624000\n",
      "processing record:  625000\n",
      "processing record:  626000\n",
      "processing record:  627000\n",
      "processing record:  628000\n",
      "processing record:  629000\n",
      "processing record:  630000\n",
      "processing record:  631000\n",
      "processing record:  632000\n",
      "processing record:  633000\n",
      "processing record:  634000\n",
      "processing record:  635000\n",
      "processing record:  636000\n",
      "processing record:  637000\n",
      "processing record:  638000\n",
      "processing record:  639000\n",
      "processing record:  640000\n",
      "processing record:  641000\n",
      "processing record:  642000\n",
      "processing record:  643000\n",
      "processing record:  644000\n",
      "processing record:  645000\n",
      "processing record:  646000\n",
      "processing record:  647000\n",
      "processing record:  648000\n",
      "processing record:  649000\n",
      "processing record:  650000\n",
      "processing record:  651000\n",
      "processing record:  652000\n",
      "processing record:  653000\n",
      "processing record:  654000\n",
      "processing record:  655000\n",
      "processing record:  656000\n",
      "processing record:  657000\n",
      "processing record:  658000\n",
      "processing record:  659000\n",
      "processing record:  660000\n",
      "processing record:  661000\n",
      "processing record:  662000\n",
      "processing record:  663000\n",
      "processing record:  664000\n",
      "processing record:  665000\n",
      "processing record:  666000\n",
      "processing record:  667000\n",
      "processing record:  668000\n",
      "processing record:  669000\n",
      "processing record:  670000\n",
      "processing record:  671000\n",
      "processing record:  672000\n",
      "processing record:  673000\n",
      "processing record:  674000\n",
      "processing record:  675000\n",
      "processing record:  676000\n",
      "processing record:  677000\n",
      "processing record:  678000\n",
      "processing record:  679000\n",
      "processing record:  680000\n",
      "processing record:  681000\n",
      "processing record:  682000\n",
      "processing record:  683000\n",
      "processing record:  684000\n",
      "processing record:  685000\n",
      "processing record:  686000\n",
      "processing record:  687000\n",
      "processing record:  688000\n",
      "processing record:  689000\n",
      "processing record:  690000\n",
      "processing record:  691000\n",
      "processing record:  692000\n",
      "processing record:  693000\n",
      "processing record:  694000\n",
      "processing record:  695000\n",
      "processing record:  696000\n",
      "processing record:  697000\n",
      "processing record:  698000\n",
      "processing record:  699000\n",
      "processing record:  700000\n",
      "processing record:  701000\n",
      "processing record:  702000\n",
      "processing record:  703000\n",
      "processing record:  704000\n",
      "processing record:  705000\n",
      "processing record:  706000\n",
      "processing record:  707000\n",
      "processing record:  708000\n",
      "processing record:  709000\n",
      "processing record:  710000\n",
      "processing record:  711000\n",
      "processing record:  712000\n",
      "processing record:  713000\n",
      "processing record:  714000\n",
      "processing record:  715000\n",
      "processing record:  716000\n",
      "processing record:  717000\n",
      "processing record:  718000\n",
      "processing record:  719000\n",
      "processing record:  720000\n",
      "processing record:  721000\n",
      "processing record:  722000\n",
      "processing record:  723000\n",
      "processing record:  724000\n",
      "processing record:  725000\n",
      "processing record:  726000\n",
      "processing record:  727000\n",
      "processing record:  728000\n",
      "processing record:  729000\n",
      "processing record:  730000\n",
      "processing record:  731000\n",
      "processing record:  732000\n",
      "processing record:  733000\n",
      "processing record:  734000\n",
      "processing record:  735000\n",
      "processing record:  736000\n",
      "processing record:  737000\n",
      "processing record:  738000\n",
      "processing record:  739000\n",
      "processing record:  740000\n",
      "processing record:  741000\n",
      "processing record:  742000\n",
      "processing record:  743000\n",
      "processing record:  744000\n",
      "processing record:  745000\n",
      "processing record:  746000\n",
      "processing record:  747000\n",
      "processing record:  748000\n",
      "processing record:  749000\n",
      "processing record:  750000\n",
      "processing record:  751000\n",
      "processing record:  752000\n",
      "processing record:  753000\n",
      "processing record:  754000\n",
      "processing record:  755000\n",
      "processing record:  756000\n",
      "processing record:  757000\n",
      "processing record:  758000\n",
      "processing record:  759000\n",
      "processing record:  760000\n",
      "processing record:  761000\n",
      "processing record:  762000\n",
      "processing record:  763000\n",
      "processing record:  764000\n",
      "processing record:  765000\n",
      "processing record:  766000\n",
      "processing record:  767000\n",
      "processing record:  768000\n",
      "processing record:  769000\n",
      "processing record:  770000\n",
      "processing record:  771000\n",
      "processing record:  772000\n",
      "processing record:  773000\n",
      "processing record:  774000\n",
      "processing record:  775000\n",
      "processing record:  776000\n",
      "processing record:  777000\n",
      "processing record:  778000\n",
      "processing record:  779000\n",
      "processing record:  780000\n",
      "processing record:  781000\n",
      "processing record:  782000\n",
      "processing record:  783000\n",
      "processing record:  784000\n",
      "processing record:  785000\n",
      "processing record:  786000\n",
      "processing record:  787000\n",
      "processing record:  788000\n",
      "processing record:  789000\n",
      "processing record:  790000\n",
      "processing record:  791000\n",
      "processing record:  792000\n",
      "processing record:  793000\n",
      "processing record:  794000\n",
      "processing record:  795000\n",
      "processing record:  796000\n",
      "processing record:  797000\n",
      "processing record:  798000\n",
      "processing record:  799000\n",
      "processing record:  800000\n",
      "processing record:  801000\n",
      "processing record:  802000\n",
      "processing record:  803000\n",
      "processing record:  804000\n",
      "processing record:  805000\n",
      "processing record:  806000\n",
      "processing record:  807000\n",
      "processing record:  808000\n",
      "processing record:  809000\n",
      "processing record:  810000\n",
      "processing record:  811000\n",
      "processing record:  812000\n",
      "processing record:  813000\n",
      "processing record:  814000\n",
      "processing record:  815000\n",
      "processing record:  816000\n",
      "processing record:  817000\n",
      "processing record:  818000\n",
      "processing record:  819000\n",
      "processing record:  820000\n",
      "processing record:  821000\n",
      "processing record:  822000\n",
      "processing record:  823000\n",
      "processing record:  824000\n",
      "processing record:  825000\n",
      "processing record:  826000\n",
      "processing record:  827000\n",
      "processing record:  828000\n",
      "processing record:  829000\n",
      "processing record:  830000\n",
      "processing record:  831000\n",
      "processing record:  832000\n",
      "processing record:  833000\n",
      "processing record:  834000\n",
      "processing record:  835000\n",
      "processing record:  836000\n",
      "processing record:  837000\n",
      "processing record:  838000\n",
      "processing record:  839000\n",
      "processing record:  840000\n",
      "processing record:  841000\n",
      "processing record:  842000\n",
      "processing record:  843000\n",
      "processing record:  844000\n",
      "processing record:  845000\n",
      "processing record:  846000\n",
      "processing record:  847000\n",
      "processing record:  848000\n",
      "processing record:  849000\n",
      "processing record:  850000\n",
      "processing record:  851000\n",
      "processing record:  852000\n",
      "processing record:  853000\n",
      "processing record:  854000\n",
      "processing record:  855000\n",
      "processing record:  856000\n",
      "processing record:  857000\n",
      "processing record:  858000\n",
      "processing record:  859000\n",
      "processing record:  860000\n",
      "processing record:  861000\n",
      "processing record:  862000\n",
      "processing record:  863000\n",
      "processing record:  864000\n",
      "processing record:  865000\n",
      "processing record:  866000\n",
      "processing record:  867000\n",
      "processing record:  868000\n",
      "processing record:  869000\n",
      "processing record:  870000\n",
      "processing record:  871000\n",
      "processing record:  872000\n",
      "processing record:  873000\n",
      "processing record:  874000\n",
      "processing record:  875000\n",
      "processing record:  876000\n",
      "processing record:  877000\n",
      "processing record:  878000\n",
      "processing record:  879000\n",
      "processing record:  880000\n",
      "processing record:  881000\n",
      "processing record:  882000\n",
      "processing record:  883000\n",
      "processing record:  884000\n",
      "processing record:  885000\n",
      "processing record:  886000\n",
      "processing record:  887000\n",
      "processing record:  888000\n",
      "processing record:  889000\n",
      "processing record:  890000\n",
      "processing record:  891000\n",
      "processing record:  892000\n",
      "processing record:  893000\n",
      "processing record:  894000\n",
      "processing record:  895000\n",
      "processing record:  896000\n",
      "processing record:  897000\n",
      "processing record:  898000\n",
      "processing record:  899000\n",
      "processing record:  900000\n",
      "processing record:  901000\n",
      "processing record:  902000\n",
      "processing record:  903000\n",
      "processing record:  904000\n",
      "processing record:  905000\n",
      "processing record:  906000\n",
      "processing record:  907000\n",
      "processing record:  908000\n",
      "processing record:  909000\n",
      "processing record:  910000\n",
      "processing record:  911000\n",
      "processing record:  912000\n",
      "processing record:  913000\n",
      "processing record:  914000\n",
      "processing record:  915000\n",
      "processing record:  916000\n",
      "processing record:  917000\n",
      "processing record:  918000\n",
      "processing record:  919000\n",
      "processing record:  920000\n",
      "processing record:  921000\n",
      "processing record:  922000\n",
      "processing record:  923000\n",
      "processing record:  924000\n",
      "processing record:  925000\n",
      "processing record:  926000\n",
      "processing record:  927000\n",
      "processing record:  928000\n",
      "processing record:  929000\n",
      "processing record:  930000\n",
      "processing record:  931000\n",
      "processing record:  932000\n",
      "processing record:  933000\n",
      "processing record:  934000\n",
      "processing record:  935000\n",
      "processing record:  936000\n",
      "processing record:  937000\n",
      "processing record:  938000\n",
      "processing record:  939000\n",
      "processing record:  940000\n",
      "processing record:  941000\n",
      "processing record:  942000\n",
      "processing record:  943000\n",
      "processing record:  944000\n",
      "processing record:  945000\n",
      "processing record:  946000\n",
      "processing record:  947000\n",
      "processing record:  948000\n",
      "processing record:  949000\n",
      "processing record:  950000\n",
      "processing record:  951000\n",
      "processing record:  952000\n",
      "processing record:  953000\n",
      "processing record:  954000\n",
      "processing record:  955000\n",
      "processing record:  956000\n",
      "processing record:  957000\n",
      "processing record:  958000\n",
      "processing record:  959000\n",
      "processing record:  960000\n",
      "processing record:  961000\n",
      "processing record:  962000\n",
      "processing record:  963000\n",
      "processing record:  964000\n",
      "processing record:  965000\n",
      "processing record:  966000\n",
      "processing record:  967000\n",
      "processing record:  968000\n",
      "processing record:  969000\n",
      "processing record:  970000\n",
      "processing record:  971000\n",
      "processing record:  972000\n",
      "processing record:  973000\n",
      "processing record:  974000\n",
      "processing record:  975000\n",
      "processing record:  976000\n",
      "processing record:  977000\n",
      "processing record:  978000\n",
      "processing record:  979000\n",
      "processing record:  980000\n",
      "processing record:  981000\n",
      "processing record:  982000\n",
      "processing record:  983000\n",
      "processing record:  984000\n",
      "processing record:  985000\n",
      "processing record:  986000\n",
      "processing record:  987000\n",
      "processing record:  988000\n",
      "processing record:  989000\n",
      "processing record:  990000\n",
      "processing record:  991000\n",
      "processing record:  992000\n",
      "processing record:  993000\n",
      "processing record:  994000\n",
      "processing record:  995000\n",
      "processing record:  996000\n",
      "processing record:  997000\n",
      "processing record:  998000\n",
      "processing record:  999000\n",
      "processing record:  1000000\n",
      "processing record:  1001000\n",
      "processing record:  1002000\n",
      "processing record:  1003000\n",
      "processing record:  1004000\n",
      "processing record:  1005000\n",
      "processing record:  1006000\n",
      "processing record:  1007000\n",
      "processing record:  1008000\n",
      "processing record:  1009000\n",
      "processing record:  1010000\n",
      "processing record:  1011000\n",
      "processing record:  1012000\n",
      "processing record:  1013000\n",
      "processing record:  1014000\n",
      "processing record:  1015000\n",
      "processing record:  1016000\n",
      "processing record:  1017000\n",
      "processing record:  1018000\n",
      "processing record:  1019000\n",
      "processing record:  1020000\n",
      "processing record:  1021000\n",
      "processing record:  1022000\n",
      "processing record:  1023000\n",
      "processing record:  1024000\n",
      "processing record:  1025000\n",
      "processing record:  1026000\n",
      "processing record:  1027000\n",
      "processing record:  1028000\n",
      "processing record:  1029000\n",
      "processing record:  1030000\n",
      "processing record:  1031000\n",
      "processing record:  1032000\n",
      "processing record:  1033000\n",
      "processing record:  1034000\n",
      "processing record:  1035000\n",
      "processing record:  1036000\n",
      "processing record:  1037000\n",
      "processing record:  1038000\n",
      "processing record:  1039000\n",
      "processing record:  1040000\n",
      "processing record:  1041000\n",
      "processing record:  1042000\n",
      "processing record:  1043000\n",
      "processing record:  1044000\n",
      "processing record:  1045000\n",
      "processing record:  1046000\n",
      "processing record:  1047000\n",
      "processing record:  1048000\n",
      "processing record:  1049000\n",
      "processing record:  1050000\n",
      "processing record:  1051000\n",
      "processing record:  1052000\n",
      "processing record:  1053000\n",
      "processing record:  1054000\n",
      "processing record:  1055000\n",
      "processing record:  1056000\n",
      "processing record:  1057000\n",
      "processing record:  1058000\n",
      "processing record:  1059000\n",
      "processing record:  1060000\n",
      "processing record:  1061000\n",
      "processing record:  1062000\n",
      "processing record:  1063000\n",
      "processing record:  1064000\n",
      "processing record:  1065000\n",
      "processing record:  1066000\n",
      "processing record:  1067000\n",
      "processing record:  1068000\n",
      "processing record:  1069000\n",
      "processing record:  1070000\n",
      "processing record:  1071000\n",
      "processing record:  1072000\n",
      "processing record:  1073000\n",
      "processing record:  1074000\n",
      "processing record:  1075000\n",
      "processing record:  1076000\n",
      "processing record:  1077000\n",
      "processing record:  1078000\n",
      "processing record:  1079000\n",
      "processing record:  1080000\n",
      "processing record:  1081000\n",
      "processing record:  1082000\n",
      "processing record:  1083000\n",
      "processing record:  1084000\n",
      "processing record:  1085000\n",
      "processing record:  1086000\n",
      "processing record:  1087000\n",
      "processing record:  1088000\n",
      "processing record:  1089000\n",
      "processing record:  1090000\n",
      "processing record:  1091000\n",
      "processing record:  1092000\n",
      "processing record:  1093000\n",
      "processing record:  1094000\n",
      "processing record:  1095000\n",
      "processing record:  1096000\n",
      "processing record:  1097000\n",
      "processing record:  1098000\n",
      "processing record:  1099000\n",
      "processing record:  1100000\n",
      "processing record:  1101000\n",
      "processing record:  1102000\n",
      "processing record:  1103000\n",
      "processing record:  1104000\n",
      "processing record:  1105000\n",
      "processing record:  1106000\n",
      "processing record:  1107000\n",
      "processing record:  1108000\n",
      "processing record:  1109000\n",
      "processing record:  1110000\n",
      "processing record:  1111000\n",
      "processing record:  1112000\n",
      "processing record:  1113000\n",
      "processing record:  1114000\n",
      "processing record:  1115000\n",
      "processing record:  1116000\n",
      "processing record:  1117000\n",
      "processing record:  1118000\n",
      "processing record:  1119000\n",
      "processing record:  1120000\n",
      "processing record:  1121000\n",
      "processing record:  1122000\n",
      "processing record:  1123000\n",
      "processing record:  1124000\n",
      "processing record:  1125000\n",
      "processing record:  1126000\n",
      "processing record:  1127000\n",
      "processing record:  1128000\n",
      "processing record:  1129000\n",
      "processing record:  1130000\n",
      "processing record:  1131000\n",
      "processing record:  1132000\n",
      "processing record:  1133000\n",
      "processing record:  1134000\n",
      "processing record:  1135000\n",
      "processing record:  1136000\n",
      "processing record:  1137000\n",
      "processing record:  1138000\n",
      "processing record:  1139000\n",
      "processing record:  1140000\n",
      "processing record:  1141000\n",
      "processing record:  1142000\n",
      "processing record:  1143000\n",
      "processing record:  1144000\n",
      "processing record:  1145000\n",
      "processing record:  1146000\n",
      "processing record:  1147000\n",
      "processing record:  1148000\n",
      "processing record:  1149000\n",
      "processing record:  1150000\n",
      "processing record:  1151000\n",
      "processing record:  1152000\n",
      "processing record:  1153000\n",
      "processing record:  1154000\n",
      "processing record:  1155000\n",
      "processing record:  1156000\n",
      "processing record:  1157000\n",
      "processing record:  1158000\n",
      "processing record:  1159000\n",
      "processing record:  1160000\n",
      "processing record:  1161000\n",
      "processing record:  1162000\n",
      "processing record:  1163000\n",
      "processing record:  1164000\n",
      "processing record:  1165000\n",
      "processing record:  1166000\n",
      "processing record:  1167000\n",
      "processing record:  1168000\n",
      "processing record:  1169000\n",
      "processing record:  1170000\n",
      "processing record:  1171000\n",
      "processing record:  1172000\n",
      "processing record:  1173000\n",
      "processing record:  1174000\n",
      "processing record:  1175000\n",
      "processing record:  1176000\n",
      "processing record:  1177000\n",
      "processing record:  1178000\n",
      "processing record:  1179000\n",
      "processing record:  1180000\n",
      "processing record:  1181000\n",
      "processing record:  1182000\n",
      "processing record:  1183000\n",
      "processing record:  1184000\n",
      "processing record:  1185000\n",
      "processing record:  1186000\n",
      "processing record:  1187000\n",
      "processing record:  1188000\n",
      "processing record:  1189000\n",
      "processing record:  1190000\n",
      "processing record:  1191000\n",
      "processing record:  1192000\n",
      "processing record:  1193000\n",
      "processing record:  1194000\n",
      "processing record:  1195000\n",
      "processing record:  1196000\n",
      "processing record:  1197000\n",
      "processing record:  1198000\n",
      "processing record:  1199000\n",
      "processing record:  1200000\n",
      "processing record:  1201000\n",
      "processing record:  1202000\n",
      "processing record:  1203000\n",
      "processing record:  1204000\n",
      "processing record:  1205000\n",
      "processing record:  1206000\n",
      "processing record:  1207000\n",
      "processing record:  1208000\n",
      "processing record:  1209000\n",
      "processing record:  1210000\n",
      "processing record:  1211000\n",
      "processing record:  1212000\n",
      "processing record:  1213000\n",
      "processing record:  1214000\n",
      "processing record:  1215000\n",
      "processing record:  1216000\n",
      "processing record:  1217000\n",
      "processing record:  1218000\n",
      "processing record:  1219000\n",
      "processing record:  1220000\n",
      "processing record:  1221000\n",
      "processing record:  1222000\n",
      "processing record:  1223000\n",
      "processing record:  1224000\n",
      "processing record:  1225000\n",
      "processing record:  1226000\n",
      "processing record:  1227000\n",
      "processing record:  1228000\n",
      "processing record:  1229000\n",
      "processing record:  1230000\n",
      "processing record:  1231000\n",
      "processing record:  1232000\n",
      "processing record:  1233000\n",
      "processing record:  1234000\n",
      "processing record:  1235000\n",
      "processing record:  1236000\n",
      "processing record:  1237000\n",
      "processing record:  1238000\n",
      "processing record:  1239000\n",
      "processing record:  1240000\n",
      "processing record:  1241000\n",
      "processing record:  1242000\n",
      "processing record:  1243000\n",
      "processing record:  1244000\n",
      "processing record:  1245000\n",
      "processing record:  1246000\n",
      "processing record:  1247000\n",
      "processing record:  1248000\n",
      "processing record:  1249000\n",
      "processing record:  1250000\n",
      "processing record:  1251000\n",
      "processing record:  1252000\n",
      "processing record:  1253000\n",
      "processing record:  1254000\n",
      "processing record:  1255000\n",
      "processing record:  1256000\n",
      "processing record:  1257000\n",
      "processing record:  1258000\n",
      "processing record:  1259000\n",
      "processing record:  1260000\n",
      "processing record:  1261000\n",
      "processing record:  1262000\n",
      "processing record:  1263000\n",
      "processing record:  1264000\n",
      "processing record:  1265000\n",
      "processing record:  1266000\n",
      "processing record:  1267000\n",
      "processing record:  1268000\n",
      "processing record:  1269000\n",
      "processing record:  1270000\n",
      "processing record:  1271000\n",
      "processing record:  1272000\n",
      "processing record:  1273000\n",
      "processing record:  1274000\n",
      "processing record:  1275000\n",
      "processing record:  1276000\n",
      "processing record:  1277000\n",
      "processing record:  1278000\n",
      "processing record:  1279000\n",
      "processing record:  1280000\n",
      "processing record:  1281000\n",
      "processing record:  1282000\n",
      "processing record:  1283000\n",
      "processing record:  1284000\n",
      "processing record:  1285000\n",
      "processing record:  1286000\n",
      "processing record:  1287000\n",
      "processing record:  1288000\n",
      "processing record:  1289000\n",
      "processing record:  1290000\n",
      "processing record:  1291000\n",
      "processing record:  1292000\n",
      "processing record:  1293000\n",
      "processing record:  1294000\n",
      "processing record:  1295000\n",
      "processing record:  1296000\n",
      "processing record:  1297000\n",
      "processing record:  1298000\n",
      "processing record:  1299000\n",
      "processing record:  1300000\n",
      "processing record:  1301000\n",
      "processing record:  1302000\n",
      "processing record:  1303000\n",
      "processing record:  1304000\n",
      "processing record:  1305000\n",
      "processing record:  1306000\n",
      "processing record:  1307000\n",
      "processing record:  1308000\n",
      "processing record:  1309000\n",
      "processing record:  1310000\n",
      "processing record:  1311000\n",
      "processing record:  1312000\n",
      "processing record:  1313000\n",
      "processing record:  1314000\n",
      "processing record:  1315000\n",
      "processing record:  1316000\n",
      "processing record:  1317000\n",
      "processing record:  1318000\n",
      "processing record:  1319000\n",
      "processing record:  1320000\n",
      "processing record:  1321000\n",
      "processing record:  1322000\n",
      "processing record:  1323000\n",
      "processing record:  1324000\n",
      "processing record:  1325000\n",
      "processing record:  1326000\n",
      "processing record:  1327000\n",
      "processing record:  1328000\n",
      "processing record:  1329000\n",
      "processing record:  1330000\n",
      "processing record:  1331000\n",
      "processing record:  1332000\n",
      "processing record:  1333000\n",
      "processing record:  1334000\n",
      "processing record:  1335000\n",
      "processing record:  1336000\n",
      "processing record:  1337000\n",
      "processing record:  1338000\n",
      "processing record:  1339000\n",
      "processing record:  1340000\n",
      "processing record:  1341000\n",
      "processing record:  1342000\n",
      "processing record:  1343000\n",
      "processing record:  1344000\n",
      "processing record:  1345000\n",
      "processing record:  1346000\n",
      "processing record:  1347000\n",
      "processing record:  1348000\n",
      "processing record:  1349000\n",
      "processing record:  1350000\n",
      "processing record:  1351000\n",
      "processing record:  1352000\n",
      "processing record:  1353000\n",
      "processing record:  1354000\n",
      "processing record:  1355000\n",
      "processing record:  1356000\n",
      "processing record:  1357000\n",
      "processing record:  1358000\n",
      "processing record:  1359000\n",
      "processing record:  1360000\n",
      "processing record:  1361000\n",
      "processing record:  1362000\n",
      "processing record:  1363000\n",
      "processing record:  1364000\n",
      "processing record:  1365000\n",
      "processing record:  1366000\n",
      "processing record:  1367000\n",
      "processing record:  1368000\n",
      "processing record:  1369000\n",
      "processing record:  1370000\n",
      "processing record:  1371000\n",
      "processing record:  1372000\n",
      "processing record:  1373000\n",
      "processing record:  1374000\n",
      "processing record:  1375000\n",
      "processing record:  1376000\n",
      "processing record:  1377000\n",
      "processing record:  1378000\n",
      "processing record:  1379000\n",
      "processing record:  1380000\n",
      "processing record:  1381000\n",
      "processing record:  1382000\n",
      "processing record:  1383000\n",
      "processing record:  1384000\n",
      "processing record:  1385000\n",
      "processing record:  1386000\n",
      "processing record:  1387000\n",
      "processing record:  1388000\n",
      "processing record:  1389000\n",
      "processing record:  1390000\n",
      "processing record:  1391000\n",
      "processing record:  1392000\n",
      "processing record:  1393000\n",
      "processing record:  1394000\n",
      "processing record:  1395000\n",
      "processing record:  1396000\n",
      "processing record:  1397000\n",
      "processing record:  1398000\n",
      "processing record:  1399000\n",
      "processing record:  1400000\n",
      "processing record:  1401000\n",
      "processing record:  1402000\n",
      "processing record:  1403000\n",
      "processing record:  1404000\n",
      "processing record:  1405000\n",
      "processing record:  1406000\n",
      "processing record:  1407000\n",
      "processing record:  1408000\n",
      "processing record:  1409000\n",
      "processing record:  1410000\n",
      "processing record:  1411000\n",
      "processing record:  1412000\n",
      "processing record:  1413000\n",
      "processing record:  1414000\n",
      "processing record:  1415000\n",
      "processing record:  1416000\n",
      "processing record:  1417000\n",
      "processing record:  1418000\n",
      "processing record:  1419000\n",
      "processing record:  1420000\n",
      "processing record:  1421000\n",
      "processing record:  1422000\n",
      "processing record:  1423000\n",
      "processing record:  1424000\n",
      "processing record:  1425000\n",
      "processing record:  1426000\n",
      "processing record:  1427000\n",
      "processing record:  1428000\n",
      "processing record:  1429000\n",
      "processing record:  1430000\n",
      "processing record:  1431000\n",
      "processing record:  1432000\n",
      "processing record:  1433000\n",
      "processing record:  1434000\n",
      "processing record:  1435000\n",
      "processing record:  1436000\n",
      "processing record:  1437000\n",
      "processing record:  1438000\n",
      "processing record:  1439000\n",
      "processing record:  1440000\n",
      "processing record:  1441000\n",
      "processing record:  1442000\n",
      "processing record:  1443000\n",
      "processing record:  1444000\n",
      "processing record:  1445000\n",
      "processing record:  1446000\n",
      "processing record:  1447000\n",
      "processing record:  1448000\n",
      "processing record:  1449000\n",
      "processing record:  1450000\n",
      "processing record:  1451000\n",
      "processing record:  1452000\n",
      "processing record:  1453000\n",
      "processing record:  1454000\n",
      "processing record:  1455000\n",
      "processing record:  1456000\n",
      "processing record:  1457000\n",
      "processing record:  1458000\n",
      "processing record:  1459000\n",
      "processing record:  1460000\n",
      "processing record:  1461000\n",
      "processing record:  1462000\n",
      "processing record:  1463000\n",
      "processing record:  1464000\n",
      "processing record:  1465000\n",
      "processing record:  1466000\n",
      "processing record:  1467000\n",
      "processing record:  1468000\n",
      "processing record:  1469000\n",
      "processing record:  1470000\n",
      "processing record:  1471000\n",
      "processing record:  1472000\n",
      "processing record:  1473000\n",
      "processing record:  1474000\n",
      "processing record:  1475000\n",
      "processing record:  1476000\n",
      "processing record:  1477000\n",
      "processing record:  1478000\n",
      "processing record:  1479000\n",
      "processing record:  1480000\n",
      "processing record:  1481000\n",
      "processing record:  1482000\n",
      "processing record:  1483000\n",
      "processing record:  1484000\n",
      "processing record:  1485000\n",
      "processing record:  1486000\n",
      "processing record:  1487000\n",
      "processing record:  1488000\n",
      "processing record:  1489000\n",
      "processing record:  1490000\n",
      "processing record:  1491000\n",
      "processing record:  1492000\n",
      "processing record:  1493000\n",
      "processing record:  1494000\n",
      "processing record:  1495000\n",
      "processing record:  1496000\n",
      "processing record:  1497000\n",
      "processing record:  1498000\n",
      "processing record:  1499000\n",
      "processing record:  1500000\n",
      "processing record:  1501000\n",
      "processing record:  1502000\n",
      "processing record:  1503000\n",
      "processing record:  1504000\n",
      "processing record:  1505000\n",
      "processing record:  1506000\n",
      "processing record:  1507000\n",
      "processing record:  1508000\n",
      "processing record:  1509000\n",
      "processing record:  1510000\n",
      "processing record:  1511000\n",
      "processing record:  1512000\n",
      "processing record:  1513000\n",
      "processing record:  1514000\n",
      "processing record:  1515000\n",
      "processing record:  1516000\n",
      "processing record:  1517000\n",
      "processing record:  1518000\n",
      "processing record:  1519000\n",
      "processing record:  1520000\n",
      "processing record:  1521000\n",
      "processing record:  1522000\n",
      "processing record:  1523000\n",
      "processing record:  1524000\n",
      "processing record:  1525000\n",
      "processing record:  1526000\n",
      "processing record:  1527000\n",
      "processing record:  1528000\n",
      "processing record:  1529000\n",
      "processing record:  1530000\n",
      "processing record:  1531000\n",
      "processing record:  1532000\n",
      "processing record:  1533000\n",
      "processing record:  1534000\n",
      "processing record:  1535000\n",
      "processing record:  1536000\n",
      "processing record:  1537000\n",
      "processing record:  1538000\n",
      "processing record:  1539000\n",
      "processing record:  1540000\n",
      "processing record:  1541000\n",
      "processing record:  1542000\n",
      "processing record:  1543000\n",
      "processing record:  1544000\n",
      "processing record:  1545000\n",
      "processing record:  1546000\n",
      "processing record:  1547000\n",
      "processing record:  1548000\n",
      "processing record:  1549000\n",
      "processing record:  1550000\n",
      "processing record:  1551000\n",
      "processing record:  1552000\n",
      "processing record:  1553000\n",
      "processing record:  1554000\n",
      "processing record:  1555000\n",
      "processing record:  1556000\n",
      "processing record:  1557000\n",
      "processing record:  1558000\n",
      "processing record:  1559000\n",
      "processing record:  1560000\n",
      "processing record:  1561000\n",
      "processing record:  1562000\n",
      "processing record:  1563000\n",
      "processing record:  1564000\n",
      "processing record:  1565000\n",
      "processing record:  1566000\n",
      "processing record:  1567000\n",
      "processing record:  1568000\n",
      "processing record:  1569000\n",
      "processing record:  1570000\n",
      "processing record:  1571000\n",
      "processing record:  1572000\n",
      "processing record:  1573000\n",
      "processing record:  1574000\n",
      "processing record:  1575000\n",
      "processing record:  1576000\n",
      "processing record:  1577000\n",
      "processing record:  1578000\n",
      "processing record:  1579000\n",
      "processing record:  1580000\n",
      "processing record:  1581000\n",
      "processing record:  1582000\n",
      "processing record:  1583000\n",
      "processing record:  1584000\n",
      "processing record:  1585000\n",
      "processing record:  1586000\n",
      "processing record:  1587000\n",
      "processing record:  1588000\n",
      "processing record:  1589000\n",
      "processing record:  1590000\n",
      "processing record:  1591000\n",
      "processing record:  1592000\n",
      "processing record:  1593000\n",
      "processing record:  1594000\n",
      "processing record:  1595000\n",
      "processing record:  1596000\n",
      "processing record:  1597000\n",
      "processing record:  1598000\n",
      "processing record:  1599000\n",
      "processing record:  1600000\n",
      "processing record:  1601000\n",
      "processing record:  1602000\n",
      "processing record:  1603000\n",
      "processing record:  1604000\n",
      "processing record:  1605000\n",
      "processing record:  1606000\n",
      "processing record:  1607000\n",
      "processing record:  1608000\n",
      "processing record:  1609000\n",
      "processing record:  1610000\n",
      "processing record:  1611000\n",
      "processing record:  1612000\n",
      "processing record:  1613000\n",
      "processing record:  1614000\n",
      "processing record:  1615000\n",
      "processing record:  1616000\n",
      "processing record:  1617000\n",
      "processing record:  1618000\n",
      "processing record:  1619000\n",
      "processing record:  1620000\n",
      "processing record:  1621000\n",
      "processing record:  1622000\n",
      "processing record:  1623000\n",
      "processing record:  1624000\n",
      "processing record:  1625000\n",
      "processing record:  1626000\n",
      "processing record:  1627000\n",
      "processing record:  1628000\n",
      "processing record:  1629000\n",
      "processing record:  1630000\n",
      "processing record:  1631000\n",
      "processing record:  1632000\n",
      "processing record:  1633000\n",
      "processing record:  1634000\n",
      "processing record:  1635000\n",
      "processing record:  1636000\n",
      "processing record:  1637000\n",
      "processing record:  1638000\n",
      "processing record:  1639000\n",
      "processing record:  1640000\n",
      "processing record:  1641000\n",
      "processing record:  1642000\n",
      "processing record:  1643000\n",
      "processing record:  1644000\n",
      "processing record:  1645000\n",
      "processing record:  1646000\n",
      "processing record:  1647000\n",
      "processing record:  1648000\n",
      "processing record:  1649000\n",
      "processing record:  1650000\n",
      "processing record:  1651000\n",
      "processing record:  1652000\n",
      "processing record:  1653000\n",
      "processing record:  1654000\n",
      "processing record:  1655000\n",
      "processing record:  1656000\n",
      "processing record:  1657000\n",
      "processing record:  1658000\n",
      "processing record:  1659000\n",
      "processing record:  1660000\n",
      "processing record:  1661000\n",
      "processing record:  1662000\n",
      "processing record:  1663000\n",
      "processing record:  1664000\n",
      "processing record:  1665000\n",
      "processing record:  1666000\n",
      "processing record:  1667000\n",
      "processing record:  1668000\n",
      "processing record:  1669000\n",
      "processing record:  1670000\n",
      "processing record:  1671000\n",
      "processing record:  1672000\n",
      "processing record:  1673000\n",
      "processing record:  1674000\n",
      "processing record:  1675000\n",
      "processing record:  1676000\n",
      "processing record:  1677000\n",
      "processing record:  1678000\n",
      "processing record:  1679000\n",
      "processing record:  1680000\n",
      "processing record:  1681000\n",
      "processing record:  1682000\n",
      "processing record:  1683000\n",
      "processing record:  1684000\n",
      "processing record:  1685000\n",
      "processing record:  1686000\n",
      "processing record:  1687000\n",
      "processing record:  1688000\n",
      "processing record:  1689000\n",
      "processing record:  1690000\n",
      "processing record:  1691000\n",
      "processing record:  1692000\n",
      "processing record:  1693000\n",
      "processing record:  1694000\n",
      "processing record:  1695000\n",
      "processing record:  1696000\n",
      "processing record:  1697000\n",
      "processing record:  1698000\n",
      "processing record:  1699000\n",
      "processing record:  1700000\n",
      "processing record:  1701000\n",
      "processing record:  1702000\n",
      "processing record:  1703000\n",
      "processing record:  1704000\n",
      "processing record:  1705000\n",
      "processing record:  1706000\n",
      "processing record:  1707000\n",
      "processing record:  1708000\n",
      "processing record:  1709000\n",
      "processing record:  1710000\n",
      "processing record:  1711000\n",
      "processing record:  1712000\n",
      "processing record:  1713000\n",
      "processing record:  1714000\n",
      "processing record:  1715000\n",
      "processing record:  1716000\n",
      "processing record:  1717000\n",
      "processing record:  1718000\n",
      "processing record:  1719000\n",
      "processing record:  1720000\n",
      "processing record:  1721000\n",
      "processing record:  1722000\n",
      "processing record:  1723000\n",
      "processing record:  1724000\n",
      "processing record:  1725000\n",
      "processing record:  1726000\n",
      "processing record:  1727000\n",
      "processing record:  1728000\n",
      "processing record:  1729000\n",
      "processing record:  1730000\n",
      "processing record:  1731000\n",
      "processing record:  1732000\n",
      "processing record:  1733000\n",
      "processing record:  1734000\n",
      "processing record:  1735000\n",
      "processing record:  1736000\n",
      "processing record:  1737000\n",
      "processing record:  1738000\n",
      "processing record:  1739000\n",
      "processing record:  1740000\n",
      "processing record:  1741000\n",
      "processing record:  1742000\n",
      "processing record:  1743000\n",
      "processing record:  1744000\n",
      "processing record:  1745000\n",
      "processing record:  1746000\n",
      "processing record:  1747000\n",
      "processing record:  1748000\n",
      "processing record:  1749000\n",
      "processing record:  1750000\n",
      "processing record:  1751000\n",
      "processing record:  1752000\n",
      "processing record:  1753000\n",
      "processing record:  1754000\n",
      "processing record:  1755000\n",
      "processing record:  1756000\n",
      "processing record:  1757000\n",
      "processing record:  1758000\n",
      "processing record:  1759000\n",
      "processing record:  1760000\n",
      "processing record:  1761000\n",
      "processing record:  1762000\n",
      "processing record:  1763000\n",
      "processing record:  1764000\n",
      "processing record:  1765000\n",
      "processing record:  1766000\n",
      "processing record:  1767000\n",
      "processing record:  1768000\n",
      "processing record:  1769000\n",
      "processing record:  1770000\n",
      "processing record:  1771000\n",
      "processing record:  1772000\n",
      "processing record:  1773000\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[168], line 32\u001b[0m\n\u001b[1;32m     21\u001b[0m mRNAs_ \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[1;32m     22\u001b[0m     {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mid\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCDS_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcds_row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mid\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m-UTR5_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mutr5_row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTranscript\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m-UTR3_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mutr3_row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTranscript\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     23\u001b[0m      \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCDS\u001b[39m\u001b[38;5;124m'\u001b[39m: cds_row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmRNA_seq\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mupper()\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mU\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mT\u001b[39m\u001b[38;5;124m'\u001b[39m),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     29\u001b[0m      \n\u001b[1;32m     30\u001b[0m      }, index\u001b[38;5;241m=\u001b[39m[v])\n\u001b[1;32m     31\u001b[0m v\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m\n\u001b[0;32m---> 32\u001b[0m mRNAs \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([mRNAs, mRNAs_])\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m v \u001b[38;5;241m%\u001b[39m \u001b[38;5;241m1000\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m     35\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprocessing record: \u001b[39m\u001b[38;5;124m'\u001b[39m, v)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "\n",
    "mRNAs = pd.DataFrame()\n",
    "\n",
    "v = 0\n",
    "for i, cds_row in LinearDesign_LAMBDA_merged_CDS_gens_unique_new.iterrows(): \n",
    "    # print(cds_row['id'], cds_row['mRNA_seq'].upper().replace('U', 'T'))\n",
    "    # print(i, cds_row)\n",
    "    # if i >3:\n",
    "    #     break\n",
    "    \n",
    "    for j, utr5_row in UTR_genes_human_ts_clean_all.reset_index().iterrows():\n",
    "        \n",
    "        # print(utr5_row['Transcript'], utr5_row['UTR5'])\n",
    "        # if j >3:\n",
    "        #     break\n",
    "        \n",
    "        for k, utr3_row in UTR_genes_human_ts_clean_all[~UTR_genes_human_ts_clean_all['UTR3'].isnull()].reset_index().iterrows():\n",
    "    \n",
    "            # print(utr3_row['Transcript'], utr5_row['UTR3'])\n",
    "            # if k >3:\n",
    "            #     break\n",
    "            mRNAs_ = pd.DataFrame(\n",
    "                {'id': f\"CDS_{cds_row['id']}-UTR5_{utr5_row['Transcript']}-UTR3_{utr3_row['Transcript']}\",\n",
    "                 'CDS': cds_row['mRNA_seq'].upper().replace('U', 'T'),\n",
    "                 'CDS_structure': cds_row['mRNA_structure'],\n",
    "                 'CDS_mfe': cds_row['mfe'],\n",
    "                 'CDS_cai': cds_row['cai'],\n",
    "                 'UTR5': utr5_row['UTR5'],\n",
    "                 'UTR3': utr3_row['UTR3'],\n",
    "                 \n",
    "                 }, index=[v])\n",
    "            v+=1\n",
    "            mRNAs = pd.concat([mRNAs, mRNAs_])\n",
    "            \n",
    "            if v % 1000 == 0:\n",
    "                print('processing record: ', v)\n",
    "            if v % 100000 == 0:\n",
    "                mRNAs.to_csv('mRNAs_for_finetune.csv', index=False)\n",
    "                \n",
    "mRNAs.to_csv('mRNAs_for_finetune.csv', index=False)  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3fc789",
   "metadata": {},
   "outputs": [],
   "source": [
    "# /data1/genomics/mRNAs_for_finetune.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fe78bdc",
   "metadata": {},
   "source": [
    "## RF 配对数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "68b9a4e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>iBAQ A549_exp</th>\n",
       "      <th>iBAQ HEK293_exp</th>\n",
       "      <th>iBAQ HeLa_exp</th>\n",
       "      <th>iBAQ K562_exp</th>\n",
       "      <th>iBAQ_TransEffect_A549_exp</th>\n",
       "      <th>iBAQ_TransEffect_HEK293_exp</th>\n",
       "      <th>iBAQ_TransEffect_HeLa_exp</th>\n",
       "      <th>iBAQ_TransEffect_K562_exp</th>\n",
       "      <th>iBAQ_TransEffect_A549_rank</th>\n",
       "      <th>iBAQ_TransEffect_HEK293_rank</th>\n",
       "      <th>...</th>\n",
       "      <th>K562_GR_2017_rank</th>\n",
       "      <th>K562_elife_rank</th>\n",
       "      <th>K562_timeLapse_rank</th>\n",
       "      <th>ESC_GR_2017_label</th>\n",
       "      <th>Hela_GSE102113_label</th>\n",
       "      <th>K562_GR_2017_label</th>\n",
       "      <th>K562_elife_label</th>\n",
       "      <th>K562_timeLapse_label</th>\n",
       "      <th>consistent_label_HalfLife</th>\n",
       "      <th>consistent_label_all</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>6.527933</td>\n",
       "      <td>6.796727</td>\n",
       "      <td>5.517671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.134571</td>\n",
       "      <td>5.195596</td>\n",
       "      <td>4.181992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1074.0</td>\n",
       "      <td>2159.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>6.989328</td>\n",
       "      <td>6.970460</td>\n",
       "      <td>6.817138</td>\n",
       "      <td>6.404391</td>\n",
       "      <td>5.275929</td>\n",
       "      <td>5.256304</td>\n",
       "      <td>4.987699</td>\n",
       "      <td>4.315847</td>\n",
       "      <td>756.0</td>\n",
       "      <td>1958.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1926.0</td>\n",
       "      <td>5030.0</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>4.331123</td>\n",
       "      <td>5.157944</td>\n",
       "      <td>3.871246</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.588284</td>\n",
       "      <td>4.523821</td>\n",
       "      <td>3.200597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5498.0</td>\n",
       "      <td>4664.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9413.0</td>\n",
       "      <td>8183.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>5.157366</td>\n",
       "      <td>5.762855</td>\n",
       "      <td>4.583522</td>\n",
       "      <td>5.264936</td>\n",
       "      <td>3.871606</td>\n",
       "      <td>4.495252</td>\n",
       "      <td>3.234262</td>\n",
       "      <td>3.624279</td>\n",
       "      <td>4839.0</td>\n",
       "      <td>4755.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5547.0</td>\n",
       "      <td>4374.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.211548</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285304</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3039.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285330</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285396</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3149.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285441</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>740.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000288825</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4981.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19868 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 iBAQ A549_exp  iBAQ HEK293_exp  iBAQ HeLa_exp  iBAQ K562_exp  \\\n",
       "ensembl_gene_id                                                                 \n",
       "ENSG00000000003       6.527933         6.796727       5.517671            NaN   \n",
       "ENSG00000000419       6.989328         6.970460       6.817138       6.404391   \n",
       "ENSG00000000457       4.331123         5.157944       3.871246            NaN   \n",
       "ENSG00000000460       5.157366         5.762855       4.583522       5.264936   \n",
       "ENSG00000000938            NaN         4.211548            NaN            NaN   \n",
       "...                        ...              ...            ...            ...   \n",
       "ENSG00000285304            NaN              NaN            NaN            NaN   \n",
       "ENSG00000285330            NaN              NaN            NaN            NaN   \n",
       "ENSG00000285396            NaN              NaN            NaN            NaN   \n",
       "ENSG00000285441            NaN              NaN            NaN            NaN   \n",
       "ENSG00000288825            NaN              NaN            NaN            NaN   \n",
       "\n",
       "                 iBAQ_TransEffect_A549_exp  iBAQ_TransEffect_HEK293_exp  \\\n",
       "ensembl_gene_id                                                           \n",
       "ENSG00000000003                   5.134571                     5.195596   \n",
       "ENSG00000000419                   5.275929                     5.256304   \n",
       "ENSG00000000457                   3.588284                     4.523821   \n",
       "ENSG00000000460                   3.871606                     4.495252   \n",
       "ENSG00000000938                        NaN                          NaN   \n",
       "...                                    ...                          ...   \n",
       "ENSG00000285304                        NaN                          NaN   \n",
       "ENSG00000285330                        NaN                          NaN   \n",
       "ENSG00000285396                        NaN                          NaN   \n",
       "ENSG00000285441                        NaN                          NaN   \n",
       "ENSG00000288825                        NaN                          NaN   \n",
       "\n",
       "                 iBAQ_TransEffect_HeLa_exp  iBAQ_TransEffect_K562_exp  \\\n",
       "ensembl_gene_id                                                         \n",
       "ENSG00000000003                   4.181992                        NaN   \n",
       "ENSG00000000419                   4.987699                   4.315847   \n",
       "ENSG00000000457                   3.200597                        NaN   \n",
       "ENSG00000000460                   3.234262                   3.624279   \n",
       "ENSG00000000938                        NaN                        NaN   \n",
       "...                                    ...                        ...   \n",
       "ENSG00000285304                        NaN                        NaN   \n",
       "ENSG00000285330                        NaN                        NaN   \n",
       "ENSG00000285396                        NaN                        NaN   \n",
       "ENSG00000285441                        NaN                        NaN   \n",
       "ENSG00000288825                        NaN                        NaN   \n",
       "\n",
       "                 iBAQ_TransEffect_A549_rank  iBAQ_TransEffect_HEK293_rank  \\\n",
       "ensembl_gene_id                                                             \n",
       "ENSG00000000003                      1074.0                        2159.0   \n",
       "ENSG00000000419                       756.0                        1958.0   \n",
       "ENSG00000000457                      5498.0                        4664.0   \n",
       "ENSG00000000460                      4839.0                        4755.0   \n",
       "ENSG00000000938                         NaN                           NaN   \n",
       "...                                     ...                           ...   \n",
       "ENSG00000285304                         NaN                           NaN   \n",
       "ENSG00000285330                         NaN                           NaN   \n",
       "ENSG00000285396                         NaN                           NaN   \n",
       "ENSG00000285441                         NaN                           NaN   \n",
       "ENSG00000288825                         NaN                           NaN   \n",
       "\n",
       "                 ...  K562_GR_2017_rank  K562_elife_rank  K562_timeLapse_rank  \\\n",
       "ensembl_gene_id  ...                                                            \n",
       "ENSG00000000003  ...                NaN              NaN                  NaN   \n",
       "ENSG00000000419  ...             1926.0           5030.0               1362.0   \n",
       "ENSG00000000457  ...             9413.0           8183.0                  NaN   \n",
       "ENSG00000000460  ...             5547.0           4374.0                  NaN   \n",
       "ENSG00000000938  ...                NaN              NaN                  NaN   \n",
       "...              ...                ...              ...                  ...   \n",
       "ENSG00000285304  ...                NaN           3039.0                  NaN   \n",
       "ENSG00000285330  ...                NaN            160.0                  NaN   \n",
       "ENSG00000285396  ...                NaN           3149.0                  NaN   \n",
       "ENSG00000285441  ...                NaN            740.0                  NaN   \n",
       "ENSG00000288825  ...                NaN              NaN               4981.0   \n",
       "\n",
       "                 ESC_GR_2017_label  Hela_GSE102113_label  K562_GR_2017_label  \\\n",
       "ensembl_gene_id                                                                \n",
       "ENSG00000000003                NaN                   1.0                 NaN   \n",
       "ENSG00000000419                1.0                   1.0                 1.0   \n",
       "ENSG00000000457               -1.0                   NaN                -1.0   \n",
       "ENSG00000000460                0.0                   0.0                 0.0   \n",
       "ENSG00000000938                NaN                   NaN                 NaN   \n",
       "...                            ...                   ...                 ...   \n",
       "ENSG00000285304                NaN                   NaN                 NaN   \n",
       "ENSG00000285330                NaN                   NaN                 NaN   \n",
       "ENSG00000285396                NaN                   NaN                 NaN   \n",
       "ENSG00000285441                NaN                   NaN                 NaN   \n",
       "ENSG00000288825                NaN                   NaN                 NaN   \n",
       "\n",
       "                 K562_elife_label  K562_timeLapse_label  \\\n",
       "ensembl_gene_id                                           \n",
       "ENSG00000000003               NaN                   NaN   \n",
       "ENSG00000000419               0.0                   0.0   \n",
       "ENSG00000000457              -1.0                   NaN   \n",
       "ENSG00000000460               0.0                   NaN   \n",
       "ENSG00000000938               NaN                   NaN   \n",
       "...                           ...                   ...   \n",
       "ENSG00000285304               0.0                   NaN   \n",
       "ENSG00000285330               1.0                   NaN   \n",
       "ENSG00000285396               0.0                   NaN   \n",
       "ENSG00000285441               1.0                   NaN   \n",
       "ENSG00000288825               NaN                  -1.0   \n",
       "\n",
       "                 consistent_label_HalfLife  consistent_label_all  \n",
       "ensembl_gene_id                                                   \n",
       "ENSG00000000003                        NaN                   NaN  \n",
       "ENSG00000000419                        NaN                   NaN  \n",
       "ENSG00000000457                        NaN                   NaN  \n",
       "ENSG00000000460                        NaN                   NaN  \n",
       "ENSG00000000938                        NaN                   NaN  \n",
       "...                                    ...                   ...  \n",
       "ENSG00000285304                        NaN                   NaN  \n",
       "ENSG00000285330                        NaN                   NaN  \n",
       "ENSG00000285396                        NaN                   NaN  \n",
       "ENSG00000285441                        NaN                   NaN  \n",
       "ENSG00000288825                        NaN                   NaN  \n",
       "\n",
       "[19868 rows x 54 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "all_genes_omics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dde3eb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "基本思路：\n",
    "1）HEK 细胞系一致表达配对\n",
    "2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "6df199da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['iBAQ A549_exp', 'iBAQ HEK293_exp', 'iBAQ HeLa_exp', 'iBAQ K562_exp',\n",
       "       'iBAQ_TransEffect_A549_exp', 'iBAQ_TransEffect_HEK293_exp',\n",
       "       'iBAQ_TransEffect_HeLa_exp', 'iBAQ_TransEffect_K562_exp',\n",
       "       'iBAQ_TransEffect_A549_rank', 'iBAQ_TransEffect_HEK293_rank',\n",
       "       'iBAQ_TransEffect_HeLa_rank', 'iBAQ_TransEffect_K562_rank',\n",
       "       'iBAQ_TransEffect_A549_label', 'iBAQ_TransEffect_HEK293_label',\n",
       "       'iBAQ_TransEffect_HeLa_label', 'iBAQ_TransEffect_K562_label',\n",
       "       'consistent_label_iBAQ', 'ProteinFamily', 'CopyNumber_Hela_Nagaraj_exp',\n",
       "       'CopyNumber_Hela_Zecha_exp', 'TransEffect_Hela_Nagaraj_exp',\n",
       "       'TransEffect_Hela_Zecha_exp', 'TransEffect_Hela_Nagaraj_rank',\n",
       "       'TransEffect_Hela_Zecha_rank', 'TransEffect_Hela_Nagaraj_label',\n",
       "       'TransEffect_Hela_Zecha_label', 'consistent_label_CopyNumber',\n",
       "       'A549-SRR611121-SRR3286543-zhanggong_exp',\n",
       "       'hela-SRR8443314-SRR8443308_exp', 'HEK293-SRR1795409_10_exp',\n",
       "       'A549-SRR611121-SRR3286543-zhanggong_rank', 'HEK293-SRR1795409_10_rank',\n",
       "       'hela-SRR8443314-SRR8443308_rank',\n",
       "       'A549-SRR611121-SRR3286543-zhanggong_label',\n",
       "       'hela-SRR8443314-SRR8443308_label', 'HEK293-SRR1795409_10_label',\n",
       "       'consistent_label_RiboSeq', 'ESC_GR_2017_exp', 'Hela_GSE102113_exp',\n",
       "       'K562_GR_2017_exp', 'K562_elife_exp', 'K562_timeLapse_exp',\n",
       "       'ESC_GR_2017_rank', 'Hela_GSE102113_rank', 'K562_GR_2017_rank',\n",
       "       'K562_elife_rank', 'K562_timeLapse_rank', 'ESC_GR_2017_label',\n",
       "       'Hela_GSE102113_label', 'K562_GR_2017_label', 'K562_elife_label',\n",
       "       'K562_timeLapse_label', 'consistent_label_HalfLife',\n",
       "       'consistent_label_all'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "all_genes_omics.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "361a96ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>iBAQ_TransEffect_HEK293_rank</th>\n",
       "      <th>HEK293-SRR1795409_10_rank</th>\n",
       "      <th>ESC_GR_2017_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>iBAQ_TransEffect_HEK293_rank</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.274423</td>\n",
       "      <td>0.214244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEK293-SRR1795409_10_rank</th>\n",
       "      <td>0.274423</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.178760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ESC_GR_2017_rank</th>\n",
       "      <td>0.214244</td>\n",
       "      <td>0.178760</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              iBAQ_TransEffect_HEK293_rank  \\\n",
       "iBAQ_TransEffect_HEK293_rank                      1.000000   \n",
       "HEK293-SRR1795409_10_rank                         0.274423   \n",
       "ESC_GR_2017_rank                                  0.214244   \n",
       "\n",
       "                              HEK293-SRR1795409_10_rank  ESC_GR_2017_rank  \n",
       "iBAQ_TransEffect_HEK293_rank                   0.274423          0.214244  \n",
       "HEK293-SRR1795409_10_rank                      1.000000          0.178760  \n",
       "ESC_GR_2017_rank                               0.178760          1.000000  "
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# all_genes_omics[\n",
    "#     [\n",
    "   \n",
    "#         'iBAQ_TransEffect_HEK293_exp',\n",
    "#         'iBAQ_TransEffect_HEK293_rank',\n",
    "#         'iBAQ_TransEffect_HEK293_label',\n",
    "#         'HEK293-SRR1795409_10_exp',\n",
    "#         'HEK293-SRR1795409_10_rank',\n",
    "#         'HEK293-SRR1795409_10_label',\n",
    "#         'ESC_GR_2017_exp',\n",
    "#         'ESC_GR_2017_label',\n",
    "#         'ESC_GR_2017_rank',\n",
    "#     ]\n",
    "# ]\n",
    "\n",
    "all_genes_omics[\n",
    "    [\n",
    "   \n",
    "        'iBAQ_TransEffect_HEK293_exp',\n",
    "        'HEK293-SRR1795409_10_exp',\n",
    "        'ESC_GR_2017_exp',\n",
    "    ]\n",
    "].corr(method='spearman')\n",
    "\n",
    "\n",
    "all_genes_omics[\n",
    "    [\n",
    "   \n",
    "        'iBAQ_TransEffect_HEK293_rank',\n",
    "        'HEK293-SRR1795409_10_rank',\n",
    "        'ESC_GR_2017_rank',\n",
    "    ]\n",
    "].corr(method='spearman')\n",
    "\n",
    "all_genes_omics['ESC_GR_2017_exp'].hist()\n",
    "\n",
    "all_genes_omics['ESC_GR_2017_exp']\n",
    "# all_genes_omics['ESC_GR_2017_exp']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "8e4be964",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       " 0.0    3955\n",
       " 1.0    1978\n",
       "-1.0    1978\n",
       "Name: iBAQ_TransEffect_HEK293_label, dtype: int64"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "all_genes_omics['iBAQ_TransEffect_HEK293_label'].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "47eb3868",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ensembl_gene_id\n",
       "ENSG00000000003    0\n",
       "ENSG00000000419    1\n",
       "ENSG00000000457    0\n",
       "ENSG00000000460    0\n",
       "ENSG00000000938    0\n",
       "                  ..\n",
       "ENSG00000285304    0\n",
       "ENSG00000285330    0\n",
       "ENSG00000285396    0\n",
       "ENSG00000285441    0\n",
       "ENSG00000288825    0\n",
       "Length: 19868, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def f(x):\n",
    "    if x >=2:\n",
    "        return 1\n",
    "    elif x<=-2:\n",
    "        return -1\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "HEK_ESC_label = all_genes_omics[\n",
    "    [\n",
    "   \n",
    "        'iBAQ_TransEffect_HEK293_label',\n",
    "        'HEK293-SRR1795409_10_label',\n",
    "        'ESC_GR_2017_label',\n",
    "    ]\n",
    "].sum(axis=1).apply(f)\n",
    "\n",
    "HEK_ESC_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "59b8920e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id_raw</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>UTR5_regs</th>\n",
       "      <th>CDS_regs</th>\n",
       "      <th>UTR3_regs</th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>UTR3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>ENST00000371588</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>&gt;ENST00000371588.10|ENSG00000000419.14|OTTHUMG...</td>\n",
       "      <td>AGTTCCGCCATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGT...</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>UTR5:1-9</td>\n",
       "      <td>CDS:10-792</td>\n",
       "      <td>UTR3:793-1054</td>\n",
       "      <td>AGTTCCGCC</td>\n",
       "      <td>ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...</td>\n",
       "      <td>AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001497</th>\n",
       "      <td>ENST00000374811</td>\n",
       "      <td>ENSG00000001497.18</td>\n",
       "      <td>&gt;ENST00000374811.8|ENSG00000001497.18|OTTHUMG0...</td>\n",
       "      <td>GGAGTGAGGGTAGCACGCTGAGCTGAAGGCTGTGCGGAGCGGCGCG...</td>\n",
       "      <td>ENST00000374811.8</td>\n",
       "      <td>ENST00000374811.8</td>\n",
       "      <td>LAS1L</td>\n",
       "      <td>UTR5:1-72</td>\n",
       "      <td>CDS:73-2277</td>\n",
       "      <td>UTR3:2278-2470</td>\n",
       "      <td>GGAGTGAGGGTAGCACGCTGAGCTGAAGGCTGTGCGGAGCGGCGCG...</td>\n",
       "      <td>ATGTCGTGGGAATCCGGGGCCGGGCCAGGTCTAGGTTCCCAGGGGA...</td>\n",
       "      <td>TGGCCATCCCTGGTGCAAGTGTTCATCCAGCCGTGCCAGGGCAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002549</th>\n",
       "      <td>ENST00000606142</td>\n",
       "      <td>ENSG00000002549.13</td>\n",
       "      <td>&gt;ENST00000606142.5|ENSG00000002549.13|OTTHUMG0...</td>\n",
       "      <td>GCGAGTAGTCGTCCGACGTCTGGCCGTGAGACGTTTCGGGAGCCGG...</td>\n",
       "      <td>ENST00000606142.5</td>\n",
       "      <td>ENST00000606142.5</td>\n",
       "      <td>LAP3</td>\n",
       "      <td>UTR5:1-64</td>\n",
       "      <td>CDS:65-1531</td>\n",
       "      <td>UTR3:1532-1876</td>\n",
       "      <td>GCGAGTAGTCGTCCGACGTCTGGCCGTGAGACGTTTCGGGAGCCGG...</td>\n",
       "      <td>ATGACGAAGGGCCTTGTTTTAGGAATCTATTCCAAAGAAAAAGAAG...</td>\n",
       "      <td>TTCAGATACTCAAAAATGTCTTCACTCTGTCTTAAATTGGACAGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002834</th>\n",
       "      <td>ENST00000318008</td>\n",
       "      <td>ENSG00000002834.18</td>\n",
       "      <td>&gt;ENST00000318008.11|ENSG00000002834.18|OTTHUMG...</td>\n",
       "      <td>GCCAGTTCCCCAGCTCCAGCCGCCGTCGCTGCTGCCTGTGTAGTTG...</td>\n",
       "      <td>ENST00000318008.11</td>\n",
       "      <td>ENST00000318008.11</td>\n",
       "      <td>LASP1</td>\n",
       "      <td>UTR5:1-132</td>\n",
       "      <td>CDS:133-918</td>\n",
       "      <td>UTR3:919-3910</td>\n",
       "      <td>GCCAGTTCCCCAGCTCCAGCCGCCGTCGCTGCTGCCTGTGTAGTTG...</td>\n",
       "      <td>ATGAACCCCAACTGCGCCCGGTGCGGCAAGATCGTGTATCCCACGG...</td>\n",
       "      <td>ACCCGCAGCGCCCCCATCTGTCTTCAGCACATTCCACGGCATCGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004478</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>&gt;ENST00000001008.6|ENSG00000004478.8|OTTHUMG00...</td>\n",
       "      <td>CCTACCCCAGCTCTCGCGCCGCGTGCAGAGGTGCTCAAGCCTCCTC...</td>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>UTR5:1-170</td>\n",
       "      <td>CDS:171-1550</td>\n",
       "      <td>UTR3:1551-3715</td>\n",
       "      <td>CCTACCCCAGCTCTCGCGCCGCGTGCAGAGGTGCTCAAGCCTCCTC...</td>\n",
       "      <td>ATGACAGCCGAGGAGATGAAGGCGACCGAGAGCGGGGCGCAGTCGG...</td>\n",
       "      <td>CCCCTCTCCACCAGCCCTACTCCTGCGGCTGCCTGCCCCCCAGTCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Transcript             gene_id  \\\n",
       "ensembl_gene_id                                        \n",
       "ENSG00000000419  ENST00000371588  ENSG00000000419.14   \n",
       "ENSG00000001497  ENST00000374811  ENSG00000001497.18   \n",
       "ENSG00000002549  ENST00000606142  ENSG00000002549.13   \n",
       "ENSG00000002834  ENST00000318008  ENSG00000002834.18   \n",
       "ENSG00000004478  ENST00000001008   ENSG00000004478.8   \n",
       "\n",
       "                                                              name  \\\n",
       "ensembl_gene_id                                                      \n",
       "ENSG00000000419  >ENST00000371588.10|ENSG00000000419.14|OTTHUMG...   \n",
       "ENSG00000001497  >ENST00000374811.8|ENSG00000001497.18|OTTHUMG0...   \n",
       "ENSG00000002549  >ENST00000606142.5|ENSG00000002549.13|OTTHUMG0...   \n",
       "ENSG00000002834  >ENST00000318008.11|ENSG00000002834.18|OTTHUMG...   \n",
       "ENSG00000004478  >ENST00000001008.6|ENSG00000004478.8|OTTHUMG00...   \n",
       "\n",
       "                                                               seq  \\\n",
       "ensembl_gene_id                                                      \n",
       "ENSG00000000419  AGTTCCGCCATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGT...   \n",
       "ENSG00000001497  GGAGTGAGGGTAGCACGCTGAGCTGAAGGCTGTGCGGAGCGGCGCG...   \n",
       "ENSG00000002549  GCGAGTAGTCGTCCGACGTCTGGCCGTGAGACGTTTCGGGAGCCGG...   \n",
       "ENSG00000002834  GCCAGTTCCCCAGCTCCAGCCGCCGTCGCTGCTGCCTGTGTAGTTG...   \n",
       "ENSG00000004478  CCTACCCCAGCTCTCGCGCCGCGTGCAGAGGTGCTCAAGCCTCCTC...   \n",
       "\n",
       "                      transcript_id         gene_id_raw gene_name   UTR5_regs  \\\n",
       "ensembl_gene_id                                                                 \n",
       "ENSG00000000419  ENST00000371588.10  ENST00000371588.10      DPM1    UTR5:1-9   \n",
       "ENSG00000001497   ENST00000374811.8   ENST00000374811.8     LAS1L   UTR5:1-72   \n",
       "ENSG00000002549   ENST00000606142.5   ENST00000606142.5      LAP3   UTR5:1-64   \n",
       "ENSG00000002834  ENST00000318008.11  ENST00000318008.11     LASP1  UTR5:1-132   \n",
       "ENSG00000004478   ENST00000001008.6   ENST00000001008.6     FKBP4  UTR5:1-170   \n",
       "\n",
       "                     CDS_regs       UTR3_regs  \\\n",
       "ensembl_gene_id                                 \n",
       "ENSG00000000419    CDS:10-792   UTR3:793-1054   \n",
       "ENSG00000001497   CDS:73-2277  UTR3:2278-2470   \n",
       "ENSG00000002549   CDS:65-1531  UTR3:1532-1876   \n",
       "ENSG00000002834   CDS:133-918   UTR3:919-3910   \n",
       "ENSG00000004478  CDS:171-1550  UTR3:1551-3715   \n",
       "\n",
       "                                                              UTR5  \\\n",
       "ensembl_gene_id                                                      \n",
       "ENSG00000000419                                          AGTTCCGCC   \n",
       "ENSG00000001497  GGAGTGAGGGTAGCACGCTGAGCTGAAGGCTGTGCGGAGCGGCGCG...   \n",
       "ENSG00000002549  GCGAGTAGTCGTCCGACGTCTGGCCGTGAGACGTTTCGGGAGCCGG...   \n",
       "ENSG00000002834  GCCAGTTCCCCAGCTCCAGCCGCCGTCGCTGCTGCCTGTGTAGTTG...   \n",
       "ENSG00000004478  CCTACCCCAGCTCTCGCGCCGCGTGCAGAGGTGCTCAAGCCTCCTC...   \n",
       "\n",
       "                                                               CDS  \\\n",
       "ensembl_gene_id                                                      \n",
       "ENSG00000000419  ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...   \n",
       "ENSG00000001497  ATGTCGTGGGAATCCGGGGCCGGGCCAGGTCTAGGTTCCCAGGGGA...   \n",
       "ENSG00000002549  ATGACGAAGGGCCTTGTTTTAGGAATCTATTCCAAAGAAAAAGAAG...   \n",
       "ENSG00000002834  ATGAACCCCAACTGCGCCCGGTGCGGCAAGATCGTGTATCCCACGG...   \n",
       "ENSG00000004478  ATGACAGCCGAGGAGATGAAGGCGACCGAGAGCGGGGCGCAGTCGG...   \n",
       "\n",
       "                                                              UTR3  \n",
       "ensembl_gene_id                                                     \n",
       "ENSG00000000419  AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...  \n",
       "ENSG00000001497  TGGCCATCCCTGGTGCAAGTGTTCATCCAGCCGTGCCAGGGCAACA...  \n",
       "ENSG00000002549  TTCAGATACTCAAAAATGTCTTCACTCTGTCTTAAATTGGACAGTT...  \n",
       "ENSG00000002834  ACCCGCAGCGCCCCCATCTGTCTTCAGCACATTCCACGGCATCGCA...  \n",
       "ENSG00000004478  CCCCTCTCCACCAGCCCTACTCCTGCGGCTGCCTGCCCCCCAGTCT...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "HEK_ESC_data = human_ts.set_index('Esemble').reindex(HEK_ESC_label[HEK_ESC_label!=0].index)\n",
    "HEK_ESC_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4bc0a043",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id_raw</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>UTR5_regs</th>\n",
       "      <th>CDS_regs</th>\n",
       "      <th>UTR3_regs</th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>UTR3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>ENST00000371588</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>&gt;ENST00000371588.10|ENSG00000000419.14|OTTHUMG...</td>\n",
       "      <td>AGTTCCGCCATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGT...</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>UTR5:1-9</td>\n",
       "      <td>CDS:10-792</td>\n",
       "      <td>UTR3:793-1054</td>\n",
       "      <td>AGTTCCGCC</td>\n",
       "      <td>ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...</td>\n",
       "      <td>AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001497</th>\n",
       "      <td>ENST00000374811</td>\n",
       "      <td>ENSG00000001497.18</td>\n",
       "      <td>&gt;ENST00000374811.8|ENSG00000001497.18|OTTHUMG0...</td>\n",
       "      <td>GGAGTGAGGGTAGCACGCTGAGCTGAAGGCTGTGCGGAGCGGCGCG...</td>\n",
       "      <td>ENST00000374811.8</td>\n",
       "      <td>ENST00000374811.8</td>\n",
       "      <td>LAS1L</td>\n",
       "      <td>UTR5:1-72</td>\n",
       "      <td>CDS:73-2277</td>\n",
       "      <td>UTR3:2278-2470</td>\n",
       "      <td>GGAGTGAGGGTAGCACGCTGAGCTGAAGGCTGTGCGGAGCGGCGCG...</td>\n",
       "      <td>ATGTCGTGGGAATCCGGGGCCGGGCCAGGTCTAGGTTCCCAGGGGA...</td>\n",
       "      <td>TGGCCATCCCTGGTGCAAGTGTTCATCCAGCCGTGCCAGGGCAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002549</th>\n",
       "      <td>ENST00000606142</td>\n",
       "      <td>ENSG00000002549.13</td>\n",
       "      <td>&gt;ENST00000606142.5|ENSG00000002549.13|OTTHUMG0...</td>\n",
       "      <td>GCGAGTAGTCGTCCGACGTCTGGCCGTGAGACGTTTCGGGAGCCGG...</td>\n",
       "      <td>ENST00000606142.5</td>\n",
       "      <td>ENST00000606142.5</td>\n",
       "      <td>LAP3</td>\n",
       "      <td>UTR5:1-64</td>\n",
       "      <td>CDS:65-1531</td>\n",
       "      <td>UTR3:1532-1876</td>\n",
       "      <td>GCGAGTAGTCGTCCGACGTCTGGCCGTGAGACGTTTCGGGAGCCGG...</td>\n",
       "      <td>ATGACGAAGGGCCTTGTTTTAGGAATCTATTCCAAAGAAAAAGAAG...</td>\n",
       "      <td>TTCAGATACTCAAAAATGTCTTCACTCTGTCTTAAATTGGACAGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002834</th>\n",
       "      <td>ENST00000318008</td>\n",
       "      <td>ENSG00000002834.18</td>\n",
       "      <td>&gt;ENST00000318008.11|ENSG00000002834.18|OTTHUMG...</td>\n",
       "      <td>GCCAGTTCCCCAGCTCCAGCCGCCGTCGCTGCTGCCTGTGTAGTTG...</td>\n",
       "      <td>ENST00000318008.11</td>\n",
       "      <td>ENST00000318008.11</td>\n",
       "      <td>LASP1</td>\n",
       "      <td>UTR5:1-132</td>\n",
       "      <td>CDS:133-918</td>\n",
       "      <td>UTR3:919-3910</td>\n",
       "      <td>GCCAGTTCCCCAGCTCCAGCCGCCGTCGCTGCTGCCTGTGTAGTTG...</td>\n",
       "      <td>ATGAACCCCAACTGCGCCCGGTGCGGCAAGATCGTGTATCCCACGG...</td>\n",
       "      <td>ACCCGCAGCGCCCCCATCTGTCTTCAGCACATTCCACGGCATCGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004478</th>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>&gt;ENST00000001008.6|ENSG00000004478.8|OTTHUMG00...</td>\n",
       "      <td>CCTACCCCAGCTCTCGCGCCGCGTGCAGAGGTGCTCAAGCCTCCTC...</td>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>UTR5:1-170</td>\n",
       "      <td>CDS:171-1550</td>\n",
       "      <td>UTR3:1551-3715</td>\n",
       "      <td>CCTACCCCAGCTCTCGCGCCGCGTGCAGAGGTGCTCAAGCCTCCTC...</td>\n",
       "      <td>ATGACAGCCGAGGAGATGAAGGCGACCGAGAGCGGGGCGCAGTCGG...</td>\n",
       "      <td>CCCCTCTCCACCAGCCCTACTCCTGCGGCTGCCTGCCCCCCAGTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000260455</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000261115</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000261326</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000266904</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000269343</th>\n",
       "      <td>ENST00000594901</td>\n",
       "      <td>ENSG00000269343.9</td>\n",
       "      <td>&gt;ENST00000594901.2|ENSG00000269343.9|OTTHUMG00...</td>\n",
       "      <td>GATTGGTGTTGGGTTTATTTGTCGGAGAGGCTCCTGAGCGCTAGGT...</td>\n",
       "      <td>ENST00000594901.2</td>\n",
       "      <td>ENST00000594901.2</td>\n",
       "      <td>ZNF587B</td>\n",
       "      <td>UTR5:1-230</td>\n",
       "      <td>CDS:231-2132</td>\n",
       "      <td>UTR3:2133-5794</td>\n",
       "      <td>GATTGGTGTTGGGTTTATTTGTCGGAGAGGCTCCTGAGCGCTAGGT...</td>\n",
       "      <td>ATGGCGGTGGTGGCCACGCTGAGGCTCTCTGCTCAGGGCACAGTGA...</td>\n",
       "      <td>GTGCAGAGAATGAGTCCAGTCTCATTAAATACAGGAGAGCACACAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2068 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Transcript             gene_id  \\\n",
       "ensembl_gene_id                                        \n",
       "ENSG00000000419  ENST00000371588  ENSG00000000419.14   \n",
       "ENSG00000001497  ENST00000374811  ENSG00000001497.18   \n",
       "ENSG00000002549  ENST00000606142  ENSG00000002549.13   \n",
       "ENSG00000002834  ENST00000318008  ENSG00000002834.18   \n",
       "ENSG00000004478  ENST00000001008   ENSG00000004478.8   \n",
       "...                          ...                 ...   \n",
       "ENSG00000260455              NaN                 NaN   \n",
       "ENSG00000261115              NaN                 NaN   \n",
       "ENSG00000261326              NaN                 NaN   \n",
       "ENSG00000266904              NaN                 NaN   \n",
       "ENSG00000269343  ENST00000594901   ENSG00000269343.9   \n",
       "\n",
       "                                                              name  \\\n",
       "ensembl_gene_id                                                      \n",
       "ENSG00000000419  >ENST00000371588.10|ENSG00000000419.14|OTTHUMG...   \n",
       "ENSG00000001497  >ENST00000374811.8|ENSG00000001497.18|OTTHUMG0...   \n",
       "ENSG00000002549  >ENST00000606142.5|ENSG00000002549.13|OTTHUMG0...   \n",
       "ENSG00000002834  >ENST00000318008.11|ENSG00000002834.18|OTTHUMG...   \n",
       "ENSG00000004478  >ENST00000001008.6|ENSG00000004478.8|OTTHUMG00...   \n",
       "...                                                            ...   \n",
       "ENSG00000260455                                                NaN   \n",
       "ENSG00000261115                                                NaN   \n",
       "ENSG00000261326                                                NaN   \n",
       "ENSG00000266904                                                NaN   \n",
       "ENSG00000269343  >ENST00000594901.2|ENSG00000269343.9|OTTHUMG00...   \n",
       "\n",
       "                                                               seq  \\\n",
       "ensembl_gene_id                                                      \n",
       "ENSG00000000419  AGTTCCGCCATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGT...   \n",
       "ENSG00000001497  GGAGTGAGGGTAGCACGCTGAGCTGAAGGCTGTGCGGAGCGGCGCG...   \n",
       "ENSG00000002549  GCGAGTAGTCGTCCGACGTCTGGCCGTGAGACGTTTCGGGAGCCGG...   \n",
       "ENSG00000002834  GCCAGTTCCCCAGCTCCAGCCGCCGTCGCTGCTGCCTGTGTAGTTG...   \n",
       "ENSG00000004478  CCTACCCCAGCTCTCGCGCCGCGTGCAGAGGTGCTCAAGCCTCCTC...   \n",
       "...                                                            ...   \n",
       "ENSG00000260455                                                NaN   \n",
       "ENSG00000261115                                                NaN   \n",
       "ENSG00000261326                                                NaN   \n",
       "ENSG00000266904                                                NaN   \n",
       "ENSG00000269343  GATTGGTGTTGGGTTTATTTGTCGGAGAGGCTCCTGAGCGCTAGGT...   \n",
       "\n",
       "                      transcript_id         gene_id_raw gene_name   UTR5_regs  \\\n",
       "ensembl_gene_id                                                                 \n",
       "ENSG00000000419  ENST00000371588.10  ENST00000371588.10      DPM1    UTR5:1-9   \n",
       "ENSG00000001497   ENST00000374811.8   ENST00000374811.8     LAS1L   UTR5:1-72   \n",
       "ENSG00000002549   ENST00000606142.5   ENST00000606142.5      LAP3   UTR5:1-64   \n",
       "ENSG00000002834  ENST00000318008.11  ENST00000318008.11     LASP1  UTR5:1-132   \n",
       "ENSG00000004478   ENST00000001008.6   ENST00000001008.6     FKBP4  UTR5:1-170   \n",
       "...                             ...                 ...       ...         ...   \n",
       "ENSG00000260455                 NaN                 NaN       NaN         NaN   \n",
       "ENSG00000261115                 NaN                 NaN       NaN         NaN   \n",
       "ENSG00000261326                 NaN                 NaN       NaN         NaN   \n",
       "ENSG00000266904                 NaN                 NaN       NaN         NaN   \n",
       "ENSG00000269343   ENST00000594901.2   ENST00000594901.2   ZNF587B  UTR5:1-230   \n",
       "\n",
       "                     CDS_regs       UTR3_regs  \\\n",
       "ensembl_gene_id                                 \n",
       "ENSG00000000419    CDS:10-792   UTR3:793-1054   \n",
       "ENSG00000001497   CDS:73-2277  UTR3:2278-2470   \n",
       "ENSG00000002549   CDS:65-1531  UTR3:1532-1876   \n",
       "ENSG00000002834   CDS:133-918   UTR3:919-3910   \n",
       "ENSG00000004478  CDS:171-1550  UTR3:1551-3715   \n",
       "...                       ...             ...   \n",
       "ENSG00000260455           NaN             NaN   \n",
       "ENSG00000261115           NaN             NaN   \n",
       "ENSG00000261326           NaN             NaN   \n",
       "ENSG00000266904           NaN             NaN   \n",
       "ENSG00000269343  CDS:231-2132  UTR3:2133-5794   \n",
       "\n",
       "                                                              UTR5  \\\n",
       "ensembl_gene_id                                                      \n",
       "ENSG00000000419                                          AGTTCCGCC   \n",
       "ENSG00000001497  GGAGTGAGGGTAGCACGCTGAGCTGAAGGCTGTGCGGAGCGGCGCG...   \n",
       "ENSG00000002549  GCGAGTAGTCGTCCGACGTCTGGCCGTGAGACGTTTCGGGAGCCGG...   \n",
       "ENSG00000002834  GCCAGTTCCCCAGCTCCAGCCGCCGTCGCTGCTGCCTGTGTAGTTG...   \n",
       "ENSG00000004478  CCTACCCCAGCTCTCGCGCCGCGTGCAGAGGTGCTCAAGCCTCCTC...   \n",
       "...                                                            ...   \n",
       "ENSG00000260455                                                NaN   \n",
       "ENSG00000261115                                                NaN   \n",
       "ENSG00000261326                                                NaN   \n",
       "ENSG00000266904                                                NaN   \n",
       "ENSG00000269343  GATTGGTGTTGGGTTTATTTGTCGGAGAGGCTCCTGAGCGCTAGGT...   \n",
       "\n",
       "                                                               CDS  \\\n",
       "ensembl_gene_id                                                      \n",
       "ENSG00000000419  ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...   \n",
       "ENSG00000001497  ATGTCGTGGGAATCCGGGGCCGGGCCAGGTCTAGGTTCCCAGGGGA...   \n",
       "ENSG00000002549  ATGACGAAGGGCCTTGTTTTAGGAATCTATTCCAAAGAAAAAGAAG...   \n",
       "ENSG00000002834  ATGAACCCCAACTGCGCCCGGTGCGGCAAGATCGTGTATCCCACGG...   \n",
       "ENSG00000004478  ATGACAGCCGAGGAGATGAAGGCGACCGAGAGCGGGGCGCAGTCGG...   \n",
       "...                                                            ...   \n",
       "ENSG00000260455                                                NaN   \n",
       "ENSG00000261115                                                NaN   \n",
       "ENSG00000261326                                                NaN   \n",
       "ENSG00000266904                                                NaN   \n",
       "ENSG00000269343  ATGGCGGTGGTGGCCACGCTGAGGCTCTCTGCTCAGGGCACAGTGA...   \n",
       "\n",
       "                                                              UTR3  \n",
       "ensembl_gene_id                                                     \n",
       "ENSG00000000419  AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...  \n",
       "ENSG00000001497  TGGCCATCCCTGGTGCAAGTGTTCATCCAGCCGTGCCAGGGCAACA...  \n",
       "ENSG00000002549  TTCAGATACTCAAAAATGTCTTCACTCTGTCTTAAATTGGACAGTT...  \n",
       "ENSG00000002834  ACCCGCAGCGCCCCCATCTGTCTTCAGCACATTCCACGGCATCGCA...  \n",
       "ENSG00000004478  CCCCTCTCCACCAGCCCTACTCCTGCGGCTGCCTGCCCCCCAGTCT...  \n",
       "...                                                            ...  \n",
       "ENSG00000260455                                                NaN  \n",
       "ENSG00000261115                                                NaN  \n",
       "ENSG00000261326                                                NaN  \n",
       "ENSG00000266904                                                NaN  \n",
       "ENSG00000269343  GTGCAGAGAATGAGTCCAGTCTCATTAAATACAGGAGAGCACACAC...  \n",
       "\n",
       "[2068 rows x 13 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# row[\"CDS\"]\n",
    "\n",
    "HEK_ESC_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "6a5455f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open('/home/yul/project/nas_project/bioSPT2/data/interiem/processed/HEK_ESC_data.fasta', 'w') as f:\n",
    "    for index, row in HEK_ESC_data.iterrows():\n",
    "        f.write(f'>{index}\\n{row[\"CDS\"]}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f13b93a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_17137/1899396817.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rf_cdhit_pd['group_item_labels'] = rf_cdhit_pd['group_items'].apply(lambda vars: [HEK_ESC_label.reindex([v])[0] for v in vars])\n",
      "/tmp/ipykernel_17137/1899396817.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rf_cdhit_pd['group_item_ESC_values'] = rf_cdhit_pd['group_items'].apply(lambda vars: [all_genes_omics.reindex([v])['ESC_GR_2017_exp'][0] for v in vars])\n",
      "/tmp/ipykernel_17137/1899396817.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rf_cdhit_pd['group_item_iBAQ_values'] = rf_cdhit_pd['group_items'].apply(lambda vars: [all_genes_omics.reindex([v])['iBAQ_TransEffect_HEK293_exp'][0] for v in vars])\n",
      "/tmp/ipykernel_17137/1899396817.py:38: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rf_cdhit_pd['group_item_TE_values'] = rf_cdhit_pd['group_items'].apply(lambda vars: [all_genes_omics.reindex([v])['HEK293-SRR1795409_10_exp'][0] for v in vars])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group_id</th>\n",
       "      <th>group_items</th>\n",
       "      <th>group_item_labels</th>\n",
       "      <th>group_item_ESC_values</th>\n",
       "      <th>group_item_iBAQ_values</th>\n",
       "      <th>group_item_TE_values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>&gt;Cluster 208</td>\n",
       "      <td>[ENSG00000160336, ENSG00000213799, ENSG0000019...</td>\n",
       "      <td>[-1, -1, -1]</td>\n",
       "      <td>[nan, 1.2009223333333334, 2.2410646666666665]</td>\n",
       "      <td>[3.026569772605642, 3.981287824628937, nan]</td>\n",
       "      <td>[0.080205, 0.256335, 0.063765]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>&gt;Cluster 243</td>\n",
       "      <td>[ENSG00000072110, ENSG00000130402]</td>\n",
       "      <td>[1, 1]</td>\n",
       "      <td>[nan, nan]</td>\n",
       "      <td>[5.638357939299184, 6.182171418362166]</td>\n",
       "      <td>[1.123115, 2.271605]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>&gt;Cluster 462</td>\n",
       "      <td>[ENSG00000178935, ENSG00000269343]</td>\n",
       "      <td>[-1, -1]</td>\n",
       "      <td>[1.7730736666666669, 1.8657866666666667]</td>\n",
       "      <td>[3.567627576953936, nan]</td>\n",
       "      <td>[0.06825, 0.012015]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>&gt;Cluster 473</td>\n",
       "      <td>[ENSG00000178229, ENSG00000083844]</td>\n",
       "      <td>[-1, -1]</td>\n",
       "      <td>[1.0260586, 1.2323676666666668]</td>\n",
       "      <td>[3.011159736350703, nan]</td>\n",
       "      <td>[0.468535, 0.075545]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>&gt;Cluster 586</td>\n",
       "      <td>[ENSG00000182141, ENSG00000237440]</td>\n",
       "      <td>[-1, -1]</td>\n",
       "      <td>[2.041819333333333, 1.306525]</td>\n",
       "      <td>[2.775471759967293, nan]</td>\n",
       "      <td>[0.0962, 0.06184]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group_id                                        group_items  \\\n",
       "208  >Cluster 208  [ENSG00000160336, ENSG00000213799, ENSG0000019...   \n",
       "243  >Cluster 243                 [ENSG00000072110, ENSG00000130402]   \n",
       "462  >Cluster 462                 [ENSG00000178935, ENSG00000269343]   \n",
       "473  >Cluster 473                 [ENSG00000178229, ENSG00000083844]   \n",
       "586  >Cluster 586                 [ENSG00000182141, ENSG00000237440]   \n",
       "\n",
       "    group_item_labels                          group_item_ESC_values  \\\n",
       "208      [-1, -1, -1]  [nan, 1.2009223333333334, 2.2410646666666665]   \n",
       "243            [1, 1]                                     [nan, nan]   \n",
       "462          [-1, -1]       [1.7730736666666669, 1.8657866666666667]   \n",
       "473          [-1, -1]                [1.0260586, 1.2323676666666668]   \n",
       "586          [-1, -1]                  [2.041819333333333, 1.306525]   \n",
       "\n",
       "                          group_item_iBAQ_values  \\\n",
       "208  [3.026569772605642, 3.981287824628937, nan]   \n",
       "243       [5.638357939299184, 6.182171418362166]   \n",
       "462                     [3.567627576953936, nan]   \n",
       "473                     [3.011159736350703, nan]   \n",
       "586                     [2.775471759967293, nan]   \n",
       "\n",
       "               group_item_TE_values  \n",
       "208  [0.080205, 0.256335, 0.063765]  \n",
       "243            [1.123115, 2.271605]  \n",
       "462             [0.06825, 0.012015]  \n",
       "473            [0.468535, 0.075545]  \n",
       "586               [0.0962, 0.06184]  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 提取聚类簇\n",
    "\n",
    "cdhit_pd = pd.DataFrame()\n",
    "with open('/data1/genomics/yul/bioSPT2/data/interiem/processed/HEK_ESC_data_db_0.8.clstr', 'r') as f:\n",
    "    \n",
    "    lines=f.readlines()\n",
    "    \n",
    "    group_ids = []\n",
    "    groups = []\n",
    "    for line in lines:\n",
    "        if line.startswith('>'):\n",
    "            group_id = line.strip()\n",
    "            group_ids.append(group_id)\n",
    "            group_items = []\n",
    "            groups.append(group_items)\n",
    "        else:\n",
    "            group_item = line.strip().split('>')[-1].strip().split('...')[0]\n",
    "            group_items.append(group_item)\n",
    "            \n",
    "    groups.append(group_items)\n",
    "    \n",
    "\n",
    "cdhit_pd = pd.DataFrame(\n",
    "    {\n",
    "        'group_id': group_ids,\n",
    "        'group_items': groups[1:]\n",
    "    }\n",
    ")\n",
    "\n",
    "rf_cdhit_pd = cdhit_pd[cdhit_pd['group_items'].apply(len) > 1]\n",
    "\n",
    "rf_cdhit_pd['group_item_labels'] = rf_cdhit_pd['group_items'].apply(lambda vars: [HEK_ESC_label.reindex([v])[0] for v in vars])\n",
    "\n",
    "rf_cdhit_pd['group_item_ESC_values'] = rf_cdhit_pd['group_items'].apply(lambda vars: [all_genes_omics.reindex([v])['ESC_GR_2017_exp'][0] for v in vars]) \n",
    "\n",
    "rf_cdhit_pd['group_item_iBAQ_values'] = rf_cdhit_pd['group_items'].apply(lambda vars: [all_genes_omics.reindex([v])['iBAQ_TransEffect_HEK293_exp'][0] for v in vars]) \n",
    "\n",
    "rf_cdhit_pd['group_item_TE_values'] = rf_cdhit_pd['group_items'].apply(lambda vars: [all_genes_omics.reindex([v])['HEK293-SRR1795409_10_exp'][0] for v in vars])\n",
    "\n",
    "rf_cdhit_pd.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d6769109",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "208\n",
      "['ENSG00000160336', 'ENSG00000213799', 0, -1, -1]\n",
      "['ENSG00000160336', 'ENSG00000196417', 0, 0, 0]\n",
      "['ENSG00000213799', 'ENSG00000196417', -1, 1, 0]\n",
      "243\n",
      "['ENSG00000072110', 'ENSG00000130402', 0, -1, 0]\n",
      "462\n",
      "['ENSG00000178935', 'ENSG00000269343', 0, 1, 0]\n",
      "473\n",
      "['ENSG00000178229', 'ENSG00000083844', 0, 1, 0]\n",
      "586\n",
      "['ENSG00000182141', 'ENSG00000237440', 1, 1, 0]\n",
      "602\n",
      "['ENSG00000083812', 'ENSG00000249471', 0, 1, 0]\n",
      "620\n",
      "['ENSG00000187626', 'ENSG00000189298', -1, 1, 0]\n",
      "757\n",
      "['ENSG00000130741', 'ENSG00000180574', 1, 0, 0]\n",
      "813\n",
      "['ENSG00000104833', 'ENSG00000137267', -1, 1, 0]\n",
      "['ENSG00000104833', 'ENSG00000137285', -1, 0, 0]\n",
      "['ENSG00000104833', 'ENSG00000188229', 0, 1, 0]\n",
      "['ENSG00000104833', 'ENSG00000196230', -1, 0, 0]\n",
      "['ENSG00000104833', 'ENSG00000258947', 0, 1, 0]\n",
      "['ENSG00000137267', 'ENSG00000137285', -1, 0, 1]\n",
      "['ENSG00000137267', 'ENSG00000188229', 1, 0, 0]\n",
      "['ENSG00000137267', 'ENSG00000196230', -1, 0, 0]\n",
      "['ENSG00000137267', 'ENSG00000258947', 0, 1, 0]\n",
      "['ENSG00000137285', 'ENSG00000188229', 1, 0, -1]\n",
      "['ENSG00000137285', 'ENSG00000196230', -1, 0, 0]\n",
      "['ENSG00000137285', 'ENSG00000258947', 0, 1, 0]\n",
      "['ENSG00000188229', 'ENSG00000196230', -1, 0, 0]\n",
      "['ENSG00000188229', 'ENSG00000258947', 0, 1, 0]\n",
      "['ENSG00000196230', 'ENSG00000258947', 0, 0, 0]\n",
      "814\n",
      "['ENSG00000126945', 'ENSG00000169045', -1, 0, 0]\n",
      "903\n",
      "['ENSG00000223572', 'ENSG00000237289', -1, 1, 0]\n",
      "954\n",
      "['ENSG00000197054', 'ENSG00000181896', -1, 1, 0]\n",
      "1031\n",
      "['ENSG00000206503', 'ENSG00000234745', 0, 0, 0]\n",
      "1588\n",
      "['ENSG00000101608', 'ENSG00000118680', 0, -1, 0]\n",
      "1695\n",
      "['ENSG00000203852', 'ENSG00000274750', 0, 0, 1]\n",
      "['ENSG00000203852', 'ENSG00000275714', 0, 1, 1]\n",
      "['ENSG00000203852', 'ENSG00000286522', 0, 0, 1]\n",
      "['ENSG00000274750', 'ENSG00000275714', 0, 1, 0]\n",
      "['ENSG00000274750', 'ENSG00000286522', 0, 0, 0]\n",
      "['ENSG00000275714', 'ENSG00000286522', 0, -1, 0]\n",
      "1705\n",
      "['ENSG00000184260', 'ENSG00000196747', 0, 0, 0]\n",
      "['ENSG00000184260', 'ENSG00000196787', 0, 0, 0]\n",
      "['ENSG00000184260', 'ENSG00000246705', 0, 0, 0]\n",
      "['ENSG00000184260', 'ENSG00000276903', 0, 0, 0]\n",
      "['ENSG00000196747', 'ENSG00000196787', 0, 0, 0]\n",
      "['ENSG00000196747', 'ENSG00000246705', -1, 0, 0]\n",
      "['ENSG00000196747', 'ENSG00000276903', 0, 1, 0]\n",
      "['ENSG00000196787', 'ENSG00000246705', -1, 0, 0]\n",
      "['ENSG00000196787', 'ENSG00000276903', 0, 0, 0]\n",
      "['ENSG00000246705', 'ENSG00000276903', 0, 0, 0]\n",
      "1717\n",
      "['ENSG00000158373', 'ENSG00000184678', 0, -1, 0]\n",
      "['ENSG00000158373', 'ENSG00000185130', 0, 1, 1]\n",
      "['ENSG00000184678', 'ENSG00000185130', 0, 1, 1]\n",
      "1768\n",
      "['ENSG00000158406', 'ENSG00000197061', 0, 1, 0]\n",
      "['ENSG00000158406', 'ENSG00000270276', 0, 1, 0]\n",
      "['ENSG00000158406', 'ENSG00000276180', 0, 1, 0]\n",
      "['ENSG00000158406', 'ENSG00000277157', 0, 1, 0]\n",
      "['ENSG00000158406', 'ENSG00000278637', 0, 1, 0]\n",
      "['ENSG00000197061', 'ENSG00000270276', 0, 0, 0]\n",
      "['ENSG00000197061', 'ENSG00000276180', 0, 1, 0]\n",
      "['ENSG00000197061', 'ENSG00000277157', 0, 0, 0]\n",
      "['ENSG00000197061', 'ENSG00000278637', 0, 0, 0]\n",
      "['ENSG00000270276', 'ENSG00000276180', 0, 1, 0]\n",
      "['ENSG00000270276', 'ENSG00000277157', 0, 0, 0]\n",
      "['ENSG00000270276', 'ENSG00000278637', 0, 0, 0]\n",
      "['ENSG00000276180', 'ENSG00000277157', 0, -1, 0]\n",
      "['ENSG00000276180', 'ENSG00000278637', 0, -1, 0]\n",
      "['ENSG00000277157', 'ENSG00000278637', 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "pairs = []\n",
    "for index, row in rf_cdhit_pd.iterrows():\n",
    "    \n",
    "    logb = 1.3\n",
    "    lb, ub  = 1/logb, logb    \n",
    "    \n",
    "    print(index)\n",
    "    for i in range(len(row['group_items'])-1):\n",
    "        for j in range(i+1, len(row['group_items'])):\n",
    "            pair = []\n",
    "            pair.extend([row['group_items'][i], row['group_items'][j]])\n",
    "            if row['group_item_ESC_values'][i] and row['group_item_ESC_values'][j]:\n",
    "                if row['group_item_ESC_values'][i]/row['group_item_ESC_values'][j] >=ub:\n",
    "                    pair.append(1)\n",
    "                    \n",
    "                elif row['group_item_ESC_values'][i]/row['group_item_ESC_values'][j] <= lb:\n",
    "                    pair.append(-1)\n",
    "                    \n",
    "                else:\n",
    "                    pair.append(0)\n",
    "            else:\n",
    "                pair.append(None)\n",
    "        \n",
    "            if row['group_item_TE_values'][i] and row['group_item_TE_values'][j]:\n",
    "                if row['group_item_TE_values'][i]/row['group_item_TE_values'][j] >=ub:\n",
    "                    pair.append(1)\n",
    "                    \n",
    "                elif row['group_item_TE_values'][i]/row['group_item_TE_values'][j] <= lb:\n",
    "                    pair.append(-1)\n",
    "                    \n",
    "                else:\n",
    "                    pair.append(0)\n",
    "            else:\n",
    "                pair.append(None)\n",
    "                    \n",
    "            if row['group_item_iBAQ_values'][i] and row['group_item_iBAQ_values'][j]:\n",
    "                if row['group_item_iBAQ_values'][i]/row['group_item_iBAQ_values'][j] >=ub:\n",
    "                    pair.append(1)\n",
    "                    \n",
    "                elif row['group_item_iBAQ_values'][i]/row['group_item_iBAQ_values'][j] <= lb:\n",
    "                    pair.append(-1)\n",
    "                    \n",
    "                else:\n",
    "                    pair.append(0)\n",
    "            else:\n",
    "                pair.append(None)\n",
    "                \n",
    "            print(pair)\n",
    "            pairs.append(pair)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "96bf92f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>ESC</th>\n",
       "      <th>TE</th>\n",
       "      <th>iBAQ</th>\n",
       "      <th>pair_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000160336</td>\n",
       "      <td>ENSG00000213799</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000160336</td>\n",
       "      <td>ENSG00000196417</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000213799</td>\n",
       "      <td>ENSG00000196417</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000072110</td>\n",
       "      <td>ENSG00000130402</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000178935</td>\n",
       "      <td>ENSG00000269343</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000270276</td>\n",
       "      <td>ENSG00000277157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ENSG00000270276</td>\n",
       "      <td>ENSG00000278637</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ENSG00000276180</td>\n",
       "      <td>ENSG00000277157</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>ENSG00000276180</td>\n",
       "      <td>ENSG00000278637</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>ENSG00000277157</td>\n",
       "      <td>ENSG00000278637</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene1            gene2  ESC  TE  iBAQ  pair_label\n",
       "0   ENSG00000160336  ENSG00000213799    0  -1    -1          -1\n",
       "1   ENSG00000160336  ENSG00000196417    0   0     0           0\n",
       "2   ENSG00000213799  ENSG00000196417   -1   1     0           0\n",
       "3   ENSG00000072110  ENSG00000130402    0  -1     0          -1\n",
       "4   ENSG00000178935  ENSG00000269343    0   1     0           1\n",
       "..              ...              ...  ...  ..   ...         ...\n",
       "59  ENSG00000270276  ENSG00000277157    0   0     0           0\n",
       "60  ENSG00000270276  ENSG00000278637    0   0     0           0\n",
       "61  ENSG00000276180  ENSG00000277157    0  -1     0          -1\n",
       "62  ENSG00000276180  ENSG00000278637    0  -1     0          -1\n",
       "63  ENSG00000277157  ENSG00000278637    0   0     0           0\n",
       "\n",
       "[64 rows x 6 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "RF_exp_pairs = pd.DataFrame(pairs, columns=['gene1', 'gene2', 'ESC', 'TE', 'iBAQ'])\n",
    "\n",
    "\n",
    "def label(x):\n",
    "    \n",
    "    cols = ['ESC', 'TE', 'iBAQ']\n",
    "    \n",
    "    if all([x[c] >=0 for c in cols]) or all([x[c] <=0 for c in cols]):\n",
    "        if sum([x[c] for c in cols]) > 0:\n",
    "            return 1\n",
    "        elif sum([x[c] for c in cols]) < 0:\n",
    "            return -1\n",
    "        else:\n",
    "            return 0\n",
    "    return 0\n",
    "    \n",
    "RF_exp_pairs['pair_label'] = RF_exp_pairs.apply(label, axis=1)\n",
    "\n",
    "RF_exp_pairs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8dca6a22",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "RF_exp_pairs.to_csv('/home/yul/project/nas_project/bioSPT2/data/interiem/processed/HEK_ESC_data_db_0.8_pairs_expression.tsv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a67bc46",
   "metadata": {},
   "source": [
    "#### utr5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2cd4e8b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>experiment_batch</th>\n",
       "      <th>experiment_enzyme</th>\n",
       "      <th>enzyme_seq</th>\n",
       "      <th>sequence_adapter</th>\n",
       "      <th>cap_seq</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq</th>\n",
       "      <th>template_seq</th>\n",
       "      <th>template_name</th>\n",
       "      <th>gene</th>\n",
       "      <th>is_template</th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>CDS_name</th>\n",
       "      <th>UTR3</th>\n",
       "      <th>expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>47</td>\n",
       "      <td>14</td>\n",
       "      <td>v1</td>\n",
       "      <td>BamHI</td>\n",
       "      <td>GGATCC</td>\n",
       "      <td>TAATACGACTCACTATA</td>\n",
       "      <td>GGG</td>\n",
       "      <td>template@Neocura_5UTR</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>Neocura_5UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>degron-No.15</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>1.650000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>64</td>\n",
       "      <td>14</td>\n",
       "      <td>v1</td>\n",
       "      <td>BamHI</td>\n",
       "      <td>GGATCC</td>\n",
       "      <td>TAATACGACTCACTATA</td>\n",
       "      <td>GGG</td>\n",
       "      <td>template@Neocura_5UTR</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>Neocura_5UTR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "      <td>degron-No.22</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>1.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>86</td>\n",
       "      <td>No.39</td>\n",
       "      <td>v2</td>\n",
       "      <td>EcoRI</td>\n",
       "      <td>GAATTC</td>\n",
       "      <td>TAATACGACTCACTATA</td>\n",
       "      <td>GGA</td>\n",
       "      <td>69_1_ENSG00000163220</td>\n",
       "      <td>AAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCTCTT...</td>\n",
       "      <td>AAACACTCTGTGTGGCTCCTCGGCTTTGACAGAGTGCAAGACG</td>\n",
       "      <td>ENSG00000163220</td>\n",
       "      <td>ENSG00000163220</td>\n",
       "      <td>False</td>\n",
       "      <td>GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...</td>\n",
       "      <td>AGCAGAGTGCAGCGGGTGACAGACAGCCTGGACACACTGCTGGAGA...</td>\n",
       "      <td>AG2-25</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>1.433288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>88</td>\n",
       "      <td>No.41</td>\n",
       "      <td>v2</td>\n",
       "      <td>EcoRI</td>\n",
       "      <td>GAATTC</td>\n",
       "      <td>TAATACGACTCACTATA</td>\n",
       "      <td>GGA</td>\n",
       "      <td>171_1_ENSG00000163220</td>\n",
       "      <td>ACTTCTCGAGTTTTGGTGTTTTGTCGGTTCCGCTGTTGCGTTGTGG...</td>\n",
       "      <td>AAACACTCTGTGTGGCTCCTCGGCTTTGACAGAGTGCAAGACG</td>\n",
       "      <td>ENSG00000163220</td>\n",
       "      <td>ENSG00000163220</td>\n",
       "      <td>False</td>\n",
       "      <td>GGAACTTCTCGAGTTTTGGTGTTTTGTCGGTTCCGCTGTTGCGTTG...</td>\n",
       "      <td>AGCAGAGTGCAGCGGGTGACAGACAGCCTGGACACACTGCTGGAGA...</td>\n",
       "      <td>AG2-25</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>1.404425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>53</td>\n",
       "      <td>No.38</td>\n",
       "      <td>v2</td>\n",
       "      <td>EcoRI</td>\n",
       "      <td>GAATTC</td>\n",
       "      <td>TAATACGACTCACTATA</td>\n",
       "      <td>GGA</td>\n",
       "      <td>65_1_ENSG00000163220</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>AAACACTCTGTGTGGCTCCTCGGCTTTGACAGAGTGCAAGACG</td>\n",
       "      <td>ENSG00000163220</td>\n",
       "      <td>ENSG00000163220</td>\n",
       "      <td>False</td>\n",
       "      <td>GGAAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>degron-No.5</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>1.389417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 experiment_id experiment_batch experiment_enzyme enzyme_seq  \\\n",
       "0          47            14               v1             BamHI     GGATCC   \n",
       "1          64            14               v1             BamHI     GGATCC   \n",
       "2          86         No.39               v2             EcoRI     GAATTC   \n",
       "3          88         No.41               v2             EcoRI     GAATTC   \n",
       "4          53         No.38               v2             EcoRI     GAATTC   \n",
       "\n",
       "    sequence_adapter cap_seq               seq_name  \\\n",
       "0  TAATACGACTCACTATA     GGG  template@Neocura_5UTR   \n",
       "1  TAATACGACTCACTATA     GGG  template@Neocura_5UTR   \n",
       "2  TAATACGACTCACTATA     GGA   69_1_ENSG00000163220   \n",
       "3  TAATACGACTCACTATA     GGA  171_1_ENSG00000163220   \n",
       "4  TAATACGACTCACTATA     GGA   65_1_ENSG00000163220   \n",
       "\n",
       "                                                 seq  \\\n",
       "0  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "1  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2  AAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCTCTT...   \n",
       "3  ACTTCTCGAGTTTTGGTGTTTTGTCGGTTCCGCTGTTGCGTTGTGG...   \n",
       "4  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "\n",
       "                                        template_seq    template_name  \\\n",
       "0  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     Neocura_5UTR   \n",
       "1  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     Neocura_5UTR   \n",
       "2        AAACACTCTGTGTGGCTCCTCGGCTTTGACAGAGTGCAAGACG  ENSG00000163220   \n",
       "3        AAACACTCTGTGTGGCTCCTCGGCTTTGACAGAGTGCAAGACG  ENSG00000163220   \n",
       "4        AAACACTCTGTGTGGCTCCTCGGCTTTGACAGAGTGCAAGACG  ENSG00000163220   \n",
       "\n",
       "              gene  is_template  \\\n",
       "0              NaN         True   \n",
       "1              NaN         True   \n",
       "2  ENSG00000163220        False   \n",
       "3  ENSG00000163220        False   \n",
       "4  ENSG00000163220        False   \n",
       "\n",
       "                                                UTR5  \\\n",
       "0  GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1  GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2  GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...   \n",
       "3  GGAACTTCTCGAGTTTTGGTGTTTTGTCGGTTCCGCTGTTGCGTTG...   \n",
       "4  GGAAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...   \n",
       "\n",
       "                                                 CDS      CDS_name  \\\n",
       "0  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  degron-No.15   \n",
       "1  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...  degron-No.22   \n",
       "2  AGCAGAGTGCAGCGGGTGACAGACAGCCTGGACACACTGCTGGAGA...        AG2-25   \n",
       "3  AGCAGAGTGCAGCGGGTGACAGACAGCCTGGACACACTGCTGGAGA...        AG2-25   \n",
       "4  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   degron-No.5   \n",
       "\n",
       "                                                UTR3  expression  \n",
       "0  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...    1.650000  \n",
       "1  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...    1.500000  \n",
       "2  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...    1.433288  \n",
       "3  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...    1.404425  \n",
       "4  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...    1.389417  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 读取 湿实验数据\n",
    "\n",
    "experiment_exp = pd.read_excel(\n",
    "    \n",
    "    '/home/yul/project/nas_project/bioSPT2/data/interiem/processed/UTR5_experiment_validation_results_summary.xlsx'\n",
    ")\n",
    "\n",
    "experiment_exp.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "52ac7d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "experiment_exp_agg = experiment_exp.groupby(by=['UTR5', 'CDS', 'UTR3']).agg(list)\n",
    "agged_cols = list(experiment_exp_agg.columns)\n",
    "experiment_exp_agg['gene_id'] = [f'gene_{g}' for g in range(experiment_exp_agg.shape[0])]\n",
    "\n",
    "experiment_exp_agg_exploded = experiment_exp_agg.explode(agged_cols)\n",
    "\n",
    "experiment_exp_agg_exploded = experiment_exp_agg_exploded.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f0838ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "pairs = []\n",
    "cutoff  = 0.3\n",
    "for i_index, i_row in experiment_exp_agg_exploded.iterrows():\n",
    "    for j_index, j_row in experiment_exp_agg_exploded.iterrows():\n",
    "    \n",
    "        if i_row['expression'] and j_row['expression'] and \\\n",
    "            i_row['gene_id'] != j_row['gene_id'] and \\\n",
    "                i_row['CDS'] == j_row['CDS']:\n",
    "            pair = [i_row['gene_id'], j_row['gene_id']]\n",
    "            if i_row['expression'] - j_row['expression'] >= cutoff:\n",
    "                pair.append(1)\n",
    "                \n",
    "            elif i_row['expression'] - j_row['expression'] <= -cutoff:\n",
    "                pair.append(-1)\n",
    "                \n",
    "            else:\n",
    "                pair.append(0)\n",
    "                \n",
    "            pairs.append(pair)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab3d0d1a-9f26-403e-8d73-fb47c65e1b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 配对\n",
    "pairs = []\n",
    "cutoff  = 0 # 0.3\n",
    "for i_index, i_row in experiment_exp_agg_exploded.iterrows():\n",
    "    for j_index, j_row in experiment_exp_agg_exploded.iterrows():\n",
    "    \n",
    "        if i_row['expression'] and j_row['expression'] and \\\n",
    "            i_row['gene_id'] != j_row['gene_id'] and \\\n",
    "                i_row['CDS'] == j_row['CDS']:\n",
    "            pair = [i_row['gene_id'], j_row['gene_id']]\n",
    "            if i_row['expression']/j_row['expression'] >= (1+cutoff):\n",
    "                pair.append(1)\n",
    "                \n",
    "            elif i_row['expression']/j_row['expression'] <= 1/(1+cutoff):\n",
    "                pair.append(-1)\n",
    "                \n",
    "            else:\n",
    "                pair.append(0)\n",
    "                \n",
    "            pair.append(i_row['expression']/j_row['expression'])\n",
    "            \n",
    "            pairs.append(pair)\n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4b7a7bbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8780"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# experiment_exp_agg.reset_index().set_index('gene_id').reindex(['gene_0'])['UTR5'].values[0]\n",
    "len(pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "03149009",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_A_UTR5</th>\n",
       "      <th>pair_A_UTR3</th>\n",
       "      <th>pair_A_CDS</th>\n",
       "      <th>pair_B_UTR5</th>\n",
       "      <th>pair_B_UTR3</th>\n",
       "      <th>pair_B_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_52</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_56</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_60</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAATCAACACAGACACCAGACCACCAGCACCAGAACACACAGAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_64</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAATCACCACCAGAACACACACACAGAAACACACAGACACCAGA...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_68</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAATTCGTCGACTCAGGAACTATTCTTCTGGTCCCCACAGACTC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8775</th>\n",
       "      <td>gene_115</td>\n",
       "      <td>gene_95</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGATTCCCCCACCACCAGAGCACCAGTCCCGCAGTCGGCGTCCAG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8776</th>\n",
       "      <td>gene_115</td>\n",
       "      <td>gene_99</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGATTCCCCGCCTCCTCCTGTCCCGCAGTCGGCGTCCAGCGGCTC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8777</th>\n",
       "      <td>gene_115</td>\n",
       "      <td>gene_103</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGGCCACCAAGTTGTTACATGAAATCTGCAGTTTCATAATTTCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8778</th>\n",
       "      <td>gene_115</td>\n",
       "      <td>gene_107</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGGCCAGCAGAAGTTGTTACAAATCTGCAGTTTCATTTCCGTGGG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8779</th>\n",
       "      <td>gene_115</td>\n",
       "      <td>gene_111</td>\n",
       "      <td>0</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGGCCCAACACACACCAGACCACCAGAGCAGACCAAGTTGTTACA...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8780 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        pair_A    pair_B  pair_label  \\\n",
       "0       gene_0   gene_52           1   \n",
       "1       gene_0   gene_56           1   \n",
       "2       gene_0   gene_60           1   \n",
       "3       gene_0   gene_64           1   \n",
       "4       gene_0   gene_68           1   \n",
       "...        ...       ...         ...   \n",
       "8775  gene_115   gene_95           0   \n",
       "8776  gene_115   gene_99           0   \n",
       "8777  gene_115  gene_103           1   \n",
       "8778  gene_115  gene_107           1   \n",
       "8779  gene_115  gene_111           0   \n",
       "\n",
       "                                            pair_A_UTR5  \\\n",
       "0     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "3     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "4     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "...                                                 ...   \n",
       "8775  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "8776  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "8777  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "8778  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "8779  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "\n",
       "                                            pair_A_UTR3  \\\n",
       "0     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "...                                                 ...   \n",
       "8775  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8776  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8777  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8778  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8779  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                             pair_A_CDS  \\\n",
       "0     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "1     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "2     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "3     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "4     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "...                                                 ...   \n",
       "8775  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "8776  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "8777  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "8778  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "8779  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "\n",
       "                                            pair_B_UTR5  \\\n",
       "0     GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...   \n",
       "1     GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2     GGGAATCAACACAGACACCAGACCACCAGCACCAGAACACACAGAC...   \n",
       "3     GGGAATCACCACCAGAACACACACACAGAAACACACAGACACCAGA...   \n",
       "4     GGGAATTCGTCGACTCAGGAACTATTCTTCTGGTCCCCACAGACTC...   \n",
       "...                                                 ...   \n",
       "8775  GGGATTCCCCCACCACCAGAGCACCAGTCCCGCAGTCGGCGTCCAG...   \n",
       "8776  GGGATTCCCCGCCTCCTCCTGTCCCGCAGTCGGCGTCCAGCGGCTC...   \n",
       "8777  GGGGCCACCAAGTTGTTACATGAAATCTGCAGTTTCATAATTTCCG...   \n",
       "8778  GGGGCCAGCAGAAGTTGTTACAAATCTGCAGTTTCATTTCCGTGGG...   \n",
       "8779  GGGGCCCAACACACACCAGACCACCAGAGCAGACCAAGTTGTTACA...   \n",
       "\n",
       "                                            pair_B_UTR3  \\\n",
       "0     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "...                                                 ...   \n",
       "8775  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8776  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8777  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8778  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8779  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                             pair_B_CDS  \n",
       "0     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "1     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "2     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "3     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "4     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "...                                                 ...  \n",
       "8775  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "8776  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "8777  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "8778  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "8779  ATGGTGTCTAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "\n",
       "[8780 rows x 9 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "experiment_pair_exp = pd.DataFrame(pairs, columns=['pair_A', 'pair_B', 'pair_label', 'pair_diff'])\n",
    "\n",
    "experiment_pair_exp['pair_A_UTR5'] = experiment_pair_exp['pair_A'].apply(lambda x: experiment_exp_agg.reset_index().set_index('gene_id').reindex([x])['UTR5'].values[0])\n",
    "experiment_pair_exp['pair_A_UTR3'] = experiment_pair_exp['pair_A'].apply(lambda x: experiment_exp_agg.reset_index().set_index('gene_id').reindex([x])['UTR3'].values[0])\n",
    "experiment_pair_exp['pair_A_CDS'] = experiment_pair_exp['pair_A'].apply(lambda x: experiment_exp_agg.reset_index().set_index('gene_id').reindex([x])['CDS'].values[0])\n",
    "\n",
    "experiment_pair_exp['pair_B_UTR5'] = experiment_pair_exp['pair_B'].apply(lambda x: experiment_exp_agg.reset_index().set_index('gene_id').reindex([x])['UTR5'].values[0])\n",
    "experiment_pair_exp['pair_B_UTR3'] = experiment_pair_exp['pair_B'].apply(lambda x: experiment_exp_agg.reset_index().set_index('gene_id').reindex([x])['UTR3'].values[0])\n",
    "experiment_pair_exp['pair_B_CDS'] = experiment_pair_exp['pair_B'].apply(lambda x: experiment_exp_agg.reset_index().set_index('gene_id').reindex([x])['CDS'].values[0])\n",
    "\n",
    "\n",
    "experiment_pair_exp\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "332c0835",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "experiment_pair_exp.to_csv(f'data/interiem/processed/experiment_UTR5_pair_exp_cutoff{cutoff}.tsv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "768345e3",
   "metadata": {},
   "source": [
    "#### UTR3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "00867b1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>No.</th>\n",
       "      <th>ID-Wet</th>\n",
       "      <th>Name-Wet</th>\n",
       "      <th>ID-Dry</th>\n",
       "      <th>Note-Dry</th>\n",
       "      <th>3-nt</th>\n",
       "      <th>5'UTR</th>\n",
       "      <th>CDS</th>\n",
       "      <th>3'UTR</th>\n",
       "      <th>5‘UTR-CDS</th>\n",
       "      <th>...</th>\n",
       "      <th>NA4.2</th>\n",
       "      <th>Test1.2</th>\n",
       "      <th>Test2.2</th>\n",
       "      <th>Test3.2</th>\n",
       "      <th>Test4.2</th>\n",
       "      <th>Ref1.2</th>\n",
       "      <th>Ref2.2</th>\n",
       "      <th>Ref3.2</th>\n",
       "      <th>Ref4.2</th>\n",
       "      <th>Ref seq (Name-Wet).2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3'UTR-1</td>\n",
       "      <td>pNC-eGFPdegron5-3'UTR-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BNT_1</td>\n",
       "      <td>GGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>CTGGTACTGGAATAAGCACGAATTAATGCTAGCTGCCCCTAAAGCT...</td>\n",
       "      <td>ggatcc</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>3'UTR-2</td>\n",
       "      <td>pNC-eGFPdegron5-3'UTR-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BNT_2</td>\n",
       "      <td>GGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>CTGGTACTGCCCCTGCTGCACCGCCATGCTAGCTGGCTGTCCCGTC...</td>\n",
       "      <td>ggatcc</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>3'UTR-3</td>\n",
       "      <td>pNC-eGFPdegron5-3'UTR-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BNT_3</td>\n",
       "      <td>GGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>CTGGGCAGTAAAGGCACGCAATGCTTGCTGCTGCCCCTCCTGTGCG...</td>\n",
       "      <td>ggatcc</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>3'UTR-4</td>\n",
       "      <td>pNC-eGFPdegron5-3'UTR-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BNT_4</td>\n",
       "      <td>GGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>CTTCCGCCGTACTGCATCCTGGCACGCAATGCTAGCTGCCCCTTTC...</td>\n",
       "      <td>ggatcc</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>3'UTR-5</td>\n",
       "      <td>pNC-eGFPdegron5-3'UTR-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Moderna_1</td>\n",
       "      <td>GGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GCTTCGTGTCTTCTGTGGCCTAGCTTCCCTGCTCTTGGGCCTCCCC...</td>\n",
       "      <td>ggatcc</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   No.   ID-Wet                 Name-Wet  ID-Dry   Note-Dry 3-nt  \\\n",
       "0    1  3'UTR-1  pNC-eGFPdegron5-3'UTR-1     NaN      BNT_1  GGG   \n",
       "1    2  3'UTR-2  pNC-eGFPdegron5-3'UTR-2     NaN      BNT_2  GGG   \n",
       "2    3  3'UTR-3  pNC-eGFPdegron5-3'UTR-3     NaN      BNT_3  GGG   \n",
       "3    4  3'UTR-4  pNC-eGFPdegron5-3'UTR-4     NaN      BNT_4  GGG   \n",
       "4    5  3'UTR-5  pNC-eGFPdegron5-3'UTR-5     NaN  Moderna_1  GGG   \n",
       "\n",
       "                                               5'UTR  \\\n",
       "0  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "1  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "\n",
       "                                                 CDS  \\\n",
       "0  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "1  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "2  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "3  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "4  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "\n",
       "                                               3'UTR 5‘UTR-CDS  ... NA4.2  \\\n",
       "0  CTGGTACTGGAATAAGCACGAATTAATGCTAGCTGCCCCTAAAGCT...    ggatcc  ...   NaN   \n",
       "1  CTGGTACTGCCCCTGCTGCACCGCCATGCTAGCTGGCTGTCCCGTC...    ggatcc  ...   NaN   \n",
       "2  CTGGGCAGTAAAGGCACGCAATGCTTGCTGCTGCCCCTCCTGTGCG...    ggatcc  ...   NaN   \n",
       "3  CTTCCGCCGTACTGCATCCTGGCACGCAATGCTAGCTGCCCCTTTC...    ggatcc  ...   NaN   \n",
       "4  GCTTCGTGTCTTCTGTGGCCTAGCTTCCCTGCTCTTGGGCCTCCCC...    ggatcc  ...   NaN   \n",
       "\n",
       "  Test1.2 Test2.2 Test3.2 Test4.2 Ref1.2  Ref2.2  Ref3.2  Ref4.2  \\\n",
       "0     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "1     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "2     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "3     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "4     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "\n",
       "   Ref seq (Name-Wet).2  \n",
       "0                   NaN  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4                   NaN  \n",
       "\n",
       "[5 rows x 60 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "experiment_UTR3_exp = pd.read_excel(\n",
    "    \n",
    "    '/home/yul/project/nas_project/bioSPT2/data/interiem/processed/Neocura_3UTR_validation_results_062024.xlsx', header=1\n",
    ")\n",
    "\n",
    "experiment_UTR3_exp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c3471e5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['No.', 'ID-Wet', 'Name-Wet', 'ID-Dry', 'Note-Dry', '3-nt', '5'UTR',\n",
       "       'CDS', '3'UTR', '5‘UTR-CDS', 'CDS-3‘UTR', '3‘UTR-polyA', 'polyA',\n",
       "       'Linearization Enzyme', 'Comeplete Seq', 'Cell-line', 'Test Date',\n",
       "       'NA1', 'NA2', 'NA3', 'NA4', 'Test1', 'Test2', 'Test3', 'Test4', 'Ref1',\n",
       "       'Ref2', 'Ref3', 'Ref4', 'Ref seq (Name-Wet)', 'Cell-line.1',\n",
       "       'Test Date.1', 'NA1.1', 'NA2.1', 'NA3.1', 'NA4.1', 'Test1.1', 'Test2.1',\n",
       "       'Test3.1', 'Test4.1', 'Ref1.1', 'Ref2.1', 'Ref3.1', 'Ref4.1',\n",
       "       'Ref seq (Name-Wet).1', 'Cell-line.2', 'Test Date.2', 'NA1.2', 'NA2.2',\n",
       "       'NA3.2', 'NA4.2', 'Test1.2', 'Test2.2', 'Test3.2', 'Test4.2', 'Ref1.2',\n",
       "       'Ref2.2', 'Ref3.2', 'Ref4.2', 'Ref seq (Name-Wet).2', 'mean_exp'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "agged_cols = experiment_UTR3_exp.columns\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a35751b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NA1</th>\n",
       "      <th>NA2</th>\n",
       "      <th>NA3</th>\n",
       "      <th>NA4</th>\n",
       "      <th>Test1</th>\n",
       "      <th>Test2</th>\n",
       "      <th>Test3</th>\n",
       "      <th>Test4</th>\n",
       "      <th>NA1.1</th>\n",
       "      <th>NA2.1</th>\n",
       "      <th>...</th>\n",
       "      <th>Test3.1</th>\n",
       "      <th>Test4.1</th>\n",
       "      <th>NA1.2</th>\n",
       "      <th>NA2.2</th>\n",
       "      <th>NA3.2</th>\n",
       "      <th>NA4.2</th>\n",
       "      <th>Test1.2</th>\n",
       "      <th>Test2.2</th>\n",
       "      <th>Test3.2</th>\n",
       "      <th>Test4.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>3202.0</td>\n",
       "      <td>3017.0</td>\n",
       "      <td>3540.0</td>\n",
       "      <td>3453.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>7390.0</td>\n",
       "      <td>7050.0</td>\n",
       "      <td>8280.0</td>\n",
       "      <td>8091.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>10709.0</td>\n",
       "      <td>9974.0</td>\n",
       "      <td>10965.0</td>\n",
       "      <td>9778.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8801.0</td>\n",
       "      <td>7830.0</td>\n",
       "      <td>9147.0</td>\n",
       "      <td>8806.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>12652.0</td>\n",
       "      <td>12386.0</td>\n",
       "      <td>13643.0</td>\n",
       "      <td>12911.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>15304.0</td>\n",
       "      <td>15034.0</td>\n",
       "      <td>14655.0</td>\n",
       "      <td>15721.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>21254.0</td>\n",
       "      <td>19505.0</td>\n",
       "      <td>21539.0</td>\n",
       "      <td>20411.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>15648.0</td>\n",
       "      <td>15528.0</td>\n",
       "      <td>17036.0</td>\n",
       "      <td>16464.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>19017.0</td>\n",
       "      <td>18885.0</td>\n",
       "      <td>19263.0</td>\n",
       "      <td>18886.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>9857.0</td>\n",
       "      <td>9756.0</td>\n",
       "      <td>11010.0</td>\n",
       "      <td>11626.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>11325.0</td>\n",
       "      <td>10841.0</td>\n",
       "      <td>11982.0</td>\n",
       "      <td>13109.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>13119.0</td>\n",
       "      <td>12626.0</td>\n",
       "      <td>13321.0</td>\n",
       "      <td>12366.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>12571.0</td>\n",
       "      <td>12485.0</td>\n",
       "      <td>14079.0</td>\n",
       "      <td>14863.0</td>\n",
       "      <td>8.66</td>\n",
       "      <td>8.66</td>\n",
       "      <td>...</td>\n",
       "      <td>12395.0</td>\n",
       "      <td>12005.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>16249.0</td>\n",
       "      <td>17704.0</td>\n",
       "      <td>19117.0</td>\n",
       "      <td>19491.0</td>\n",
       "      <td>8.66</td>\n",
       "      <td>8.66</td>\n",
       "      <td>...</td>\n",
       "      <td>16034.0</td>\n",
       "      <td>17524.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>14163.0</td>\n",
       "      <td>14613.0</td>\n",
       "      <td>15219.0</td>\n",
       "      <td>14986.0</td>\n",
       "      <td>8.66</td>\n",
       "      <td>8.66</td>\n",
       "      <td>...</td>\n",
       "      <td>12899.0</td>\n",
       "      <td>13200.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>12746.0</td>\n",
       "      <td>13240.0</td>\n",
       "      <td>12712.0</td>\n",
       "      <td>14407.0</td>\n",
       "      <td>8.66</td>\n",
       "      <td>8.66</td>\n",
       "      <td>...</td>\n",
       "      <td>11647.0</td>\n",
       "      <td>12249.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>8.31</td>\n",
       "      <td>9574.0</td>\n",
       "      <td>10914.0</td>\n",
       "      <td>11712.0</td>\n",
       "      <td>11222.0</td>\n",
       "      <td>8.66</td>\n",
       "      <td>8.66</td>\n",
       "      <td>...</td>\n",
       "      <td>10047.0</td>\n",
       "      <td>10205.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.66</td>\n",
       "      <td>8.66</td>\n",
       "      <td>...</td>\n",
       "      <td>10596.0</td>\n",
       "      <td>11241.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.66</td>\n",
       "      <td>8.66</td>\n",
       "      <td>...</td>\n",
       "      <td>10273.0</td>\n",
       "      <td>10392.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.66</td>\n",
       "      <td>8.66</td>\n",
       "      <td>...</td>\n",
       "      <td>6406.0</td>\n",
       "      <td>5711.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     NA1   NA2   NA3   NA4    Test1    Test2    Test3    Test4  NA1.1  NA2.1  \\\n",
       "0   8.31  8.31  8.31  8.31   3202.0   3017.0   3540.0   3453.0    NaN    NaN   \n",
       "1   8.31  8.31  8.31  8.31   7390.0   7050.0   8280.0   8091.0    NaN    NaN   \n",
       "2   8.31  8.31  8.31  8.31  10709.0   9974.0  10965.0   9778.0    NaN    NaN   \n",
       "3   8.31  8.31  8.31  8.31   8801.0   7830.0   9147.0   8806.0    NaN    NaN   \n",
       "4   8.31  8.31  8.31  8.31  12652.0  12386.0  13643.0  12911.0    NaN    NaN   \n",
       "5   8.31  8.31  8.31  8.31  15304.0  15034.0  14655.0  15721.0    NaN    NaN   \n",
       "6   8.31  8.31  8.31  8.31  21254.0  19505.0  21539.0  20411.0    NaN    NaN   \n",
       "7   8.31  8.31  8.31  8.31  15648.0  15528.0  17036.0  16464.0    NaN    NaN   \n",
       "8   8.31  8.31  8.31  8.31  19017.0  18885.0  19263.0  18886.0    NaN    NaN   \n",
       "9   8.31  8.31  8.31  8.31   9857.0   9756.0  11010.0  11626.0    NaN    NaN   \n",
       "10  8.31  8.31  8.31  8.31  11325.0  10841.0  11982.0  13109.0    NaN    NaN   \n",
       "11  8.31  8.31  8.31  8.31  13119.0  12626.0  13321.0  12366.0    NaN    NaN   \n",
       "12  8.31  8.31  8.31  8.31  12571.0  12485.0  14079.0  14863.0   8.66   8.66   \n",
       "13  8.31  8.31  8.31  8.31  16249.0  17704.0  19117.0  19491.0   8.66   8.66   \n",
       "14  8.31  8.31  8.31  8.31  14163.0  14613.0  15219.0  14986.0   8.66   8.66   \n",
       "15  8.31  8.31  8.31  8.31  12746.0  13240.0  12712.0  14407.0   8.66   8.66   \n",
       "16  8.31  8.31  8.31  8.31   9574.0  10914.0  11712.0  11222.0   8.66   8.66   \n",
       "17   NaN   NaN   NaN   NaN      NaN      NaN      NaN      NaN   8.66   8.66   \n",
       "18   NaN   NaN   NaN   NaN      NaN      NaN      NaN      NaN   8.66   8.66   \n",
       "19   NaN   NaN   NaN   NaN      NaN      NaN      NaN      NaN   8.66   8.66   \n",
       "\n",
       "    ...  Test3.1  Test4.1  NA1.2  NA2.2  NA3.2  NA4.2  Test1.2  Test2.2  \\\n",
       "0   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "1   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "2   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "3   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "4   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "5   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "6   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "7   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "8   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "9   ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "10  ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "11  ...      NaN      NaN    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "12  ...  12395.0  12005.0    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "13  ...  16034.0  17524.0    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "14  ...  12899.0  13200.0    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "15  ...  11647.0  12249.0    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "16  ...  10047.0  10205.0    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "17  ...  10596.0  11241.0    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "18  ...  10273.0  10392.0    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "19  ...   6406.0   5711.0    NaN    NaN    NaN    NaN      NaN      NaN   \n",
       "\n",
       "    Test3.2  Test4.2  \n",
       "0       NaN      NaN  \n",
       "1       NaN      NaN  \n",
       "2       NaN      NaN  \n",
       "3       NaN      NaN  \n",
       "4       NaN      NaN  \n",
       "5       NaN      NaN  \n",
       "6       NaN      NaN  \n",
       "7       NaN      NaN  \n",
       "8       NaN      NaN  \n",
       "9       NaN      NaN  \n",
       "10      NaN      NaN  \n",
       "11      NaN      NaN  \n",
       "12      NaN      NaN  \n",
       "13      NaN      NaN  \n",
       "14      NaN      NaN  \n",
       "15      NaN      NaN  \n",
       "16      NaN      NaN  \n",
       "17      NaN      NaN  \n",
       "18      NaN      NaN  \n",
       "19      NaN      NaN  \n",
       "\n",
       "[20 rows x 24 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "experiment_UTR3_exp[\n",
    "    ['NA1', 'NA2', 'NA3', 'NA4',\n",
    "     'Test1', 'Test2', 'Test3', 'Test4',\n",
    "     'NA1.1', 'NA2.1', 'NA3.1', 'NA4.1', 'Test1.1', 'Test2.1',\n",
    "       'Test3.1', 'Test4.1',\n",
    "     'NA1.2', 'NA2.2', 'NA3.2', 'NA4.2',\n",
    "     'Test1.2', 'Test2.2', 'Test3.2', 'Test4.2',\n",
    "     ]\n",
    "                    ].head(20)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d0c7fecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "experiment_UTR3_exp['mean_exp'] =  experiment_UTR3_exp[\n",
    "    [\n",
    "     'Test1', 'Test2', 'Test3', 'Test4',\n",
    "      'Test1.1', 'Test2.1','Test3.1', 'Test4.1',\n",
    "     'Test1.2', 'Test2.2', 'Test3.2', 'Test4.2',\n",
    "     ]\n",
    "                    ].mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "af8a9034",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3-nt</th>\n",
       "      <th>5'UTR</th>\n",
       "      <th>CDS</th>\n",
       "      <th>3'UTR</th>\n",
       "      <th>5‘UTR-CDS</th>\n",
       "      <th>CDS-3‘UTR</th>\n",
       "      <th>No.</th>\n",
       "      <th>ID-Wet</th>\n",
       "      <th>Name-Wet</th>\n",
       "      <th>ID-Dry</th>\n",
       "      <th>...</th>\n",
       "      <th>Test2.2</th>\n",
       "      <th>Test3.2</th>\n",
       "      <th>Test4.2</th>\n",
       "      <th>Ref1.2</th>\n",
       "      <th>Ref2.2</th>\n",
       "      <th>Ref3.2</th>\n",
       "      <th>Ref4.2</th>\n",
       "      <th>Ref seq (Name-Wet).2</th>\n",
       "      <th>mean_exp</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>151</td>\n",
       "      <td>3'UTR-58</td>\n",
       "      <td>018-3‘UTR-58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1288.2</td>\n",
       "      <td>gene_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>ACTGTACCATTTACCACTGTGACTAAATAAAGGGTGTGTCTGTTGT...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>157</td>\n",
       "      <td>3'UTR-64</td>\n",
       "      <td>018-3‘UTR-64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1209.2</td>\n",
       "      <td>gene_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>CCCCCAATATCTGCTCCTCTGGTTTGGCTTCCCTCAGAGCCTGGGT...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>124</td>\n",
       "      <td>3'UTR-31</td>\n",
       "      <td>018-3‘UTR-31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3076.6</td>\n",
       "      <td>gene_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>CCCCGTCTGCTCACCTGCAAAATAAAAGGCTTTTCGCCCCTTGTCC...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>126</td>\n",
       "      <td>3'UTR-33</td>\n",
       "      <td>018-3‘UTR-33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2961.4</td>\n",
       "      <td>gene_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>CCCCGTCTGCTCCTCTAGGCTGGCCCCTTGTCCACCCCTCTCCACA...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>138</td>\n",
       "      <td>3'UTR-45</td>\n",
       "      <td>018-3‘UTR-45</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2973.0</td>\n",
       "      <td>gene_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  3-nt                                              5'UTR  \\\n",
       "0  AGG  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "1  AGG  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2  AGG  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3  AGG  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4  AGG  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "\n",
       "                                                 CDS  \\\n",
       "0  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "1  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "2  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "3  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "4  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "\n",
       "                                               3'UTR 5‘UTR-CDS CDS-3‘UTR  No.  \\\n",
       "0  ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...         -         -  151   \n",
       "1  ACTGTACCATTTACCACTGTGACTAAATAAAGGGTGTGTCTGTTGT...         -         -  157   \n",
       "2  CCCCCAATATCTGCTCCTCTGGTTTGGCTTCCCTCAGAGCCTGGGT...         -         -  124   \n",
       "3  CCCCGTCTGCTCACCTGCAAAATAAAAGGCTTTTCGCCCCTTGTCC...         -         -  126   \n",
       "4  CCCCGTCTGCTCCTCTAGGCTGGCCCCTTGTCCACCCCTCTCCACA...         -         -  138   \n",
       "\n",
       "     ID-Wet      Name-Wet ID-Dry  ... Test2.2 Test3.2 Test4.2 Ref1.2 Ref2.2  \\\n",
       "0  3'UTR-58  018-3‘UTR-58    NaN  ...     NaN     NaN     NaN    NaN    NaN   \n",
       "1  3'UTR-64  018-3‘UTR-64    NaN  ...     NaN     NaN     NaN    NaN    NaN   \n",
       "2  3'UTR-31  018-3‘UTR-31    NaN  ...     NaN     NaN     NaN    NaN    NaN   \n",
       "3  3'UTR-33  018-3‘UTR-33    NaN  ...     NaN     NaN     NaN    NaN    NaN   \n",
       "4  3'UTR-45  018-3‘UTR-45    NaN  ...     NaN     NaN     NaN    NaN    NaN   \n",
       "\n",
       "  Ref3.2 Ref4.2 Ref seq (Name-Wet).2 mean_exp gene_id  \n",
       "0    NaN    NaN                  NaN   1288.2  gene_0  \n",
       "1    NaN    NaN                  NaN   1209.2  gene_1  \n",
       "2    NaN    NaN                  NaN   3076.6  gene_2  \n",
       "3    NaN    NaN                  NaN   2961.4  gene_3  \n",
       "4    NaN    NaN                  NaN   2973.0  gene_4  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "experiment_UTR3_exp_agg = experiment_UTR3_exp.groupby(by=['3-nt', \"5'UTR\", 'CDS', \"3'UTR\", '5‘UTR-CDS', 'CDS-3‘UTR']).agg(list)\n",
    "\n",
    "agged_cols = list(experiment_UTR3_exp_agg.columns)\n",
    "\n",
    "experiment_UTR3_exp_agg['gene_id'] = [f'gene_{g}' for g in range(experiment_UTR3_exp_agg.shape[0])]\n",
    "\n",
    "experiment_UTR3_exp_agg_exploded = experiment_UTR3_exp_agg.explode(agged_cols)\n",
    "\n",
    "experiment_UTR3_exp_agg_exploded = experiment_UTR3_exp_agg_exploded.reset_index()\n",
    "\n",
    "experiment_UTR3_exp_agg_exploded.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "216f3181",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "pairs = []\n",
    "for i_index, i_row in experiment_UTR3_exp_agg_exploded.iterrows():\n",
    "    for j_index, j_row in experiment_UTR3_exp_agg_exploded.iterrows():\n",
    "    \n",
    "        if i_row['mean_exp'] and j_row['mean_exp'] and \\\n",
    "            i_row['gene_id'] != j_row['gene_id'] and \\\n",
    "                i_row['CDS'] == j_row['CDS']:\n",
    "      \n",
    "            pair = [i_row['gene_id'], j_row['gene_id']]\n",
    "            if i_row['mean_exp']/j_row['mean_exp'] >= (1+cutoff):\n",
    "                pair.append(1)\n",
    "                \n",
    "            elif i_row['mean_exp']/j_row['mean_exp'] <= 1/(1+cutoff):\n",
    "                pair.append(-1)\n",
    "                \n",
    "            else:\n",
    "                pair.append(0)\n",
    "                \n",
    "            pair.append(i_row['mean_exp']/j_row['mean_exp'])\n",
    "                \n",
    "            pairs.append(pair)\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "68a90aeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6846"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "988c0fec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_A_UTR5</th>\n",
       "      <th>pair_A_UTR3</th>\n",
       "      <th>pair_A_CDS</th>\n",
       "      <th>pair_B_UTR5</th>\n",
       "      <th>pair_B_UTR3</th>\n",
       "      <th>pair_B_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>0</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTGTACCATTTACCACTGTGACTAAATAAAGGGTGTGTCTGTTGT...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_2</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CCCCCAATATCTGCTCCTCTGGTTTGGCTTCCCTCAGAGCCTGGGT...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_3</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CCCCGTCTGCTCACCTGCAAAATAAAAGGCTTTTCGCCCCTTGTCC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_4</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CCCCGTCTGCTCCTCTAGGCTGGCCCCTTGTCCACCCCTCTCCACA...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_5</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CCCCTTCTGTTCTGCTCCTCTAATAAAGGCTCTGATTTTCCCCCAT...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6841</th>\n",
       "      <td>gene_156</td>\n",
       "      <td>gene_151</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>GCTTCGTGTCTTCTGTGGCCTAGCTTCCCTGCTCTTGGGCCTCCCC...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6842</th>\n",
       "      <td>gene_156</td>\n",
       "      <td>gene_152</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTAACTCCTGCGTTAAATAAAAAGAAGGAACTCGTAAAAACTCT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6843</th>\n",
       "      <td>gene_156</td>\n",
       "      <td>gene_153</td>\n",
       "      <td>1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTAAAAACTAATCTT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6844</th>\n",
       "      <td>gene_156</td>\n",
       "      <td>gene_154</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTAAAAACTCAATGT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6845</th>\n",
       "      <td>gene_156</td>\n",
       "      <td>gene_155</td>\n",
       "      <td>-1</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTAGGTTGCTGATTT...</td>\n",
       "      <td>ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6846 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        pair_A    pair_B  pair_label  \\\n",
       "0       gene_0    gene_1           0   \n",
       "1       gene_0    gene_2          -1   \n",
       "2       gene_0    gene_3          -1   \n",
       "3       gene_0    gene_4          -1   \n",
       "4       gene_0    gene_5          -1   \n",
       "...        ...       ...         ...   \n",
       "6841  gene_156  gene_151          -1   \n",
       "6842  gene_156  gene_152          -1   \n",
       "6843  gene_156  gene_153           1   \n",
       "6844  gene_156  gene_154          -1   \n",
       "6845  gene_156  gene_155          -1   \n",
       "\n",
       "                                            pair_A_UTR5  \\\n",
       "0     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "1     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "...                                                 ...   \n",
       "6841  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "6842  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "6843  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "6844  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "6845  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "\n",
       "                                            pair_A_UTR3  \\\n",
       "0     ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "1     ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "2     ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "3     ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "4     ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "...                                                 ...   \n",
       "6841  TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...   \n",
       "6842  TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...   \n",
       "6843  TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...   \n",
       "6844  TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...   \n",
       "6845  TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTATTGTGTATGCGT...   \n",
       "\n",
       "                                             pair_A_CDS  \\\n",
       "0     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "1     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "2     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "3     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "4     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "...                                                 ...   \n",
       "6841  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...   \n",
       "6842  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...   \n",
       "6843  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...   \n",
       "6844  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...   \n",
       "6845  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...   \n",
       "\n",
       "                                            pair_B_UTR5  \\\n",
       "0     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "1     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "...                                                 ...   \n",
       "6841  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "6842  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "6843  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "6844  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "6845  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "\n",
       "                                            pair_B_UTR3  \\\n",
       "0     ACTGTACCATTTACCACTGTGACTAAATAAAGGGTGTGTCTGTTGT...   \n",
       "1     CCCCCAATATCTGCTCCTCTGGTTTGGCTTCCCTCAGAGCCTGGGT...   \n",
       "2     CCCCGTCTGCTCACCTGCAAAATAAAAGGCTTTTCGCCCCTTGTCC...   \n",
       "3     CCCCGTCTGCTCCTCTAGGCTGGCCCCTTGTCCACCCCTCTCCACA...   \n",
       "4     CCCCTTCTGTTCTGCTCCTCTAATAAAGGCTCTGATTTTCCCCCAT...   \n",
       "...                                                 ...   \n",
       "6841  GCTTCGTGTCTTCTGTGGCCTAGCTTCCCTGCTCTTGGGCCTCCCC...   \n",
       "6842  TTGTAACTCCTGCGTTAAATAAAAAGAAGGAACTCGTAAAAACTCT...   \n",
       "6843  TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTAAAAACTAATCTT...   \n",
       "6844  TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTAAAAACTCAATGT...   \n",
       "6845  TTGTGTATGCGTTAATAAAAAGAAGGAACTCGTAGGTTGCTGATTT...   \n",
       "\n",
       "                                             pair_B_CDS  \n",
       "0     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "1     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "2     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "3     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "4     ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "...                                                 ...  \n",
       "6841  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...  \n",
       "6842  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...  \n",
       "6843  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...  \n",
       "6844  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...  \n",
       "6845  ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATCC...  \n",
       "\n",
       "[6846 rows x 9 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "experiment_UTR3_pair_exp = pd.DataFrame(pairs, columns=['pair_A', 'pair_B', 'pair_label', 'pair_diff'])\n",
    "\n",
    "experiment_UTR3_pair_exp['pair_A_UTR5'] = experiment_UTR3_pair_exp['pair_A'].apply(lambda x: experiment_UTR3_exp_agg.reset_index().set_index('gene_id').reindex([x])[\"5'UTR\"].values[0])\n",
    "experiment_UTR3_pair_exp['pair_A_UTR3'] = experiment_UTR3_pair_exp['pair_A'].apply(lambda x: experiment_UTR3_exp_agg.reset_index().set_index('gene_id').reindex([x])[\"3'UTR\"].values[0])\n",
    "experiment_UTR3_pair_exp['pair_A_CDS'] = experiment_UTR3_pair_exp['pair_A'].apply(lambda x: experiment_UTR3_exp_agg.reset_index().set_index('gene_id').reindex([x])['CDS'].values[0])\n",
    "\n",
    "experiment_UTR3_pair_exp['pair_B_UTR5'] = experiment_UTR3_pair_exp['pair_B'].apply(lambda x: experiment_UTR3_exp_agg.reset_index().set_index('gene_id').reindex([x])[\"5'UTR\"].values[0])\n",
    "experiment_UTR3_pair_exp['pair_B_UTR3'] = experiment_UTR3_pair_exp['pair_B'].apply(lambda x: experiment_UTR3_exp_agg.reset_index().set_index('gene_id').reindex([x])[\"3'UTR\"].values[0])\n",
    "experiment_UTR3_pair_exp['pair_B_CDS'] = experiment_UTR3_pair_exp['pair_B'].apply(lambda x: experiment_UTR3_exp_agg.reset_index().set_index('gene_id').reindex([x])['CDS'].values[0])\n",
    "\n",
    "\n",
    "experiment_UTR3_pair_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8c60ca26",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "experiment_UTR3_pair_exp.to_csv(f'/data1/genomics/yul/bioSPT2/data/interiem/processed/experiment_UTR3_pair_exp_cutoff{cutoff}.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d489ddcc",
   "metadata": {},
   "source": [
    "#### CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49db9633",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "CDS bioSPT 项目实验反馈数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bb4a491e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>No.</th>\n",
       "      <th>ID-Wet</th>\n",
       "      <th>Name-Wet</th>\n",
       "      <th>ID-Dry</th>\n",
       "      <th>Note-Dry</th>\n",
       "      <th>3-nt</th>\n",
       "      <th>5'UTR</th>\n",
       "      <th>CDS</th>\n",
       "      <th>3'UTR</th>\n",
       "      <th>5‘UTR-CDS</th>\n",
       "      <th>...</th>\n",
       "      <th>NA4.2</th>\n",
       "      <th>Test1.2</th>\n",
       "      <th>Test2.2</th>\n",
       "      <th>Test3.2</th>\n",
       "      <th>Test4.2</th>\n",
       "      <th>Ref1.2</th>\n",
       "      <th>Ref2.2</th>\n",
       "      <th>Ref3.2</th>\n",
       "      <th>Ref4.2</th>\n",
       "      <th>Ref seq (Name-Wet).2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>CompleteSeq-1</td>\n",
       "      <td>pNeoCS-eGFPdegron-1</td>\n",
       "      <td>12.0</td>\n",
       "      <td>High_RSCU_cds1-1</td>\n",
       "      <td>AGG</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>GCAGCGAAGGACGCTACGAAGCTCAGTCATTACCAGGAAGCTCCCT...</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>CompleteSeq-2</td>\n",
       "      <td>pNeoCS-eGFPdegron-2</td>\n",
       "      <td>13.0</td>\n",
       "      <td>High_RSCU_cds1-2</td>\n",
       "      <td>AGG</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>GCTGGTCCCAATTCTGGTCGAGCTCGGTCAACGGACAGTTCTGTCC...</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>CompleteSeq-3</td>\n",
       "      <td>pNeoCS-eGFPdegron-3</td>\n",
       "      <td>14.0</td>\n",
       "      <td>High_RSCU_cds1-3</td>\n",
       "      <td>AGG</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>GCCCGAAGGACGCTACGAAGCTCAGTCATTACCAGGAAGCTCCCTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>CompleteSeq-4</td>\n",
       "      <td>pNeoCS-eGFPdegron-4</td>\n",
       "      <td>15.0</td>\n",
       "      <td>High_RSCU_cds2-1</td>\n",
       "      <td>AGG</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CompleteSeq-5</td>\n",
       "      <td>pNeoCS-eGFPdegron-5</td>\n",
       "      <td>16.0</td>\n",
       "      <td>High_RSCU_cds2-2</td>\n",
       "      <td>AGG</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGCTCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>TCGAGCTCGGACAGTTCTGTACCGTGCGGGCCAAGGGCGAAGCGTC...</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   No.         ID-Wet             Name-Wet  ID-Dry          Note-Dry 3-nt  \\\n",
       "0    1  CompleteSeq-1  pNeoCS-eGFPdegron-1    12.0  High_RSCU_cds1-1  AGG   \n",
       "1    2  CompleteSeq-2  pNeoCS-eGFPdegron-2    13.0  High_RSCU_cds1-2  AGG   \n",
       "2    3  CompleteSeq-3  pNeoCS-eGFPdegron-3    14.0  High_RSCU_cds1-3  AGG   \n",
       "3    4  CompleteSeq-4  pNeoCS-eGFPdegron-4    15.0  High_RSCU_cds2-1  AGG   \n",
       "4    5  CompleteSeq-5  pNeoCS-eGFPdegron-5    16.0  High_RSCU_cds2-2  AGG   \n",
       "\n",
       "                                               5'UTR  \\\n",
       "0  GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...   \n",
       "1  GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...   \n",
       "2  GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...   \n",
       "3  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "4  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGCTCG...   \n",
       "\n",
       "                                                 CDS  \\\n",
       "0  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "1  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "2  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "3  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "4  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "\n",
       "                                               3'UTR 5‘UTR-CDS  ... NA4.2  \\\n",
       "0  GCAGCGAAGGACGCTACGAAGCTCAGTCATTACCAGGAAGCTCCCT...         -  ...   NaN   \n",
       "1  GCTGGTCCCAATTCTGGTCGAGCTCGGTCAACGGACAGTTCTGTCC...         -  ...   NaN   \n",
       "2  GCCCGAAGGACGCTACGAAGCTCAGTCATTACCAGGAAGCTCCCTG...         -  ...   NaN   \n",
       "3  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...         -  ...   NaN   \n",
       "4  TCGAGCTCGGACAGTTCTGTACCGTGCGGGCCAAGGGCGAAGCGTC...         -  ...   NaN   \n",
       "\n",
       "  Test1.2 Test2.2 Test3.2 Test4.2 Ref1.2  Ref2.2  Ref3.2  Ref4.2  \\\n",
       "0     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "1     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "2     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "3     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "4     NaN     NaN     NaN     NaN    NaN     NaN     NaN     NaN   \n",
       "\n",
       "   Ref seq (Name-Wet).2  \n",
       "0                   NaN  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4                   NaN  \n",
       "\n",
       "[5 rows x 60 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "experiment_CDS_exp = pd.read_excel('/home/yul/project/nas_project/bioSPT2/data/interiem/processed/mRNA全序列优化_实验验证结果_03192024.xlsx', header=1)\n",
    "\n",
    "experiment_CDS_exp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "966c0a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "experiment_CDS_exp['mean_exp'] =  experiment_CDS_exp[\n",
    "    [\n",
    "     'Test1', 'Test2', 'Test3', 'Test4',\n",
    "      'Test1.1', 'Test2.1','Test3.1', 'Test4.1',\n",
    "     'Test1.2', 'Test2.2', 'Test3.2', 'Test4.2',\n",
    "     ]\n",
    "                    ].mean(axis=1)\n",
    "\n",
    "experiment_CDS_exp = experiment_CDS_exp[ \n",
    "                   ~experiment_CDS_exp['mean_exp'].isnull()\n",
    "                   ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "eb055b65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3-nt</th>\n",
       "      <th>5'UTR</th>\n",
       "      <th>CDS</th>\n",
       "      <th>3'UTR</th>\n",
       "      <th>5‘UTR-CDS</th>\n",
       "      <th>CDS-3‘UTR</th>\n",
       "      <th>No.</th>\n",
       "      <th>ID-Wet</th>\n",
       "      <th>Name-Wet</th>\n",
       "      <th>ID-Dry</th>\n",
       "      <th>...</th>\n",
       "      <th>Test2.2</th>\n",
       "      <th>Test3.2</th>\n",
       "      <th>Test4.2</th>\n",
       "      <th>Ref1.2</th>\n",
       "      <th>Ref2.2</th>\n",
       "      <th>Ref3.2</th>\n",
       "      <th>Ref4.2</th>\n",
       "      <th>Ref seq (Name-Wet).2</th>\n",
       "      <th>mean_exp</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>19</td>\n",
       "      <td>CompleteSeq-19</td>\n",
       "      <td>pNeoCS-PUMCH_S001-8</td>\n",
       "      <td>23.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4603.0</td>\n",
       "      <td>gene_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>21</td>\n",
       "      <td>CompleteSeq-21</td>\n",
       "      <td>pNeoCS-PUMCH_S001-10</td>\n",
       "      <td>27.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5534.716667</td>\n",
       "      <td>gene_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>30</td>\n",
       "      <td>CompleteSeq-30</td>\n",
       "      <td>pNeoCS-301_CMQI-8</td>\n",
       "      <td>21.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.628571</td>\n",
       "      <td>gene_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>ATGCGGGTGACGGCCCCTAGAACCCTGATTCTGCTTCTGTCGGGTG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>32</td>\n",
       "      <td>CompleteSeq-32</td>\n",
       "      <td>pNeoCS-301_CMQI-10</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>81.871429</td>\n",
       "      <td>gene_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGG</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>10</td>\n",
       "      <td>CompleteSeq-10</td>\n",
       "      <td>pNeoCS-eGFPdegron-10</td>\n",
       "      <td>29.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8808.75</td>\n",
       "      <td>gene_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  3-nt                                              5'UTR  \\\n",
       "0  AGG  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "1  AGG  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "2  AGG  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "3  AGG  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "4  AGG  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "\n",
       "                                                 CDS  \\\n",
       "0  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "1  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...   \n",
       "2  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...   \n",
       "3  ATGCGGGTGACGGCCCCTAGAACCCTGATTCTGCTTCTGTCGGGTG...   \n",
       "4  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "\n",
       "                                               3'UTR 5‘UTR-CDS CDS-3‘UTR No.  \\\n",
       "0  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...         -         -  19   \n",
       "1  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...         -         -  21   \n",
       "2  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...         -         -  30   \n",
       "3  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...         -         -  32   \n",
       "4  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...         -         -  10   \n",
       "\n",
       "           ID-Wet              Name-Wet ID-Dry  ... Test2.2 Test3.2 Test4.2  \\\n",
       "0  CompleteSeq-19   pNeoCS-PUMCH_S001-8   23.0  ...     NaN     NaN     NaN   \n",
       "1  CompleteSeq-21  pNeoCS-PUMCH_S001-10   27.0  ...     NaN     NaN     NaN   \n",
       "2  CompleteSeq-30     pNeoCS-301_CMQI-8   21.0  ...     NaN     NaN     NaN   \n",
       "3  CompleteSeq-32    pNeoCS-301_CMQI-10   25.0  ...     NaN     NaN     NaN   \n",
       "4  CompleteSeq-10  pNeoCS-eGFPdegron-10   29.0  ...     NaN     NaN     NaN   \n",
       "\n",
       "  Ref1.2 Ref2.2 Ref3.2 Ref4.2 Ref seq (Name-Wet).2     mean_exp gene_id  \n",
       "0    NaN    NaN    NaN    NaN                  NaN       4603.0  gene_0  \n",
       "1    NaN    NaN    NaN    NaN                  NaN  5534.716667  gene_1  \n",
       "2    NaN    NaN    NaN    NaN                  NaN    71.628571  gene_2  \n",
       "3    NaN    NaN    NaN    NaN                  NaN    81.871429  gene_3  \n",
       "4    NaN    NaN    NaN    NaN                  NaN      8808.75  gene_4  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "experiment_CDS_exp_agg = experiment_CDS_exp.groupby(by=['3-nt', \"5'UTR\", 'CDS', \"3'UTR\", '5‘UTR-CDS', 'CDS-3‘UTR']).agg(list)\n",
    "\n",
    "agged_cols = list(experiment_CDS_exp_agg.columns)\n",
    "\n",
    "experiment_CDS_exp_agg['gene_id'] = [f'gene_{g}' for g in range(experiment_CDS_exp_agg.shape[0])]\n",
    "\n",
    "experiment_CDS_exp_agg_exploded = experiment_CDS_exp_agg.explode(agged_cols)\n",
    "\n",
    "experiment_CDS_exp_agg_exploded = experiment_CDS_exp_agg_exploded.reset_index()\n",
    "\n",
    "experiment_CDS_exp_agg_exploded.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "6f531759",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "pairs = []\n",
    "for i_index, i_row in experiment_CDS_exp_agg_exploded.iterrows():\n",
    "    for j_index, j_row in experiment_CDS_exp_agg_exploded.iterrows():\n",
    "    \n",
    "        if i_row['mean_exp'] and j_row['mean_exp'] and \\\n",
    "            i_row['gene_id'] != j_row['gene_id'] and \\\n",
    "                str(Seq(i_row['CDS']).translate()) == str(Seq(j_row['CDS']).translate()):\n",
    "            \n",
    "            # print('...')\n",
    "            pair = [i_row['gene_id'], j_row['gene_id']]\n",
    "            if i_row['mean_exp']/j_row['mean_exp'] >= (1+cutoff):\n",
    "                pair.append(1)\n",
    "                \n",
    "            elif i_row['mean_exp']/j_row['mean_exp'] <= 1/(1+cutoff):\n",
    "                pair.append(-1)\n",
    "                \n",
    "            else:\n",
    "                pair.append(0)\n",
    "                \n",
    "            pairs.append(pair)\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "cf016d66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_A_UTR5</th>\n",
       "      <th>pair_A_UTR3</th>\n",
       "      <th>pair_A_CDS</th>\n",
       "      <th>pair_B_UTR5</th>\n",
       "      <th>pair_B_UTR3</th>\n",
       "      <th>pair_B_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>0</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_7</td>\n",
       "      <td>-1</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTCCTGCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_8</td>\n",
       "      <td>0</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_9</td>\n",
       "      <td>0</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_15</td>\n",
       "      <td>-1</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>CGGGTCACGGCCAACTTCTGTCGGGAGCCGGCGAGCGGTCGCCCTC...</td>\n",
       "      <td>GCGGCAGGCGGGAGGGCGAGCGGGAGGGAAGGGAGTGGCGGGTGCG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>gene_25</td>\n",
       "      <td>gene_14</td>\n",
       "      <td>0</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>gene_25</td>\n",
       "      <td>gene_21</td>\n",
       "      <td>1</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...</td>\n",
       "      <td>GCAGCGAAGGACGCTACGAAGCTCAGTCATTACCAGGAAGCTCCCT...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>gene_25</td>\n",
       "      <td>gene_22</td>\n",
       "      <td>1</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...</td>\n",
       "      <td>GCTGGTCCCAATTCTGGTCGAGCTCGGTCAACGGACAGTTCTGTCC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>gene_25</td>\n",
       "      <td>gene_23</td>\n",
       "      <td>1</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGCTCG...</td>\n",
       "      <td>TCGAGCTCGGACAGTTCTGTACCGTGCGGGCCAAGGGCGAAGCGTC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>gene_25</td>\n",
       "      <td>gene_24</td>\n",
       "      <td>0</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>220 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      pair_A   pair_B  pair_label  \\\n",
       "0     gene_0   gene_1           0   \n",
       "1     gene_0   gene_7          -1   \n",
       "2     gene_0   gene_8           0   \n",
       "3     gene_0   gene_9           0   \n",
       "4     gene_0  gene_15          -1   \n",
       "..       ...      ...         ...   \n",
       "215  gene_25  gene_14           0   \n",
       "216  gene_25  gene_21           1   \n",
       "217  gene_25  gene_22           1   \n",
       "218  gene_25  gene_23           1   \n",
       "219  gene_25  gene_24           0   \n",
       "\n",
       "                                           pair_A_UTR5  \\\n",
       "0    AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "1    AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "2    AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "3    AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "4    AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "..                                                 ...   \n",
       "215  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "216  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "217  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "218  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "219  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "\n",
       "                                           pair_A_UTR3  \\\n",
       "0    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "..                                                 ...   \n",
       "215  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "216  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "217  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "218  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "219  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "\n",
       "                                            pair_A_CDS  \\\n",
       "0    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "1    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "2    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "3    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "4    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "..                                                 ...   \n",
       "215  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "216  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "217  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "218  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "219  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "\n",
       "                                           pair_B_UTR5  \\\n",
       "0    AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "1    AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2    AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3    AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4    CGGGTCACGGCCAACTTCTGTCGGGAGCCGGCGAGCGGTCGCCCTC...   \n",
       "..                                                 ...   \n",
       "215  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "216  GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...   \n",
       "217  GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...   \n",
       "218  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGCTCG...   \n",
       "219  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "\n",
       "                                           pair_B_UTR3  \\\n",
       "0    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3    CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4    GCGGCAGGCGGGAGGGCGAGCGGGAGGGAAGGGAGTGGCGGGTGCG...   \n",
       "..                                                 ...   \n",
       "215  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "216  GCAGCGAAGGACGCTACGAAGCTCAGTCATTACCAGGAAGCTCCCT...   \n",
       "217  GCTGGTCCCAATTCTGGTCGAGCTCGGTCAACGGACAGTTCTGTCC...   \n",
       "218  TCGAGCTCGGACAGTTCTGTACCGTGCGGGCCAAGGGCGAAGCGTC...   \n",
       "219  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "\n",
       "                                            pair_B_CDS  \n",
       "0    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...  \n",
       "1    ATGAGAGTGACCGCCCCAAGGACCCTGATTCTCCTGCTGTCTGGAG...  \n",
       "2    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...  \n",
       "3    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...  \n",
       "4    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...  \n",
       "..                                                 ...  \n",
       "215  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "216  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "217  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "218  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "219  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "\n",
       "[220 rows x 9 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "experiment_CDS_pair_exp = pd.DataFrame(pairs, columns=['pair_A', 'pair_B', 'pair_label'])\n",
    "\n",
    "experiment_CDS_pair_exp['pair_A_UTR5'] = experiment_CDS_pair_exp['pair_A'].apply(lambda x: experiment_CDS_exp_agg.reset_index().set_index('gene_id').reindex([x])[\"5'UTR\"].values[0])\n",
    "experiment_CDS_pair_exp['pair_A_UTR3'] = experiment_CDS_pair_exp['pair_A'].apply(lambda x: experiment_CDS_exp_agg.reset_index().set_index('gene_id').reindex([x])[\"3'UTR\"].values[0])\n",
    "experiment_CDS_pair_exp['pair_A_CDS'] = experiment_CDS_pair_exp['pair_A'].apply(lambda x: experiment_CDS_exp_agg.reset_index().set_index('gene_id').reindex([x])['CDS'].values[0])\n",
    "\n",
    "experiment_CDS_pair_exp['pair_B_UTR5'] = experiment_CDS_pair_exp['pair_B'].apply(lambda x: experiment_CDS_exp_agg.reset_index().set_index('gene_id').reindex([x])[\"5'UTR\"].values[0])\n",
    "experiment_CDS_pair_exp['pair_B_UTR3'] = experiment_CDS_pair_exp['pair_B'].apply(lambda x: experiment_CDS_exp_agg.reset_index().set_index('gene_id').reindex([x])[\"3'UTR\"].values[0])\n",
    "experiment_CDS_pair_exp['pair_B_CDS'] = experiment_CDS_pair_exp['pair_B'].apply(lambda x: experiment_CDS_exp_agg.reset_index().set_index('gene_id').reindex([x])['CDS'].values[0])\n",
    "\n",
    "\n",
    "experiment_CDS_pair_exp\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e3ad0093",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "experiment_CDS_pair_exp.to_csv(f'/data1/genomics/yul/bioSPT2/data/interiem/processed/experiment_CDS_pair_exp_cutoff{cutoff}.tsv', sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2db46aa1-fd30-4e8f-b1ed-e45c52e9a1b2",
   "metadata": {},
   "source": [
    "#### CDS2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f64f32c1-cf18-4f42-99f3-c2d4e6be98e7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# 规范化fasta文件\n",
    "\n",
    "def fasta2df(in_fna):\n",
    "    \"\"\"\n",
    "    transform fasta file to dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    transcript_cdna = pd.DataFrame(columns=['gene_id', 'seq'])\n",
    "    with open(in_fna, 'rt', encoding='utf-8') as f:\n",
    "        all_line = f.readlines()\n",
    "\n",
    "        for line in all_line:\n",
    "            if line.startswith('>'):\n",
    "\n",
    "                transcript_id =  line.lstrip('>').strip()\n",
    "\n",
    "                transcript_cdna.loc[transcript_id, 'gene_id'] = transcript_id\n",
    "                lines =\"\"\n",
    "            else:\n",
    "                lines += line.strip()\n",
    "                transcript_cdna.loc[transcript_id, 'seq'] = lines\n",
    "\n",
    "    transcript_cdna['seq'] = transcript_cdna['seq'].apply(lambda x: x.upper().replace('U', 'T'))\n",
    "\n",
    "    return transcript_cdna\n",
    "                                               \n",
    "                                               "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b8c5297-48a6-4bd5-91bb-899f2571c232",
   "metadata": {},
   "source": [
    "##### Fluc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "6dcf9645-9d84-4e80-a00d-3227d93eed35",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp01</th>\n",
       "      <td>Neocura-Fluc_Exp01</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp02</th>\n",
       "      <td>Neocura-Fluc_Exp02</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp03</th>\n",
       "      <td>Neocura-Fluc_Exp03</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp04</th>\n",
       "      <td>Neocura-Fluc_Exp04</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp05</th>\n",
       "      <td>Neocura-Fluc_Exp05</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               gene_id  \\\n",
       "Neocura-Fluc_Exp01  Neocura-Fluc_Exp01   \n",
       "Neocura-Fluc_Exp02  Neocura-Fluc_Exp02   \n",
       "Neocura-Fluc_Exp03  Neocura-Fluc_Exp03   \n",
       "Neocura-Fluc_Exp04  Neocura-Fluc_Exp04   \n",
       "Neocura-Fluc_Exp05  Neocura-Fluc_Exp05   \n",
       "\n",
       "                                                                  seq  \n",
       "Neocura-Fluc_Exp01  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...  \n",
       "Neocura-Fluc_Exp02  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...  \n",
       "Neocura-Fluc_Exp03  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...  \n",
       "Neocura-Fluc_Exp04  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...  \n",
       "Neocura-Fluc_Exp05  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...  "
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "Fluc_CDS = fasta2df('data/eGFP_Fluc/Fluc_Exp.Collect.CDS.fasta')\n",
    "\n",
    "Fluc_mRNA = fasta2df('data/eGFP_Fluc/Fluc_Exp.Collect.addUTR.fasta')\n",
    "\n",
    "Fluc_mRNA.head()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "85a09c0a-4f73-45a1-b1c0-9602a49e8703",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>UTR3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp01</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp02</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp03</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp04</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp05</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTTTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp06</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTTTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp07</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp08</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCCCCATTTTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp09</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp10</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp11</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp12</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp13</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCTCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp14</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp15</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGACCAGCCCCTTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp16</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp17</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCTCCCTTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp18</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp19</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp20</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGACCAGCTCCCTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp21</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp22</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp23</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp24</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp25</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp26</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCTCCATTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp27</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTTTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp28</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAATATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp29</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAATATCAAGAAGGGCCCCGCCCCTTTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp30</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp31</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGACCCGCCCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp32</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp33</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp34</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp35</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp36</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp37</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCTCCATTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promega-Fluc</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAAGATGCCAAAAACATTAAGAAGGGCCCAGCGCCATTCTACC...</td>\n",
       "      <td>TGATGACTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                 UTR5  \\\n",
       "Neocura-Fluc_Exp01  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp02  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp03  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp04  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp05  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp06  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp07  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp08  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp09  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp10  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp11  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp12  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp13  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp14  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp15  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp16  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp17  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp18  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp19  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp20  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp21  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp22  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp23  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp24  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp25  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp26  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp27  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp28  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp29  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp30  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp31  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp32  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp33  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp34  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp35  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp36  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp37  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Promega-Fluc        AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "\n",
       "                                                                  CDS  \\\n",
       "Neocura-Fluc_Exp01  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...   \n",
       "Neocura-Fluc_Exp02  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTATC...   \n",
       "Neocura-Fluc_Exp03  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...   \n",
       "Neocura-Fluc_Exp04  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...   \n",
       "Neocura-Fluc_Exp05  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTTTACC...   \n",
       "Neocura-Fluc_Exp06  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTTTACC...   \n",
       "Neocura-Fluc_Exp07  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...   \n",
       "Neocura-Fluc_Exp08  ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCCCCATTTTACC...   \n",
       "Neocura-Fluc_Exp09  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTATC...   \n",
       "Neocura-Fluc_Exp10  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...   \n",
       "Neocura-Fluc_Exp11  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "Neocura-Fluc_Exp12  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTCTACC...   \n",
       "Neocura-Fluc_Exp13  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCTCCATTCTACC...   \n",
       "Neocura-Fluc_Exp14  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTACC...   \n",
       "Neocura-Fluc_Exp15  ATGGAGGACGCCAAGAACATCAAGAAGGGACCAGCCCCTTTCTACC...   \n",
       "Neocura-Fluc_Exp16  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...   \n",
       "Neocura-Fluc_Exp17  ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCTCCCTTCTATC...   \n",
       "Neocura-Fluc_Exp18  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...   \n",
       "Neocura-Fluc_Exp19  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTACC...   \n",
       "Neocura-Fluc_Exp20  ATGGAGGACGCCAAGAACATCAAGAAGGGACCAGCTCCCTTCTACC...   \n",
       "Neocura-Fluc_Exp21  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...   \n",
       "Neocura-Fluc_Exp22  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...   \n",
       "Neocura-Fluc_Exp23  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTACC...   \n",
       "Neocura-Fluc_Exp24  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...   \n",
       "Neocura-Fluc_Exp25  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTACC...   \n",
       "Neocura-Fluc_Exp26  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCTCCATTCTATC...   \n",
       "Neocura-Fluc_Exp27  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTTTACC...   \n",
       "Neocura-Fluc_Exp28  ATGGAGGACGCCAAGAATATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "Neocura-Fluc_Exp29  ATGGAGGACGCCAAGAATATCAAGAAGGGCCCCGCCCCTTTCTATC...   \n",
       "Neocura-Fluc_Exp30  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...   \n",
       "Neocura-Fluc_Exp31  ATGGAGGACGCCAAGAACATCAAGAAGGGACCCGCCCCATTCTACC...   \n",
       "Neocura-Fluc_Exp32  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTCTATC...   \n",
       "Neocura-Fluc_Exp33  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...   \n",
       "Neocura-Fluc_Exp34  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTATC...   \n",
       "Neocura-Fluc_Exp35  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTATC...   \n",
       "Neocura-Fluc_Exp36  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...   \n",
       "Neocura-Fluc_Exp37  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCTCCATTCTATC...   \n",
       "Promega-Fluc        ATGGAAGATGCCAAAAACATTAAGAAGGGCCCAGCGCCATTCTACC...   \n",
       "\n",
       "                                                                 UTR3  \n",
       "Neocura-Fluc_Exp01  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp02  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp03  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp04  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp05  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp06  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp07  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp08  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp09  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp10  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp11  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp12  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp13  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp14  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp15  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp16  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp17  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp18  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp19  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp20  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp21  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp22  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp23  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp24  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp25  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp26  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp27  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp28  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp29  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp30  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp31  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp32  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp33  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp34  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp35  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp36  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp37  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Promega-Fluc        TGATGACTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...  "
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "def extract(row):\n",
    "    \n",
    "    CDS = Fluc_eGFP_CDS.reindex([row['gene_id']])['seq'].values[0]\n",
    "    UTR5 = row['seq'].split(CDS)[0]\n",
    "    UTR3 = row['seq'].split(CDS)[-1]\n",
    "    \n",
    "    # return pd.Series([UTR5, UTR5, UTR3])\n",
    "    return pd.Series({\n",
    "        'UTR5': UTR5, 'CDS':CDS, 'UTR3':UTR3\n",
    "    })\n",
    "\n",
    "    \n",
    "Fluc_mRNA = Fluc_mRNA.apply(lambda row: extract(row), axis=1)\n",
    "\n",
    "Fluc_mRNA\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "7aceff54-f73c-41d5-a906-e6a259b2ae35",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>No.</th>\n",
       "      <th>Value</th>\n",
       "      <th>ExpID</th>\n",
       "      <th>TaowaScore</th>\n",
       "      <th>TaowaRank</th>\n",
       "      <th>StemCount</th>\n",
       "      <th>StemRank</th>\n",
       "      <th>三等分</th>\n",
       "      <th>TaowaScore_7500</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>No.11</td>\n",
       "      <td>1.614245</td>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>14172</td>\n",
       "      <td>1.0</td>\n",
       "      <td>616</td>\n",
       "      <td>6.0</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>No.28</td>\n",
       "      <td>1.055267</td>\n",
       "      <td>Neocura-Fluc_Exp28</td>\n",
       "      <td>14104</td>\n",
       "      <td>2.0</td>\n",
       "      <td>678</td>\n",
       "      <td>2.5</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>Neocura-Fluc_Exp28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>No.17</td>\n",
       "      <td>1.511719</td>\n",
       "      <td>Neocura-Fluc_Exp17</td>\n",
       "      <td>9970</td>\n",
       "      <td>3.0</td>\n",
       "      <td>604</td>\n",
       "      <td>9.5</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>Neocura-Fluc_Exp17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>No.23</td>\n",
       "      <td>1.672987</td>\n",
       "      <td>Neocura-Fluc_Exp23</td>\n",
       "      <td>9362</td>\n",
       "      <td>4.0</td>\n",
       "      <td>700</td>\n",
       "      <td>1.0</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>Neocura-Fluc_Exp23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>No.33</td>\n",
       "      <td>1.479376</td>\n",
       "      <td>Neocura-Fluc_Exp33</td>\n",
       "      <td>9198</td>\n",
       "      <td>5.0</td>\n",
       "      <td>642</td>\n",
       "      <td>4.0</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>Neocura-Fluc_Exp33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     No.     Value               ExpID  TaowaScore  TaowaRank  StemCount  \\\n",
       "0  No.11  1.614245  Neocura-Fluc_Exp11       14172        1.0        616   \n",
       "1  No.28  1.055267  Neocura-Fluc_Exp28       14104        2.0        678   \n",
       "2  No.17  1.511719  Neocura-Fluc_Exp17        9970        3.0        604   \n",
       "3  No.23  1.672987  Neocura-Fluc_Exp23        9362        4.0        700   \n",
       "4  No.33  1.479376  Neocura-Fluc_Exp33        9198        5.0        642   \n",
       "\n",
       "   StemRank   三等分 TaowaScore_7500             gene_id  \n",
       "0       6.0  High            High  Neocura-Fluc_Exp11  \n",
       "1       2.5  High            High  Neocura-Fluc_Exp28  \n",
       "2       9.5  High            High  Neocura-Fluc_Exp17  \n",
       "3       1.0  High            High  Neocura-Fluc_Exp23  \n",
       "4       4.0  High            High  Neocura-Fluc_Exp33  "
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "Fluc_Exp = pd.read_excel('data/eGFP_Fluc/Fluc_Exp.Result.Normalized.xlsx')\n",
    "Fluc_Exp['gene_id'] = Fluc_Exp['ExpID']\n",
    "Fluc_Exp.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9407bac8-ef38-44ad-bb7d-5dffdcaf268a",
   "metadata": {},
   "source": [
    "##### eGFP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "71fe98cf-9ca3-4c65-9244-18ebc895edd7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>UTR3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>eGFP-degron_Test2013</th>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTTACAGGCGTGGTGCCCATCC...</td>\n",
       "      <td>GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eGFP-degron_Test3690</th>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eGFP-degron_Test3006</th>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eGFP-degron_Test2124</th>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTTACAGGAGTGGTGCCCATCC...</td>\n",
       "      <td>GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eGFP-degron_Test3768</th>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-7</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-8</th>\n",
       "      <td>AGGAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-9</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-10</th>\n",
       "      <td>AGGAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-11</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                   UTR5  \\\n",
       "eGFP-degron_Test2013  GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "eGFP-degron_Test3690  GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "eGFP-degron_Test3006  GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "eGFP-degron_Test2124  GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "eGFP-degron_Test3768  GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "...                                                                 ...   \n",
       "pNeoCS-eGFPdegron-7   AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "pNeoCS-eGFPdegron-8   AGGAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...   \n",
       "pNeoCS-eGFPdegron-9   AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "pNeoCS-eGFPdegron-10  AGGAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...   \n",
       "pNeoCS-eGFPdegron-11  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "\n",
       "                                                                    CDS  \\\n",
       "eGFP-degron_Test2013  ATGGTGAGCAAGGGCGAGGAGCTGTTTACAGGCGTGGTGCCCATCC...   \n",
       "eGFP-degron_Test3690  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "eGFP-degron_Test3006  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "eGFP-degron_Test2124  ATGGTGAGCAAGGGCGAGGAGCTGTTTACAGGAGTGGTGCCCATCC...   \n",
       "eGFP-degron_Test3768  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "...                                                                 ...   \n",
       "pNeoCS-eGFPdegron-7   ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "pNeoCS-eGFPdegron-8   ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "pNeoCS-eGFPdegron-9   ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "pNeoCS-eGFPdegron-10  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "pNeoCS-eGFPdegron-11  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "\n",
       "                                                                   UTR3  \n",
       "eGFP-degron_Test2013  GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...  \n",
       "eGFP-degron_Test3690  GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...  \n",
       "eGFP-degron_Test3006  GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...  \n",
       "eGFP-degron_Test2124  GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...  \n",
       "eGFP-degron_Test3768  GAGCTCCTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCG...  \n",
       "...                                                                 ...  \n",
       "pNeoCS-eGFPdegron-7   CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "pNeoCS-eGFPdegron-8   CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "pNeoCS-eGFPdegron-9   CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "pNeoCS-eGFPdegron-10  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "pNeoCS-eGFPdegron-11  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "\n",
       "[88 rows x 3 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "eGFP_CDS = fasta2df('data/eGFP_Fluc/eGFP-Tested.Collect.CDS.fasta')\n",
    "\n",
    "eGFP_mRNA = fasta2df('data/eGFP_Fluc/eGFP-Tested.Collect.addUTR.fasta')\n",
    "\n",
    "    \n",
    "eGFP_mRNA = eGFP_mRNA.apply(lambda row: extract(row), axis=1)\n",
    "\n",
    "\n",
    "eGFP_mRNA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "01fd57ae-ed02-4dda-9a03-7406348674be",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp01</th>\n",
       "      <td>Neocura-Fluc_Exp01</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp02</th>\n",
       "      <td>Neocura-Fluc_Exp02</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp03</th>\n",
       "      <td>Neocura-Fluc_Exp03</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp04</th>\n",
       "      <td>Neocura-Fluc_Exp04</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp05</th>\n",
       "      <td>Neocura-Fluc_Exp05</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTTTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-7</th>\n",
       "      <td>pNeoCS-eGFPdegron-7</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-8</th>\n",
       "      <td>pNeoCS-eGFPdegron-8</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-9</th>\n",
       "      <td>pNeoCS-eGFPdegron-9</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-10</th>\n",
       "      <td>pNeoCS-eGFPdegron-10</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pNeoCS-eGFPdegron-11</th>\n",
       "      <td>pNeoCS-eGFPdegron-11</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   gene_id  \\\n",
       "Neocura-Fluc_Exp01      Neocura-Fluc_Exp01   \n",
       "Neocura-Fluc_Exp02      Neocura-Fluc_Exp02   \n",
       "Neocura-Fluc_Exp03      Neocura-Fluc_Exp03   \n",
       "Neocura-Fluc_Exp04      Neocura-Fluc_Exp04   \n",
       "Neocura-Fluc_Exp05      Neocura-Fluc_Exp05   \n",
       "...                                    ...   \n",
       "pNeoCS-eGFPdegron-7    pNeoCS-eGFPdegron-7   \n",
       "pNeoCS-eGFPdegron-8    pNeoCS-eGFPdegron-8   \n",
       "pNeoCS-eGFPdegron-9    pNeoCS-eGFPdegron-9   \n",
       "pNeoCS-eGFPdegron-10  pNeoCS-eGFPdegron-10   \n",
       "pNeoCS-eGFPdegron-11  pNeoCS-eGFPdegron-11   \n",
       "\n",
       "                                                                    seq  \n",
       "Neocura-Fluc_Exp01    ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...  \n",
       "Neocura-Fluc_Exp02    ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTATC...  \n",
       "Neocura-Fluc_Exp03    ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...  \n",
       "Neocura-Fluc_Exp04    ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...  \n",
       "Neocura-Fluc_Exp05    ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTTTACC...  \n",
       "...                                                                 ...  \n",
       "pNeoCS-eGFPdegron-7   ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "pNeoCS-eGFPdegron-8   ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "pNeoCS-eGFPdegron-9   ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "pNeoCS-eGFPdegron-10  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "pNeoCS-eGFPdegron-11  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "\n",
       "[126 rows x 2 columns]"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "Fluc_eGFP_CDS = pd.concat([Fluc_CDS, eGFP_CDS])\n",
    "\n",
    "Fluc_eGFP_CDS\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "038ed008-8d8f-4174-9120-63406efdcbfa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Value</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>High_RSCU_high_struc_3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>High_RSCU_high_struc_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mid_RSCU_high_struc_1</th>\n",
       "      <td>0.527020</td>\n",
       "      <td>Mid_RSCU_high_struc_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mid_RSCU_high_struc_2</th>\n",
       "      <td>0.488853</td>\n",
       "      <td>Mid_RSCU_high_struc_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mid_RSCU_high_struc_3</th>\n",
       "      <td>0.408157</td>\n",
       "      <td>Mid_RSCU_high_struc_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mid_RSCU_high_struc_4</th>\n",
       "      <td>0.443380</td>\n",
       "      <td>Mid_RSCU_high_struc_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Value                 gene_id\n",
       "High_RSCU_high_struc_3  1.000000  High_RSCU_high_struc_3\n",
       "Mid_RSCU_high_struc_1   0.527020   Mid_RSCU_high_struc_1\n",
       "Mid_RSCU_high_struc_2   0.488853   Mid_RSCU_high_struc_2\n",
       "Mid_RSCU_high_struc_3   0.408157   Mid_RSCU_high_struc_3\n",
       "Mid_RSCU_high_struc_4   0.443380   Mid_RSCU_high_struc_4"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import yaml\n",
    "with open('data/eGFP_Fluc/eGFP-degron_CDS-Exp_New.yaml') as fn:\n",
    "    eGFP_Exp = pd.DataFrame.from_dict(\n",
    "        yaml.load(fn, Loader=yaml.FullLoader), orient='index', columns=['Value'])\n",
    "    \n",
    "eGFP_Exp['gene_id'] = eGFP_Exp.index\n",
    "eGFP_Exp.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "e546b1b7-9bb7-4605-8314-7af3034235af",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>UTR3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp01</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp02</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp03</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp04</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neocura-Fluc_Exp05</th>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTTTACC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                 UTR5  \\\n",
       "Neocura-Fluc_Exp01  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp02  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp03  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp04  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "Neocura-Fluc_Exp05  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "\n",
       "                                                                  CDS  \\\n",
       "Neocura-Fluc_Exp01  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTCTACC...   \n",
       "Neocura-Fluc_Exp02  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCATTTTATC...   \n",
       "Neocura-Fluc_Exp03  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTATC...   \n",
       "Neocura-Fluc_Exp04  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...   \n",
       "Neocura-Fluc_Exp05  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCTCCCTTTTACC...   \n",
       "\n",
       "                                                                 UTR3  \n",
       "Neocura-Fluc_Exp01  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp02  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp03  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp04  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  \n",
       "Neocura-Fluc_Exp05  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...  "
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "eGFP_Fluc_Exp = pd.concat([Fluc_Exp, eGFP_Exp], axis=0)\n",
    "\n",
    "eGFP_Fluc_Exp\n",
    "\n",
    "\n",
    "eGFP_Fluc_fasta = pd.concat([Fluc_mRNA, eGFP_mRNA], axis=0)\n",
    "\n",
    "\n",
    "eGFP_Fluc_fasta.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "5edde5eb-aa78-4daf-82a0-296d222f8c44",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                       Neocura-Fluc_Exp11\n",
       "1                       Neocura-Fluc_Exp28\n",
       "2                       Neocura-Fluc_Exp17\n",
       "3                       Neocura-Fluc_Exp23\n",
       "4                       Neocura-Fluc_Exp33\n",
       "                              ...         \n",
       "pNeoCS-eGFPdegron-5    pNeoCS-eGFPdegron-5\n",
       "pNeoCS-eGFPdegron-6    pNeoCS-eGFPdegron-6\n",
       "pNeoCS-eGFPdegron-7    pNeoCS-eGFPdegron-7\n",
       "pNeoCS-eGFPdegron-8    pNeoCS-eGFPdegron-8\n",
       "pNeoCS-eGFPdegron-9    pNeoCS-eGFPdegron-9\n",
       "Name: gene_id, Length: 126, dtype: object"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "eGFP_Fluc_Exp['gene_id']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "656fdf02-6e3a-4892-ad0d-3fab79fb6df5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "High_RSCU_high_struc_3    [ATGGTGTCCAAGGGGGAGGAGCTGTTCACCGGCGTGGTGCCCATC...\n",
       "Mid_RSCU_high_struc_1     [ATGGTCTCGAAGGGCGAGGAGTTGTTCACCGGGGTGGTGCCGATC...\n",
       "Mid_RSCU_high_struc_2     [ATGGTCTCGAAGGGCGAGGAGCTGTTCACGGGAGTCGTCCCGATC...\n",
       "Mid_RSCU_high_struc_3                                   [ATGGTCTCGAAGGGTGA]\n",
       "Mid_RSCU_high_struc_4     [ATGGTCTCGAAGGGAGAGGAGCTCTTCACGGGGGTCGTCCCGATC...\n",
       "                                                ...                        \n",
       "pNeoCS-eGFPdegron-5                                               [ATGGTGA]\n",
       "pNeoCS-eGFPdegron-6                                               [ATGGTGA]\n",
       "pNeoCS-eGFPdegron-7                                               [ATGGTGA]\n",
       "pNeoCS-eGFPdegron-8                                               [ATGGTGA]\n",
       "pNeoCS-eGFPdegron-9                                               [ATGGTGA]\n",
       "Name: gene_id, Length: 88, dtype: object"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "eGFP_Exp['gene_id'].apply(lambda x: eGFP_mRNA.reindex([x])['CDS'].values)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "1dc15c7e-9172-468a-9713-d04577771ab8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pNeoCS-eGFPdegron-5    ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...\n",
       "pNeoCS-eGFPdegron-6    ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...\n",
       "pNeoCS-eGFPdegron-7    ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...\n",
       "pNeoCS-eGFPdegron-8    ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...\n",
       "pNeoCS-eGFPdegron-9    ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...\n",
       "Name: CDS, dtype: object"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "eGFP_Fluc_Exp['CDS'] = eGFP_Fluc_Exp['gene_id'].apply(lambda x: eGFP_Fluc_fasta.reindex([x])['CDS'].values[0])\n",
    "\n",
    "eGFP_Fluc_Exp['CDS'].tail()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "0be6e491-980b-4eef-af04-23dd7a9cf69c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                      1.614245\n",
       "1                      1.055267\n",
       "2                      1.511719\n",
       "3                      1.672987\n",
       "4                      1.479376\n",
       "                         ...   \n",
       "pNeoCS-eGFPdegron-5    0.100000\n",
       "pNeoCS-eGFPdegron-6    1.200000\n",
       "pNeoCS-eGFPdegron-7    1.100000\n",
       "pNeoCS-eGFPdegron-8    1.300000\n",
       "pNeoCS-eGFPdegron-9    1.500000\n",
       "Name: Value, Length: 126, dtype: float64"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "eGFP_Fluc_Exp['Value']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "d19849dc-2d07-47fb-afca-e90e678df7fe",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "cutoff = 0\n",
    "\n",
    "pairs = []\n",
    "for i_index, i_row in eGFP_Fluc_Exp.iterrows():\n",
    "    for j_index, j_row in eGFP_Fluc_Exp.iterrows():\n",
    "    \n",
    "        if i_row['Value'] and j_row['Value'] and \\\n",
    "            i_row['gene_id'] != j_row['gene_id'] and \\\n",
    "                str(Seq(i_row['CDS']).translate()) == str(Seq(j_row['CDS']).translate()):\n",
    "            \n",
    "            pair = [i_row['gene_id'], j_row['gene_id']]\n",
    "            if i_row['Value']/j_row['Value'] >= (1+cutoff):\n",
    "                pair.append(1)\n",
    "                \n",
    "            elif i_row['Value']/j_row['Value'] <= 1/(1+cutoff):\n",
    "                pair.append(-1)\n",
    "                \n",
    "            else:\n",
    "                pair.append(0)\n",
    "                \n",
    "                \n",
    "            pair.append(i_row['Value']/j_row['Value'])\n",
    "                \n",
    "            pairs.append(pair)\n",
    "                  \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "61f5d6b6-6376-46d8-8fc1-0b28617153f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>pair_A_UTR5</th>\n",
       "      <th>pair_A_UTR3</th>\n",
       "      <th>pair_A_CDS</th>\n",
       "      <th>pair_B_UTR5</th>\n",
       "      <th>pair_B_UTR3</th>\n",
       "      <th>pair_B_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp28</td>\n",
       "      <td>1</td>\n",
       "      <td>1.529703</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAATATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp17</td>\n",
       "      <td>1</td>\n",
       "      <td>1.067821</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCTCCCTTCTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp23</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.964888</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp33</td>\n",
       "      <td>1</td>\n",
       "      <td>1.091166</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp04</td>\n",
       "      <td>1</td>\n",
       "      <td>1.040731</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8809</th>\n",
       "      <td>pNeoCS-eGFPdegron-9</td>\n",
       "      <td>pNeoCS-eGFPdegron-4</td>\n",
       "      <td>1</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGC...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8810</th>\n",
       "      <td>pNeoCS-eGFPdegron-9</td>\n",
       "      <td>pNeoCS-eGFPdegron-5</td>\n",
       "      <td>1</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGC...</td>\n",
       "      <td>TCGAGCTCGGACAGTTCTGTACCGTGCGGGCCAAGGGCGAAGCGTC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8811</th>\n",
       "      <td>pNeoCS-eGFPdegron-9</td>\n",
       "      <td>pNeoCS-eGFPdegron-6</td>\n",
       "      <td>1</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGC...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8812</th>\n",
       "      <td>pNeoCS-eGFPdegron-9</td>\n",
       "      <td>pNeoCS-eGFPdegron-7</td>\n",
       "      <td>1</td>\n",
       "      <td>1.363636</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8813</th>\n",
       "      <td>pNeoCS-eGFPdegron-9</td>\n",
       "      <td>pNeoCS-eGFPdegron-8</td>\n",
       "      <td>1</td>\n",
       "      <td>1.153846</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>AGGAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8814 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   pair_A               pair_B  pair_label  pair_diff  \\\n",
       "0      Neocura-Fluc_Exp11   Neocura-Fluc_Exp28           1   1.529703   \n",
       "1      Neocura-Fluc_Exp11   Neocura-Fluc_Exp17           1   1.067821   \n",
       "2      Neocura-Fluc_Exp11   Neocura-Fluc_Exp23          -1   0.964888   \n",
       "3      Neocura-Fluc_Exp11   Neocura-Fluc_Exp33           1   1.091166   \n",
       "4      Neocura-Fluc_Exp11   Neocura-Fluc_Exp04           1   1.040731   \n",
       "...                   ...                  ...         ...        ...   \n",
       "8809  pNeoCS-eGFPdegron-9  pNeoCS-eGFPdegron-4           1   1.500000   \n",
       "8810  pNeoCS-eGFPdegron-9  pNeoCS-eGFPdegron-5           1  15.000000   \n",
       "8811  pNeoCS-eGFPdegron-9  pNeoCS-eGFPdegron-6           1   1.250000   \n",
       "8812  pNeoCS-eGFPdegron-9  pNeoCS-eGFPdegron-7           1   1.363636   \n",
       "8813  pNeoCS-eGFPdegron-9  pNeoCS-eGFPdegron-8           1   1.153846   \n",
       "\n",
       "                                            pair_A_UTR5  \\\n",
       "0     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "3     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "4     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "...                                                 ...   \n",
       "8809  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "8810  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "8811  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "8812  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "8813  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "\n",
       "                                            pair_A_UTR3  \\\n",
       "0     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "...                                                 ...   \n",
       "8809  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8810  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8811  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8812  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8813  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                             pair_A_CDS  \\\n",
       "0     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "1     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "2     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "3     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "4     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "...                                                 ...   \n",
       "8809  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "8810  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "8811  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "8812  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "8813  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "\n",
       "                                            pair_B_UTR5  \\\n",
       "0     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "3     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "4     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "...                                                 ...   \n",
       "8809  AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGC...   \n",
       "8810  AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGC...   \n",
       "8811  AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGC...   \n",
       "8812  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "8813  AGGAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...   \n",
       "\n",
       "                                            pair_B_UTR3  \\\n",
       "0     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "...                                                 ...   \n",
       "8809  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "8810  TCGAGCTCGGACAGTTCTGTACCGTGCGGGCCAAGGGCGAAGCGTC...   \n",
       "8811  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "8812  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "8813  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                             pair_B_CDS  \n",
       "0     ATGGAGGACGCCAAGAATATCAAGAAGGGCCCAGCCCCATTCTATC...  \n",
       "1     ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCTCCCTTCTATC...  \n",
       "2     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTACC...  \n",
       "3     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...  \n",
       "4     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...  \n",
       "...                                                 ...  \n",
       "8809  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "8810  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "8811  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "8812  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "8813  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...  \n",
       "\n",
       "[8814 rows x 10 columns]"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "experiment_CDS_pair_exp = pd.DataFrame(pairs, columns=['pair_A', 'pair_B',  'pair_label', 'pair_diff'])\n",
    "\n",
    "experiment_CDS_pair_exp['pair_A_UTR5'] = experiment_CDS_pair_exp['pair_A'].apply(lambda x: eGFP_Fluc_fasta.reindex([x])[\"UTR5\"].values[0])\n",
    "experiment_CDS_pair_exp['pair_A_UTR3'] = experiment_CDS_pair_exp['pair_A'].apply(lambda x: eGFP_Fluc_fasta.reindex([x])[\"UTR3\"].values[0])\n",
    "experiment_CDS_pair_exp['pair_A_CDS'] = experiment_CDS_pair_exp['pair_A'].apply(lambda x: eGFP_Fluc_fasta.reindex([x])['CDS'].values[0])\n",
    "\n",
    "experiment_CDS_pair_exp['pair_B_UTR5'] = experiment_CDS_pair_exp['pair_B'].apply(lambda x: eGFP_Fluc_fasta.reindex([x])[\"UTR5\"].values[0])\n",
    "experiment_CDS_pair_exp['pair_B_UTR3'] = experiment_CDS_pair_exp['pair_B'].apply(lambda x: eGFP_Fluc_fasta.reindex([x])[\"UTR3\"].values[0])\n",
    "experiment_CDS_pair_exp['pair_B_CDS'] = experiment_CDS_pair_exp['pair_B'].apply(lambda x: eGFP_Fluc_fasta.reindex([x])['CDS'].values[0])\n",
    "\n",
    "\n",
    "experiment_CDS_pair_exp\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "ad95378e-21b8-4d96-b1f2-627917d2f702",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       " 1    3020\n",
       "-1    3020\n",
       " 0    2774\n",
       "Name: pair_label, dtype: int64"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "experiment_CDS_pair_exp['pair_label'].value_counts()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "727f3cef-d1bb-4d81-920b-46ffb36ffcdc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "experiment_CDS_pair_exp.to_csv(f'data/interim/experiment_CDS2_pair_exp_cutoff{cutoff}.tsv', sep='\\t')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c610f9ca",
   "metadata": {},
   "source": [
    "####  合并去重"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a22379d",
   "metadata": {},
   "source": [
    "##### cellines_pair_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6d3203f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>ESC</th>\n",
       "      <th>TE</th>\n",
       "      <th>iBAQ</th>\n",
       "      <th>pair_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000160336</td>\n",
       "      <td>ENSG00000213799</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000160336</td>\n",
       "      <td>ENSG00000196417</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000213799</td>\n",
       "      <td>ENSG00000196417</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000072110</td>\n",
       "      <td>ENSG00000130402</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000178935</td>\n",
       "      <td>ENSG00000269343</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0            gene1            gene2  ESC  TE  iBAQ  pair_label\n",
       "0           0  ENSG00000160336  ENSG00000213799    0  -1    -1          -1\n",
       "1           1  ENSG00000160336  ENSG00000196417    0   0     0           0\n",
       "2           2  ENSG00000213799  ENSG00000196417   -1   1     0           0\n",
       "3           3  ENSG00000072110  ENSG00000130402    0  -1     0          -1\n",
       "4           4  ENSG00000178935  ENSG00000269343    0   1     0           1"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "cellines_pair_exp = pd.read_csv('/home/yul/project/nas_project/bioSPT2/data/interiem/processed/HEK_ESC_data_db_0.8_pairs_expression.tsv', sep='\\t')\n",
    "\n",
    "cellines_pair_exp.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "2a2621b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000160336</td>\n",
       "      <td>ENSG00000213799</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000160336</td>\n",
       "      <td>ENSG00000196417</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000213799</td>\n",
       "      <td>ENSG00000196417</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000072110</td>\n",
       "      <td>ENSG00000130402</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000178935</td>\n",
       "      <td>ENSG00000269343</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENSG00000270276</td>\n",
       "      <td>ENSG00000277157</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ENSG00000270276</td>\n",
       "      <td>ENSG00000278637</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ENSG00000276180</td>\n",
       "      <td>ENSG00000277157</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>ENSG00000276180</td>\n",
       "      <td>ENSG00000278637</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>ENSG00000277157</td>\n",
       "      <td>ENSG00000278637</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             pair_A           pair_B  pair_label\n",
       "0   ENSG00000160336  ENSG00000213799          -1\n",
       "1   ENSG00000160336  ENSG00000196417           0\n",
       "2   ENSG00000213799  ENSG00000196417           0\n",
       "3   ENSG00000072110  ENSG00000130402          -1\n",
       "4   ENSG00000178935  ENSG00000269343           1\n",
       "..              ...              ...         ...\n",
       "59  ENSG00000270276  ENSG00000277157           0\n",
       "60  ENSG00000270276  ENSG00000278637           0\n",
       "61  ENSG00000276180  ENSG00000277157          -1\n",
       "62  ENSG00000276180  ENSG00000278637          -1\n",
       "63  ENSG00000277157  ENSG00000278637           0\n",
       "\n",
       "[64 rows x 3 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "cellines_pair_exp['pair_A'] = cellines_pair_exp['gene1']\n",
    "cellines_pair_exp['pair_B'] = cellines_pair_exp['gene2']\n",
    "\n",
    "cellines_pair_exp = cellines_pair_exp[[ 'pair_A', 'pair_B',  'pair_label']]\n",
    "cellines_pair_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3c018ada",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "cellines_pair_exp = cellines_pair_exp[cellines_pair_exp['pair_label'] != 0]\n",
    "\n",
    "cellines_pair_exp.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "27eb78c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id_raw</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>UTR5_regs</th>\n",
       "      <th>CDS_regs</th>\n",
       "      <th>UTR3_regs</th>\n",
       "      <th>UTR5</th>\n",
       "      <th>CDS</th>\n",
       "      <th>UTR3</th>\n",
       "      <th>Esemble</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>ENSG00000000003.15</td>\n",
       "      <td>&gt;ENST00000373020.9|ENSG00000000003.15|OTTHUMG0...</td>\n",
       "      <td>AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>ENST00000373020.9</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>UTR5:1-112</td>\n",
       "      <td>CDS:113-850</td>\n",
       "      <td>UTR3:851-3768</td>\n",
       "      <td>AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...</td>\n",
       "      <td>ATGGCGTCCCCGTCTCGGAGACTGCAGACTAAACCAGTCATTACTT...</td>\n",
       "      <td>CCCAATGTATCTGTGGGCCTATTCCTCTCTACCTTTAAGGACATTT...</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000373031</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>&gt;ENST00000373031.5|ENSG00000000005.6|OTTHUMG00...</td>\n",
       "      <td>AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...</td>\n",
       "      <td>ENST00000373031.5</td>\n",
       "      <td>ENST00000373031.5</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>UTR5:1-83</td>\n",
       "      <td>CDS:84-1037</td>\n",
       "      <td>UTR3:1038-1205</td>\n",
       "      <td>AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...</td>\n",
       "      <td>ATGGCAAAGAATCCTCCAGAGAATTGTGAAGACTGTCACATTCTAA...</td>\n",
       "      <td>TAGGAGGTTTGAGCTCAAATGCTTAAACTGCTGGCAACATATAATA...</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000371588</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>&gt;ENST00000371588.10|ENSG00000000419.14|OTTHUMG...</td>\n",
       "      <td>AGTTCCGCCATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGT...</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>ENST00000371588.10</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>UTR5:1-9</td>\n",
       "      <td>CDS:10-792</td>\n",
       "      <td>UTR3:793-1054</td>\n",
       "      <td>AGTTCCGCC</td>\n",
       "      <td>ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...</td>\n",
       "      <td>AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...</td>\n",
       "      <td>ENSG00000000419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000367772</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>&gt;ENST00000367772.8|ENSG00000000457.14|OTTHUMG0...</td>\n",
       "      <td>TGTCCCGTTTCCGGACCCGTCTCTATGGTGTAGGAGAAACCCGGCC...</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "      <td>ENST00000367772.8</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>UTR5:1-222</td>\n",
       "      <td>CDS:223-2451</td>\n",
       "      <td>UTR3:2452-3090</td>\n",
       "      <td>TGTCCCGTTTCCGGACCCGTCTCTATGGTGTAGGAGAAACCCGGCC...</td>\n",
       "      <td>ATGGGATCAGAGAACAGTGCTTTAAAGAGCTATACACTGAGAGAAC...</td>\n",
       "      <td>CAATAGATGTGAGTTAAACTTTAGGAAAAAGGATTCCCTTTTTTTA...</td>\n",
       "      <td>ENSG00000000457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000413811</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>&gt;ENST00000413811.3|ENSG00000000460.17|OTTHUMG0...</td>\n",
       "      <td>ACATTCTTGATTGGTCTCATGCCTTTGTGGTTGTAAATGTGCCTGG...</td>\n",
       "      <td>ENST00000413811.3</td>\n",
       "      <td>ENST00000413811.3</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>UTR5:1-1018</td>\n",
       "      <td>CDS:1019-2611</td>\n",
       "      <td>UTR3:2612-2661</td>\n",
       "      <td>ACATTCTTGATTGGTCTCATGCCTTTGTGGTTGTAAATGTGCCTGG...</td>\n",
       "      <td>ATGCAGGTGGTTTTGGACAGTAAATTAGACCTGCCATGTGAACTGC...</td>\n",
       "      <td>AACTTATCACTAGGCAGAACTGGGTTTGATGCTTTGTCAACTGAAA...</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Transcript             gene_id  \\\n",
       "0  ENST00000373020  ENSG00000000003.15   \n",
       "1  ENST00000373031   ENSG00000000005.6   \n",
       "2  ENST00000371588  ENSG00000000419.14   \n",
       "3  ENST00000367772  ENSG00000000457.14   \n",
       "4  ENST00000413811  ENSG00000000460.17   \n",
       "\n",
       "                                                name  \\\n",
       "0  >ENST00000373020.9|ENSG00000000003.15|OTTHUMG0...   \n",
       "1  >ENST00000373031.5|ENSG00000000005.6|OTTHUMG00...   \n",
       "2  >ENST00000371588.10|ENSG00000000419.14|OTTHUMG...   \n",
       "3  >ENST00000367772.8|ENSG00000000457.14|OTTHUMG0...   \n",
       "4  >ENST00000413811.3|ENSG00000000460.17|OTTHUMG0...   \n",
       "\n",
       "                                                 seq       transcript_id  \\\n",
       "0  AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...   ENST00000373020.9   \n",
       "1  AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...   ENST00000373031.5   \n",
       "2  AGTTCCGCCATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGT...  ENST00000371588.10   \n",
       "3  TGTCCCGTTTCCGGACCCGTCTCTATGGTGTAGGAGAAACCCGGCC...   ENST00000367772.8   \n",
       "4  ACATTCTTGATTGGTCTCATGCCTTTGTGGTTGTAAATGTGCCTGG...   ENST00000413811.3   \n",
       "\n",
       "          gene_id_raw gene_name    UTR5_regs       CDS_regs       UTR3_regs  \\\n",
       "0   ENST00000373020.9    TSPAN6   UTR5:1-112    CDS:113-850   UTR3:851-3768   \n",
       "1   ENST00000373031.5      TNMD    UTR5:1-83    CDS:84-1037  UTR3:1038-1205   \n",
       "2  ENST00000371588.10      DPM1     UTR5:1-9     CDS:10-792   UTR3:793-1054   \n",
       "3   ENST00000367772.8     SCYL3   UTR5:1-222   CDS:223-2451  UTR3:2452-3090   \n",
       "4   ENST00000413811.3  C1orf112  UTR5:1-1018  CDS:1019-2611  UTR3:2612-2661   \n",
       "\n",
       "                                                UTR5  \\\n",
       "0  AGTTGTGGACGCTCGTAAGTTTTCGGCAGTTTCCGGGGAGACTCGG...   \n",
       "1  AGCCGACTCACTTGCAACTCCACCTCAGCAGTGGTCTCTCAGTCCT...   \n",
       "2                                          AGTTCCGCC   \n",
       "3  TGTCCCGTTTCCGGACCCGTCTCTATGGTGTAGGAGAAACCCGGCC...   \n",
       "4  ACATTCTTGATTGGTCTCATGCCTTTGTGGTTGTAAATGTGCCTGG...   \n",
       "\n",
       "                                                 CDS  \\\n",
       "0  ATGGCGTCCCCGTCTCGGAGACTGCAGACTAAACCAGTCATTACTT...   \n",
       "1  ATGGCAAAGAATCCTCCAGAGAATTGTGAAGACTGTCACATTCTAA...   \n",
       "2  ATGGCCTCCTTGGAAGTCAGTCGTAGTCCTCGCAGGTCTCGGCGGG...   \n",
       "3  ATGGGATCAGAGAACAGTGCTTTAAAGAGCTATACACTGAGAGAAC...   \n",
       "4  ATGCAGGTGGTTTTGGACAGTAAATTAGACCTGCCATGTGAACTGC...   \n",
       "\n",
       "                                                UTR3          Esemble  \n",
       "0  CCCAATGTATCTGTGGGCCTATTCCTCTCTACCTTTAAGGACATTT...  ENSG00000000003  \n",
       "1  TAGGAGGTTTGAGCTCAAATGCTTAAACTGCTGGCAACATATAATA...  ENSG00000000005  \n",
       "2  AAGAAAGATACTCATTTATAGTTACGTTCATTTCAGGTTAAACATG...  ENSG00000000419  \n",
       "3  CAATAGATGTGAGTTAAACTTTAGGAAAAAGGATTCCCTTTTTTTA...  ENSG00000000457  \n",
       "4  AACTTATCACTAGGCAGAACTGGGTTTGATGCTTTGTCAACTGAAA...  ENSG00000000460  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_ts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "09699636",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_126156/2100682807.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cellines_pair_exp['pair_A_UTR5'] = cellines_pair_exp['pair_A'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])[\"UTR5\"].values[0])\n",
      "/tmp/ipykernel_126156/2100682807.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cellines_pair_exp['pair_A_UTR3'] = cellines_pair_exp['pair_A'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])[\"UTR3\"].values[0])\n",
      "/tmp/ipykernel_126156/2100682807.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cellines_pair_exp['pair_A_CDS'] = cellines_pair_exp['pair_A'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])['CDS'].values[0])\n",
      "/tmp/ipykernel_126156/2100682807.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cellines_pair_exp['pair_B_UTR5'] = cellines_pair_exp['pair_B'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])[\"UTR5\"].values[0])\n",
      "/tmp/ipykernel_126156/2100682807.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cellines_pair_exp['pair_B_UTR3'] = cellines_pair_exp['pair_B'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])[\"UTR3\"].values[0])\n",
      "/tmp/ipykernel_126156/2100682807.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cellines_pair_exp['pair_B_CDS'] = cellines_pair_exp['pair_B'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])['CDS'].values[0])\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "cellines_pair_exp['pair_A_UTR5'] = cellines_pair_exp['pair_A'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])[\"UTR5\"].values[0])\n",
    "cellines_pair_exp['pair_A_UTR3'] = cellines_pair_exp['pair_A'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])[\"UTR3\"].values[0])\n",
    "cellines_pair_exp['pair_A_CDS'] = cellines_pair_exp['pair_A'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])['CDS'].values[0])\n",
    "\n",
    "cellines_pair_exp['pair_B_UTR5'] = cellines_pair_exp['pair_B'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])[\"UTR5\"].values[0])\n",
    "cellines_pair_exp['pair_B_UTR3'] = cellines_pair_exp['pair_B'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])[\"UTR3\"].values[0])\n",
    "cellines_pair_exp['pair_B_CDS'] = cellines_pair_exp['pair_B'].apply(lambda x: human_ts.reset_index().set_index('Esemble').reindex([x])['CDS'].values[0])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "994026ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "cellines_pair_exp['pair_B_UTR3'].isnull().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "80006a58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 生成prompts\n",
    "\n",
    "prompts = []\n",
    "for index, row in cellines_pair_exp.iterrows():\n",
    "    \n",
    "    for item in ['pair_A_CDS', 'pair_B_CDS']:\n",
    "        \n",
    "        ppt = str(Seq(row[item]).translate())\n",
    "        for jtem in ['UTR5', 'UTR3','CDS']:\n",
    "            \n",
    "            pairA_token = row[f'pair_A_{jtem}']\n",
    "            pairB_token = row[f'pair_B_{jtem}']\n",
    "            label = row['pair_label']\n",
    "            g_type=jtem\n",
    "            origin = 'cellines_cluster'\n",
    "            \n",
    "            prompts.append([ppt, pairA_token, pairB_token, g_type, origin, label])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "d0a33028",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>pair_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAFSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...</td>\n",
       "      <td>ATGGAGTTTCGCTCTTGTTGCTCAGGCTGGAGTGCAATGGTGTGAT...</td>\n",
       "      <td>AAACCCGGAAGCGGATCGCATGGAGTGATGGTCCCACCGCAGCGCC...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAFSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...</td>\n",
       "      <td>TGAATGTGGTGAGGTTTTTAATCAACAAGCACACCTTGCAGGTCAT...</td>\n",
       "      <td>AAATATGAAGAATGTGACAAAGTTTACAGTTGTAAATCAAGTCTTG...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAFSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...</td>\n",
       "      <td>ATGGCTTTTTCTCAGGGTCTATTGACATTCAGGGATGTGGCCATAG...</td>\n",
       "      <td>ATGGCTCTTTCTCAGGGTCTATTGACATTCAGGGATGTGGCCATAG...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MALSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...</td>\n",
       "      <td>ATGGAGTTTCGCTCTTGTTGCTCAGGCTGGAGTGCAATGGTGTGAT...</td>\n",
       "      <td>AAACCCGGAAGCGGATCGCATGGAGTGATGGTCCCACCGCAGCGCC...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MALSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...</td>\n",
       "      <td>TGAATGTGGTGAGGTTTTTAATCAACAAGCACACCTTGCAGGTCAT...</td>\n",
       "      <td>AAATATGAAGAATGTGACAAAGTTTACAGTTGTAAATCAAGTCTTG...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td>ATGGCATTTTGAAGCCCAGTCATTCTCTAAAAAGGCCCTTTTTAGG...</td>\n",
       "      <td>GCTTACCTCTACAGTACACTACCGCAAAACCAACGGCCCTTTTCAG...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td>ATGTCAGGACGCGGCAAAGGAGGTAAGGGCCTGGGGAAAGGGGGTG...</td>\n",
       "      <td>ATGTCTGGCCGCGGTAAGGGCGGAAAGGGTCTAGGTAAGGGTGGCG...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td>AGACCTTTGTTCTCTGACCACTTGATA</td>\n",
       "      <td>ACTTGCTCTTGGTTCACTTCTTGGGAAGTC</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td>ATGGCATTTTGAAGCCCAGTCATTCTCTAAAAAGGCCCTTTTTAGG...</td>\n",
       "      <td>GGTTGCTGATTTCTCCACAGCTTGCATTTCTGAACCAAAGGCCCTT...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...</td>\n",
       "      <td>ATGTCAGGACGCGGCAAAGGAGGTAAGGGCCTGGGGAAAGGGGGTG...</td>\n",
       "      <td>ATGTCTGGACGTGGTAAGGGCGGGAAGGGTTTGGGTAAGGGGGGTG...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>cellines_cluster</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                prompt  \\\n",
       "0    MAFSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...   \n",
       "1    MAFSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...   \n",
       "2    MAFSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...   \n",
       "3    MALSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...   \n",
       "4    MALSQGLLTFRDVAIEFSQEEWKCLDPAQRTLYRDVMLENYRNLVS...   \n",
       "..                                                 ...   \n",
       "229  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "230  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "234  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "235  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "236  MSGRGKGGKGLGKGGAKRHRKVLRDNIQGITKPAIRRLARRGGVKR...   \n",
       "\n",
       "                                                pair_A  \\\n",
       "0    ATGGAGTTTCGCTCTTGTTGCTCAGGCTGGAGTGCAATGGTGTGAT...   \n",
       "1    TGAATGTGGTGAGGTTTTTAATCAACAAGCACACCTTGCAGGTCAT...   \n",
       "2    ATGGCTTTTTCTCAGGGTCTATTGACATTCAGGGATGTGGCCATAG...   \n",
       "3    ATGGAGTTTCGCTCTTGTTGCTCAGGCTGGAGTGCAATGGTGTGAT...   \n",
       "4    TGAATGTGGTGAGGTTTTTAATCAACAAGCACACCTTGCAGGTCAT...   \n",
       "..                                                 ...   \n",
       "229  ATGGCATTTTGAAGCCCAGTCATTCTCTAAAAAGGCCCTTTTTAGG...   \n",
       "230  ATGTCAGGACGCGGCAAAGGAGGTAAGGGCCTGGGGAAAGGGGGTG...   \n",
       "234                        AGACCTTTGTTCTCTGACCACTTGATA   \n",
       "235  ATGGCATTTTGAAGCCCAGTCATTCTCTAAAAAGGCCCTTTTTAGG...   \n",
       "236  ATGTCAGGACGCGGCAAAGGAGGTAAGGGCCTGGGGAAAGGGGGTG...   \n",
       "\n",
       "                                                pair_B seq_type  \\\n",
       "0    AAACCCGGAAGCGGATCGCATGGAGTGATGGTCCCACCGCAGCGCC...     UTR5   \n",
       "1    AAATATGAAGAATGTGACAAAGTTTACAGTTGTAAATCAAGTCTTG...     UTR3   \n",
       "2    ATGGCTCTTTCTCAGGGTCTATTGACATTCAGGGATGTGGCCATAG...      CDS   \n",
       "3    AAACCCGGAAGCGGATCGCATGGAGTGATGGTCCCACCGCAGCGCC...     UTR5   \n",
       "4    AAATATGAAGAATGTGACAAAGTTTACAGTTGTAAATCAAGTCTTG...     UTR3   \n",
       "..                                                 ...      ...   \n",
       "229  GCTTACCTCTACAGTACACTACCGCAAAACCAACGGCCCTTTTCAG...     UTR3   \n",
       "230  ATGTCTGGCCGCGGTAAGGGCGGAAAGGGTCTAGGTAAGGGTGGCG...      CDS   \n",
       "234                     ACTTGCTCTTGGTTCACTTCTTGGGAAGTC     UTR5   \n",
       "235  GGTTGCTGATTTCTCCACAGCTTGCATTTCTGAACCAAAGGCCCTT...     UTR3   \n",
       "236  ATGTCTGGACGTGGTAAGGGCGGGAAGGGTTTGGGTAAGGGGGGTG...      CDS   \n",
       "\n",
       "               origin  pair_label  \n",
       "0    cellines_cluster          -1  \n",
       "1    cellines_cluster          -1  \n",
       "2    cellines_cluster          -1  \n",
       "3    cellines_cluster          -1  \n",
       "4    cellines_cluster          -1  \n",
       "..                ...         ...  \n",
       "229  cellines_cluster          -1  \n",
       "230  cellines_cluster          -1  \n",
       "234  cellines_cluster          -1  \n",
       "235  cellines_cluster          -1  \n",
       "236  cellines_cluster          -1  \n",
       "\n",
       "[204 rows x 6 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "cellines_prompts = pd.DataFrame(prompts, columns=['prompt', 'pair_A', 'pair_B', 'seq_type', 'origin', 'pair_label'])\n",
    "\n",
    "cellines_prompts = cellines_prompts.drop_duplicates(subset=['prompt', 'pair_A', 'pair_B', 'seq_type'])\n",
    "\n",
    "cellines_prompts = cellines_prompts[ \n",
    "                                    cellines_prompts['pair_A']  != cellines_prompts['pair_B']\n",
    "                                    ]\n",
    "  \n",
    "cellines_prompts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e7d09c4",
   "metadata": {},
   "source": [
    "##### pair_utr5_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "86fb8523",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "cutoff = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "eb37d9f4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>pair_A_UTR5</th>\n",
       "      <th>pair_A_UTR3</th>\n",
       "      <th>pair_A_CDS</th>\n",
       "      <th>pair_B_UTR5</th>\n",
       "      <th>pair_B_UTR3</th>\n",
       "      <th>pair_B_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_52</td>\n",
       "      <td>1</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_56</td>\n",
       "      <td>1</td>\n",
       "      <td>1.650000</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_60</td>\n",
       "      <td>1</td>\n",
       "      <td>1.650000</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAATCAACACAGACACCAGACCACCAGCACCAGAACACACAGAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_64</td>\n",
       "      <td>1</td>\n",
       "      <td>3.666667</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAATCACCACCAGAACACACACACAGAAACACACAGACACCAGA...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_68</td>\n",
       "      <td>1</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>GGGAATTCGTCGACTCAGGAACTATTCTTCTGGTCCCCACAGACTC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  pair_A   pair_B  pair_label  pair_diff  \\\n",
       "0           0  gene_0  gene_52           1   3.300000   \n",
       "1           1  gene_0  gene_56           1   1.650000   \n",
       "2           2  gene_0  gene_60           1   1.650000   \n",
       "3           3  gene_0  gene_64           1   3.666667   \n",
       "4           4  gene_0  gene_68           1   5.500000   \n",
       "\n",
       "                                         pair_A_UTR5  \\\n",
       "0  GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1  GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2  GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "3  GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "4  GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "\n",
       "                                         pair_A_UTR3  \\\n",
       "0  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                          pair_A_CDS  \\\n",
       "0  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "1  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "2  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "3  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "4  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "\n",
       "                                         pair_B_UTR5  \\\n",
       "0  GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...   \n",
       "1  GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2  GGGAATCAACACAGACACCAGACCACCAGCACCAGAACACACAGAC...   \n",
       "3  GGGAATCACCACCAGAACACACACACAGAAACACACAGACACCAGA...   \n",
       "4  GGGAATTCGTCGACTCAGGAACTATTCTTCTGGTCCCCACAGACTC...   \n",
       "\n",
       "                                         pair_B_UTR3  \\\n",
       "0  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                          pair_B_CDS  \n",
       "0  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "1  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "2  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "3  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  \n",
       "4  ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...  "
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "pair_utr5_exp = pd.read_csv(f'data/experiment_UTR5_pair_exp_cutoff{cutoff}.tsv', sep='\\t')\n",
    "pair_utr5_exp = pair_utr5_exp[pair_utr5_exp['pair_label'] != 0]\n",
    "pair_utr5_exp.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "417daeac-4ad1-47c5-92ff-3e438000db15",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "all([row[f'pair_A_{tem}']==row[f'pair_B_{tem}'] or row[f'pair_A_{tem}'][4:-4]==row[f'pair_B_{tem}'][4:-4] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "\n",
    "# jtem\n",
    "\n",
    "# row[f'pair_A_UTR5']==row[f'pair_B_UTR5']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "ffcde664",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 生成prompts\n",
    "\n",
    "prompts = []\n",
    "for index, row in pair_utr5_exp.iterrows():\n",
    "    \n",
    "    for item in ['pair_A_CDS', 'pair_B_CDS']:\n",
    "        \n",
    "        ppt = str(Seq(row[item]).translate())\n",
    "        \n",
    "        for jtem in [ 'mRNA', 'UTR5','CDS', 'UTR3']:\n",
    "            \n",
    "            if jtem == 'mRNA':\n",
    "                pairA_token = '_'.join([row[f'pair_A_UTR5'], row[f'pair_A_CDS'], row[f'pair_A_UTR3']  ])\n",
    "                pairB_token = '_'.join([row[f'pair_B_UTR5'], row[f'pair_B_CDS'], row[f'pair_B_UTR3']  ])\n",
    "                exact_diff = (pairA_token == pairB_token)\n",
    "                roughly_diff = (pairA_token == pairB_token) \n",
    "            else:\n",
    "                exact_diff = all([row[f'pair_A_{jtem}']==row[f'pair_B_{jtem}'] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "                roughly_diff = all([row[f'pair_A_{tem}']==row[f'pair_B_{tem}'] or row[f'pair_A_{tem}'][4:-4]==row[f'pair_B_{tem}'][4:-4] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "                pairA_token = row[f'pair_A_{jtem}']\n",
    "                pairB_token = row[f'pair_B_{jtem}']\n",
    "                \n",
    "            label = row['pair_label']\n",
    "            value = row['pair_diff']\n",
    "            g_type=jtem\n",
    "            origin = 'utr5_exp'\n",
    "            \n",
    "            prompts.append([ppt, pairA_token, pairB_token, g_type, origin, label, value, exact_diff, roughly_diff])\n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "93f69280-6477-4ebe-8233-c650d559ffb3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    17560\n",
       "Name: exact_diff, dtype: int64"
      ]
     },
     "execution_count": 289,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR5_prompts_raw[UTR5_prompts_raw['seq_type']=='UTR5']['exact_diff'].value_counts()\n",
    "\n",
    "# prompts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "21c07253",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>exact_diff</th>\n",
       "      <th>roughly_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.30</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.30</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.65</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.65</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAATCAACACAGACACCAGACCACCAGCACCAGAACACACAGAC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.65</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70200</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>GGGATTCCCCCACCACCAGAGCACCAGTCCCGCAGTCGGCGTCCAG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.05</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70208</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>GGGATTCCCCGCCTCCTCCTGTCCCGCAGTCGGCGTCCAGCGGCTC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.84</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70216</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>GGGGCCACCAAGTTGTTACATGAAATCTGCAGTTTCATAATTTCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.15</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70224</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>GGGGCCAGCAGAAGTTGTTACAAATCTGCAGTTTCATTTCCGTGGG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>6.30</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70232</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...</td>\n",
       "      <td>GGGGCCCAACACACACCAGACCACCAGAGCAGACCAAGTTGTTACA...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.84</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4838 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  prompt  \\\n",
       "0      MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1      MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "8      MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "9      MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "16     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "...                                                  ...   \n",
       "70200  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "70208  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "70216  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "70224  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "70232  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "\n",
       "                                                  pair_A  \\\n",
       "0      GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1      GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "8      GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "9      GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "16     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "...                                                  ...   \n",
       "70200  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "70208  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "70216  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "70224  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "70232  GGGTCTCAACACAACATATACAAAACAAACGAATCTCAAGCAATCA...   \n",
       "\n",
       "                                                  pair_B seq_type    origin  \\\n",
       "0      GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...     mRNA  utr5_exp   \n",
       "1      GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...     UTR5  utr5_exp   \n",
       "8      GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...     mRNA  utr5_exp   \n",
       "9      GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...     UTR5  utr5_exp   \n",
       "16     GGGAATCAACACAGACACCAGACCACCAGCACCAGAACACACAGAC...     mRNA  utr5_exp   \n",
       "...                                                  ...      ...       ...   \n",
       "70200  GGGATTCCCCCACCACCAGAGCACCAGTCCCGCAGTCGGCGTCCAG...     mRNA  utr5_exp   \n",
       "70208  GGGATTCCCCGCCTCCTCCTGTCCCGCAGTCGGCGTCCAGCGGCTC...     mRNA  utr5_exp   \n",
       "70216  GGGGCCACCAAGTTGTTACATGAAATCTGCAGTTTCATAATTTCCG...     mRNA  utr5_exp   \n",
       "70224  GGGGCCAGCAGAAGTTGTTACAAATCTGCAGTTTCATTTCCGTGGG...     mRNA  utr5_exp   \n",
       "70232  GGGGCCCAACACACACCAGACCACCAGAGCAGACCAAGTTGTTACA...     mRNA  utr5_exp   \n",
       "\n",
       "       pair_label  pair_diff  exact_diff  roughly_diff  \n",
       "0               1       3.30       False         False  \n",
       "1               1       3.30       False          True  \n",
       "8               1       1.65       False         False  \n",
       "9               1       1.65       False          True  \n",
       "16              1       1.65       False         False  \n",
       "...           ...        ...         ...           ...  \n",
       "70200           1       1.05       False         False  \n",
       "70208          -1       0.84       False         False  \n",
       "70216           1       3.15       False         False  \n",
       "70224           1       6.30       False         False  \n",
       "70232          -1       0.84       False         False  \n",
       "\n",
       "[4838 rows x 9 columns]"
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "UTR5_prompts = pd.DataFrame(prompts, columns=['prompt', 'pair_A', 'pair_B', 'seq_type', 'origin', 'pair_label', 'pair_diff', 'exact_diff', 'roughly_diff'])\n",
    "\n",
    "UTR5_prompts_raw = UTR5_prompts\n",
    "\n",
    "UTR5_prompts = UTR5_prompts.drop_duplicates(subset=['prompt', 'pair_A', 'pair_B', 'seq_type'])\n",
    "\n",
    "UTR5_prompts = UTR5_prompts[ \n",
    "                                    UTR5_prompts['pair_A']  != UTR5_prompts['pair_B']\n",
    "                                    ]\n",
    "\n",
    "UTR5_prompts\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe9fa898",
   "metadata": {},
   "source": [
    "#####  pair_utr3_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "a53774e7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>pair_A_UTR5</th>\n",
       "      <th>pair_A_UTR3</th>\n",
       "      <th>pair_A_CDS</th>\n",
       "      <th>pair_B_UTR5</th>\n",
       "      <th>pair_B_UTR3</th>\n",
       "      <th>pair_B_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.065332</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTGTACCATTTACCACTGTGACTAAATAAAGGGTGTGTCTGTTGT...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_2</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.418709</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CCCCCAATATCTGCTCCTCTGGTTTGGCTTCCCTCAGAGCCTGGGT...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_3</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.434997</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CCCCGTCTGCTCACCTGCAAAATAAAAGGCTTTTCGCCCCTTGTCC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_4</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.433300</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CCCCGTCTGCTCCTCTAGGCTGGCCCCTTGTCCACCCCTCTCCACA...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_5</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.421146</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CCCCTTCTGTTCTGCTCCTCTAATAAAGGCTCTGATTTTCCCCCAT...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  pair_A  pair_B  pair_label  pair_diff  \\\n",
       "0           0  gene_0  gene_1           1   1.065332   \n",
       "1           1  gene_0  gene_2          -1   0.418709   \n",
       "2           2  gene_0  gene_3          -1   0.434997   \n",
       "3           3  gene_0  gene_4          -1   0.433300   \n",
       "4           4  gene_0  gene_5          -1   0.421146   \n",
       "\n",
       "                                         pair_A_UTR5  \\\n",
       "0  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "1  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "\n",
       "                                         pair_A_UTR3  \\\n",
       "0  ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "1  ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "2  ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "3  ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "4  ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "\n",
       "                                          pair_A_CDS  \\\n",
       "0  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "1  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "2  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "3  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "4  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...   \n",
       "\n",
       "                                         pair_B_UTR5  \\\n",
       "0  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "1  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "\n",
       "                                         pair_B_UTR3  \\\n",
       "0  ACTGTACCATTTACCACTGTGACTAAATAAAGGGTGTGTCTGTTGT...   \n",
       "1  CCCCCAATATCTGCTCCTCTGGTTTGGCTTCCCTCAGAGCCTGGGT...   \n",
       "2  CCCCGTCTGCTCACCTGCAAAATAAAAGGCTTTTCGCCCCTTGTCC...   \n",
       "3  CCCCGTCTGCTCCTCTAGGCTGGCCCCTTGTCCACCCCTCTCCACA...   \n",
       "4  CCCCTTCTGTTCTGCTCCTCTAATAAAGGCTCTGATTTTCCCCCAT...   \n",
       "\n",
       "                                          pair_B_CDS  \n",
       "0  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "1  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "2  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "3  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  \n",
       "4  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTGCTCCTGTCTGGAG...  "
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "pair_utr3_exp = pd.read_csv(f'data/experiment_UTR3_pair_exp_cutoff{cutoff}.tsv', sep='\\t')\n",
    "\n",
    "pair_utr3_exp = pair_utr3_exp[pair_utr3_exp['pair_label'] != 0]\n",
    "pair_utr3_exp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "id": "f22d4017-d170-409d-a8e7-76f63f497858",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# 生成prompts\n",
    "\n",
    "prompts = []\n",
    "for index, row in pair_utr3_exp.iterrows():\n",
    "    \n",
    "    for item in ['pair_A_CDS', 'pair_B_CDS']:\n",
    "        \n",
    "        ppt = str(Seq(row[item]).translate())\n",
    "        \n",
    "        for jtem in [ 'mRNA', 'UTR5','CDS', 'UTR3']:\n",
    "            \n",
    "            if jtem == 'mRNA':\n",
    "                pairA_token = '_'.join([row[f'pair_A_UTR5'], row[f'pair_A_CDS'], row[f'pair_A_UTR3']  ])\n",
    "                pairB_token = '_'.join([row[f'pair_B_UTR5'], row[f'pair_B_CDS'], row[f'pair_B_UTR3']  ])\n",
    "                exact_diff = (pairA_token == pairB_token)\n",
    "                roughly_diff = (pairA_token == pairB_token) \n",
    "            else:\n",
    "                exact_diff = all([row[f'pair_A_{jtem}']==row[f'pair_B_{jtem}'] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "                roughly_diff = all([row[f'pair_A_{tem}']==row[f'pair_B_{tem}'] or row[f'pair_A_{tem}'][4:-4]==row[f'pair_B_{tem}'][4:-4] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "                pairA_token = row[f'pair_A_{jtem}']\n",
    "                pairB_token = row[f'pair_B_{jtem}']\n",
    "                \n",
    "            label = row['pair_label']\n",
    "            value = row['pair_diff']\n",
    "            g_type=jtem\n",
    "            origin = 'utr3_exp'\n",
    "            \n",
    "            prompts.append([ppt, pairA_token, pairB_token, g_type, origin, label, value, exact_diff, roughly_diff])\n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "id": "1792d7ce",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>exact_diff</th>\n",
       "      <th>roughly_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.065332</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>ACTGTACCATTTACCACTGTGACTAAATAAAGGGTGTGTCTGTTGT...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.065332</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.418709</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...</td>\n",
       "      <td>CCCCCAATATCTGCTCCTCTGGTTTGGCTTCCCTCAGAGCCTGGGT...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.418709</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.434997</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54728</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.472478</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54736</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.726330</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54744</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.318426</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54752</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.653769</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54760</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.460665</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12628 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  prompt  \\\n",
       "0      MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "3      MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "8      MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "11     MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "16     MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "...                                                  ...   \n",
       "54728  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "54736  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "54744  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "54752  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "54760  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "\n",
       "                                                  pair_A  \\\n",
       "0      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3      ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "8      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "11     ACTAATACTGGTCTGCGGCAGCTAAATAAGATCAATAAATTAATTC...   \n",
       "16     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "...                                                  ...   \n",
       "54728  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "54736  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "54744  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "54752  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "54760  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "\n",
       "                                                  pair_B seq_type    origin  \\\n",
       "0      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "3      ACTGTACCATTTACCACTGTGACTAAATAAAGGGTGTGTCTGTTGT...     UTR3  utr3_exp   \n",
       "8      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "11     CCCCCAATATCTGCTCCTCTGGTTTGGCTTCCCTCAGAGCCTGGGT...     UTR3  utr3_exp   \n",
       "16     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "...                                                  ...      ...       ...   \n",
       "54728  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "54736  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "54744  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "54752  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "54760  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "\n",
       "       pair_label  pair_diff  exact_diff  roughly_diff  \n",
       "0               1   1.065332       False         False  \n",
       "3               1   1.065332       False          True  \n",
       "8              -1   0.418709       False         False  \n",
       "11             -1   0.418709       False          True  \n",
       "16             -1   0.434997       False         False  \n",
       "...           ...        ...         ...           ...  \n",
       "54728          -1   0.472478       False         False  \n",
       "54736          -1   0.726330       False         False  \n",
       "54744           1   1.318426       False         False  \n",
       "54752          -1   0.653769       False         False  \n",
       "54760          -1   0.460665       False         False  \n",
       "\n",
       "[12628 rows x 9 columns]"
      ]
     },
     "execution_count": 302,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "UTR3_prompts = pd.DataFrame(prompts, columns=['prompt', 'pair_A', 'pair_B', 'seq_type', 'origin', 'pair_label', 'pair_diff', 'exact_diff', 'roughly_diff'])\n",
    "\n",
    "UTR3_prompts_raw = UTR3_prompts\n",
    "\n",
    "UTR3_prompts = UTR3_prompts.drop_duplicates(subset=['prompt', 'pair_A', 'pair_B', 'seq_type'])\n",
    "\n",
    "UTR3_prompts = UTR3_prompts[ \n",
    "                                    UTR3_prompts['pair_A']  != UTR3_prompts['pair_B']\n",
    "                                    ]\n",
    "\n",
    "UTR3_prompts\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef381c23",
   "metadata": {},
   "source": [
    "##### pair_CDS_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "d6e3ea2a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>pair_A_UTR5</th>\n",
       "      <th>pair_A_UTR3</th>\n",
       "      <th>pair_A_CDS</th>\n",
       "      <th>pair_B_UTR5</th>\n",
       "      <th>pair_B_UTR3</th>\n",
       "      <th>pair_B_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.831660</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_7</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.751856</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTCCTGCTGTCTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_8</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.987945</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_9</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.835812</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>gene_0</td>\n",
       "      <td>gene_15</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.739491</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>CGGGTCACGGCCAACTTCTGTCGGGAGCCGGCGAGCGGTCGCCCTC...</td>\n",
       "      <td>GCGGCAGGCGGGAGGGCGAGCGGGAGGGAAGGGAGTGGCGGGTGCG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  pair_A   pair_B  pair_label  pair_diff  \\\n",
       "0           0  gene_0   gene_1          -1   0.831660   \n",
       "1           1  gene_0   gene_7          -1   0.751856   \n",
       "2           2  gene_0   gene_8          -1   0.987945   \n",
       "3           3  gene_0   gene_9          -1   0.835812   \n",
       "4           4  gene_0  gene_15          -1   0.739491   \n",
       "\n",
       "                                         pair_A_UTR5  \\\n",
       "0  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "1  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "2  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "3  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "4  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "\n",
       "                                         pair_A_UTR3  \\\n",
       "0  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                          pair_A_CDS  \\\n",
       "0  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "1  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "2  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "3  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "4  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "\n",
       "                                         pair_B_UTR5  \\\n",
       "0  AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "1  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4  CGGGTCACGGCCAACTTCTGTCGGGAGCCGGCGAGCGGTCGCCCTC...   \n",
       "\n",
       "                                         pair_B_UTR3  \\\n",
       "0  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4  GCGGCAGGCGGGAGGGCGAGCGGGAGGGAAGGGAGTGGCGGGTGCG...   \n",
       "\n",
       "                                          pair_B_CDS  \n",
       "0  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...  \n",
       "1  ATGAGAGTGACCGCCCCAAGGACCCTGATTCTCCTGCTGTCTGGAG...  \n",
       "2  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...  \n",
       "3  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...  \n",
       "4  ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...  "
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "pair_CDS_exp = pd.read_csv(f'data/experiment_CDS_pair_exp_cutoff{cutoff}.tsv', sep='\\t')\n",
    "pair_CDS_exp = pair_CDS_exp[pair_CDS_exp['pair_label'] != 0]\n",
    "pair_CDS_exp.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "14da875c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 生成prompts\n",
    "\n",
    "prompts = []\n",
    "for index, row in pair_CDS_exp.iterrows():\n",
    "    \n",
    "    for item in ['pair_A_CDS', 'pair_B_CDS']:\n",
    "        \n",
    "        ppt = str(Seq(row[item]).translate())\n",
    "        \n",
    "        for jtem in [ 'mRNA', 'UTR5','CDS', 'UTR3']:\n",
    "            \n",
    "            if jtem == 'mRNA':\n",
    "                pairA_token = '_'.join([row[f'pair_A_UTR5'], row[f'pair_A_CDS'], row[f'pair_A_UTR3']  ])\n",
    "                pairB_token = '_'.join([row[f'pair_B_UTR5'], row[f'pair_B_CDS'], row[f'pair_B_UTR3']  ])\n",
    "                exact_diff = (pairA_token == pairB_token)\n",
    "                roughly_diff = (pairA_token == pairB_token) \n",
    "            else:\n",
    "                exact_diff = all([row[f'pair_A_{jtem}']==row[f'pair_B_{jtem}'] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "                roughly_diff = all([row[f'pair_A_{tem}']==row[f'pair_B_{tem}'] or row[f'pair_A_{tem}'][4:-4]==row[f'pair_B_{tem}'][4:-4] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "                pairA_token = row[f'pair_A_{jtem}']\n",
    "                pairB_token = row[f'pair_B_{jtem}']\n",
    "                \n",
    "            label = row['pair_label']\n",
    "            value = row['pair_diff']\n",
    "            g_type=jtem\n",
    "            origin = 'cds_exp'\n",
    "            \n",
    "            prompts.append([ppt, pairA_token, pairB_token, g_type, origin, label, value, exact_diff, roughly_diff])\n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "9f298c52",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>exact_diff</th>\n",
       "      <th>roughly_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.831660</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.831660</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.751856</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...</td>\n",
       "      <td>AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.751856</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...</td>\n",
       "      <td>ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...</td>\n",
       "      <td>ATGAGAGTGACCGCCCCAAGGACCCTGATTCTCCTGCTGTCTGGAG...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.751856</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>GCTGGTCCCAATTCTGGTCGAGCTCGGTCAACGGACAGTTCTGTCC...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.160315</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGCTCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>20.071805</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1747</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>TCGAGCTCGGACAGTTCTGTACCGTGCGGGCCAAGGGCGAAGCGTC...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>20.071805</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1752</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.164065</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1755</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.164065</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>400 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 prompt  \\\n",
       "0     MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...   \n",
       "2     MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...   \n",
       "8     MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...   \n",
       "9     MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...   \n",
       "10    MRVTAPRTLILLLSGALALTETWAGSGGGGSGGIDEILDFFYGYQI...   \n",
       "...                                                 ...   \n",
       "1739  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1744  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1747  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1752  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1755  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "\n",
       "                                                 pair_A  \\\n",
       "0     AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "2     ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "8     AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "9     AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...   \n",
       "10    ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTGCTGTCGGGTG...   \n",
       "...                                                 ...   \n",
       "1739  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "1744  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "1747  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "1752  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "1755  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "\n",
       "                                                 pair_B seq_type   origin  \\\n",
       "0     AAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTTTCG...     mRNA  cds_exp   \n",
       "2     ATGCGGGTCACGGCCCCTAGAACACTGATTCTTCTTCTGTCGGGTG...      CDS  cds_exp   \n",
       "8     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  cds_exp   \n",
       "9     AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     UTR5  cds_exp   \n",
       "10    ATGAGAGTGACCGCCCCAAGGACCCTGATTCTCCTGCTGTCTGGAG...      CDS  cds_exp   \n",
       "...                                                 ...      ...      ...   \n",
       "1739  GCTGGTCCCAATTCTGGTCGAGCTCGGTCAACGGACAGTTCTGTCC...     UTR3  cds_exp   \n",
       "1744  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGCTCG...     mRNA  cds_exp   \n",
       "1747  TCGAGCTCGGACAGTTCTGTACCGTGCGGGCCAAGGGCGAAGCGTC...     UTR3  cds_exp   \n",
       "1752  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...     mRNA  cds_exp   \n",
       "1755  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...     UTR3  cds_exp   \n",
       "\n",
       "      pair_label  pair_diff  exact_diff  roughly_diff  \n",
       "0             -1   0.831660       False         False  \n",
       "2             -1   0.831660       False          True  \n",
       "8             -1   0.751856       False         False  \n",
       "9             -1   0.751856       False         False  \n",
       "10            -1   0.751856       False         False  \n",
       "...          ...        ...         ...           ...  \n",
       "1739           1   3.160315       False         False  \n",
       "1744           1  20.071805       False         False  \n",
       "1747           1  20.071805       False         False  \n",
       "1752           1   1.164065       False         False  \n",
       "1755           1   1.164065       False          True  \n",
       "\n",
       "[400 rows x 9 columns]"
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "CDS_prompts = pd.DataFrame(prompts, columns=['prompt', 'pair_A', 'pair_B', 'seq_type', 'origin', 'pair_label', 'pair_diff', 'exact_diff', 'roughly_diff'])\n",
    "\n",
    "CDS_prompts_raw = CDS_prompts\n",
    "\n",
    "CDS_prompts = CDS_prompts.drop_duplicates(subset=['prompt', 'pair_A', 'pair_B', 'seq_type'])\n",
    "\n",
    "CDS_prompts = CDS_prompts[ \n",
    "                                    CDS_prompts['pair_A']  != CDS_prompts['pair_B']\n",
    "                                    ]\n",
    "\n",
    "CDS_prompts\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bd0a2bf-5eab-4497-ba86-c4b1bb130c7b",
   "metadata": {},
   "source": [
    "##### pair_CDS2_expt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "c1d0af4f-75b1-43d3-bc11-79860418e4c8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>pair_A_UTR5</th>\n",
       "      <th>pair_A_UTR3</th>\n",
       "      <th>pair_A_CDS</th>\n",
       "      <th>pair_B_UTR5</th>\n",
       "      <th>pair_B_UTR3</th>\n",
       "      <th>pair_B_CDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp28</td>\n",
       "      <td>1</td>\n",
       "      <td>1.529703</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAATATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp17</td>\n",
       "      <td>1</td>\n",
       "      <td>1.067821</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCTCCCTTCTATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp23</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.964888</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp33</td>\n",
       "      <td>1</td>\n",
       "      <td>1.091166</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Neocura-Fluc_Exp11</td>\n",
       "      <td>Neocura-Fluc_Exp04</td>\n",
       "      <td>1</td>\n",
       "      <td>1.040731</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0              pair_A              pair_B  pair_label  pair_diff  \\\n",
       "0           0  Neocura-Fluc_Exp11  Neocura-Fluc_Exp28           1   1.529703   \n",
       "1           1  Neocura-Fluc_Exp11  Neocura-Fluc_Exp17           1   1.067821   \n",
       "2           2  Neocura-Fluc_Exp11  Neocura-Fluc_Exp23          -1   0.964888   \n",
       "3           3  Neocura-Fluc_Exp11  Neocura-Fluc_Exp33           1   1.091166   \n",
       "4           4  Neocura-Fluc_Exp11  Neocura-Fluc_Exp04           1   1.040731   \n",
       "\n",
       "                                         pair_A_UTR5  \\\n",
       "0  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "3  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "4  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "\n",
       "                                         pair_A_UTR3  \\\n",
       "0  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                          pair_A_CDS  \\\n",
       "0  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "1  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "2  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "3  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "4  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "\n",
       "                                         pair_B_UTR5  \\\n",
       "0  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "3  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "4  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "\n",
       "                                         pair_B_UTR3  \\\n",
       "0  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "1  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "2  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "3  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4  CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "\n",
       "                                          pair_B_CDS  \n",
       "0  ATGGAGGACGCCAAGAATATCAAGAAGGGCCCAGCCCCATTCTATC...  \n",
       "1  ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCTCCCTTCTATC...  \n",
       "2  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTACC...  \n",
       "3  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCTTTCTACC...  \n",
       "4  ATGGAGGACGCCAAGAACATCAAGAAGGGCCCCGCCCCTTTCTACC...  "
      ]
     },
     "execution_count": 298,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "pair_CDS2_exp = pd.read_csv(f'data/interim/experiment_CDS2_pair_exp_cutoff{cutoff}.tsv', sep='\\t')\n",
    "pair_CDS2_exp = pair_CDS2_exp[pair_CDS2_exp['pair_label'] != 0]\n",
    "pair_CDS2_exp.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "0d5f81c7-d26b-4012-8939-c051dbdaceeb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 生成prompts\n",
    "\n",
    "prompts = []\n",
    "for index, row in pair_CDS2_exp.iterrows():\n",
    "    \n",
    "    for item in ['pair_A_CDS', 'pair_B_CDS']:\n",
    "        \n",
    "        ppt = str(Seq(row[item]).translate())\n",
    "        \n",
    "        for jtem in [ 'mRNA', 'UTR5','CDS', 'UTR3']:\n",
    "            \n",
    "            if jtem == 'mRNA':\n",
    "                pairA_token = '_'.join([row[f'pair_A_UTR5'], row[f'pair_A_CDS'], row[f'pair_A_UTR3']  ])\n",
    "                pairB_token = '_'.join([row[f'pair_B_UTR5'], row[f'pair_B_CDS'], row[f'pair_B_UTR3']  ])\n",
    "                exact_diff = (pairA_token == pairB_token)\n",
    "                roughly_diff = (pairA_token == pairB_token) \n",
    "            else:\n",
    "                exact_diff = all([row[f'pair_A_{jtem}']==row[f'pair_B_{jtem}'] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "                roughly_diff = all([row[f'pair_A_{tem}']==row[f'pair_B_{tem}'] or row[f'pair_A_{tem}'][4:-4]==row[f'pair_B_{tem}'][4:-4] for tem in ['UTR5', 'UTR3','CDS'] if tem!=jtem])\n",
    "                pairA_token = row[f'pair_A_{jtem}']\n",
    "                pairB_token = row[f'pair_B_{jtem}']\n",
    "                \n",
    "            label = row['pair_label']\n",
    "            value = row['pair_diff']\n",
    "            g_type=jtem\n",
    "            origin = 'cds2_exp'\n",
    "            \n",
    "            prompts.append([ppt, pairA_token, pairB_token, g_type, origin, label, value, exact_diff, roughly_diff])\n",
    "            \n",
    "            \n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "id": "75e4f258-9649-4f17-a620-c183a5874670",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>exact_diff</th>\n",
       "      <th>roughly_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.529703</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAATATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.529703</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.067821</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...</td>\n",
       "      <td>ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCTCCCTTCTATC...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.067821</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.964888</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70472</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70480</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70488</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.250000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70496</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.363636</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70504</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>AGGAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds2_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.153846</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16380 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  prompt  \\\n",
       "0      MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...   \n",
       "2      MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...   \n",
       "8      MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...   \n",
       "10     MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...   \n",
       "16     MEDAKNIKKGPAPFYPLEDGTAGEQLHKAMKRYALVPGTIAFTDAH...   \n",
       "...                                                  ...   \n",
       "70472  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "70480  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "70488  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "70496  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "70504  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "\n",
       "                                                  pair_A  \\\n",
       "0      AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2      ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "8      AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "10     ATGGAGGACGCCAAGAACATCAAGAAGGGCCCAGCCCCATTCTATC...   \n",
       "16     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "...                                                  ...   \n",
       "70472  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "70480  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "70488  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "70496  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "70504  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "\n",
       "                                                  pair_B seq_type    origin  \\\n",
       "0      AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...     mRNA  cds2_exp   \n",
       "2      ATGGAGGACGCCAAGAATATCAAGAAGGGCCCAGCCCCATTCTATC...      CDS  cds2_exp   \n",
       "8      AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...     mRNA  cds2_exp   \n",
       "10     ATGGAGGACGCCAAGAACATTAAGAAGGGCCCCGCTCCCTTCTATC...      CDS  cds2_exp   \n",
       "16     AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...     mRNA  cds2_exp   \n",
       "...                                                  ...      ...       ...   \n",
       "70472  AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGC...     mRNA  cds2_exp   \n",
       "70480  AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGC...     mRNA  cds2_exp   \n",
       "70488  AGGGGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGC...     mRNA  cds2_exp   \n",
       "70496  AGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...     mRNA  cds2_exp   \n",
       "70504  AGGAAGCGGTACCGCGGTTGTTACGCTGCGTTTGTTCGTTGCTTTT...     mRNA  cds2_exp   \n",
       "\n",
       "       pair_label  pair_diff  exact_diff  roughly_diff  \n",
       "0               1   1.529703       False         False  \n",
       "2               1   1.529703       False          True  \n",
       "8               1   1.067821       False         False  \n",
       "10              1   1.067821       False          True  \n",
       "16             -1   0.964888       False         False  \n",
       "...           ...        ...         ...           ...  \n",
       "70472           1   1.500000       False         False  \n",
       "70480           1  15.000000       False         False  \n",
       "70488           1   1.250000       False         False  \n",
       "70496           1   1.363636       False         False  \n",
       "70504           1   1.153846       False         False  \n",
       "\n",
       "[16380 rows x 9 columns]"
      ]
     },
     "execution_count": 306,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "CDS2_prompts = pd.DataFrame(prompts, columns=['prompt', 'pair_A', 'pair_B', 'seq_type', 'origin', 'pair_label', 'pair_diff', 'exact_diff', 'roughly_diff'])\n",
    "\n",
    "CDS2_prompts_raw = CDS2_prompts\n",
    "\n",
    "CDS2_prompts = CDS2_prompts.drop_duplicates(subset=['prompt', 'pair_A', 'pair_B', 'seq_type'])\n",
    "\n",
    "CDS2_prompts = CDS2_prompts[ \n",
    "                                    CDS2_prompts['pair_A']  != CDS2_prompts['pair_B']\n",
    "                                    ]\n",
    "\n",
    "CDS2_prompts\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adfee92a",
   "metadata": {},
   "source": [
    "##### merge all prompts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "d85e2eaa-a6ca-44ed-a6a4-7e77e07e5cfe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>exact_diff</th>\n",
       "      <th>roughly_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1755</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.164065</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1756</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.164065</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1757</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.164065</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1758</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.164065</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1759</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.164065</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197280 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 prompt  \\\n",
       "0     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "2     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "3     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "4     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "...                                                 ...   \n",
       "1755  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1756  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1757  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1758  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1759  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "\n",
       "                                                 pair_A  \\\n",
       "0     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "2     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...   \n",
       "3     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...   \n",
       "4     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "...                                                 ...   \n",
       "1755  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "1756  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "1757  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "1758  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...   \n",
       "1759  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...   \n",
       "\n",
       "                                                 pair_B seq_type    origin  \\\n",
       "0     GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...     mRNA  utr5_exp   \n",
       "1     GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...     UTR5  utr5_exp   \n",
       "2     ATGGTGAGCAAGGGCGAGGAGCTGTTCACAGGCGTGGTGCCTATCC...      CDS  utr5_exp   \n",
       "3     CTGGTACTGCATGCACGCAATGCTAGCTGCCCCTTTCCCGTCCTGG...     UTR3  utr5_exp   \n",
       "4     GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...     mRNA  utr5_exp   \n",
       "...                                                 ...      ...       ...   \n",
       "1755  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...     UTR3   cds_exp   \n",
       "1756  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...     mRNA   cds_exp   \n",
       "1757  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...     UTR5   cds_exp   \n",
       "1758  ATGGTGAGCAAGGGCGAGGAGTTGTTCACCGGCGTGGTCCCAATTC...      CDS   cds_exp   \n",
       "1759  GTGGTCCCAATTCTGGTCGAGCTCACGGACAGTTCTGTACCAAACC...     UTR3   cds_exp   \n",
       "\n",
       "      pair_label  pair_diff  exact_diff  roughly_diff  \n",
       "0              1   3.300000       False         False  \n",
       "1              1   3.300000       False          True  \n",
       "2              1   3.300000        True         False  \n",
       "3              1   3.300000        True         False  \n",
       "4              1   3.300000       False         False  \n",
       "...          ...        ...         ...           ...  \n",
       "1755           1   1.164065       False          True  \n",
       "1756           1   1.164065       False         False  \n",
       "1757           1   1.164065        True         False  \n",
       "1758           1   1.164065        True         False  \n",
       "1759           1   1.164065       False          True  \n",
       "\n",
       "[197280 rows x 9 columns]"
      ]
     },
     "execution_count": 307,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# raw prompots\n",
    "\n",
    "# merge all prompts\n",
    "\n",
    "merged_prompts_raw = pd.concat([UTR5_prompts_raw, UTR3_prompts_raw, CDS2_prompts_raw, CDS_prompts_raw, ])\n",
    "\n",
    "merged_prompts_raw\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "ea1f795b-d48c-4d95-9574-df087b8e6ce2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "merged_prompts_raw.to_csv(f'data/rf_mRNA_small_cutoff{cutoff}_prompts_raw.csv', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "2a6175d6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>exact_diff</th>\n",
       "      <th>roughly_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.300000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.650000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>UTR5</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.650000</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...</td>\n",
       "      <td>GGGAATCAACACAGACACCAGACCACCAGCACCAGAACACACAGAC...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.650000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1720</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.021720</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1728</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.415057</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1736</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>3.160315</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGCTCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>20.071805</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1752</th>\n",
       "      <td>MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>cds_exp</td>\n",
       "      <td>1</td>\n",
       "      <td>1.164065</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34210 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 prompt  \\\n",
       "0     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "8     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "9     MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "16    MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "...                                                 ...   \n",
       "1720  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1728  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1736  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1744  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "1752  MVSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLK...   \n",
       "\n",
       "                                                 pair_A  \\\n",
       "0     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "1     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "8     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "9     GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "16    GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...   \n",
       "...                                                 ...   \n",
       "1720  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "1728  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "1736  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "1744  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "1752  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...   \n",
       "\n",
       "                                                 pair_B seq_type    origin  \\\n",
       "0     GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...     mRNA  utr5_exp   \n",
       "1     GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCG...     UTR5  utr5_exp   \n",
       "8     GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...     mRNA  utr5_exp   \n",
       "9     GGGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAAC...     UTR5  utr5_exp   \n",
       "16    GGGAATCAACACAGACACCAGACCACCAGCACCAGAACACACAGAC...     mRNA  utr5_exp   \n",
       "...                                                 ...      ...       ...   \n",
       "1720  AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA   cds_exp   \n",
       "1728  GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...     mRNA   cds_exp   \n",
       "1736  GGCAAGGGCGAGTTCACCGGCGTGGTCCCAATTCTGGTCGAGCTCG...     mRNA   cds_exp   \n",
       "1744  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGCCGAGCTCG...     mRNA   cds_exp   \n",
       "1752  GGCAAGGGCGAGTTCACCGGGCTGGTCCCAATTCTGGTCGAGCTCA...     mRNA   cds_exp   \n",
       "\n",
       "      pair_label  pair_diff  exact_diff  roughly_diff  \n",
       "0              1   3.300000       False         False  \n",
       "1              1   3.300000       False          True  \n",
       "8              1   1.650000       False         False  \n",
       "9              1   1.650000       False          True  \n",
       "16             1   1.650000       False         False  \n",
       "...          ...        ...         ...           ...  \n",
       "1720           1   1.021720       False         False  \n",
       "1728           1   3.415057       False         False  \n",
       "1736           1   3.160315       False         False  \n",
       "1744           1  20.071805       False         False  \n",
       "1752           1   1.164065       False         False  \n",
       "\n",
       "[34210 rows x 9 columns]"
      ]
     },
     "execution_count": 308,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge all prompts\n",
    "\n",
    "# merged_prompts = pd.concat([UTR5_prompts, UTR3_prompts, CDS_prompts, cellines_prompts])\n",
    "\n",
    "\n",
    "merged_prompts = pd.concat([UTR5_prompts, UTR3_prompts, CDS2_prompts, CDS_prompts])\n",
    "\n",
    "merged_prompts = merged_prompts.drop_duplicates(subset=['prompt', 'pair_A', 'pair_B', 'seq_type'])\n",
    "\n",
    "merged_prompts\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "b70968b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UTR3    5428\n",
       "UTR5    2242\n",
       "CDS       86\n",
       "Name: seq_type, dtype: int64"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "merged_prompts['pair_label'].value_counts()\n",
    "merged_prompts['seq_type'].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "9282a862",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "merged_prompts.to_csv(f'data/rf_mRNA_small_cutoff{cutoff}_prompts.csv', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a17948a-f850-4070-adba-f259ad1596f2",
   "metadata": {},
   "source": [
    "##### 去重 & 分析"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d163ccc9-9f61-4a3b-8140-7c2c6c07a57c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "merged_prompts_raw = pd.read_csv(f'data/rf_mRNA_small_cutoff0_prompts_raw.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2429105f-f989-435b-8fe9-591340831828",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(197280, 9)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "merged_prompts_raw.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3ab09a69-9d20-4585-91e5-eeb67db6fa3c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(197280, 10)\n",
      "(34251, 6)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "rank = 2\n",
    "\n",
    "merged_prompts_ranked = merged_prompts_raw\n",
    "merged_prompts_ranked.loc[merged_prompts_ranked['seq_type']=='mRNA', 'exact_diff'] = True # 预处理出错，此处矫正\n",
    "merged_prompts_ranked.loc[merged_prompts_ranked['seq_type']=='mRNA', 'roughly_diff'] = True # 预处理出错，此处矫正\n",
    "\n",
    "def ranked(x):\n",
    "    \n",
    "    if x > rank:\n",
    "        return 1\n",
    "    elif x <1/rank:\n",
    "        return -1\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "merged_prompts_ranked['pair_diff_label'] = merged_prompts_ranked['pair_diff'].apply(ranked) \n",
    "\n",
    "print(merged_prompts_ranked.shape)\n",
    "\n",
    "# vote to get the concensus value\n",
    "from scipy import stats\n",
    "\n",
    "merged_prompts_ranked_grouped = merged_prompts_ranked.groupby(by=['prompt', 'pair_A', 'pair_B', 'seq_type']).agg(list)\n",
    "\n",
    "print(merged_prompts_ranked_grouped.shape)\n",
    "\n",
    "merged_prompts_ranked_grouped['pair_diff_label'] = merged_prompts_ranked_grouped['pair_diff_label'].apply(lambda x: stats.mode(x)[0][0])\n",
    "merged_prompts_ranked_grouped['exact_diff'] = merged_prompts_ranked_grouped['exact_diff'].apply(lambda x: stats.mode(x)[0][0])\n",
    "\n",
    "merged_prompts_ranked_grouped['roughly_diff'] = merged_prompts_ranked_grouped['roughly_diff'].apply(lambda x: stats.mode(x)[0][0])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cbdb2d9f-b088-4490-9728-513cf7ab3185",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>pair_A</th>\n",
       "      <th>pair_B</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>origin</th>\n",
       "      <th>pair_label</th>\n",
       "      <th>pair_diff</th>\n",
       "      <th>exact_diff</th>\n",
       "      <th>roughly_diff</th>\n",
       "      <th>pair_diff_label</th>\n",
       "      <th>pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.418709</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.418709</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.434997</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.434997</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr3_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.4333</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13001</th>\n",
       "      <td>SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...</td>\n",
       "      <td>GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...</td>\n",
       "      <td>GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.53103</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13002</th>\n",
       "      <td>SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...</td>\n",
       "      <td>GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...</td>\n",
       "      <td>GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.422966</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13003</th>\n",
       "      <td>SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...</td>\n",
       "      <td>GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...</td>\n",
       "      <td>GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.422966</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13004</th>\n",
       "      <td>SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...</td>\n",
       "      <td>GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...</td>\n",
       "      <td>GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.457813</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13005</th>\n",
       "      <td>SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...</td>\n",
       "      <td>GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...</td>\n",
       "      <td>GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>utr5_exp</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.457813</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1</td>\n",
       "      <td>[GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12998 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  prompt  \\\n",
       "0      MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "1      MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "2      MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "3      MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "4      MRVTAPRTLILLLSGALALTETWAGSGGGGSGGCSVDNCRLFISGI...   \n",
       "...                                                  ...   \n",
       "13001  SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...   \n",
       "13002  SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...   \n",
       "13003  SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...   \n",
       "13004  SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...   \n",
       "13005  SRVQRVTDSLDTLLEISSNKQNHSYEYFMKQMNDARHGGWTTKMDW...   \n",
       "\n",
       "                                                  pair_A  \\\n",
       "0      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "1      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "2      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "3      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "4      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...   \n",
       "...                                                  ...   \n",
       "13001  GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...   \n",
       "13002  GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...   \n",
       "13003  GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...   \n",
       "13004  GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...   \n",
       "13005  GGAGTCTCCCCTCGCCGCATCCACTCTCCGGCCGGCCGCCTGCCCG...   \n",
       "\n",
       "                                                  pair_B seq_type    origin  \\\n",
       "0      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "1      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "2      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "3      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "4      AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCG...     mRNA  utr3_exp   \n",
       "...                                                  ...      ...       ...   \n",
       "13001  GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...     mRNA  utr5_exp   \n",
       "13002  GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...     mRNA  utr5_exp   \n",
       "13003  GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...     mRNA  utr5_exp   \n",
       "13004  GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...     mRNA  utr5_exp   \n",
       "13005  GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGCT...     mRNA  utr5_exp   \n",
       "\n",
       "      pair_label pair_diff  exact_diff  roughly_diff  pair_diff_label  \\\n",
       "0             -1  0.418709        True          True               -1   \n",
       "1             -1  0.418709        True          True               -1   \n",
       "2             -1  0.434997        True          True               -1   \n",
       "3             -1  0.434997        True          True               -1   \n",
       "4             -1    0.4333        True          True               -1   \n",
       "...          ...       ...         ...           ...              ...   \n",
       "13001         -1   0.53103        True          True               -1   \n",
       "13002         -1  0.422966        True          True               -1   \n",
       "13003         -1  0.422966        True          True               -1   \n",
       "13004         -1  0.457813        True          True               -1   \n",
       "13005         -1  0.457813        True          True               -1   \n",
       "\n",
       "                                                   pairs  \n",
       "0      [AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...  \n",
       "1      [AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...  \n",
       "2      [AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...  \n",
       "3      [AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...  \n",
       "4      [AATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCC...  \n",
       "...                                                  ...  \n",
       "13001  [GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...  \n",
       "13002  [GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...  \n",
       "13003  [GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...  \n",
       "13004  [GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...  \n",
       "13005  [GGAAAGCGGTATCGGTGTGCGAGTCTTGCGGTTGTCTCTCTTCGC...  \n",
       "\n",
       "[12998 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#v过滤\n",
    "\n",
    "merged_prompts_ranked_grouped_explode = merged_prompts_ranked_grouped[\n",
    "    (merged_prompts_ranked_grouped['pair_diff_label']==-1) &\n",
    "    (merged_prompts_ranked_grouped['roughly_diff']==True)\n",
    "].explode(['origin', 'pair_label','pair_diff'])\n",
    "\n",
    "# merged_prompts_ranked_grouped_explode = merged_prompts_ranked_grouped[\n",
    "#     (merged_prompts_ranked_grouped['pair_diff_label']==-1)\n",
    "# ].explode(['origin', 'pair_label', 'pair_diff'])\n",
    "\n",
    "\n",
    "\n",
    "merged_prompts_ranked_grouped_explode = merged_prompts_ranked_grouped_explode.reset_index()\n",
    "\n",
    "merged_prompts_ranked_grouped_explode = merged_prompts_ranked_grouped_explode[\n",
    "        merged_prompts_ranked_grouped_explode['pair_A'] != merged_prompts_ranked_grouped_explode['pair_B'] \n",
    "                                                                                        ]\n",
    "\n",
    "import numpy as np\n",
    "     \n",
    "# 去除重复对    \n",
    "merged_prompts_ranked_grouped_explode['pairs'] = merged_prompts_ranked_grouped_explode.apply(lambda x: np.sort([x['prompt'], x['pair_A'], x['pair_B']]), axis=1)\n",
    "\n",
    "merged_prompts_ranked_grouped_explode_dropped = merged_prompts_ranked_grouped_explode.drop_duplicates(subset= ['pairs'])\n",
    "\n",
    "merged_prompts_ranked_grouped_explode_dropped\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "78963f90-6681-42b1-987a-ec3e3fd9f568",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mRNA    7598\n",
       "UTR3    3514\n",
       "UTR5    1142\n",
       "CDS      744\n",
       "Name: seq_type, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "merged_prompts_ranked_grouped_explode_dropped['seq_type'].value_counts()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "72a8eed5-517f-4d2c-82a5-909e4c94af25",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mRNA    11986\n",
       "UTR3     4670\n",
       "CDS      2320\n",
       "UTR5     2028\n",
       "Name: seq_type, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "merged_prompts_ranked_grouped_explode_dropped['seq_type'].value_counts()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dd6ef42c-940f-4dff-aaf3-6929a8d60086",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mRNA    18596\n",
       "UTR3     5684\n",
       "UTR5     4524\n",
       "CDS      4414\n",
       "Name: seq_type, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# analysis\n",
    "\n",
    "merged_prompts_ranked_grouped_explode_dropped['seq_type'].value_counts()\n",
    "# merged_prompts_ranked_grouped_explode_dropped['pair_diff_label'].value_counts()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a78e2921-c910-47f4-9bde-6d6ebb11dbc0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "rank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c0492f4b-8096-40f0-9e6f-f8fcd6c4117b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# save to file\n",
    "\n",
    "merged_prompts_ranked_grouped_explode_dropped.to_csv(f'data/rf_mRNA_large_rank{rank}_roughly_diff_prompts.csv')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3b1d89f5-1775-4d37-93f9-a730a605a497",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21004, 12)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load\n",
    "\n",
    "rank = 1.5\n",
    "\n",
    "merged_prompts_ranked_preference = pd.read_csv(f'data/rf_mRNA_large_rank{rank}_roughly_diff_prompts.csv')\n",
    "\n",
    "merged_prompts_ranked_preference.shape\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8f746192-03dc-4af4-a1cb-0b903f6c35c2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt_len</th>\n",
       "      <th>pair_A_len</th>\n",
       "      <th>pair_B_len</th>\n",
       "      <th>gen_seq_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>21004.00</td>\n",
       "      <td>21004.00</td>\n",
       "      <td>21004.00</td>\n",
       "      <td>21004.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>305.44</td>\n",
       "      <td>870.07</td>\n",
       "      <td>853.57</td>\n",
       "      <td>1175.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>100.50</td>\n",
       "      <td>530.27</td>\n",
       "      <td>546.86</td>\n",
       "      <td>565.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>125.00</td>\n",
       "      <td>46.00</td>\n",
       "      <td>46.00</td>\n",
       "      <td>171.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>286.00</td>\n",
       "      <td>292.00</td>\n",
       "      <td>214.00</td>\n",
       "      <td>586.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>286.00</td>\n",
       "      <td>1035.00</td>\n",
       "      <td>1022.00</td>\n",
       "      <td>1314.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>286.00</td>\n",
       "      <td>1221.00</td>\n",
       "      <td>1221.00</td>\n",
       "      <td>1507.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>811.00</td>\n",
       "      <td>2791.00</td>\n",
       "      <td>2791.00</td>\n",
       "      <td>3602.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       prompt_len  pair_A_len  pair_B_len  gen_seq_len\n",
       "count    21004.00    21004.00    21004.00     21004.00\n",
       "mean       305.44      870.07      853.57      1175.51\n",
       "std        100.50      530.27      546.86       565.44\n",
       "min        125.00       46.00       46.00       171.00\n",
       "25%        286.00      292.00      214.00       586.00\n",
       "50%        286.00     1035.00     1022.00      1314.00\n",
       "75%        286.00     1221.00     1221.00      1507.00\n",
       "max        811.00     2791.00     2791.00      3602.00"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "merged_prompts_ranked_preference['seq_type'].value_counts()\n",
    "\n",
    "for col in ['prompt',\t'pair_A',\t'pair_B']:\n",
    "    merged_prompts_ranked_preference[col+'_len'] = merged_prompts_ranked_preference[col].apply(len)\n",
    "    \n",
    "merged_prompts_ranked_preference['gen_seq'] = merged_prompts_ranked_preference['prompt'] + merged_prompts_ranked_preference['pair_A']\n",
    "merged_prompts_ranked_preference['gen_seq_len'] = merged_prompts_ranked_preference['gen_seq'].apply(len)\n",
    "\n",
    "merged_prompts_ranked_preference[[col for col in merged_prompts_ranked_preference.columns if '_len' in col]].describe().round(2)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
